{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import multiprocessing\n",
    "import os\n",
    "import re\n",
    "import signal\n",
    "from math import ceil\n",
    "from os.path import join\n",
    "\n",
    "import numpy as np\n",
    "import torch\n",
    "from numpy.random import choice\n",
    "from torchtext.data import Field, TabularDataset\n",
    "\n",
    "import glob\n",
    "from io import StringIO\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists('model_data/data'):\n",
    "    os.mkdir('model_data/data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = 'model_data/data/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## sklearn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "sk_glob = sorted(glob.glob('sklearn/*'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sklearn/sklearn_1_10_decision_trees_1_10_1_classification.txt',\n",
       " 'sklearn/sklearn_1_10_decision_trees_1_10_2_regression.txt',\n",
       " 'sklearn/sklearn_1_10_decision_trees_1_10_3_multi-output_problems.txt',\n",
       " 'sklearn/sklearn_1_10_decision_trees_1_10_4_complexity.txt',\n",
       " 'sklearn/sklearn_1_10_decision_trees_1_10_5_tips_on_practical_use.txt',\n",
       " 'sklearn/sklearn_1_10_decision_trees_1_10_6_tree_algorithms_id3_c4_5_c5_0_and_cart.txt',\n",
       " 'sklearn/sklearn_1_10_decision_trees_1_10_7_mathematical_formulation.txt',\n",
       " 'sklearn/sklearn_1_10_decision_trees_1_10_8_minimal_cost-complexity_pruning.txt',\n",
       " 'sklearn/sklearn_1_11_ensemble_methods_1_11_1_bagging_meta-estimator.txt',\n",
       " 'sklearn/sklearn_1_11_ensemble_methods_1_11_2_forests_of_randomized_trees.txt',\n",
       " 'sklearn/sklearn_1_11_ensemble_methods_1_11_3_adaboost.txt',\n",
       " 'sklearn/sklearn_1_11_ensemble_methods_1_11_4_gradient_tree_boosting.txt',\n",
       " 'sklearn/sklearn_1_11_ensemble_methods_1_11_5_histogram-based_gradient_boosting.txt',\n",
       " 'sklearn/sklearn_1_11_ensemble_methods_1_11_6_voting_classifier.txt',\n",
       " 'sklearn/sklearn_1_11_ensemble_methods_1_11_7_voting_regressor.txt',\n",
       " 'sklearn/sklearn_1_11_ensemble_methods_1_11_8_stacked_generalization.txt',\n",
       " 'sklearn/sklearn_1_12_multiclass_and_multilabel_algorithms_1_12_1_multilabel_classification_format.txt',\n",
       " 'sklearn/sklearn_1_12_multiclass_and_multilabel_algorithms_1_12_2_one-vs-the-rest.txt',\n",
       " 'sklearn/sklearn_1_12_multiclass_and_multilabel_algorithms_1_12_3_one-vs-one.txt',\n",
       " 'sklearn/sklearn_1_12_multiclass_and_multilabel_algorithms_1_12_4_error-correcting_output-codes.txt',\n",
       " 'sklearn/sklearn_1_12_multiclass_and_multilabel_algorithms_1_12_5_multioutput_regression.txt',\n",
       " 'sklearn/sklearn_1_12_multiclass_and_multilabel_algorithms_1_12_6_multioutput_classification.txt',\n",
       " 'sklearn/sklearn_1_12_multiclass_and_multilabel_algorithms_1_12_7_classifier_chain.txt',\n",
       " 'sklearn/sklearn_1_12_multiclass_and_multilabel_algorithms_1_12_8_regressor_chain.txt',\n",
       " 'sklearn/sklearn_1_13_feature_selection_1_13_1_removing_features_with_low_variance.txt',\n",
       " 'sklearn/sklearn_1_13_feature_selection_1_13_2_univariate_feature_selection.txt',\n",
       " 'sklearn/sklearn_1_13_feature_selection_1_13_3_recursive_feature_elimination.txt',\n",
       " 'sklearn/sklearn_1_13_feature_selection_1_13_4_feature_selection_using_selectfrommodel.txt',\n",
       " 'sklearn/sklearn_1_13_feature_selection_1_13_5_feature_selection_as_part_of_a_pipeline.txt',\n",
       " 'sklearn/sklearn_1_14_semi-supervised_1_14_1_label_propagation.txt',\n",
       " 'sklearn/sklearn_1_16_probability_calibration_1_16_1_calibration_curves.txt',\n",
       " 'sklearn/sklearn_1_16_probability_calibration_1_16_2_calibrating_a_classifier.txt',\n",
       " 'sklearn/sklearn_1_16_probability_calibration_1_16_3_usage.txt',\n",
       " 'sklearn/sklearn_1_17_neural_network_models_supervised_1_17_1_multi-layer_perceptron.txt',\n",
       " 'sklearn/sklearn_1_17_neural_network_models_supervised_1_17_2_classification.txt',\n",
       " 'sklearn/sklearn_1_17_neural_network_models_supervised_1_17_3_regression.txt',\n",
       " 'sklearn/sklearn_1_17_neural_network_models_supervised_1_17_4_regularization.txt',\n",
       " 'sklearn/sklearn_1_17_neural_network_models_supervised_1_17_5_algorithms.txt',\n",
       " 'sklearn/sklearn_1_17_neural_network_models_supervised_1_17_6_complexity.txt',\n",
       " 'sklearn/sklearn_1_17_neural_network_models_supervised_1_17_7_mathematical_formulation.txt',\n",
       " 'sklearn/sklearn_1_17_neural_network_models_supervised_1_17_8_tips_on_practical_use.txt',\n",
       " 'sklearn/sklearn_1_17_neural_network_models_supervised_1_17_9_more_control_with_warm_start.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_10_bayesian_regression.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_11_logistic_regression.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_12_generalized_linear_regression.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_13_stochastic_gradient_descent_-_sgd.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_14_perceptron.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_15_passive_aggressive_algorithms.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_16_robustness_regression_outliers_and_modeling_errors.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_17_polynomial_regression_extending_linear_models_with_basis_functions.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_1_ordinary_least_squares.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_2_ridge_regression_and_classification.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_3_lasso.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_4_multi-task_lasso.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_5_elastic-net.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_6_multi-task_elastic-net.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_7_least_angle_regression.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_8_lars_lasso.txt',\n",
       " 'sklearn/sklearn_1_1_linear_models_1_1_9_orthogonal_matching_pursuit_omp.txt',\n",
       " 'sklearn/sklearn_1_2_linear_and_quadratic_discriminant_analysis_1_2_1_dimensionality_reduction_using_linear_discriminant_analysis.txt',\n",
       " 'sklearn/sklearn_1_2_linear_and_quadratic_discriminant_analysis_1_2_2_mathematical_formulation_of_the_lda_and_qda_classifiers.txt',\n",
       " 'sklearn/sklearn_1_2_linear_and_quadratic_discriminant_analysis_1_2_3_mathematical_formulation_of_lda_dimensionality_reduction.txt',\n",
       " 'sklearn/sklearn_1_2_linear_and_quadratic_discriminant_analysis_1_2_4_shrinkage.txt',\n",
       " 'sklearn/sklearn_1_2_linear_and_quadratic_discriminant_analysis_1_2_5_estimation_algorithms.txt',\n",
       " 'sklearn/sklearn_1_4_support_vector_machines_1_4_1_classification.txt',\n",
       " 'sklearn/sklearn_1_4_support_vector_machines_1_4_2_regression.txt',\n",
       " 'sklearn/sklearn_1_4_support_vector_machines_1_4_3_density_estimation_novelty_detection.txt',\n",
       " 'sklearn/sklearn_1_4_support_vector_machines_1_4_4_complexity.txt',\n",
       " 'sklearn/sklearn_1_4_support_vector_machines_1_4_5_tips_on_practical_use.txt',\n",
       " 'sklearn/sklearn_1_4_support_vector_machines_1_4_6_kernel_functions.txt',\n",
       " 'sklearn/sklearn_1_4_support_vector_machines_1_4_7_mathematical_formulation.txt',\n",
       " 'sklearn/sklearn_1_4_support_vector_machines_1_4_8_implementation_details.txt',\n",
       " 'sklearn/sklearn_1_5_stochastic_gradient_descent_1_5_1_classification.txt',\n",
       " 'sklearn/sklearn_1_5_stochastic_gradient_descent_1_5_2_regression.txt',\n",
       " 'sklearn/sklearn_1_5_stochastic_gradient_descent_1_5_3_stochastic_gradient_descent_for_sparse_data.txt',\n",
       " 'sklearn/sklearn_1_5_stochastic_gradient_descent_1_5_4_complexity.txt',\n",
       " 'sklearn/sklearn_1_5_stochastic_gradient_descent_1_5_5_stopping_criterion.txt',\n",
       " 'sklearn/sklearn_1_5_stochastic_gradient_descent_1_5_6_tips_on_practical_use.txt',\n",
       " 'sklearn/sklearn_1_5_stochastic_gradient_descent_1_5_7_mathematical_formulation.txt',\n",
       " 'sklearn/sklearn_1_5_stochastic_gradient_descent_1_5_8_implementation_details.txt',\n",
       " 'sklearn/sklearn_1_6_nearest_neighbors_1_6_1_unsupervised_nearest_neighbors.txt',\n",
       " 'sklearn/sklearn_1_6_nearest_neighbors_1_6_2_nearest_neighbors_classification.txt',\n",
       " 'sklearn/sklearn_1_6_nearest_neighbors_1_6_3_nearest_neighbors_regression.txt',\n",
       " 'sklearn/sklearn_1_6_nearest_neighbors_1_6_4_nearest_neighbor_algorithms.txt',\n",
       " 'sklearn/sklearn_1_6_nearest_neighbors_1_6_5_nearest_centroid_classifier.txt',\n",
       " 'sklearn/sklearn_1_6_nearest_neighbors_1_6_6_nearest_neighbors_transformer.txt',\n",
       " 'sklearn/sklearn_1_6_nearest_neighbors_1_6_7_neighborhood_components_analysis.txt',\n",
       " 'sklearn/sklearn_1_7_gaussian_processes_1_7_1_gaussian_process_regression_gpr.txt',\n",
       " 'sklearn/sklearn_1_7_gaussian_processes_1_7_2_gpr_examples.txt',\n",
       " 'sklearn/sklearn_1_7_gaussian_processes_1_7_3_gaussian_process_classification_gpc.txt',\n",
       " 'sklearn/sklearn_1_7_gaussian_processes_1_7_4_gpc_examples.txt',\n",
       " 'sklearn/sklearn_1_7_gaussian_processes_1_7_5_kernels_for_gaussian_processes.txt',\n",
       " 'sklearn/sklearn_1_9_naive_bayes_1_9_1_gaussian_naive_bayes.txt',\n",
       " 'sklearn/sklearn_1_9_naive_bayes_1_9_2_multinomial_naive_bayes.txt',\n",
       " 'sklearn/sklearn_1_9_naive_bayes_1_9_3_complement_naive_bayes.txt',\n",
       " 'sklearn/sklearn_1_9_naive_bayes_1_9_4_bernoulli_naive_bayes.txt',\n",
       " 'sklearn/sklearn_1_9_naive_bayes_1_9_5_categorical_naive_bayes.txt',\n",
       " 'sklearn/sklearn_1_9_naive_bayes_1_9_6_out-of-core_naive_bayes_model_fitting.txt',\n",
       " 'sklearn/sklearn_2_1_gaussian_mixture_models_2_1_1_gaussian_mixture.txt',\n",
       " 'sklearn/sklearn_2_1_gaussian_mixture_models_2_1_2_variational_bayesian_gaussian_mixture.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_10_tips_on_practical_use.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_1_introduction.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_2_isomap.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_3_locally_linear_embedding.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_4_modified_locally_linear_embedding.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_5_hessian_eigenmapping.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_6_spectral_embedding.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_7_local_tangent_space_alignment.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_8_multi-dimensional_scaling_mds.txt',\n",
       " 'sklearn/sklearn_2_2_manifold_learning_2_2_9_t-distributed_stochastic_neighbor_embedding_t-sne.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_10_clustering_performance_evaluation.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_1_overview_of_clustering_methods.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_2_k-means.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_3_affinity_propagation.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_4_mean_shift.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_5_spectral_clustering.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_6_hierarchical_clustering.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_7_dbscan.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_8_optics.txt',\n",
       " 'sklearn/sklearn_2_3_clustering_2_3_9_birch.txt',\n",
       " 'sklearn/sklearn_2_4_biclustering_2_4_1_spectral_co-clustering.txt',\n",
       " 'sklearn/sklearn_2_4_biclustering_2_4_2_spectral_biclustering.txt',\n",
       " 'sklearn/sklearn_2_4_biclustering_2_4_3_biclustering_evaluation.txt',\n",
       " 'sklearn/sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems_2_5_1_principal_component_analysis_pca.txt',\n",
       " 'sklearn/sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems_2_5_2_truncated_singular_value_decomposition_and_latent_semantic_analysis.txt',\n",
       " 'sklearn/sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems_2_5_3_dictionary_learning.txt',\n",
       " 'sklearn/sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems_2_5_4_factor_analysis.txt',\n",
       " 'sklearn/sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems_2_5_5_independent_component_analysis_ica.txt',\n",
       " 'sklearn/sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems_2_5_6_non-negative_matrix_factorization_nmf_or_nnmf.txt',\n",
       " 'sklearn/sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems_2_5_7_latent_dirichlet_allocation_lda.txt',\n",
       " 'sklearn/sklearn_2_6_covariance_estimation_2_6_1_empirical_covariance.txt',\n",
       " 'sklearn/sklearn_2_6_covariance_estimation_2_6_2_shrunk_covariance.txt',\n",
       " 'sklearn/sklearn_2_6_covariance_estimation_2_6_3_sparse_inverse_covariance.txt',\n",
       " 'sklearn/sklearn_2_6_covariance_estimation_2_6_4_robust_covariance_estimation.txt',\n",
       " 'sklearn/sklearn_2_7_novelty_and_outlier_detection_2_7_1_overview_of_outlier_detection_methods.txt',\n",
       " 'sklearn/sklearn_2_7_novelty_and_outlier_detection_2_7_2_novelty_detection.txt',\n",
       " 'sklearn/sklearn_2_7_novelty_and_outlier_detection_2_7_3_outlier_detection.txt',\n",
       " 'sklearn/sklearn_2_7_novelty_and_outlier_detection_2_7_4_novelty_detection_with_local_outlier_factor.txt',\n",
       " 'sklearn/sklearn_2_8_density_estimation_2_8_1_density_estimation_histograms.txt',\n",
       " 'sklearn/sklearn_2_8_density_estimation_2_8_2_kernel_density_estimation.txt',\n",
       " 'sklearn/sklearn_2_9_neural_network_models_unsupervised_2_9_1_restricted_boltzmann_machines.txt',\n",
       " 'sklearn/sklearn_3_1_cross-validation_evaluating_estimator_performance_3_1_1_computing_cross-validated_metrics.txt',\n",
       " 'sklearn/sklearn_3_1_cross-validation_evaluating_estimator_performance_3_1_2_cross_validation_iterators.txt',\n",
       " 'sklearn/sklearn_3_1_cross-validation_evaluating_estimator_performance_3_1_3_a_note_on_shuffling.txt',\n",
       " 'sklearn/sklearn_3_1_cross-validation_evaluating_estimator_performance_3_1_4_cross_validation_and_model_selection.txt',\n",
       " 'sklearn/sklearn_3_2_tuning_the_hyper-parameters_of_an_estimator_3_2_1_exhaustive_grid_search.txt',\n",
       " 'sklearn/sklearn_3_2_tuning_the_hyper-parameters_of_an_estimator_3_2_2_randomized_parameter_optimization.txt',\n",
       " 'sklearn/sklearn_3_2_tuning_the_hyper-parameters_of_an_estimator_3_2_3_tips_for_parameter_search.txt',\n",
       " 'sklearn/sklearn_3_2_tuning_the_hyper-parameters_of_an_estimator_3_2_4_alternatives_to_brute_force_parameter_search.txt',\n",
       " 'sklearn/sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions_3_3_1_the.txt',\n",
       " 'sklearn/sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions_3_3_2_classification_metrics.txt',\n",
       " 'sklearn/sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions_3_3_3_multilabel_ranking_metrics.txt',\n",
       " 'sklearn/sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions_3_3_4_regression_metrics.txt',\n",
       " 'sklearn/sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions_3_3_5_clustering_metrics.txt',\n",
       " 'sklearn/sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions_3_3_6_dummy_estimators.txt',\n",
       " 'sklearn/sklearn_3_4_model_persistence_3_4_1_persistence_example.txt',\n",
       " 'sklearn/sklearn_3_4_model_persistence_3_4_2_security_&_maintainability_limitations.txt',\n",
       " 'sklearn/sklearn_3_5_validation_curves_plotting_scores_to_evaluate_models_3_5_1_validation_curve.txt',\n",
       " 'sklearn/sklearn_3_5_validation_curves_plotting_scores_to_evaluate_models_3_5_2_learning_curve.txt',\n",
       " 'sklearn/sklearn_4_1_partial_dependence_plots_4_1_1_mathematical_definition.txt',\n",
       " 'sklearn/sklearn_4_1_partial_dependence_plots_4_1_2_computation_methods.txt',\n",
       " 'sklearn/sklearn_4_2_permutation_feature_importance_4_2_1_outline_of_the_permutation_importance_algorithm.txt',\n",
       " 'sklearn/sklearn_4_2_permutation_feature_importance_4_2_2_relation_to_impurity-based_importance_in_trees.txt',\n",
       " 'sklearn/sklearn_4_2_permutation_feature_importance_4_2_3_misleading_values_on_strongly_correlated_features.txt',\n",
       " 'sklearn/sklearn_5_1_available_plotting_utilities_5_1_available_plotting_utilities.txt',\n",
       " 'sklearn/sklearn_6_1_pipelines_and_composite_estimators_6_1_1_pipeline_chaining_estimators.txt',\n",
       " 'sklearn/sklearn_6_1_pipelines_and_composite_estimators_6_1_2_transforming_target_in_regression.txt',\n",
       " 'sklearn/sklearn_6_1_pipelines_and_composite_estimators_6_1_3_featureunion_composite_feature_spaces.txt',\n",
       " 'sklearn/sklearn_6_1_pipelines_and_composite_estimators_6_1_4_columntransformer_for_heterogeneous_data.txt',\n",
       " 'sklearn/sklearn_6_1_pipelines_and_composite_estimators_6_1_5_visualizing_composite_estimators.txt',\n",
       " 'sklearn/sklearn_6_2_feature_extraction_6_2_1_loading_features_from_dicts.txt',\n",
       " 'sklearn/sklearn_6_2_feature_extraction_6_2_2_feature_hashing.txt',\n",
       " 'sklearn/sklearn_6_2_feature_extraction_6_2_3_text_feature_extraction.txt',\n",
       " 'sklearn/sklearn_6_2_feature_extraction_6_2_4_image_feature_extraction.txt',\n",
       " 'sklearn/sklearn_6_3_preprocessing_data_6_3_1_standardization_or_mean_removal_and_variance_scaling.txt',\n",
       " 'sklearn/sklearn_6_3_preprocessing_data_6_3_2_non-linear_transformation.txt',\n",
       " 'sklearn/sklearn_6_3_preprocessing_data_6_3_3_normalization.txt',\n",
       " 'sklearn/sklearn_6_3_preprocessing_data_6_3_4_encoding_categorical_features.txt',\n",
       " 'sklearn/sklearn_6_3_preprocessing_data_6_3_5_discretization.txt',\n",
       " 'sklearn/sklearn_6_3_preprocessing_data_6_3_6_imputation_of_missing_values.txt',\n",
       " 'sklearn/sklearn_6_3_preprocessing_data_6_3_7_generating_polynomial_features.txt',\n",
       " 'sklearn/sklearn_6_3_preprocessing_data_6_3_8_custom_transformers.txt',\n",
       " 'sklearn/sklearn_6_4_imputation_of_missing_values_6_4_1_univariate_vs_multivariate_imputation.txt',\n",
       " 'sklearn/sklearn_6_4_imputation_of_missing_values_6_4_2_univariate_feature_imputation.txt',\n",
       " 'sklearn/sklearn_6_4_imputation_of_missing_values_6_4_3_multivariate_feature_imputation.txt',\n",
       " 'sklearn/sklearn_6_4_imputation_of_missing_values_6_4_4_references.txt',\n",
       " 'sklearn/sklearn_6_4_imputation_of_missing_values_6_4_5_nearest_neighbors_imputation.txt',\n",
       " 'sklearn/sklearn_6_4_imputation_of_missing_values_6_4_6_marking_imputed_values.txt',\n",
       " 'sklearn/sklearn_6_5_unsupervised_dimensionality_reduction_6_5_1_pca_principal_component_analysis.txt',\n",
       " 'sklearn/sklearn_6_5_unsupervised_dimensionality_reduction_6_5_2_random_projections.txt',\n",
       " 'sklearn/sklearn_6_5_unsupervised_dimensionality_reduction_6_5_3_feature_agglomeration.txt',\n",
       " 'sklearn/sklearn_6_6_random_projection_6_6_1_the_johnson-lindenstrauss_lemma.txt',\n",
       " 'sklearn/sklearn_6_6_random_projection_6_6_2_gaussian_random_projection.txt',\n",
       " 'sklearn/sklearn_6_6_random_projection_6_6_3_sparse_random_projection.txt',\n",
       " 'sklearn/sklearn_6_7_kernel_approximation_6_7_1_nystroem_method_for_kernel_approximation.txt',\n",
       " 'sklearn/sklearn_6_7_kernel_approximation_6_7_2_radial_basis_function_kernel.txt',\n",
       " 'sklearn/sklearn_6_7_kernel_approximation_6_7_3_additive_chi_squared_kernel.txt',\n",
       " 'sklearn/sklearn_6_7_kernel_approximation_6_7_4_skewed_chi_squared_kernel.txt',\n",
       " 'sklearn/sklearn_6_7_kernel_approximation_6_7_5_mathematical_details.txt',\n",
       " 'sklearn/sklearn_6_8_pairwise_metrics_affinities_and_kernels_6_8_1_cosine_similarity.txt',\n",
       " 'sklearn/sklearn_6_8_pairwise_metrics_affinities_and_kernels_6_8_2_linear_kernel.txt',\n",
       " 'sklearn/sklearn_6_8_pairwise_metrics_affinities_and_kernels_6_8_3_polynomial_kernel.txt',\n",
       " 'sklearn/sklearn_6_8_pairwise_metrics_affinities_and_kernels_6_8_4_sigmoid_kernel.txt',\n",
       " 'sklearn/sklearn_6_8_pairwise_metrics_affinities_and_kernels_6_8_5_rbf_kernel.txt',\n",
       " 'sklearn/sklearn_6_8_pairwise_metrics_affinities_and_kernels_6_8_6_laplacian_kernel.txt',\n",
       " 'sklearn/sklearn_6_8_pairwise_metrics_affinities_and_kernels_6_8_7_chi-squared_kernel.txt',\n",
       " 'sklearn/sklearn_6_9_transforming_the_prediction_target_y_6_9_1_label_binarization.txt',\n",
       " 'sklearn/sklearn_6_9_transforming_the_prediction_target_y_6_9_2_label_encoding.txt',\n",
       " 'sklearn/sklearn_7_1_general_dataset_api_7_1_general_dataset_api.txt',\n",
       " 'sklearn/sklearn_7_1_general_dataset_api_7_2_toy_datasets.txt',\n",
       " 'sklearn/sklearn_7_1_general_dataset_api_7_3_real_world_datasets.txt',\n",
       " 'sklearn/sklearn_7_1_general_dataset_api_7_4_generated_datasets.txt',\n",
       " 'sklearn/sklearn_7_1_general_dataset_api_7_5_loading_other_datasets.txt',\n",
       " 'sklearn/sklearn_7_2_toy_datasets_7_1_general_dataset_api.txt',\n",
       " 'sklearn/sklearn_7_2_toy_datasets_7_2_toy_datasets.txt',\n",
       " 'sklearn/sklearn_7_2_toy_datasets_7_3_real_world_datasets.txt',\n",
       " 'sklearn/sklearn_7_2_toy_datasets_7_4_generated_datasets.txt',\n",
       " 'sklearn/sklearn_7_2_toy_datasets_7_5_loading_other_datasets.txt',\n",
       " 'sklearn/sklearn_7_3_real_world_datasets_7_1_general_dataset_api.txt',\n",
       " 'sklearn/sklearn_7_3_real_world_datasets_7_2_toy_datasets.txt',\n",
       " 'sklearn/sklearn_7_3_real_world_datasets_7_3_real_world_datasets.txt',\n",
       " 'sklearn/sklearn_7_3_real_world_datasets_7_4_generated_datasets.txt',\n",
       " 'sklearn/sklearn_7_3_real_world_datasets_7_5_loading_other_datasets.txt',\n",
       " 'sklearn/sklearn_7_4_generated_datasets_7_1_general_dataset_api.txt',\n",
       " 'sklearn/sklearn_7_4_generated_datasets_7_2_toy_datasets.txt',\n",
       " 'sklearn/sklearn_7_4_generated_datasets_7_3_real_world_datasets.txt',\n",
       " 'sklearn/sklearn_7_4_generated_datasets_7_4_generated_datasets.txt',\n",
       " 'sklearn/sklearn_7_4_generated_datasets_7_5_loading_other_datasets.txt',\n",
       " 'sklearn/sklearn_7_5_loading_other_datasets_7_1_general_dataset_api.txt',\n",
       " 'sklearn/sklearn_7_5_loading_other_datasets_7_2_toy_datasets.txt',\n",
       " 'sklearn/sklearn_7_5_loading_other_datasets_7_3_real_world_datasets.txt',\n",
       " 'sklearn/sklearn_7_5_loading_other_datasets_7_4_generated_datasets.txt',\n",
       " 'sklearn/sklearn_7_5_loading_other_datasets_7_5_loading_other_datasets.txt',\n",
       " 'sklearn/sklearn_8_1_strategies_to_scale_computationally_bigger_data_8_1_strategies_to_scale_computationally_bigger_data.txt',\n",
       " 'sklearn/sklearn_8_1_strategies_to_scale_computationally_bigger_data_8_2_computational_performance.txt',\n",
       " 'sklearn/sklearn_8_1_strategies_to_scale_computationally_bigger_data_8_3_parallelism_resource_management_and_configuration.txt',\n",
       " 'sklearn/sklearn_8_2_computational_performance_8_1_strategies_to_scale_computationally_bigger_data.txt',\n",
       " 'sklearn/sklearn_8_2_computational_performance_8_2_computational_performance.txt',\n",
       " 'sklearn/sklearn_8_2_computational_performance_8_3_parallelism_resource_management_and_configuration.txt',\n",
       " 'sklearn/sklearn_8_3_parallelism_resource_management_and_configuration_8_1_strategies_to_scale_computationally_bigger_data.txt',\n",
       " 'sklearn/sklearn_8_3_parallelism_resource_management_and_configuration_8_2_computational_performance.txt',\n",
       " 'sklearn/sklearn_8_3_parallelism_resource_management_and_configuration_8_3_parallelism_resource_management_and_configuration.txt']"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sk_glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = []\n",
    "for file in sk_glob:\n",
    "    with open(file) as f:\n",
    "        samples.append(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "242"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open(f'{base_path}sklearn.csv', 'w') as f:\n",
    "    doc_text = \"text\\n\" + \"\\n\".join(['\"' + sample.replace('\\n', ' ') + '\"' for sample in samples])\n",
    "    f.write(doc_text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sklearn_1_10_decision_trees\\n1.10. Decision Trees\\nmodules/tree.html\\n 1.10.1. Classification  DecisionTreeClassifier is a class capable of performing multi-class classification on a dataset. As with other classifiers, DecisionTreeClassifier takes as input two arrays: an array X, sparse or dense, of size holding the training samples, and an array Y of integer values, size , holding the class labels for the training samples: After being fitted, the model can then be used to predict the class of samples: Alternatively, the probability of each class can be predicted, which is the fraction of training samples of the same class in a leaf: DecisionTreeClassifier is capable of both binary (where the labels are [-1, 1]) classification and multiclass (where the labels are [0, …, K-1]) classification. Using the Iris dataset, we can construct a tree as follows: Once trained, you can plot the tree with the plot_tree function: We can also export the tree in Graphviz format using the export_graphviz exporter. If you use the conda package manager, the graphviz binaries and the python package can be installed with . Alternatively binaries for graphviz can be downloaded from the graphviz project homepage, and the Python wrapper installed from pypi with . Below is an example graphviz export of the above tree trained on the entire iris dataset; the results are saved in an output file : The export_graphviz exporter also supports a variety of aesthetic options, including coloring nodes by their class (or value for regression) and using explicit variable and class names if desired. Jupyter notebooks also render these plots inline automatically: Alternatively, the tree can also be exported in textual format with the function export_text . This method doesn’t require the installation of external libraries and is more compact: Examples: Plot the decision surface of a decision tree on the iris dataset Understanding the decision tree structure ',\n",
       " 'sklearn_1_10_decision_trees\\n1.10. Decision Trees\\nmodules/tree.html\\n 1.10.2. Regression  Decision trees can also be applied to regression problems, using the DecisionTreeRegressor class. As in the classification setting, the fit method will take as argument arrays X and y, only that in this case y is expected to have floating point values instead of integer values: Examples: Decision Tree Regression ',\n",
       " 'sklearn_1_10_decision_trees\\n1.10. Decision Trees\\nmodules/tree.html\\n 1.10.3. Multi-output problems  A multi-output problem is a supervised learning problem with several outputs to predict, that is when Y is a 2d array of size . When there is no correlation between the outputs, a very simple way to solve this kind of problem is to build n independent models, i.e. one for each output, and then to use those models to independently predict each one of the n outputs. However, because it is likely that the output values related to the same input are themselves correlated, an often better way is to build a single model capable of predicting simultaneously all n outputs. First, it requires lower training time since only a single estimator is built. Second, the generalization accuracy of the resulting estimator may often be increased. With regard to decision trees, this strategy can readily be used to support multi-output problems. This requires the following changes: Store n output values in leaves, instead of 1; Use splitting criteria that compute the average reduction across all n outputs. This module offers support for multi-output problems by implementing this strategy in both DecisionTreeClassifier and DecisionTreeRegressor . If a decision tree is fit on an output array Y of size then the resulting estimator will: Output n_output values upon ; Output a list of n_output arrays of class probabilities upon . The use of multi-output trees for regression is demonstrated in Multi-output Decision Tree Regression . In this example, the input X is a single real value and the outputs Y are the sine and cosine of X. The use of multi-output trees for classification is demonstrated in Face completion with a multi-output estimators . In this example, the inputs X are the pixels of the upper half of faces and the outputs Y are the pixels of the lower half of those faces. Examples: Multi-output Decision Tree Regression Face completion with a multi-output estimators References: M. Dumont et al, Fast multi-class image annotation with random subwindows and multiple output randomized trees , International Conference on Computer Vision Theory and Applications 2009 ',\n",
       " 'sklearn_1_10_decision_trees\\n1.10. Decision Trees\\nmodules/tree.html\\n 1.10.4. Complexity  In general, the run time cost to construct a balanced binary tree is and query time . Although the tree construction algorithm attempts to generate balanced trees, they will not always be balanced. Assuming that the subtrees remain approximately balanced, the cost at each node consists of searching through to find the feature that offers the largest reduction in entropy. This has a cost of at each node, leading to a total cost over the entire trees (by summing the cost at each node) of . ',\n",
       " 'sklearn_1_10_decision_trees\\n1.10. Decision Trees\\nmodules/tree.html\\n 1.10.5. Tips on practical use  Decision trees tend to overfit on data with a large number of features. Getting the right ratio of samples to number of features is important, since a tree with few samples in high dimensional space is very likely to overfit. Consider performing dimensionality reduction ( PCA , ICA , or Feature selection ) beforehand to give your tree a better chance of finding features that are discriminative. Understanding the decision tree structure will help in gaining more insights about how the decision tree makes predictions, which is important for understanding the important features in the data. Visualise your tree as you are training by using the function. Use as an initial tree depth to get a feel for how the tree is fitting to your data, and then increase the depth. Remember that the number of samples required to populate the tree doubles for each additional level the tree grows to. Use to control the size of the tree to prevent overfitting. Use or to ensure that multiple samples inform every decision in the tree, by controlling which splits will be considered. A very small number will usually mean the tree will overfit, whereas a large number will prevent the tree from learning the data. Try as an initial value. If the sample size varies greatly, a float number can be used as percentage in these two parameters. While can create arbitrarily small leaves, guarantees that each leaf has a minimum size, avoiding low-variance, over-fit leaf nodes in regression problems. For classification with few classes, is often the best choice. Balance your dataset before training to prevent the tree from being biased toward the classes that are dominant. Class balancing can be done by sampling an equal number of samples from each class, or preferably by normalizing the sum of the sample weights ( ) for each class to the same value. Also note that weight-based pre-pruning criteria, such as , will then be less biased toward dominant classes than criteria that are not aware of the sample weights, like . If the samples are weighted, it will be easier to optimize the tree structure using weight-based pre-pruning criterion such as , which ensure that leaf nodes contain at least a fraction of the overall sum of the sample weights. All decision trees use arrays internally. If training data is not in this format, a copy of the dataset will be made. If the input matrix X is very sparse, it is recommended to convert to sparse before calling fit and sparse before calling predict. Training time can be orders of magnitude faster for a sparse matrix input compared to a dense matrix when features have zero values in most of the samples. ',\n",
       " 'sklearn_1_10_decision_trees\\n1.10. Decision Trees\\nmodules/tree.html\\n 1.10.6. Tree algorithms: ID3, C4.5, C5.0 and CART  What are all the various decision tree algorithms and how do they differ from each other? Which one is implemented in scikit-learn? ID3 (Iterative Dichotomiser 3) was developed in 1986 by Ross Quinlan. The algorithm creates a multiway tree, finding for each node (i.e. in a greedy manner) the categorical feature that will yield the largest information gain for categorical targets. Trees are grown to their maximum size and then a pruning step is usually applied to improve the ability of the tree to generalise to unseen data. C4.5 is the successor to ID3 and removed the restriction that features must be categorical by dynamically defining a discrete attribute (based on numerical variables) that partitions the continuous attribute value into a discrete set of intervals. C4.5 converts the trained trees (i.e. the output of the ID3 algorithm) into sets of if-then rules. These accuracy of each rule is then evaluated to determine the order in which they should be applied. Pruning is done by removing a rule’s precondition if the accuracy of the rule improves without it. C5.0 is Quinlan’s latest version release under a proprietary license. It uses less memory and builds smaller rulesets than C4.5 while being more accurate. CART (Classification and Regression Trees) is very similar to C4.5, but it differs in that it supports numerical target variables (regression) and does not compute rule sets. CART constructs binary trees using the feature and threshold that yield the largest information gain at each node. scikit-learn uses an optimised version of the CART algorithm; however, scikit-learn implementation does not support categorical variables for now. ',\n",
       " 'sklearn_1_10_decision_trees\\n1.10. Decision Trees\\nmodules/tree.html\\n 1.10.7. Mathematical formulation  Given training vectors , i1,…, l and a label vector , a decision tree recursively partitions the space such that the samples with the same labels are grouped together. Let the data at node be represented by . For each candidate split consisting of a feature and threshold , partition the data into and subsets \\\\[ \\\\begin{align}\\\\begin{aligned}Q_{left}(\\\\theta)  {(x, y) | x_j < t_m}\\\\\\\\Q_{right}(\\\\theta)  Q \\\\setminus Q_{left}(\\\\theta)\\\\end{aligned}\\\\end{align} \\\\] The impurity at is computed using an impurity function , the choice of which depends on the task being solved (classification or regression) \\\\[G(Q, \\\\theta)  \\\\frac{n_{left}}{N_m} H(Q_{left}(\\\\theta)) + \\\\frac{n_{right}}{N_m} H(Q_{right}(\\\\theta))\\\\] Select the parameters that minimises the impurity \\\\[\\\\theta^*  \\\\operatorname{argmin}_\\\\theta G(Q, \\\\theta)\\\\] Recurse for subsets and until the maximum allowable depth is reached, or . 1.10.7.1. Classification criteria  If a target is a classification outcome taking on values 0,1,…,K-1, for node , representing a region with observations, let \\\\[p_{mk}  1/ N_m \\\\sum_{x_i \\\\in R_m} I(y_i  k)\\\\] be the proportion of class k observations in node Common measures of impurity are Gini \\\\[H(X_m)  \\\\sum_k p_{mk} (1 - p_{mk})\\\\] Entropy \\\\[H(X_m)  - \\\\sum_k p_{mk} \\\\log(p_{mk})\\\\] and Misclassification \\\\[H(X_m)  1 - \\\\max(p_{mk})\\\\] where is the training data in node 1.10.7.2. Regression criteria  If the target is a continuous value, then for node , representing a region with observations, common criteria to minimise as for determining locations for future splits are Mean Squared Error, which minimizes the L2 error using mean values at terminal nodes, and Mean Absolute Error, which minimizes the L1 error using median values at terminal nodes. Mean Squared Error: \\\\[ \\\\begin{align}\\\\begin{aligned}\\\\bar{y}_m  \\\\frac{1}{N_m} \\\\sum_{i \\\\in N_m} y_i\\\\\\\\H(X_m)  \\\\frac{1}{N_m} \\\\sum_{i \\\\in N_m} (y_i - \\\\bar{y}_m)^2\\\\end{aligned}\\\\end{align} \\\\] Mean Absolute Error: \\\\[ \\\\begin{align}\\\\begin{aligned}median(y)_m  \\\\underset{i \\\\in N_m}{\\\\mathrm{median}}(y_i)\\\\\\\\H(X_m)  \\\\frac{1}{N_m} \\\\sum_{i \\\\in N_m} |y_i - median(y)_m|\\\\end{aligned}\\\\end{align} \\\\] where is the training data in node ',\n",
       " 'sklearn_1_10_decision_trees\\n1.10. Decision Trees\\nmodules/tree.html\\n 1.10.8. Minimal Cost-Complexity Pruning  Minimal cost-complexity pruning is an algorithm used to prune a tree to avoid over-fitting, described in Chapter 3 of [BRE] . This algorithm is parameterized by known as the complexity parameter. The complexity parameter is used to define the cost-complexity measure, of a given tree : \\\\[R_\\\\alpha(T)  R(T) + \\\\alpha|T|\\\\] where is the number of terminal nodes in and is traditionally defined as the total misclassification rate of the terminal nodes. Alternatively, scikit-learn uses the total sample weighted impurity of the terminal nodes for . As shown above, the impurity of a node depends on the criterion. Minimal cost-complexity pruning finds the subtree of that minimizes . The cost complexity measure of a single node is . The branch, , is defined to be a tree where node is its root. In general, the impurity of a node is greater than the sum of impurities of its terminal nodes, . However, the cost complexity measure of a node, , and its branch, , can be equal depending on . We define the effective of a node to be the value where they are equal, or . A non-terminal node with the smallest value of is the weakest link and will be pruned. This process stops when the pruned tree’s minimal is greater than the parameter. Examples: Post pruning decision trees with cost complexity pruning References: BRE L. Breiman, J. Friedman, R. Olshen, and C. Stone. Classification and Regression Trees. Wadsworth, Belmont, CA, 1984. https://en.wikipedia.org/wiki/Decision_tree_learning https://en.wikipedia.org/wiki/Predictive_analytics J.R. Quinlan. C4. 5: programs for machine learning. Morgan Kaufmann, 1993. T. Hastie, R. Tibshirani and J. Friedman. Elements of Statistical Learning, Springer, 2009. ',\n",
       " 'sklearn_1_11_ensemble_methods\\n1.11. Ensemble methods\\nmodules/ensemble.html\\n 1.11.1. Bagging meta-estimator  In ensemble algorithms, bagging methods form a class of algorithms which build several instances of a black-box estimator on random subsets of the original training set and then aggregate their individual predictions to form a final prediction. These methods are used as a way to reduce the variance of a base estimator (e.g., a decision tree), by introducing randomization into its construction procedure and then making an ensemble out of it. In many cases, bagging methods constitute a very simple way to improve with respect to a single model, without making it necessary to adapt the underlying base algorithm. As they provide a way to reduce overfitting, bagging methods work best with strong and complex models (e.g., fully developed decision trees), in contrast with boosting methods which usually work best with weak models (e.g., shallow decision trees). Bagging methods come in many flavours but mostly differ from each other by the way they draw random subsets of the training set: When random subsets of the dataset are drawn as random subsets of the samples, then this algorithm is known as Pasting [B1999] . When samples are drawn with replacement, then the method is known as Bagging [B1996] . When random subsets of the dataset are drawn as random subsets of the features, then the method is known as Random Subspaces [H1998] . Finally, when base estimators are built on subsets of both samples and features, then the method is known as Random Patches [LG2012] . In scikit-learn, bagging methods are offered as a unified BaggingClassifier meta-estimator (resp. BaggingRegressor ), taking as input a user-specified base estimator along with parameters specifying the strategy to draw random subsets. In particular, and control the size of the subsets (in terms of samples and features), while and control whether samples and features are drawn with or without replacement. When using a subset of the available samples the generalization accuracy can be estimated with the out-of-bag samples by setting . As an example, the snippet below illustrates how to instantiate a bagging ensemble of KNeighborsClassifier base estimators, each built on random subsets of 50% of the samples and 50% of the features. Examples: Single estimator versus bagging: bias-variance decomposition References B1999 L. Breiman, “Pasting small votes for classification in large databases and on-line”, Machine Learning, 36(1), 85-103, 1999. B1996 L. Breiman, “Bagging predictors”, Machine Learning, 24(2), 123-140, 1996. H1998 T. Ho, “The random subspace method for constructing decision forests”, Pattern Analysis and Machine Intelligence, 20(8), 832-844, 1998. LG2012 G. Louppe and P. Geurts, “Ensembles on Random Patches”, Machine Learning and Knowledge Discovery in Databases, 346-361, 2012. ',\n",
       " 'sklearn_1_11_ensemble_methods\\n1.11. Ensemble methods\\nmodules/ensemble.html\\n 1.11.2. Forests of randomized trees  The sklearn.ensemble module includes two averaging algorithms based on randomized decision trees : the RandomForest algorithm and the Extra-Trees method. Both algorithms are perturb-and-combine techniques [B1998] specifically designed for trees. This means a diverse set of classifiers is created by introducing randomness in the classifier construction. The prediction of the ensemble is given as the averaged prediction of the individual classifiers. As other classifiers, forest classifiers have to be fitted with two arrays: a sparse or dense array X of size holding the training samples, and an array Y of size holding the target values (class labels) for the training samples: Like decision trees , forests of trees also extend to multi-output problems (if Y is an array of size ). 1.11.2.1. Random Forests  In random forests (see RandomForestClassifier and RandomForestRegressor classes), each tree in the ensemble is built from a sample drawn with replacement (i.e., a bootstrap sample) from the training set. Furthermore, when splitting each node during the construction of a tree, the best split is found either from all input features or a random subset of size . (See the parameter tuning guidelines for more details). The purpose of these two sources of randomness is to decrease the variance of the forest estimator. Indeed, individual decision trees typically exhibit high variance and tend to overfit. The injected randomness in forests yield decision trees with somewhat decoupled prediction errors. By taking an average of those predictions, some errors can cancel out. Random forests achieve a reduced variance by combining diverse trees, sometimes at the cost of a slight increase in bias. In practice the variance reduction is often significant hence yielding an overall better model. In contrast to the original publication [B2001] , the scikit-learn implementation combines classifiers by averaging their probabilistic prediction, instead of letting each classifier vote for a single class. 1.11.2.2. Extremely Randomized Trees  In extremely randomized trees (see ExtraTreesClassifier and ExtraTreesRegressor classes), randomness goes one step further in the way splits are computed. As in random forests, a random subset of candidate features is used, but instead of looking for the most discriminative thresholds, thresholds are drawn at random for each candidate feature and the best of these randomly-generated thresholds is picked as the splitting rule. This usually allows to reduce the variance of the model a bit more, at the expense of a slightly greater increase in bias: 1.11.2.3. Parameters  The main parameters to adjust when using these methods is and . The former is the number of trees in the forest. The larger the better, but also the longer it will take to compute. In addition, note that results will stop getting significantly better beyond a critical number of trees. The latter is the size of the random subsets of features to consider when splitting a node. The lower the greater the reduction of variance, but also the greater the increase in bias. Empirical good default values are (always considering all features instead of a random subset) for regression problems, and (using a random subset of size ) for classification tasks (where is the number of features in the data). Good results are often achieved when setting in combination with (i.e., when fully developing the trees). Bear in mind though that these values are usually not optimal, and might result in models that consume a lot of RAM. The best parameter values should always be cross-validated. In addition, note that in random forests, bootstrap samples are used by default ( ) while the default strategy for extra-trees is to use the whole dataset ( ). When using bootstrap sampling the generalization accuracy can be estimated on the left out or out-of-bag samples. This can be enabled by setting . Note The size of the model with the default parameters is , where is the number of trees and is the number of samples. In order to reduce the size of the model, you can change these parameters: , , and . 1.11.2.4. Parallelization  Finally, this module also features the parallel construction of the trees and the parallel computation of the predictions through the parameter. If then computations are partitioned into jobs, and run on cores of the machine. If then all cores available on the machine are used. Note that because of inter-process communication overhead, the speedup might not be linear (i.e., using jobs will unfortunately not be times as fast). Significant speedup can still be achieved though when building a large number of trees, or when building a single tree requires a fair amount of time (e.g., on large datasets). Examples: Plot the decision surfaces of ensembles of trees on the iris dataset Pixel importances with a parallel forest of trees Face completion with a multi-output estimators References B2001 Breiman, “Random Forests”, Machine Learning, 45(1), 5-32, 2001. B1998 Breiman, “Arcing Classifiers”, Annals of Statistics 1998. P. Geurts, D. Ernst., and L. Wehenkel, “Extremely randomized trees”, Machine Learning, 63(1), 3-42, 2006. 1.11.2.5. Feature importance evaluation  The relative rank (i.e. depth) of a feature used as a decision node in a tree can be used to assess the relative importance of that feature with respect to the predictability of the target variable. Features used at the top of the tree contribute to the final prediction decision of a larger fraction of the input samples. The expected fraction of the samples they contribute to can thus be used as an estimate of the relative importance of the features . In scikit-learn, the fraction of samples a feature contributes to is combined with the decrease in impurity from splitting them to create a normalized estimate of the predictive power of that feature. By averaging the estimates of predictive ability over several randomized trees one can reduce the variance of such an estimate and use it for feature selection. This is known as the mean decrease in impurity, or MDI. Refer to [L2014] for more information on MDI and feature importance evaluation with Random Forests. Warning The impurity-based feature importances computed on tree-based models suffer from two flaws that can lead to misleading conclusions. First they are computed on statistics derived from the training dataset and therefore do not necessarily inform us on which features are most important to make good predictions on held-out dataset . Secondly, they favor high cardinality features , that is features with many unique values. Permutation feature importance is an alternative to impurity-based feature importance that does not suffer from these flaws. These two methods of obtaining feature importance are explored in: Permutation Importance vs Random Forest Feature Importance (MDI) . The following example shows a color-coded representation of the relative importances of each individual pixel for a face recognition task using a ExtraTreesClassifier model. In practice those estimates are stored as an attribute named on the fitted model. This is an array with shape whose values are positive and sum to 1.0. The higher the value, the more important is the contribution of the matching feature to the prediction function. Examples: Pixel importances with a parallel forest of trees Feature importances with forests of trees References L2014 G. Louppe, “Understanding Random Forests: From Theory to Practice”, PhD Thesis, U. of Liege, 2014. 1.11.2.6. Totally Random Trees Embedding  RandomTreesEmbedding implements an unsupervised transformation of the data. Using a forest of completely random trees, RandomTreesEmbedding encodes the data by the indices of the leaves a data point ends up in. This index is then encoded in a one-of-K manner, leading to a high dimensional, sparse binary coding. This coding can be computed very efficiently and can then be used as a basis for other learning tasks. The size and sparsity of the code can be influenced by choosing the number of trees and the maximum depth per tree. For each tree in the ensemble, the coding contains one entry of one. The size of the coding is at most , the maximum number of leaves in the forest. As neighboring data points are more likely to lie within the same leaf of a tree, the transformation performs an implicit, non-parametric density estimation. Examples: Hashing feature transformation using Totally Random Trees Manifold learning on handwritten digits: Locally Linear Embedding, Isomap… compares non-linear dimensionality reduction techniques on handwritten digits. Feature transformations with ensembles of trees compares supervised and unsupervised tree based feature transformations. See also Manifold learning techniques can also be useful to derive non-linear representations of feature space, also these approaches focus also on dimensionality reduction. ',\n",
       " 'sklearn_1_11_ensemble_methods\\n1.11. Ensemble methods\\nmodules/ensemble.html\\n 1.11.3. AdaBoost  The module sklearn.ensemble includes the popular boosting algorithm AdaBoost, introduced in 1995 by Freund and Schapire [FS1995] . The core principle of AdaBoost is to fit a sequence of weak learners (i.e., models that are only slightly better than random guessing, such as small decision trees) on repeatedly modified versions of the data. The predictions from all of them are then combined through a weighted majority vote (or sum) to produce the final prediction. The data modifications at each so-called boosting iteration consist of applying weights , , …, to each of the training samples. Initially, those weights are all set to , so that the first step simply trains a weak learner on the original data. For each successive iteration, the sample weights are individually modified and the learning algorithm is reapplied to the reweighted data. At a given step, those training examples that were incorrectly predicted by the boosted model induced at the previous step have their weights increased, whereas the weights are decreased for those that were predicted correctly. As iterations proceed, examples that are difficult to predict receive ever-increasing influence. Each subsequent weak learner is thereby forced to concentrate on the examples that are missed by the previous ones in the sequence [HTF] . AdaBoost can be used both for classification and regression problems: For multi-class classification, AdaBoostClassifier implements AdaBoost-SAMME and AdaBoost-SAMME.R [ZZRH2009] . For regression, AdaBoostRegressor implements AdaBoost.R2 [D1997] . 1.11.3.1. Usage  The following example shows how to fit an AdaBoost classifier with 100 weak learners: The number of weak learners is controlled by the parameter . The parameter controls the contribution of the weak learners in the final combination. By default, weak learners are decision stumps. Different weak learners can be specified through the parameter. The main parameters to tune to obtain good results are and the complexity of the base estimators (e.g., its depth or minimum required number of samples to consider a split ). Examples: Discrete versus Real AdaBoost compares the classification error of a decision stump, decision tree, and a boosted decision stump using AdaBoost-SAMME and AdaBoost-SAMME.R. Multi-class AdaBoosted Decision Trees shows the performance of AdaBoost-SAMME and AdaBoost-SAMME.R on a multi-class problem. Two-class AdaBoost shows the decision boundary and decision function values for a non-linearly separable two-class problem using AdaBoost-SAMME. Decision Tree Regression with AdaBoost demonstrates regression with the AdaBoost.R2 algorithm. References FS1995 Y. Freund, and R. Schapire, “A Decision-Theoretic Generalization of On-Line Learning and an Application to Boosting”, 1997. ZZRH2009 J. Zhu, H. Zou, S. Rosset, T. Hastie. “Multi-class AdaBoost”, 2009. D1997 Drucker. “Improving Regressors using Boosting Techniques”, 1997. HTF ( 1 , 2 , 3 ) T. Hastie, R. Tibshirani and J. Friedman, “Elements of Statistical Learning Ed. 2”, Springer, 2009. ',\n",
       " 'sklearn_1_11_ensemble_methods\\n1.11. Ensemble methods\\nmodules/ensemble.html\\n 1.11.4. Gradient Tree Boosting  Gradient Tree Boosting or Gradient Boosted Decision Trees (GBDT) is a generalization of boosting to arbitrary differentiable loss functions. GBDT is an accurate and effective off-the-shelf procedure that can be used for both regression and classification problems in a variety of areas including Web search ranking and ecology. The module sklearn.ensemble provides methods for both classification and regression via gradient boosted decision trees. Note Scikit-learn 0.21 introduces two new experimental implementations of gradient boosting trees, namely HistGradientBoostingClassifier and HistGradientBoostingRegressor , inspired by LightGBM (See [LightGBM] ). These histogram-based estimators can be orders of magnitude faster than GradientBoostingClassifier and GradientBoostingRegressor when the number of samples is larger than tens of thousands of samples. They also have built-in support for missing values, which avoids the need for an imputer. These estimators are described in more detail below in Histogram-Based Gradient Boosting . The following guide focuses on GradientBoostingClassifier and GradientBoostingRegressor , which might be preferred for small sample sizes since binning may lead to split points that are too approximate in this setting. The usage and the parameters of GradientBoostingClassifier and GradientBoostingRegressor are described below. The 2 most important parameters of these estimators are and . 1.11.4.1. Classification  GradientBoostingClassifier supports both binary and multi-class classification. The following example shows how to fit a gradient boosting classifier with 100 decision stumps as weak learners: The number of weak learners (i.e. regression trees) is controlled by the parameter ; The size of each tree can be controlled either by setting the tree depth via or by setting the number of leaf nodes via . The is a hyper-parameter in the range (0.0, 1.0] that controls overfitting via shrinkage . Note Classification with more than 2 classes requires the induction of regression trees at each iteration, thus, the total number of induced trees equals . For datasets with a large number of classes we strongly recommend to use HistGradientBoostingClassifier as an alternative to GradientBoostingClassifier . 1.11.4.2. Regression  GradientBoostingRegressor supports a number of different loss functions for regression which can be specified via the argument ; the default loss function for regression is least squares ( ). The figure below shows the results of applying GradientBoostingRegressor with least squares loss and 500 base learners to the Boston house price dataset ( sklearn.datasets.load_boston ). The plot on the left shows the train and test error at each iteration. The train error at each iteration is stored in the train_score_ attribute of the gradient boosting model. The test error at each iterations can be obtained via the staged_predict method which returns a generator that yields the predictions at each stage. Plots like these can be used to determine the optimal number of trees (i.e. ) by early stopping. The plot on the right shows the impurity-based feature importances which can be obtained via the property. Examples: Gradient Boosting regression Gradient Boosting Out-of-Bag estimates 1.11.4.3. Fitting additional weak-learners  Both GradientBoostingRegressor and GradientBoostingClassifier support which allows you to add more estimators to an already fitted model. 1.11.4.4. Controlling the tree size  The size of the regression tree base learners defines the level of variable interactions that can be captured by the gradient boosting model. In general, a tree of depth can capture interactions of order . There are two ways in which the size of the individual regression trees can be controlled. If you specify then complete binary trees of depth will be grown. Such trees will have (at most) leaf nodes and split nodes. Alternatively, you can control the tree size by specifying the number of leaf nodes via the parameter . In this case, trees will be grown using best-first search where nodes with the highest improvement in impurity will be expanded first. A tree with has split nodes and thus can model interactions of up to order . We found that gives comparable results to but is significantly faster to train at the expense of a slightly higher training error. The parameter corresponds to the variable in the chapter on gradient boosting in [F2001] and is related to the parameter in R’s gbm package where . 1.11.4.5. Mathematical formulation  We first present GBRT for regression, and then detail the classification case. 1.11.4.5.1. Regression  GBRT regressors are additive models whose prediction for a given input is of the following form: \\\\[\\\\hat{y_i}  F_M(x_i)  \\\\sum_{m1}^{M} h_m(x_i)\\\\] where the are estimators called weak learners in the context of boosting. Gradient Tree Boosting uses decision tree regressors of fixed size as weak learners. The constant M corresponds to the parameter. Similar to other boosting algorithms, a GBRT is built in a greedy fashion: \\\\[F_m(x)  F_{m-1}(x) + h_m(x),\\\\] where the newly added tree is fitted in order to minimize a sum of losses , given the previous ensemble : \\\\[h_m  \\\\arg\\\\min_{h} L_m  \\\\arg\\\\min_{h} \\\\sum_{i1}^{n} l(y_i, F_{m-1}(x_i) + h(x_i)),\\\\] where is defined by the parameter, detailed in the next section. By default, the initial model is chosen as the constant that minimizes the loss: for a least-squares loss, this is the empirical mean of the target values. The initial model can also be specified via the argument. Using a first-order Taylor approximation, the value of can be approximated as follows: \\\\[l(y_i, F_{m-1}(x_i) + h_m(x_i)) \\\\approx l(y_i, F_{m-1}(x_i)) + h_m(x_i) \\\\left[ \\\\frac{\\\\partial l(y_i, F(x_i))}{\\\\partial F(x_i)} \\\\right]_{FF_{m - 1}}.\\\\] Note Briefly, a first-order Taylor approximation says that . Here, corresponds to , and corresponds to The quantity is the derivative of the loss with respect to its second parameter, evaluated at . It is easy to compute for any given in a closed form since the loss is differentiable. We will denote it by . Removing the constant terms, we have: \\\\[h_m \\\\approx \\\\arg\\\\min_{h} \\\\sum_{i1}^{n} h(x_i) g_i\\\\] This is minimized if is fitted to predict a value that is proportional to the negative gradient . Therefore, at each iteration, the estimator is fitted to predict the negative gradients of the samples . The gradients are updated at each iteration. This can be considered as some kind of gradient descent in a functional space. Note For some losses, e.g. the least absolute deviation (LAD) where the gradients are , the values predicted by a fitted are not accurate enough: the tree can only output integer values. As a result, the leaves values of the tree are modified once the tree is fitted, such that the leaves values minimize the loss . The update is loss-dependent: for the LAD loss, the value of a leaf is updated to the median of the samples in that leaf. 1.11.4.5.2. Classification  Gradient boosting for classification is very similar to the regression case. However, the sum of the trees is not homogeneous to a prediction: it cannot be a class, since the trees predict continuous values. The mapping from the value to a class or a probability is loss-dependent. For the deviance (or log-loss), the probability that belongs to the positive class is modeled as where is the sigmoid function. For multiclass classification, K trees (for K classes) are built at each of the iterations. The probability that belongs to class k is modeled as a softmax of the values. Note that even for a classification task, the sub-estimator is still a regressor, not a classifier. This is because the sub-estimators are trained to predict (negative) gradients , which are always continuous quantities. 1.11.4.6. Loss Functions  The following loss functions are supported and can be specified using the parameter : Regression Least squares ( ): The natural choice for regression due to its superior computational properties. The initial model is given by the mean of the target values. Least absolute deviation ( ): A robust loss function for regression. The initial model is given by the median of the target values. Huber ( ): Another robust loss function that combines least squares and least absolute deviation; use to control the sensitivity with regards to outliers (see [F2001] for more details). Quantile ( ): A loss function for quantile regression. Use to specify the quantile. This loss function can be used to create prediction intervals (see Prediction Intervals for Gradient Boosting Regression ). Classification Binomial deviance ( ): The negative binomial log-likelihood loss function for binary classification (provides probability estimates). The initial model is given by the log odds-ratio. Multinomial deviance ( ): The negative multinomial log-likelihood loss function for multi-class classification with mutually exclusive classes. It provides probability estimates. The initial model is given by the prior probability of each class. At each iteration regression trees have to be constructed which makes GBRT rather inefficient for data sets with a large number of classes. Exponential loss ( ): The same loss function as AdaBoostClassifier . Less robust to mislabeled examples than ; can only be used for binary classification. 1.11.4.7. Shrinkage via learning rate  [F2001] proposed a simple regularization strategy that scales the contribution of each weak learner by a constant factor : \\\\[F_m(x)  F_{m-1}(x) + \\\\nu h_m(x)\\\\] The parameter is also called the learning rate because it scales the step length the gradient descent procedure; it can be set via the parameter. The parameter strongly interacts with the parameter , the number of weak learners to fit. Smaller values of require larger numbers of weak learners to maintain a constant training error. Empirical evidence suggests that small values of favor better test error. [HTF] recommend to set the learning rate to a small constant (e.g. ) and choose by early stopping. For a more detailed discussion of the interaction between and see [R2007] . 1.11.4.8. Subsampling  [F1999] proposed stochastic gradient boosting, which combines gradient boosting with bootstrap averaging (bagging). At each iteration the base classifier is trained on a fraction of the available training data. The subsample is drawn without replacement. A typical value of is 0.5. The figure below illustrates the effect of shrinkage and subsampling on the goodness-of-fit of the model. We can clearly see that shrinkage outperforms no-shrinkage. Subsampling with shrinkage can further increase the accuracy of the model. Subsampling without shrinkage, on the other hand, does poorly. Another strategy to reduce the variance is by subsampling the features analogous to the random splits in RandomForestClassifier . The number of subsampled features can be controlled via the parameter. Note Using a small value can significantly decrease the runtime. Stochastic gradient boosting allows to compute out-of-bag estimates of the test deviance by computing the improvement in deviance on the examples that are not included in the bootstrap sample (i.e. the out-of-bag examples). The improvements are stored in the attribute oob_improvement_ . holds the improvement in terms of the loss on the OOB samples if you add the i-th stage to the current predictions. Out-of-bag estimates can be used for model selection, for example to determine the optimal number of iterations. OOB estimates are usually very pessimistic thus we recommend to use cross-validation instead and only use OOB if cross-validation is too time consuming. Examples: Gradient Boosting regularization Gradient Boosting Out-of-Bag estimates OOB Errors for Random Forests 1.11.4.9. Interpretation with feature importance  Individual decision trees can be interpreted easily by simply visualizing the tree structure. Gradient boosting models, however, comprise hundreds of regression trees thus they cannot be easily interpreted by visual inspection of the individual trees. Fortunately, a number of techniques have been proposed to summarize and interpret gradient boosting models. Often features do not contribute equally to predict the target response; in many situations the majority of the features are in fact irrelevant. When interpreting a model, the first question usually is: what are those important features and how do they contributing in predicting the target response? Individual decision trees intrinsically perform feature selection by selecting appropriate split points. This information can be used to measure the importance of each feature; the basic idea is: the more often a feature is used in the split points of a tree the more important that feature is. This notion of importance can be extended to decision tree ensembles by simply averaging the impurity-based feature importance of each tree (see Feature importance evaluation for more details). The feature importance scores of a fit gradient boosting model can be accessed via the property: Note that this computation of feature importance is based on entropy, and it is distinct from sklearn.inspection.permutation_importance which is based on permutation of the features. Examples: Gradient Boosting regression ',\n",
       " 'sklearn_1_11_ensemble_methods\\n1.11. Ensemble methods\\nmodules/ensemble.html\\n 1.11.5. Histogram-Based Gradient Boosting  Scikit-learn 0.21 introduced two new experimental implementations of gradient boosting trees, namely HistGradientBoostingClassifier and HistGradientBoostingRegressor , inspired by LightGBM (See [LightGBM] ). These histogram-based estimators can be orders of magnitude faster than GradientBoostingClassifier and GradientBoostingRegressor when the number of samples is larger than tens of thousands of samples. They also have built-in support for missing values, which avoids the need for an imputer. These fast estimators first bin the input samples into integer-valued bins (typically 256 bins) which tremendously reduces the number of splitting points to consider, and allows the algorithm to leverage integer-based data structures (histograms) instead of relying on sorted continuous values when building the trees. The API of these estimators is slightly different, and some of the features from GradientBoostingClassifier and GradientBoostingRegressor are not yet supported, for instance some loss functions. These estimators are still experimental : their predictions and their API might change without any deprecation cycle. To use them, you need to explicitly import : Examples: Partial Dependence Plots 1.11.5.1. Usage  Most of the parameters are unchanged from GradientBoostingClassifier and GradientBoostingRegressor . One exception is the parameter that replaces , and controls the number of iterations of the boosting process: Available losses for regression are ‘least_squares’, ‘least_absolute_deviation’, which is less sensitive to outliers, and ‘poisson’, which is well suited to model counts and frequencies. For classification, ‘binary_crossentropy’ is used for binary classification and ‘categorical_crossentropy’ is used for multiclass classification. By default the loss is ‘auto’ and will select the appropriate loss depending on y passed to fit . The size of the trees can be controlled through the , , and parameters. The number of bins used to bin the data is controlled with the parameter. Using less bins acts as a form of regularization. It is generally recommended to use as many bins as possible, which is the default. The parameter is a regularizer on the loss function and corresponds to in equation (2) of [XGBoost] . Note that early-stopping is enabled by default if the number of samples is larger than 10,000 . The early-stopping behaviour is controlled via the , , , , and parameters. It is possible to early-stop using an arbitrary scorer , or just the training or validation loss. Note that for technical reasons, using a scorer is significantly slower than using the loss. By default, early-stopping is performed if there are at least 10,000 samples in the training set, using the validation loss. 1.11.5.2. Missing values support  HistGradientBoostingClassifier and HistGradientBoostingRegressor have built-in support for missing values (NaNs). During training, the tree grower learns at each split point whether samples with missing values should go to the left or right child, based on the potential gain. When predicting, samples with missing values are assigned to the left or right child consequently: When the missingness pattern is predictive, the splits can be done on whether the feature value is missing or not: If no missing values were encountered for a given feature during training, then samples with missing values are mapped to whichever child has the most samples. 1.11.5.3. Sample weight support  HistGradientBoostingClassifier and HistGradientBoostingRegressor sample support weights during fit . The following toy example demonstrates how the model ignores the samples with zero sample weights: As you can see, the is comfortably classified as since the first two samples are ignored due to their sample weights. Implementation detail: taking sample weights into account amounts to multiplying the gradients (and the hessians) by the sample weights. Note that the binning stage (specifically the quantiles computation) does not take the weights into account. 1.11.5.4. Monotonic Constraints  Depending on the problem at hand, you may have prior knowledge indicating that a given feature should in general have a positive (or negative) effect on the target value. For example, all else being equal, a higher credit score should increase the probability of getting approved for a loan. Monotonic constraints allow you to incorporate such prior knowledge into the model. A positive monotonic constraint is a constraint of the form: , where is the predictor with two features. Similarly, a negative monotonic constraint is of the form: . Note that monotonic constraints only constraint the output “all else being equal”. Indeed, the following relation is not enforced by a positive constraint: . You can specify a monotonic constraint on each feature using the parameter. For each feature, a value of 0 indicates no constraint, while -1 and 1 indicate a negative and positive constraint, respectively: In a binary classification context, imposing a monotonic constraint means that the feature is supposed to have a positive / negative effect on the probability to belong to the positive class. Monotonic constraints are not supported for multiclass context. Examples: Monotonic Constraints 1.11.5.5. Low-level parallelism  HistGradientBoostingClassifier and HistGradientBoostingRegressor have implementations that use OpenMP for parallelization through Cython. For more details on how to control the number of threads, please refer to our Parallelism notes. The following parts are parallelized: mapping samples from real values to integer-valued bins (finding the bin thresholds is however sequential) building histograms is parallelized over features finding the best split point at a node is parallelized over features during fit, mapping samples into the left and right children is parallelized over samples gradient and hessians computations are parallelized over samples predicting is parallelized over samples 1.11.5.6. Why it’s faster  The bottleneck of a gradient boosting procedure is building the decision trees. Building a traditional decision tree (as in the other GBDTs GradientBoostingClassifier and GradientBoostingRegressor ) requires sorting the samples at each node (for each feature). Sorting is needed so that the potential gain of a split point can be computed efficiently. Splitting a single node has thus a complexity of where is the number of samples at the node. HistGradientBoostingClassifier and HistGradientBoostingRegressor , in contrast, do not require sorting the feature values and instead use a data-structure called a histogram, where the samples are implicitly ordered. Building a histogram has a complexity, so the node splitting procedure has a complexity, much smaller than the previous one. In addition, instead of considering split points, we here consider only split points, which is much smaller. In order to build histograms, the input data needs to be binned into integer-valued bins. This binning procedure does require sorting the feature values, but it only happens once at the very beginning of the boosting process (not at each node, like in GradientBoostingClassifier and GradientBoostingRegressor ). Finally, many parts of the implementation of HistGradientBoostingClassifier and HistGradientBoostingRegressor are parallelized. References F1999 Friedmann, Jerome H., 2007, “Stochastic Gradient Boosting” R2007 G. Ridgeway, “Generalized Boosted Models: A guide to the gbm package”, 2007 XGBoost Tianqi Chen, Carlos Guestrin, “XGBoost: A Scalable Tree Boosting System” LightGBM ( 1 , 2 ) Ke et. al. “LightGBM: A Highly Efficient Gradient BoostingDecision Tree” ',\n",
       " 'sklearn_1_11_ensemble_methods\\n1.11. Ensemble methods\\nmodules/ensemble.html\\n 1.11.6. Voting Classifier  The idea behind the VotingClassifier is to combine conceptually different machine learning classifiers and use a majority vote or the average predicted probabilities (soft vote) to predict the class labels. Such a classifier can be useful for a set of equally well performing model in order to balance out their individual weaknesses. 1.11.6.1. Majority Class Labels (Majority/Hard Voting)  In majority voting, the predicted class label for a particular sample is the class label that represents the majority (mode) of the class labels predicted by each individual classifier. E.g., if the prediction for a given sample is classifier 1 -> class 1 classifier 2 -> class 1 classifier 3 -> class 2 the VotingClassifier (with ) would classify the sample as “class 1” based on the majority class label. In the cases of a tie, the VotingClassifier will select the class based on the ascending sort order. E.g., in the following scenario classifier 1 -> class 2 classifier 2 -> class 1 the class label 1 will be assigned to the sample. 1.11.6.2. Usage  The following example shows how to fit the majority rule classifier: 1.11.6.3. Weighted Average Probabilities (Soft Voting)  In contrast to majority voting (hard voting), soft voting returns the class label as argmax of the sum of predicted probabilities. Specific weights can be assigned to each classifier via the parameter. When weights are provided, the predicted class probabilities for each classifier are collected, multiplied by the classifier weight, and averaged. The final class label is then derived from the class label with the highest average probability. To illustrate this with a simple example, let’s assume we have 3 classifiers and a 3-class classification problems where we assign equal weights to all classifiers: w11, w21, w31. The weighted average probabilities for a sample would then be calculated as follows: classifier class 1 class 2 class 3 classifier 1 w1 * 0.2 w1 * 0.5 w1 * 0.3 classifier 2 w2 * 0.6 w2 * 0.3 w2 * 0.1 classifier 3 w3 * 0.3 w3 * 0.4 w3 * 0.3 weighted average 0.37 0.4 0.23 Here, the predicted class label is 2, since it has the highest average probability. The following example illustrates how the decision regions may change when a soft VotingClassifier is used based on an linear Support Vector Machine, a Decision Tree, and a K-nearest neighbor classifier: 1.11.6.4. Using the with  The VotingClassifier can also be used together with GridSearchCV in order to tune the hyperparameters of the individual estimators: 1.11.6.5. Usage  In order to predict the class labels based on the predicted class-probabilities (scikit-learn estimators in the VotingClassifier must support method): Optionally, weights can be provided for the individual classifiers: ',\n",
       " 'sklearn_1_11_ensemble_methods\\n1.11. Ensemble methods\\nmodules/ensemble.html\\n 1.11.7. Voting Regressor  The idea behind the VotingRegressor is to combine conceptually different machine learning regressors and return the average predicted values. Such a regressor can be useful for a set of equally well performing models in order to balance out their individual weaknesses. 1.11.7.1. Usage  The following example shows how to fit the VotingRegressor: Examples: Plot individual and voting regression predictions ',\n",
       " 'sklearn_1_11_ensemble_methods\\n1.11. Ensemble methods\\nmodules/ensemble.html\\n 1.11.8. Stacked generalization  Stacked generalization is a method for combining estimators to reduce their biases [W1992] [HTF] . More precisely, the predictions of each individual estimator are stacked together and used as input to a final estimator to compute the prediction. This final estimator is trained through cross-validation. The StackingClassifier and StackingRegressor provide such strategies which can be applied to classification and regression problems. The parameter corresponds to the list of the estimators which are stacked together in parallel on the input data. It should be given as a list of names and estimators: The will use the predictions of the as input. It needs to be a classifier or a regressor when using StackingClassifier or StackingRegressor , respectively: To train the and , the method needs to be called on the training data: During training, the are fitted on the whole training data . They will be used when calling or . To generalize and avoid over-fitting, the is trained on out-samples using sklearn.model_selection.cross_val_predict internally. For StackingClassifier , note that the output of the is controlled by the parameter and it is called by each estimator. This parameter is either a string, being estimator method names, or which will automatically identify an available method depending on the availability, tested in the order of preference: , and . A StackingRegressor and StackingClassifier can be used as any other regressor or classifier, exposing a , , and methods, e.g.: Note that it is also possible to get the output of the stacked using the method: In practise, a stacking predictor predict as good as the best predictor of the base layer and even sometimes outputperform it by combining the different strength of the these predictors. However, training a stacking predictor is computationally expensive. Note For StackingClassifier , when using , the first column is dropped when the problem is a binary classification problem. Indeed, both probability columns predicted by each estimator are perfectly collinear. Note Multiple stacking layers can be achieved by assigning to a StackingClassifier or StackingRegressor : References W1992 Wolpert, David H. “Stacked generalization.” Neural networks 5.2 (1992): 241-259. ',\n",
       " 'sklearn_1_12_multiclass_and_multilabel_algorithms\\n1.12. Multiclass and multilabel algorithms\\nmodules/multiclass.html\\n 1.12.1. Multilabel classification format  In multilabel learning, the joint set of binary classification tasks is expressed with label binary indicator array: each sample is one row of a 2d array of shape (n_samples, n_classes) with binary values: the one, i.e. the non zero elements, corresponds to the subset of labels. An array such as represents label 0 in the first sample, labels 1 and 2 in the second sample, and no labels in the third sample. Producing multilabel data as a list of sets of labels may be more intuitive. The MultiLabelBinarizer transformer can be used to convert between a collection of collections of labels and the indicator format. ',\n",
       " 'sklearn_1_12_multiclass_and_multilabel_algorithms\\n1.12. Multiclass and multilabel algorithms\\nmodules/multiclass.html\\n 1.12.2. One-Vs-The-Rest  This strategy, also known as one-vs-all , is implemented in OneVsRestClassifier . The strategy consists in fitting one classifier per class. For each classifier, the class is fitted against all the other classes. In addition to its computational efficiency (only classifiers are needed), one advantage of this approach is its interpretability. Since each class is represented by one and only one classifier, it is possible to gain knowledge about the class by inspecting its corresponding classifier. This is the most commonly used strategy and is a fair default choice. 1.12.2.1. Multiclass learning  Below is an example of multiclass learning using OvR: 1.12.2.2. Multilabel learning  OneVsRestClassifier also supports multilabel classification. To use this feature, feed the classifier an indicator matrix, in which cell [i, j] indicates the presence of label j in sample i. Examples: Multilabel classification ',\n",
       " 'sklearn_1_12_multiclass_and_multilabel_algorithms\\n1.12. Multiclass and multilabel algorithms\\nmodules/multiclass.html\\n 1.12.3. One-Vs-One  OneVsOneClassifier constructs one classifier per pair of classes. At prediction time, the class which received the most votes is selected. In the event of a tie (among two classes with an equal number of votes), it selects the class with the highest aggregate classification confidence by summing over the pair-wise classification confidence levels computed by the underlying binary classifiers. Since it requires to fit classifiers, this method is usually slower than one-vs-the-rest, due to its O(n_classes^2) complexity. However, this method may be advantageous for algorithms such as kernel algorithms which don’t scale well with . This is because each individual learning problem only involves a small subset of the data whereas, with one-vs-the-rest, the complete dataset is used times. The decision function is the result of a monotonic transformation of the one-versus-one classification. 1.12.3.1. Multiclass learning  Below is an example of multiclass learning using OvO: References: “Pattern Recognition and Machine Learning. Springer”, Christopher M. Bishop, page 183, (First Edition) ',\n",
       " 'sklearn_1_12_multiclass_and_multilabel_algorithms\\n1.12. Multiclass and multilabel algorithms\\nmodules/multiclass.html\\n 1.12.4. Error-Correcting Output-Codes  Output-code based strategies are fairly different from one-vs-the-rest and one-vs-one. With these strategies, each class is represented in a Euclidean space, where each dimension can only be 0 or 1. Another way to put it is that each class is represented by a binary code (an array of 0 and 1). The matrix which keeps track of the location/code of each class is called the code book. The code size is the dimensionality of the aforementioned space. Intuitively, each class should be represented by a code as unique as possible and a good code book should be designed to optimize classification accuracy. In this implementation, we simply use a randomly-generated code book as advocated in 3 although more elaborate methods may be added in the future. At fitting time, one binary classifier per bit in the code book is fitted. At prediction time, the classifiers are used to project new points in the class space and the class closest to the points is chosen. In OutputCodeClassifier , the attribute allows the user to control the number of classifiers which will be used. It is a percentage of the total number of classes. A number between 0 and 1 will require fewer classifiers than one-vs-the-rest. In theory, is sufficient to represent each class unambiguously. However, in practice, it may not lead to good accuracy since is much smaller than n_classes. A number greater than 1 will require more classifiers than one-vs-the-rest. In this case, some classifiers will in theory correct for the mistakes made by other classifiers, hence the name “error-correcting”. In practice, however, this may not happen as classifier mistakes will typically be correlated. The error-correcting output codes have a similar effect to bagging. 1.12.4.1. Multiclass learning  Below is an example of multiclass learning using Output-Codes: References: “Solving multiclass learning problems via error-correcting output codes”, Dietterich T., Bakiri G., Journal of Artificial Intelligence Research 2, 1995. 3 “The error coding method and PICTs”, James G., Hastie T., Journal of Computational and Graphical statistics 7, 1998. “The Elements of Statistical Learning”, Hastie T., Tibshirani R., Friedman J., page 606 (second-edition) 2008. ',\n",
       " 'sklearn_1_12_multiclass_and_multilabel_algorithms\\n1.12. Multiclass and multilabel algorithms\\nmodules/multiclass.html\\n 1.12.5. Multioutput regression  Multioutput regression support can be added to any regressor with MultiOutputRegressor . This strategy consists of fitting one regressor per target. Since each target is represented by exactly one regressor it is possible to gain knowledge about the target by inspecting its corresponding regressor. As MultiOutputRegressor fits one regressor per target it can not take advantage of correlations between targets. Below is an example of multioutput regression: ',\n",
       " 'sklearn_1_12_multiclass_and_multilabel_algorithms\\n1.12. Multiclass and multilabel algorithms\\nmodules/multiclass.html\\n 1.12.6. Multioutput classification  Multioutput classification support can be added to any classifier with MultiOutputClassifier . This strategy consists of fitting one classifier per target. This allows multiple target variable classifications. The purpose of this class is to extend estimators to be able to estimate a series of target functions (f1,f2,f3…,fn) that are trained on a single X predictor matrix to predict a series of responses (y1,y2,y3…,yn). Below is an example of multioutput classification: ',\n",
       " 'sklearn_1_12_multiclass_and_multilabel_algorithms\\n1.12. Multiclass and multilabel algorithms\\nmodules/multiclass.html\\n 1.12.7. Classifier Chain  Classifier chains (see ClassifierChain ) are a way of combining a number of binary classifiers into a single multi-label model that is capable of exploiting correlations among targets. For a multi-label classification problem with N classes, N binary classifiers are assigned an integer between 0 and N-1. These integers define the order of models in the chain. Each classifier is then fit on the available training data plus the true labels of the classes whose models were assigned a lower number. When predicting, the true labels will not be available. Instead the predictions of each model are passed on to the subsequent models in the chain to be used as features. Clearly the order of the chain is important. The first model in the chain has no information about the other labels while the last model in the chain has features indicating the presence of all of the other labels. In general one does not know the optimal ordering of the models in the chain so typically many randomly ordered chains are fit and their predictions are averaged together. References: Jesse Read, Bernhard Pfahringer, Geoff Holmes, Eibe Frank, “Classifier Chains for Multi-label Classification”, 2009. ',\n",
       " 'sklearn_1_12_multiclass_and_multilabel_algorithms\\n1.12. Multiclass and multilabel algorithms\\nmodules/multiclass.html\\n 1.12.8. Regressor Chain  Regressor chains (see RegressorChain ) is analogous to ClassifierChain as a way of combining a number of regressions into a single multi-target model that is capable of exploiting correlations among targets. ',\n",
       " 'sklearn_1_13_feature_selection\\n1.13. Feature selection\\nmodules/feature_selection.html\\n 1.13.1. Removing features with low variance  VarianceThreshold is a simple baseline approach to feature selection. It removes all features whose variance doesn’t meet some threshold. By default, it removes all zero-variance features, i.e. features that have the same value in all samples. As an example, suppose that we have a dataset with boolean features, and we want to remove all features that are either one or zero (on or off) in more than 80% of the samples. Boolean features are Bernoulli random variables, and the variance of such variables is given by \\\\[\\\\mathrm{Var}[X]  p(1 - p)\\\\] so we can select using the threshold : As expected, has removed the first column, which has a probability of containing a zero. ',\n",
       " 'sklearn_1_13_feature_selection\\n1.13. Feature selection\\nmodules/feature_selection.html\\n 1.13.2. Univariate feature selection  Univariate feature selection works by selecting the best features based on univariate statistical tests. It can be seen as a preprocessing step to an estimator. Scikit-learn exposes feature selection routines as objects that implement the method: SelectKBest removes all but the highest scoring features SelectPercentile removes all but a user-specified highest scoring percentage of features using common univariate statistical tests for each feature: false positive rate SelectFpr , false discovery rate SelectFdr , or family wise error SelectFwe . GenericUnivariateSelect allows to perform univariate feature selection with a configurable strategy. This allows to select the best univariate selection strategy with hyper-parameter search estimator. For instance, we can perform a test to the samples to retrieve only the two best features as follows: These objects take as input a scoring function that returns univariate scores and p-values (or only scores for SelectKBest and SelectPercentile ): For regression: f_regression , mutual_info_regression For classification: chi2 , f_classif , mutual_info_classif The methods based on F-test estimate the degree of linear dependency between two random variables. On the other hand, mutual information methods can capture any kind of statistical dependency, but being nonparametric, they require more samples for accurate estimation. Feature selection with sparse data If you use sparse data (i.e. data represented as sparse matrices), chi2 , mutual_info_regression , mutual_info_classif will deal with the data without making it dense. Warning Beware not to use a regression scoring function with a classification problem, you will get useless results. Examples: Univariate Feature Selection Comparison of F-test and mutual information ',\n",
       " 'sklearn_1_13_feature_selection\\n1.13. Feature selection\\nmodules/feature_selection.html\\n 1.13.3. Recursive feature elimination  Given an external estimator that assigns weights to features (e.g., the coefficients of a linear model), recursive feature elimination ( RFE ) is to select features by recursively considering smaller and smaller sets of features. First, the estimator is trained on the initial set of features and the importance of each feature is obtained either through a attribute or through a attribute. Then, the least important features are pruned from current set of features.That procedure is recursively repeated on the pruned set until the desired number of features to select is eventually reached. RFECV performs RFE in a cross-validation loop to find the optimal number of features. Examples: Recursive feature elimination : A recursive feature elimination example showing the relevance of pixels in a digit classification task. Recursive feature elimination with cross-validation : A recursive feature elimination example with automatic tuning of the number of features selected with cross-validation. ',\n",
       " 'sklearn_1_13_feature_selection\\n1.13. Feature selection\\nmodules/feature_selection.html\\n 1.13.4. Feature selection using SelectFromModel  SelectFromModel is a meta-transformer that can be used along with any estimator that has a or attribute after fitting. The features are considered unimportant and removed, if the corresponding or values are below the provided parameter. Apart from specifying the threshold numerically, there are built-in heuristics for finding a threshold using a string argument. Available heuristics are “mean”, “median” and float multiples of these like “0.1*mean”. In combination with the criteria, one can use the parameter to set a limit on the number of features to select. For examples on how it is to be used refer to the sections below. Examples Feature selection using SelectFromModel and LassoCV : Selecting the two most important features from the diabetes dataset without knowing the threshold beforehand. 1.13.4.1. L1-based feature selection  Linear models penalized with the L1 norm have sparse solutions: many of their estimated coefficients are zero. When the goal is to reduce the dimensionality of the data to use with another classifier, they can be used along with feature_selection.SelectFromModel to select the non-zero coefficients. In particular, sparse estimators useful for this purpose are the linear_model.Lasso for regression, and of linear_model.LogisticRegression and svm.LinearSVC for classification: With SVMs and logistic-regression, the parameter C controls the sparsity: the smaller C the fewer features selected. With Lasso, the higher the alpha parameter, the fewer features selected. Examples: Classification of text documents using sparse features : Comparison of different algorithms for document classification including L1-based feature selection. L1-recovery and compressive sensing For a good choice of alpha, the Lasso can fully recover the exact set of non-zero variables using only few observations, provided certain specific conditions are met. In particular, the number of samples should be “sufficiently large”, or L1 models will perform at random, where “sufficiently large” depends on the number of non-zero coefficients, the logarithm of the number of features, the amount of noise, the smallest absolute value of non-zero coefficients, and the structure of the design matrix X. In addition, the design matrix must display certain specific properties, such as not being too correlated. There is no general rule to select an alpha parameter for recovery of non-zero coefficients. It can by set by cross-validation ( LassoCV or LassoLarsCV ), though this may lead to under-penalized models: including a small number of non-relevant variables is not detrimental to prediction score. BIC ( LassoLarsIC ) tends, on the opposite, to set high values of alpha. Reference Richard G. Baraniuk “Compressive Sensing”, IEEE Signal Processing Magazine [120] July 2007 http://users.isr.ist.utl.pt/~aguiar/CS_notes.pdf 1.13.4.2. Tree-based feature selection  Tree-based estimators (see the sklearn.tree module and forest of trees in the sklearn.ensemble module) can be used to compute impurity-based feature importances, which in turn can be used to discard irrelevant features (when coupled with the sklearn.feature_selection.SelectFromModel meta-transformer): Examples: Feature importances with forests of trees : example on synthetic data showing the recovery of the actually meaningful features. Pixel importances with a parallel forest of trees : example on face recognition data. ',\n",
       " 'sklearn_1_13_feature_selection\\n1.13. Feature selection\\nmodules/feature_selection.html\\n 1.13.5. Feature selection as part of a pipeline  Feature selection is usually used as a pre-processing step before doing the actual learning. The recommended way to do this in scikit-learn is to use a sklearn.pipeline.Pipeline : In this snippet we make use of a sklearn.svm.LinearSVC coupled with sklearn.feature_selection.SelectFromModel to evaluate feature importances and select the most relevant features. Then, a sklearn.ensemble.RandomForestClassifier is trained on the transformed output, i.e. using only relevant features. You can perform similar operations with the other feature selection methods and also classifiers that provide a way to evaluate feature importances of course. See the sklearn.pipeline.Pipeline examples for more details. ',\n",
       " 'sklearn_1_14_semi-supervised\\n1.14. Semi-Supervised\\nmodules/label_propagation.html\\n 1.14.1. Label Propagation  Label propagation denotes a few variations of semi-supervised graph inference algorithms. A few features available in this model: Can be used for classification and regression tasks Kernel methods to project data into alternate dimensional spaces provides two label propagation models: LabelPropagation and LabelSpreading . Both work by constructing a similarity graph over all items in the input dataset. An illustration of label-propagation: the structure of unlabeled observations is consistent with the class structure, and thus the class label can be propagated to the unlabeled observations of the training set.  LabelPropagation and LabelSpreading differ in modifications to the similarity matrix that graph and the clamping effect on the label distributions. Clamping allows the algorithm to change the weight of the true ground labeled data to some degree. The LabelPropagation algorithm performs hard clamping of input labels, which means . This clamping factor can be relaxed, to say , which means that we will always retain 80 percent of our original label distribution, but the algorithm gets to change its confidence of the distribution within 20 percent. LabelPropagation uses the raw similarity matrix constructed from the data with no modifications. In contrast, LabelSpreading minimizes a loss function that has regularization properties, as such it is often more robust to noise. The algorithm iterates on a modified version of the original graph and normalizes the edge weights by computing the normalized graph Laplacian matrix. This procedure is also used in Spectral clustering . Label propagation models have two built-in kernel methods. Choice of kernel effects both scalability and performance of the algorithms. The following are available: rbf ( ). is specified by keyword gamma. knn ( ). is specified by keyword n_neighbors. The RBF kernel will produce a fully connected graph which is represented in memory by a dense matrix. This matrix may be very large and combined with the cost of performing a full matrix multiplication calculation for each iteration of the algorithm can lead to prohibitively long running times. On the other hand, the KNN kernel will produce a much more memory-friendly sparse matrix which can drastically reduce running times. Examples Decision boundary of label propagation versus SVM on the Iris dataset Label Propagation learning a complex structure Label Propagation digits: Demonstrating performance Label Propagation digits active learning References [1] Yoshua Bengio, Olivier Delalleau, Nicolas Le Roux. In Semi-Supervised Learning (2006), pp. 193-216 [2] Olivier Delalleau, Yoshua Bengio, Nicolas Le Roux. Efficient Non-Parametric Function Induction in Semi-Supervised Learning. AISTAT 2005 https://research.microsoft.com/en-us/people/nicolasl/efficient_ssl.pdf ',\n",
       " 'sklearn_1_16_probability_calibration\\n1.16. Probability calibration\\nmodules/calibration.html\\n 1.16.1. Calibration curves  The following plot compares how well the probabilistic predictions of different classifiers are calibrated, using calibration_curve . The x axis represents the average predicted probability in each bin. The y axis is the fraction of positives , i.e. the proportion of samples whose class is the positive class (in each bin). LogisticRegression returns well calibrated predictions by default as it directly optimizes log-loss. In contrast, the other methods return biased probabilities; with different biases per method: GaussianNB tends to push probabilities to 0 or 1 (note the counts in the histograms). This is mainly because it makes the assumption that features are conditionally independent given the class, which is not the case in this dataset which contains 2 redundant features. RandomForestClassifier shows the opposite behavior: the histograms show peaks at approximately 0.2 and 0.9 probability, while probabilities close to 0 or 1 are very rare. An explanation for this is given by Niculescu-Mizil and Caruana 1 : “Methods such as bagging and random forests that average predictions from a base set of models can have difficulty making predictions near 0 and 1 because variance in the underlying base models will bias predictions that should be near zero or one away from these values. Because predictions are restricted to the interval [0,1], errors caused by variance tend to be one-sided near zero and one. For example, if a model should predict p  0 for a case, the only way bagging can achieve this is if all bagged trees predict zero. If we add noise to the trees that bagging is averaging over, this noise will cause some trees to predict values larger than 0 for this case, thus moving the average prediction of the bagged ensemble away from 0. We observe this effect most strongly with random forests because the base-level trees trained with random forests have relatively high variance due to feature subsetting.” As a result, the calibration curve also referred to as the reliability diagram (Wilks 1995 2 ) shows a characteristic sigmoid shape, indicating that the classifier could trust its “intuition” more and return probabilities closer to 0 or 1 typically. Linear Support Vector Classification ( LinearSVC ) shows an even more sigmoid curve as the RandomForestClassifier, which is typical for maximum-margin methods (compare Niculescu-Mizil and Caruana 1 ), which focus on hard samples that are close to the decision boundary (the support vectors). ',\n",
       " 'sklearn_1_16_probability_calibration\\n1.16. Probability calibration\\nmodules/calibration.html\\n 1.16.2. Calibrating a classifier  Calibrating a classifier consists in fitting a regressor (called a calibrator ) that maps the output of the classifier (as given by predict or predict_proba ) to a calibrated probability in [0, 1]. Denoting the output of the classifier for a given sample by , the calibrator tries to predict . The samples that are used to train the calibrator should not be used to train the target classifier. ',\n",
       " 'sklearn_1_16_probability_calibration\\n1.16. Probability calibration\\nmodules/calibration.html\\n 1.16.3. Usage  The CalibratedClassifierCV class is used to calibrate a classifier. CalibratedClassifierCV uses a cross-validation approach to fit both the classifier and the regressor. For each of the k couple, a classifier is trained on the train set, and its predictions on the test set are used to fit a regressor. We end up with k couples where each regressor maps the output of its corresponding classifier into [0, 1]. Each couple is exposed in the attribute, where each entry is a calibrated classifier with a predict_proba method that outputs calibrated probabilities. The output of predict_proba for the main CalibratedClassifierCV instance corresponds to the average of the predicted probabilities of the estimators in the list. The output of predict is the class that has the highest probability. The regressor that is used for calibration depends on the parameter. corresponds to a parametric approach based on Platt’s logistic model 3 , i.e. is modeled as where is the logistic function, and and are real numbers to be determined when fitting the regressor via maximum likelihood. will instead fit a non-parametric isotonic regressor, which outputs a step-wise non-decreasing function (see sklearn.isotonic ). An already fitted classifier can be calibrated by setting . In this case, the data is only used to fit the regressor. It is up to the user make sure that the data used for fitting the classifier is disjoint from the data used for fitting the regressor. CalibratedClassifierCV can calibrate probabilities in a multiclass setting if the base estimator supports multiclass predictions. The classifier is calibrated first for each class separately in a one-vs-rest fashion 4 . When predicting probabilities, the calibrated probabilities for each class are predicted separately. As those probabilities do not necessarily sum to one, a postprocessing is performed to normalize them. The sklearn.metrics.brier_score_loss may be used to evaluate how well a classifier is calibrated. Examples: Probability Calibration curves Probability Calibration for 3-class classification Probability calibration of classifiers Comparison of Calibration of Classifiers References: 1 ( 1 , 2 ) Predicting Good Probabilities with Supervised Learning, A. Niculescu-Mizil & R. Caruana, ICML 2005 2 On the combination of forecast probabilities for consecutive precipitation periods. Wea. Forecasting, 5, 640–650., Wilks, D. S., 1990a 3 Probabilistic Outputs for Support Vector Machines and Comparisons to Regularized Likelihood Methods, J. Platt, (1999) 4 Transforming Classifier Scores into Accurate Multiclass Probability Estimates, B. Zadrozny & C. Elkan, (KDD 2002) ',\n",
       " 'sklearn_1_17_neural_network_models_supervised\\n1.17. Neural network models (supervised)\\nmodules/neural_networks_supervised.html\\n 1.17.1. Multi-layer Perceptron  Multi-layer Perceptron (MLP) is a supervised learning algorithm that learns a function by training on a dataset, where is the number of dimensions for input and is the number of dimensions for output. Given a set of features and a target , it can learn a non-linear function approximator for either classification or regression. It is different from logistic regression, in that between the input and the output layer, there can be one or more non-linear layers, called hidden layers. Figure 1 shows a one hidden layer MLP with scalar output. Figure 1 : One hidden layer MLP.  The leftmost layer, known as the input layer, consists of a set of neurons representing the input features. Each neuron in the hidden layer transforms the values from the previous layer with a weighted linear summation , followed by a non-linear activation function - like the hyperbolic tan function. The output layer receives the values from the last hidden layer and transforms them into output values. The module contains the public attributes and . is a list of weight matrices, where weight matrix at index represents the weights between layer and layer . is a list of bias vectors, where the vector at index represents the bias values added to layer . The advantages of Multi-layer Perceptron are: Capability to learn non-linear models. Capability to learn models in real-time (on-line learning) using . The disadvantages of Multi-layer Perceptron (MLP) include: MLP with hidden layers have a non-convex loss function where there exists more than one local minimum. Therefore different random weight initializations can lead to different validation accuracy. MLP requires tuning a number of hyperparameters such as the number of hidden neurons, layers, and iterations. MLP is sensitive to feature scaling. Please see Tips on Practical Use section that addresses some of these disadvantages. ',\n",
       " 'sklearn_1_17_neural_network_models_supervised\\n1.17. Neural network models (supervised)\\nmodules/neural_networks_supervised.html\\n 1.17.2. Classification  Class MLPClassifier implements a multi-layer perceptron (MLP) algorithm that trains using Backpropagation . MLP trains on two arrays: array X of size (n_samples, n_features), which holds the training samples represented as floating point feature vectors; and array y of size (n_samples,), which holds the target values (class labels) for the training samples: After fitting (training), the model can predict labels for new samples: MLP can fit a non-linear model to the training data. contains the weight matrices that constitute the model parameters: Currently, MLPClassifier supports only the Cross-Entropy loss function, which allows probability estimates by running the method. MLP trains using Backpropagation. More precisely, it trains using some form of gradient descent and the gradients are calculated using Backpropagation. For classification, it minimizes the Cross-Entropy loss function, giving a vector of probability estimates per sample : MLPClassifier supports multi-class classification by applying Softmax as the output function. Further, the model supports multi-label classification in which a sample can belong to more than one class. For each class, the raw output passes through the logistic function. Values larger or equal to are rounded to , otherwise to . For a predicted output of a sample, the indices where the value is represents the assigned classes of that sample: See the examples below and the docstring of MLPClassifier.fit for further information. Examples: Compare Stochastic learning strategies for MLPClassifier Visualization of MLP weights on MNIST ',\n",
       " 'sklearn_1_17_neural_network_models_supervised\\n1.17. Neural network models (supervised)\\nmodules/neural_networks_supervised.html\\n 1.17.3. Regression  Class MLPRegressor implements a multi-layer perceptron (MLP) that trains using backpropagation with no activation function in the output layer, which can also be seen as using the identity function as activation function. Therefore, it uses the square error as the loss function, and the output is a set of continuous values. MLPRegressor also supports multi-output regression, in which a sample can have more than one target. ',\n",
       " 'sklearn_1_17_neural_network_models_supervised\\n1.17. Neural network models (supervised)\\nmodules/neural_networks_supervised.html\\n 1.17.4. Regularization  Both MLPRegressor and MLPClassifier use parameter for regularization (L2 regularization) term which helps in avoiding overfitting by penalizing weights with large magnitudes. Following plot displays varying decision function with value of alpha. See the examples below for further information. Examples: Varying regularization in Multi-layer Perceptron ',\n",
       " 'sklearn_1_17_neural_network_models_supervised\\n1.17. Neural network models (supervised)\\nmodules/neural_networks_supervised.html\\n 1.17.5. Algorithms  MLP trains using Stochastic Gradient Descent , Adam , or L-BFGS . Stochastic Gradient Descent (SGD) updates parameters using the gradient of the loss function with respect to a parameter that needs adaptation, i.e. \\\\[w \\\\leftarrow w - \\\\eta (\\\\alpha \\\\frac{\\\\partial R(w)}{\\\\partial w} + \\\\frac{\\\\partial Loss}{\\\\partial w})\\\\] where is the learning rate which controls the step-size in the parameter space search. is the loss function used for the network. More details can be found in the documentation of SGD Adam is similar to SGD in a sense that it is a stochastic optimizer, but it can automatically adjust the amount to update parameters based on adaptive estimates of lower-order moments. With SGD or Adam, training supports online and mini-batch learning. L-BFGS is a solver that approximates the Hessian matrix which represents the second-order partial derivative of a function. Further it approximates the inverse of the Hessian matrix to perform parameter updates. The implementation uses the Scipy version of L-BFGS . If the selected solver is ‘L-BFGS’, training does not support online nor mini-batch learning. ',\n",
       " 'sklearn_1_17_neural_network_models_supervised\\n1.17. Neural network models (supervised)\\nmodules/neural_networks_supervised.html\\n 1.17.6. Complexity  Suppose there are training samples, features, hidden layers, each containing neurons - for simplicity, and output neurons. The time complexity of backpropagation is , where is the number of iterations. Since backpropagation has a high time complexity, it is advisable to start with smaller number of hidden neurons and few hidden layers for training. ',\n",
       " 'sklearn_1_17_neural_network_models_supervised\\n1.17. Neural network models (supervised)\\nmodules/neural_networks_supervised.html\\n 1.17.7. Mathematical formulation  Given a set of training examples where and , a one hidden layer one hidden neuron MLP learns the function where and are model parameters. represent the weights of the input layer and hidden layer, respectively; and represent the bias added to the hidden layer and the output layer, respectively. is the activation function, set by default as the hyperbolic tan. It is given as, \\\\[g(z) \\\\frac{e^z-e^{-z}}{e^z+e^{-z}}\\\\] For binary classification, passes through the logistic function to obtain output values between zero and one. A threshold, set to 0.5, would assign samples of outputs larger or equal 0.5 to the positive class, and the rest to the negative class. If there are more than two classes, itself would be a vector of size (n_classes,). Instead of passing through logistic function, it passes through the softmax function, which is written as, \\\\[\\\\text{softmax}(z)_i  \\\\frac{\\\\exp(z_i)}{\\\\sum_{l1}^k\\\\exp(z_l)}\\\\] where represents the th element of the input to softmax, which corresponds to class , and is the number of classes. The result is a vector containing the probabilities that sample belong to each class. The output is the class with the highest probability. In regression, the output remains as ; therefore, output activation function is just the identity function. MLP uses different loss functions depending on the problem type. The loss function for classification is Cross-Entropy, which in binary case is given as, \\\\[Loss(\\\\hat{y},y,W)  -y \\\\ln {\\\\hat{y}} - (1-y) \\\\ln{(1-\\\\hat{y})} + \\\\alpha ||W||_2^2\\\\] where is an L2-regularization term (aka penalty) that penalizes complex models; and is a non-negative hyperparameter that controls the magnitude of the penalty. For regression, MLP uses the Square Error loss function; written as, \\\\[Loss(\\\\hat{y},y,W)  \\\\frac{1}{2}||\\\\hat{y} - y ||_2^2 + \\\\frac{\\\\alpha}{2} ||W||_2^2\\\\] Starting from initial random weights, multi-layer perceptron (MLP) minimizes the loss function by repeatedly updating these weights. After computing the loss, a backward pass propagates it from the output layer to the previous layers, providing each weight parameter with an update value meant to decrease the loss. In gradient descent, the gradient of the loss with respect to the weights is computed and deducted from . More formally, this is expressed as, \\\\[W^{i+1}  W^i - \\\\epsilon \\\\nabla {Loss}_{W}^{i}\\\\] where is the iteration step, and is the learning rate with a value larger than 0. The algorithm stops when it reaches a preset maximum number of iterations; or when the improvement in loss is below a certain, small number. ',\n",
       " 'sklearn_1_17_neural_network_models_supervised\\n1.17. Neural network models (supervised)\\nmodules/neural_networks_supervised.html\\n 1.17.8. Tips on Practical Use  Multi-layer Perceptron is sensitive to feature scaling, so it is highly recommended to scale your data. For example, scale each attribute on the input vector X to [0, 1] or [-1, +1], or standardize it to have mean 0 and variance 1. Note that you must apply the same scaling to the test set for meaningful results. You can use StandardScaler for standardization. An alternative and recommended approach is to use StandardScaler in a Pipeline Finding a reasonable regularization parameter is best done using GridSearchCV , usually in the range . Empirically, we observed that converges faster and with better solutions on small datasets. For relatively large datasets, however, is very robust. It usually converges quickly and gives pretty good performance. with momentum or nesterov’s momentum, on the other hand, can perform better than those two algorithms if learning rate is correctly tuned. ',\n",
       " 'sklearn_1_17_neural_network_models_supervised\\n1.17. Neural network models (supervised)\\nmodules/neural_networks_supervised.html\\n 1.17.9. More control with warm_start  If you want more control over stopping criteria or learning rate in SGD, or want to do additional monitoring, using and and iterating yourself can be helpful: References: “Learning representations by back-propagating errors.” Rumelhart, David E., Geoffrey E. Hinton, and Ronald J. Williams. “Stochastic Gradient Descent” L. Bottou - Website, 2010. “Backpropagation” Andrew Ng, Jiquan Ngiam, Chuan Yu Foo, Yifan Mai, Caroline Suen - Website, 2011. “Efficient BackProp” Y. LeCun, L. Bottou, G. Orr, K. Müller - In Neural Networks: Tricks of the Trade 1998. “Adam: A method for stochastic optimization.” Kingma, Diederik, and Jimmy Ba. arXiv preprint arXiv:1412.6980 (2014). ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.10. Bayesian Regression  Bayesian regression techniques can be used to include regularization parameters in the estimation procedure: the regularization parameter is not set in a hard sense but tuned to the data at hand. This can be done by introducing uninformative priors over the hyper parameters of the model. The regularization used in Ridge regression and classification is equivalent to finding a maximum a posteriori estimation under a Gaussian prior over the coefficients with precision . Instead of setting manually, it is possible to treat it as a random variable to be estimated from the data. To obtain a fully probabilistic model, the output is assumed to be Gaussian distributed around : \\\\[p(y|X,w,\\\\alpha)  \\\\mathcal{N}(y|X w,\\\\alpha)\\\\] where is again treated as a random variable that is to be estimated from the data. The advantages of Bayesian Regression are: It adapts to the data at hand. It can be used to include regularization parameters in the estimation procedure. The disadvantages of Bayesian regression include: Inference of the model can be time consuming. References A good introduction to Bayesian methods is given in C. Bishop: Pattern Recognition and Machine learning Original Algorithm is detailed in the book by Radford M. Neal 1.1.10.1. Bayesian Ridge Regression  BayesianRidge estimates a probabilistic model of the regression problem as described above. The prior for the coefficient is given by a spherical Gaussian: \\\\[p(w|\\\\lambda)  \\\\mathcal{N}(w|0,\\\\lambda^{-1}\\\\mathbf{I}_{p})\\\\] The priors over and are chosen to be gamma distributions , the conjugate prior for the precision of the Gaussian. The resulting model is called Bayesian Ridge Regression , and is similar to the classical Ridge . The parameters , and are estimated jointly during the fit of the model, the regularization parameters and being estimated by maximizing the log marginal likelihood . The scikit-learn implementation is based on the algorithm described in Appendix A of (Tipping, 2001) where the update of the parameters and is done as suggested in (MacKay, 1992). The initial value of the maximization procedure can be set with the hyperparameters and . There are four more hyperparameters, , , and of the gamma prior distributions over and . These are usually chosen to be non-informative . By default . Bayesian Ridge Regression is used for regression: After being fitted, the model can then be used to predict new values: The coefficients of the model can be accessed: Due to the Bayesian framework, the weights found are slightly different to the ones found by Ordinary Least Squares . However, Bayesian Ridge Regression is more robust to ill-posed problems. Examples: Bayesian Ridge Regression Curve Fitting with Bayesian Ridge Regression References: Section 3.3 in Christopher M. Bishop: Pattern Recognition and Machine Learning, 2006 David J. C. MacKay, Bayesian Interpolation , 1992. Michael E. Tipping, Sparse Bayesian Learning and the Relevance Vector Machine , 2001. 1.1.10.2. Automatic Relevance Determination - ARD  ARDRegression is very similar to Bayesian Ridge Regression , but can lead to sparser coefficients 1 2 . ARDRegression poses a different prior over , by dropping the assumption of the Gaussian being spherical. Instead, the distribution over is assumed to be an axis-parallel, elliptical Gaussian distribution. This means each coefficient is drawn from a Gaussian distribution, centered on zero and with a precision : \\\\[p(w|\\\\lambda)  \\\\mathcal{N}(w|0,A^{-1})\\\\] with . In contrast to Bayesian Ridge Regression , each coordinate of has its own standard deviation . The prior over all is chosen to be the same gamma distribution given by hyperparameters and . ARD is also known in the literature as Sparse Bayesian Learning and Relevance Vector Machine 3 4 . Examples: Automatic Relevance Determination Regression (ARD) References: 1 Christopher M. Bishop: Pattern Recognition and Machine Learning, Chapter 7.2.1 2 David Wipf and Srikantan Nagarajan: A new view of automatic relevance determination 3 Michael E. Tipping: Sparse Bayesian Learning and the Relevance Vector Machine 4 Tristan Fletcher: Relevance Vector Machines explained ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.11. Logistic regression  Logistic regression, despite its name, is a linear model for classification rather than regression. Logistic regression is also known in the literature as logit regression, maximum-entropy classification (MaxEnt) or the log-linear classifier. In this model, the probabilities describing the possible outcomes of a single trial are modeled using a logistic function . Logistic regression is implemented in LogisticRegression . This implementation can fit binary, One-vs-Rest, or multinomial logistic regression with optional , or Elastic-Net regularization. Note Regularization is applied by default, which is common in machine learning but not in statistics. Another advantage of regularization is that it improves numerical stability. No regularization amounts to setting C to a very high value. As an optimization problem, binary class penalized logistic regression minimizes the following cost function: \\\\[\\\\min_{w, c} \\\\frac{1}{2}w^T w + C \\\\sum_{i1}^n \\\\log(\\\\exp(- y_i (X_i^T w + c)) + 1) .\\\\] Similarly, regularized logistic regression solves the following optimization problem: \\\\[\\\\min_{w, c} \\\\|w\\\\|_1 + C \\\\sum_{i1}^n \\\\log(\\\\exp(- y_i (X_i^T w + c)) + 1).\\\\] Elastic-Net regularization is a combination of and , and minimizes the following cost function: \\\\[\\\\min_{w, c} \\\\frac{1 - \\\\rho}{2}w^T w + \\\\rho \\\\|w\\\\|_1 + C \\\\sum_{i1}^n \\\\log(\\\\exp(- y_i (X_i^T w + c)) + 1),\\\\] where controls the strength of regularization vs. regularization (it corresponds to the parameter). Note that, in this notation, it’s assumed that the target takes values in the set at trial . We can also see that Elastic-Net is equivalent to when and equivalent to when . The solvers implemented in the class LogisticRegression are “liblinear”, “newton-cg”, “lbfgs”, “sag” and “saga”: The solver “liblinear” uses a coordinate descent (CD) algorithm, and relies on the excellent C++ LIBLINEAR library , which is shipped with scikit-learn. However, the CD algorithm implemented in liblinear cannot learn a true multinomial (multiclass) model; instead, the optimization problem is decomposed in a “one-vs-rest” fashion so separate binary classifiers are trained for all classes. This happens under the hood, so LogisticRegression instances using this solver behave as multiclass classifiers. For regularization sklearn.svm.l1_min_c allows to calculate the lower bound for C in order to get a non “null” (all feature weights to zero) model. The “lbfgs”, “sag” and “newton-cg” solvers only support regularization or no regularization, and are found to converge faster for some high-dimensional data. Setting to “multinomial” with these solvers learns a true multinomial logistic regression model 5 , which means that its probability estimates should be better calibrated than the default “one-vs-rest” setting. The “sag” solver uses Stochastic Average Gradient descent 6 . It is faster than other solvers for large datasets, when both the number of samples and the number of features are large. The “saga” solver 7 is a variant of “sag” that also supports the non-smooth . This is therefore the solver of choice for sparse multinomial logistic regression. It is also the only solver that supports . The “lbfgs” is an optimization algorithm that approximates the Broyden–Fletcher–Goldfarb–Shanno algorithm 8 , which belongs to quasi-Newton methods. The “lbfgs” solver is recommended for use for small data-sets but for larger datasets its performance suffers. 9 The following table summarizes the penalties supported by each solver: Solvers Penalties ‘liblinear’ ‘lbfgs’ ‘newton-cg’ ‘sag’ ‘saga’ Multinomial + L2 penalty no yes yes yes yes OVR + L2 penalty yes yes yes yes yes Multinomial + L1 penalty no no no no yes OVR + L1 penalty yes no no no yes Elastic-Net no no no no yes No penalty (‘none’) no yes yes yes yes Behaviors Penalize the intercept (bad) yes no no no no Faster for large datasets no no no yes yes Robust to unscaled datasets yes yes yes no no The “lbfgs” solver is used by default for its robustness. For large datasets the “saga” solver is usually faster. For large dataset, you may also consider using SGDClassifier with ‘log’ loss, which might be even faster but requires more tuning. Examples: L1 Penalty and Sparsity in Logistic Regression Regularization path of L1- Logistic Regression Plot multinomial and One-vs-Rest Logistic Regression Multiclass sparse logistic regression on 20newgroups MNIST classification using multinomial logistic + L1 Differences from liblinear: There might be a difference in the scores obtained between LogisticRegression with or LinearSVC and the external liblinear library directly, when and the fit (or) the data to be predicted are zeroes. This is because for the sample(s) with zero, LogisticRegression and LinearSVC predict the negative class, while liblinear predicts the positive class. Note that a model with and having many samples with zero, is likely to be a underfit, bad model and you are advised to set and increase the intercept_scaling. Note Feature selection with sparse logistic regression A logistic regression with penalty yields sparse models, and can thus be used to perform feature selection, as detailed in L1-based feature selection . Note P-value estimation It is possible to obtain the p-values and confidence intervals for coefficients in cases of regression without penalization. The natively supports this. Within sklearn, one could use bootstrapping instead as well. LogisticRegressionCV implements Logistic Regression with built-in cross-validation support, to find the optimal and parameters according to the attribute. The “newton-cg”, “sag”, “saga” and “lbfgs” solvers are found to be faster for high-dimensional dense data, due to warm-starting (see Glossary ). References: 5 Christopher M. Bishop: Pattern Recognition and Machine Learning, Chapter 4.3.4 6 Mark Schmidt, Nicolas Le Roux, and Francis Bach: Minimizing Finite Sums with the Stochastic Average Gradient. 7 Aaron Defazio, Francis Bach, Simon Lacoste-Julien: SAGA: A Fast Incremental Gradient Method With Support for Non-Strongly Convex Composite Objectives. 8 https://en.wikipedia.org/wiki/Broyden%E2%80%93Fletcher%E2%80%93Goldfarb%E2%80%93Shanno_algorithm 9 “Performance Evaluation of Lbfgs vs other solvers” ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.12. Generalized Linear Regression  Generalized Linear Models (GLM) extend linear models in two ways 10 . First, the predicted values are linked to a linear combination of the input variables via an inverse link function as \\\\[\\\\hat{y}(w, X)  h(Xw).\\\\] Secondly, the squared loss function is replaced by the unit deviance of a distribution in the exponential family (or more precisely, a reproductive exponential dispersion model (EDM) 11 ). The minimization problem becomes: \\\\[\\\\min_{w} \\\\frac{1}{2 n_{\\\\text{samples}}} \\\\sum_i d(y_i, \\\\hat{y}_i) + \\\\frac{\\\\alpha}{2} ||w||_2,\\\\] where is the L2 regularization penalty. When sample weights are provided, the average becomes a weighted average. The following table lists some specific EDMs and their unit deviance (all of these are instances of the Tweedie family): Distribution Target Domain Unit Deviance Normal Poisson Gamma Inverse Gaussian The Probability Density Functions (PDF) of these distributions are illustrated in the following figure, PDF of a random variable Y following Poisson, Tweedie (power1.5) and Gamma distributions with different mean values ( ). Observe the point mass at for the Poisson distribution and the Tweedie (power1.5) distribution, but not for the Gamma distribution which has a strictly positive target domain.  The choice of the distribution depends on the problem at hand: If the target values are counts (non-negative integer valued) or relative frequencies (non-negative), you might use a Poisson deviance with log-link. If the target values are positive valued and skewed, you might try a Gamma deviance with log-link. If the target values seem to be heavier tailed than a Gamma distribution, you might try an Inverse Gaussian deviance (or even higher variance powers of the Tweedie family). Examples of use cases include: Agriculture / weather modeling: number of rain events per year (Poisson), amount of rainfall per event (Gamma), total rainfall per year (Tweedie / Compound Poisson Gamma). Risk modeling / insurance policy pricing: number of claim events / policyholder per year (Poisson), cost per event (Gamma), total cost per policyholder per year (Tweedie / Compound Poisson Gamma). Predictive maintenance: number of production interruption events per year (Poisson), duration of interruption (Gamma), total interruption time per year (Tweedie / Compound Poisson Gamma). References: 10 McCullagh, Peter; Nelder, John (1989). Generalized Linear Models, Second Edition. Boca Raton: Chapman and Hall/CRC. ISBN 0-412-31760-5. 11 Jørgensen, B. (1992). The theory of exponential dispersion models and analysis of deviance. Monografias de matemática, no. 51. See also Exponential dispersion model. 1.1.12.1. Usage  TweedieRegressor implements a generalized linear model for the Tweedie distribution, that allows to model any of the above mentioned distributions using the appropriate parameter. In particular: : Normal distribution. Specific estimators such as Ridge , ElasticNet are generally more appropriate in this case. : Poisson distribution. PoissonRegressor is exposed for convenience. However, it is strictly equivalent to . : Gamma distribution. GammaRegressor is exposed for convenience. However, it is strictly equivalent to . : Inverse Gaussian distribution. The link function is determined by the parameter. Usage example: Examples: Poisson regression and non-normal loss Tweedie regression on insurance claims 1.1.12.2. Practical considerations  The feature matrix should be standardized before fitting. This ensures that the penalty treats features equally. Since the linear predictor can be negative and Poisson, Gamma and Inverse Gaussian distributions don’t support negative values, it is necessary to apply an inverse link function that guarantees the non-negativeness. For example with , the inverse link function becomes . If you want to model a relative frequency, i.e. counts per exposure (time, volume, …) you can do so by using a Poisson distribution and passing as target values together with as sample weights. For a concrete example see e.g. Tweedie regression on insurance claims . When performing cross-validation for the parameter of , it is advisable to specify an explicit function, because the default scorer TweedieRegressor.score is a function of itself. ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.13. Stochastic Gradient Descent - SGD  Stochastic gradient descent is a simple yet very efficient approach to fit linear models. It is particularly useful when the number of samples (and the number of features) is very large. The method allows online/out-of-core learning. The classes SGDClassifier and SGDRegressor provide functionality to fit linear models for classification and regression using different (convex) loss functions and different penalties. E.g., with , SGDClassifier fits a logistic regression model, while with it fits a linear support vector machine (SVM). References Stochastic Gradient Descent ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.14. Perceptron  The Perceptron is another simple classification algorithm suitable for large scale learning. By default: It does not require a learning rate. It is not regularized (penalized). It updates its model only on mistakes. The last characteristic implies that the Perceptron is slightly faster to train than SGD with the hinge loss and that the resulting models are sparser. ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.15. Passive Aggressive Algorithms  The passive-aggressive algorithms are a family of algorithms for large-scale learning. They are similar to the Perceptron in that they do not require a learning rate. However, contrary to the Perceptron, they include a regularization parameter . For classification, PassiveAggressiveClassifier can be used with (PA-I) or (PA-II). For regression, PassiveAggressiveRegressor can be used with (PA-I) or (PA-II). References: “Online Passive-Aggressive Algorithms” K. Crammer, O. Dekel, J. Keshat, S. Shalev-Shwartz, Y. Singer - JMLR 7 (2006) ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.16. Robustness regression: outliers and modeling errors  Robust regression aims to fit a regression model in the presence of corrupt data: either outliers, or error in the model. 1.1.16.1. Different scenario and useful concepts  There are different things to keep in mind when dealing with data corrupted by outliers: Outliers in X or in y ? Outliers in the y direction Outliers in the X direction Fraction of outliers versus amplitude of error The number of outlying points matters, but also how much they are outliers. Small outliers Large outliers An important notion of robust fitting is that of breakdown point: the fraction of data that can be outlying for the fit to start missing the inlying data. Note that in general, robust fitting in high-dimensional setting (large ) is very hard. The robust models here will probably not work in these settings. Trade-offs: which estimator? Scikit-learn provides 3 robust regression estimators: RANSAC , Theil Sen and HuberRegressor . HuberRegressor should be faster than RANSAC and Theil Sen unless the number of samples are very large, i.e >> . This is because RANSAC and Theil Sen fit on smaller subsets of the data. However, both Theil Sen and RANSAC are unlikely to be as robust as HuberRegressor for the default parameters. RANSAC is faster than Theil Sen and scales much better with the number of samples. RANSAC will deal better with large outliers in the y direction (most common situation). Theil Sen will cope better with medium-size outliers in the X direction, but this property will disappear in high-dimensional settings. When in doubt, use RANSAC . 1.1.16.2. RANSAC: RANdom SAmple Consensus  RANSAC (RANdom SAmple Consensus) fits a model from random subsets of inliers from the complete data set. RANSAC is a non-deterministic algorithm producing only a reasonable result with a certain probability, which is dependent on the number of iterations (see parameter). It is typically used for linear and non-linear regression problems and is especially popular in the field of photogrammetric computer vision. The algorithm splits the complete input sample data into a set of inliers, which may be subject to noise, and outliers, which are e.g. caused by erroneous measurements or invalid hypotheses about the data. The resulting model is then estimated only from the determined inliers. 1.1.16.2.1. Details of the algorithm  Each iteration performs the following steps: Select random samples from the original data and check whether the set of data is valid (see ). Fit a model to the random subset ( ) and check whether the estimated model is valid (see ). Classify all data as inliers or outliers by calculating the residuals to the estimated model ( ) - all data samples with absolute residuals smaller than the are considered as inliers. Save fitted model as best model if number of inlier samples is maximal. In case the current estimated model has the same number of inliers, it is only considered as the best model if it has better score. These steps are performed either a maximum number of times ( ) or until one of the special stop criteria are met (see and ). The final model is estimated using all inlier samples (consensus set) of the previously determined best model. The and functions allow to identify and reject degenerate combinations of random sub-samples. If the estimated model is not needed for identifying degenerate cases, should be used as it is called prior to fitting the model and thus leading to better computational performance. Examples: Robust linear model estimation using RANSAC Robust linear estimator fitting References: https://en.wikipedia.org/wiki/RANSAC “Random Sample Consensus: A Paradigm for Model Fitting with Applications to Image Analysis and Automated Cartography” Martin A. Fischler and Robert C. Bolles - SRI International (1981) “Performance Evaluation of RANSAC Family” Sunglok Choi, Taemin Kim and Wonpil Yu - BMVC (2009) 1.1.16.3. Theil-Sen estimator: generalized-median-based estimator  The TheilSenRegressor estimator uses a generalization of the median in multiple dimensions. It is thus robust to multivariate outliers. Note however that the robustness of the estimator decreases quickly with the dimensionality of the problem. It loses its robustness properties and becomes no better than an ordinary least squares in high dimension. Examples: Theil-Sen Regression Robust linear estimator fitting References: https://en.wikipedia.org/wiki/Theil%E2%80%93Sen_estimator 1.1.16.3.1. Theoretical considerations  TheilSenRegressor is comparable to the Ordinary Least Squares (OLS) in terms of asymptotic efficiency and as an unbiased estimator. In contrast to OLS, Theil-Sen is a non-parametric method which means it makes no assumption about the underlying distribution of the data. Since Theil-Sen is a median-based estimator, it is more robust against corrupted data aka outliers. In univariate setting, Theil-Sen has a breakdown point of about 29.3% in case of a simple linear regression which means that it can tolerate arbitrary corrupted data of up to 29.3%. The implementation of TheilSenRegressor in scikit-learn follows a generalization to a multivariate linear regression model 12 using the spatial median which is a generalization of the median to multiple dimensions 13 . In terms of time and space complexity, Theil-Sen scales according to \\\\[\\\\binom{n_{\\\\text{samples}}}{n_{\\\\text{subsamples}}}\\\\] which makes it infeasible to be applied exhaustively to problems with a large number of samples and features. Therefore, the magnitude of a subpopulation can be chosen to limit the time and space complexity by considering only a random subset of all possible combinations. Examples: Theil-Sen Regression References: 12 Xin Dang, Hanxiang Peng, Xueqin Wang and Heping Zhang: Theil-Sen Estimators in a Multiple Linear Regression Model. 13 Kärkkäinen and S. Äyrämö: On Computation of Spatial Median for Robust Data Mining. 1.1.16.4. Huber Regression  The HuberRegressor is different to Ridge because it applies a linear loss to samples that are classified as outliers. A sample is classified as an inlier if the absolute error of that sample is lesser than a certain threshold. It differs from TheilSenRegressor and RANSACRegressor because it does not ignore the effect of the outliers but gives a lesser weight to them. The loss function that HuberRegressor minimizes is given by \\\\[\\\\min_{w, \\\\sigma} {\\\\sum_{i1}^n\\\\left(\\\\sigma + H_{\\\\epsilon}\\\\left(\\\\frac{X_{i}w - y_{i}}{\\\\sigma}\\\\right)\\\\sigma\\\\right) + \\\\alpha {||w||_2}^2}\\\\] where \\\\[\\\\begin{split}H_{\\\\epsilon}(z)  \\\\begin{cases} z^2, & \\\\text {if } |z| < \\\\epsilon, \\\\\\\\ 2\\\\epsilon|z| - \\\\epsilon^2, & \\\\text{otherwise} \\\\end{cases}\\\\end{split}\\\\] It is advised to set the parameter to 1.35 to achieve 95% statistical efficiency. 1.1.16.5. Notes  The HuberRegressor differs from using SGDRegressor with loss set to in the following ways. HuberRegressor is scaling invariant. Once is set, scaling and down or up by different values would produce the same robustness to outliers as before. as compared to SGDRegressor where has to be set again when and are scaled. HuberRegressor should be more efficient to use on data with small number of samples while SGDRegressor needs a number of passes on the training data to produce the same robustness. Examples: HuberRegressor vs Ridge on dataset with strong outliers References: Peter J. Huber, Elvezio M. Ronchetti: Robust Statistics, Concomitant scale estimates, pg 172 Note that this estimator is different from the R implementation of Robust Regression ( http://www.ats.ucla.edu/stat/r/dae/rreg.htm ) because the R implementation does a weighted least squares implementation with weights given to each sample on the basis of how much the residual is greater than a certain threshold. ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.17. Polynomial regression: extending linear models with basis functions  One common pattern within machine learning is to use linear models trained on nonlinear functions of the data. This approach maintains the generally fast performance of linear methods, while allowing them to fit a much wider range of data. For example, a simple linear regression can be extended by constructing polynomial features from the coefficients. In the standard linear regression case, you might have a model that looks like this for two-dimensional data: \\\\[\\\\hat{y}(w, x)  w_0 + w_1 x_1 + w_2 x_2\\\\] If we want to fit a paraboloid to the data instead of a plane, we can combine the features in second-order polynomials, so that the model looks like this: \\\\[\\\\hat{y}(w, x)  w_0 + w_1 x_1 + w_2 x_2 + w_3 x_1 x_2 + w_4 x_1^2 + w_5 x_2^2\\\\] The (sometimes surprising) observation is that this is still a linear model : to see this, imagine creating a new set of features \\\\[z  [x_1, x_2, x_1 x_2, x_1^2, x_2^2]\\\\] With this re-labeling of the data, our problem can be written \\\\[\\\\hat{y}(w, z)  w_0 + w_1 z_1 + w_2 z_2 + w_3 z_3 + w_4 z_4 + w_5 z_5\\\\] We see that the resulting polynomial regression is in the same class of linear models we considered above (i.e. the model is linear in ) and can be solved by the same techniques. By considering linear fits within a higher-dimensional space built with these basis functions, the model has the flexibility to fit a much broader range of data. Here is an example of applying this idea to one-dimensional data, using polynomial features of varying degrees: This figure is created using the PolynomialFeatures transformer, which transforms an input data matrix into a new data matrix of a given degree. It can be used as follows: The features of have been transformed from to , and can now be used within any linear model. This sort of preprocessing can be streamlined with the Pipeline tools. A single object representing a simple polynomial regression can be created and used as follows: The linear model trained on polynomial features is able to exactly recover the input polynomial coefficients. In some cases it’s not necessary to include higher powers of any single feature, but only the so-called interaction features that multiply together at most distinct features. These can be gotten from PolynomialFeatures with the setting . For example, when dealing with boolean features, for all and is therefore useless; but represents the conjunction of two booleans. This way, we can solve the XOR problem with a linear classifier: And the classifier “predictions” are perfect: ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.1. Ordinary Least Squares  LinearRegression fits a linear model with coefficients to minimize the residual sum of squares between the observed targets in the dataset, and the targets predicted by the linear approximation. Mathematically it solves a problem of the form: \\\\[\\\\min_{w} || X w - y||_2^2\\\\] LinearRegression will take in its method arrays X, y and will store the coefficients of the linear model in its member: The coefficient estimates for Ordinary Least Squares rely on the independence of the features. When features are correlated and the columns of the design matrix have an approximate linear dependence, the design matrix becomes close to singular and as a result, the least-squares estimate becomes highly sensitive to random errors in the observed target, producing a large variance. This situation of multicollinearity can arise, for example, when data are collected without an experimental design. Examples: Linear Regression Example 1.1.1.1. Ordinary Least Squares Complexity  The least squares solution is computed using the singular value decomposition of X. If X is a matrix of shape this method has a cost of , assuming that . ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.2. Ridge regression and classification  1.1.2.1. Regression  Ridge regression addresses some of the problems of Ordinary Least Squares by imposing a penalty on the size of the coefficients. The ridge coefficients minimize a penalized residual sum of squares: \\\\[\\\\min_{w} || X w - y||_2^2 + \\\\alpha ||w||_2^2\\\\] The complexity parameter controls the amount of shrinkage: the larger the value of , the greater the amount of shrinkage and thus the coefficients become more robust to collinearity. As with other linear models, Ridge will take in its method arrays X, y and will store the coefficients of the linear model in its member: 1.1.2.2. Classification  The Ridge regressor has a classifier variant: RidgeClassifier . This classifier first converts binary targets to and then treats the problem as a regression task, optimizing the same objective as above. The predicted class corresponds to the sign of the regressor’s prediction. For multiclass classification, the problem is treated as multi-output regression, and the predicted class corresponds to the output with the highest value. It might seem questionable to use a (penalized) Least Squares loss to fit a classification model instead of the more traditional logistic or hinge losses. However in practice all those models can lead to similar cross-validation scores in terms of accuracy or precision/recall, while the penalized least squares loss used by the RidgeClassifier allows for a very different choice of the numerical solvers with distinct computational performance profiles. The RidgeClassifier can be significantly faster than e.g. LogisticRegression with a high number of classes, because it is able to compute the projection matrix only once. This classifier is sometimes referred to as a Least Squares Support Vector Machines with a linear kernel. Examples: Plot Ridge coefficients as a function of the regularization Classification of text documents using sparse features Common pitfalls in interpretation of coefficients of linear models 1.1.2.3. Ridge Complexity  This method has the same order of complexity as Ordinary Least Squares . 1.1.2.4. Setting the regularization parameter: generalized Cross-Validation  RidgeCV implements ridge regression with built-in cross-validation of the alpha parameter. The object works in the same way as GridSearchCV except that it defaults to Generalized Cross-Validation (GCV), an efficient form of leave-one-out cross-validation: Specifying the value of the cv attribute will trigger the use of cross-validation with GridSearchCV , for example for 10-fold cross-validation, rather than Generalized Cross-Validation. References “Notes on Regularized Least Squares”, Rifkin & Lippert ( technical report , course slides ). ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.3. Lasso  The Lasso is a linear model that estimates sparse coefficients. It is useful in some contexts due to its tendency to prefer solutions with fewer non-zero coefficients, effectively reducing the number of features upon which the given solution is dependent. For this reason Lasso and its variants are fundamental to the field of compressed sensing. Under certain conditions, it can recover the exact set of non-zero coefficients (see Compressive sensing: tomography reconstruction with L1 prior (Lasso) ). Mathematically, it consists of a linear model with an added regularization term. The objective function to minimize is: \\\\[\\\\min_{w} { \\\\frac{1}{2n_{\\\\text{samples}}} ||X w - y||_2 ^ 2 + \\\\alpha ||w||_1}\\\\] The lasso estimate thus solves the minimization of the least-squares penalty with added, where is a constant and is the -norm of the coefficient vector. The implementation in the class Lasso uses coordinate descent as the algorithm to fit the coefficients. See Least Angle Regression for another implementation: The function lasso_path is useful for lower-level tasks, as it computes the coefficients along the full path of possible values. Examples: Lasso and Elastic Net for Sparse Signals Compressive sensing: tomography reconstruction with L1 prior (Lasso) Common pitfalls in interpretation of coefficients of linear models Note Feature selection with Lasso As the Lasso regression yields sparse models, it can thus be used to perform feature selection, as detailed in L1-based feature selection . The following two references explain the iterations used in the coordinate descent solver of scikit-learn, as well as the duality gap computation used for convergence control. References “Regularization Path For Generalized linear Models by Coordinate Descent”, Friedman, Hastie & Tibshirani, J Stat Softw, 2010 ( Paper ). “An Interior-Point Method for Large-Scale L1-Regularized Least Squares,” S. J. Kim, K. Koh, M. Lustig, S. Boyd and D. Gorinevsky, in IEEE Journal of Selected Topics in Signal Processing, 2007 ( Paper ) 1.1.3.1. Setting regularization parameter  The parameter controls the degree of sparsity of the estimated coefficients. 1.1.3.1.1. Using cross-validation  scikit-learn exposes objects that set the Lasso parameter by cross-validation: LassoCV and LassoLarsCV . LassoLarsCV is based on the Least Angle Regression algorithm explained below. For high-dimensional datasets with many collinear features, LassoCV is most often preferable. However, LassoLarsCV has the advantage of exploring more relevant values of parameter, and if the number of samples is very small compared to the number of features, it is often faster than LassoCV . 1.1.3.1.2. Information-criteria based model selection  Alternatively, the estimator LassoLarsIC proposes to use the Akaike information criterion (AIC) and the Bayes Information criterion (BIC). It is a computationally cheaper alternative to find the optimal value of alpha as the regularization path is computed only once instead of k+1 times when using k-fold cross-validation. However, such criteria needs a proper estimation of the degrees of freedom of the solution, are derived for large samples (asymptotic results) and assume the model is correct, i.e. that the data are actually generated by this model. They also tend to break when the problem is badly conditioned (more features than samples). Examples: Lasso model selection: Cross-Validation / AIC / BIC 1.1.3.1.3. Comparison with the regularization parameter of SVM  The equivalence between and the regularization parameter of SVM, is given by or , depending on the estimator and the exact objective function optimized by the model. ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.4. Multi-task Lasso  The MultiTaskLasso is a linear model that estimates sparse coefficients for multiple regression problems jointly: is a 2D array, of shape . The constraint is that the selected features are the same for all the regression problems, also called tasks. The following figure compares the location of the non-zero entries in the coefficient matrix W obtained with a simple Lasso or a MultiTaskLasso. The Lasso estimates yield scattered non-zeros while the non-zeros of the MultiTaskLasso are full columns. Fitting a time-series model, imposing that any active feature be active at all times. Examples: Joint feature selection with multi-task Lasso Mathematically, it consists of a linear model trained with a mixed -norm for regularization. The objective function to minimize is: \\\\[\\\\min_{w} { \\\\frac{1}{2n_{\\\\text{samples}}} ||X W - Y||_{\\\\text{Fro}} ^ 2 + \\\\alpha ||W||_{21}}\\\\] where indicates the Frobenius norm \\\\[||A||_{\\\\text{Fro}}  \\\\sqrt{\\\\sum_{ij} a_{ij}^2}\\\\] and reads \\\\[||A||_{2 1}  \\\\sum_i \\\\sqrt{\\\\sum_j a_{ij}^2}.\\\\] The implementation in the class MultiTaskLasso uses coordinate descent as the algorithm to fit the coefficients. ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.5. Elastic-Net  ElasticNet is a linear regression model trained with both and -norm regularization of the coefficients. This combination allows for learning a sparse model where few of the weights are non-zero like Lasso , while still maintaining the regularization properties of Ridge . We control the convex combination of and using the parameter. Elastic-net is useful when there are multiple features which are correlated with one another. Lasso is likely to pick one of these at random, while elastic-net is likely to pick both. A practical advantage of trading-off between Lasso and Ridge is that it allows Elastic-Net to inherit some of Ridge’s stability under rotation. The objective function to minimize is in this case \\\\[\\\\min_{w} { \\\\frac{1}{2n_{\\\\text{samples}}} ||X w - y||_2 ^ 2 + \\\\alpha \\\\rho ||w||_1 + \\\\frac{\\\\alpha(1-\\\\rho)}{2} ||w||_2 ^ 2}\\\\] The class ElasticNetCV can be used to set the parameters ( ) and ( ) by cross-validation. Examples: Lasso and Elastic Net for Sparse Signals Lasso and Elastic Net The following two references explain the iterations used in the coordinate descent solver of scikit-learn, as well as the duality gap computation used for convergence control. References “Regularization Path For Generalized linear Models by Coordinate Descent”, Friedman, Hastie & Tibshirani, J Stat Softw, 2010 ( Paper ). “An Interior-Point Method for Large-Scale L1-Regularized Least Squares,” S. J. Kim, K. Koh, M. Lustig, S. Boyd and D. Gorinevsky, in IEEE Journal of Selected Topics in Signal Processing, 2007 ( Paper ) ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.6. Multi-task Elastic-Net  The MultiTaskElasticNet is an elastic-net model that estimates sparse coefficients for multiple regression problems jointly: is a 2D array of shape . The constraint is that the selected features are the same for all the regression problems, also called tasks. Mathematically, it consists of a linear model trained with a mixed -norm and -norm for regularization. The objective function to minimize is: \\\\[\\\\min_{W} { \\\\frac{1}{2n_{\\\\text{samples}}} ||X W - Y||_{\\\\text{Fro}}^2 + \\\\alpha \\\\rho ||W||_{2 1} + \\\\frac{\\\\alpha(1-\\\\rho)}{2} ||W||_{\\\\text{Fro}}^2}\\\\] The implementation in the class MultiTaskElasticNet uses coordinate descent as the algorithm to fit the coefficients. The class MultiTaskElasticNetCV can be used to set the parameters ( ) and ( ) by cross-validation. ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.7. Least Angle Regression  Least-angle regression (LARS) is a regression algorithm for high-dimensional data, developed by Bradley Efron, Trevor Hastie, Iain Johnstone and Robert Tibshirani. LARS is similar to forward stepwise regression. At each step, it finds the feature most correlated with the target. When there are multiple features having equal correlation, instead of continuing along the same feature, it proceeds in a direction equiangular between the features. The advantages of LARS are: It is numerically efficient in contexts where the number of features is significantly greater than the number of samples. It is computationally just as fast as forward selection and has the same order of complexity as ordinary least squares. It produces a full piecewise linear solution path, which is useful in cross-validation or similar attempts to tune the model. If two features are almost equally correlated with the target, then their coefficients should increase at approximately the same rate. The algorithm thus behaves as intuition would expect, and also is more stable. It is easily modified to produce solutions for other estimators, like the Lasso. The disadvantages of the LARS method include: Because LARS is based upon an iterative refitting of the residuals, it would appear to be especially sensitive to the effects of noise. This problem is discussed in detail by Weisberg in the discussion section of the Efron et al. (2004) Annals of Statistics article. The LARS model can be used using estimator Lars , or its low-level implementation lars_path or lars_path_gram . ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.8. LARS Lasso  LassoLars is a lasso model implemented using the LARS algorithm, and unlike the implementation based on coordinate descent, this yields the exact solution, which is piecewise linear as a function of the norm of its coefficients. Examples: Lasso path using LARS The Lars algorithm provides the full path of the coefficients along the regularization parameter almost for free, thus a common operation is to retrieve the path with one of the functions lars_path or lars_path_gram . 1.1.8.1. Mathematical formulation  The algorithm is similar to forward stepwise regression, but instead of including features at each step, the estimated coefficients are increased in a direction equiangular to each one’s correlations with the residual. Instead of giving a vector result, the LARS solution consists of a curve denoting the solution for each value of the norm of the parameter vector. The full coefficients path is stored in the array , which has size (n_features, max_features+1). The first column is always zero. References: Original Algorithm is detailed in the paper Least Angle Regression by Hastie et al. ',\n",
       " 'sklearn_1_1_linear_models\\n1.1. Linear Models\\nmodules/linear_model.html\\n 1.1.9. Orthogonal Matching Pursuit (OMP)  OrthogonalMatchingPursuit and orthogonal_mp implements the OMP algorithm for approximating the fit of a linear model with constraints imposed on the number of non-zero coefficients (ie. the pseudo-norm). Being a forward feature selection method like Least Angle Regression , orthogonal matching pursuit can approximate the optimum solution vector with a fixed number of non-zero elements: \\\\[\\\\underset{w}{\\\\operatorname{arg\\\\,min\\\\,}} ||y - Xw||_2^2 \\\\text{ subject to } ||w||_0 \\\\leq n_{\\\\text{nonzero\\\\_coefs}}\\\\] Alternatively, orthogonal matching pursuit can target a specific error instead of a specific number of non-zero coefficients. This can be expressed as: \\\\[\\\\underset{w}{\\\\operatorname{arg\\\\,min\\\\,}} ||w||_0 \\\\text{ subject to } ||y-Xw||_2^2 \\\\leq \\\\text{tol}\\\\] OMP is based on a greedy algorithm that includes at each step the atom most highly correlated with the current residual. It is similar to the simpler matching pursuit (MP) method, but better in that at each iteration, the residual is recomputed using an orthogonal projection on the space of the previously chosen dictionary elements. Examples: Orthogonal Matching Pursuit References: https://www.cs.technion.ac.il/~ronrubin/Publications/KSVD-OMP-v2.pdf Matching pursuits with time-frequency dictionaries , S. G. Mallat, Z. Zhang, ',\n",
       " 'sklearn_1_2_linear_and_quadratic_discriminant_analysis\\n1.2. Linear and Quadratic Discriminant Analysis\\nmodules/lda_qda.html\\n 1.2.1. Dimensionality reduction using Linear Discriminant Analysis  LinearDiscriminantAnalysis can be used to perform supervised dimensionality reduction, by projecting the input data to a linear subspace consisting of the directions which maximize the separation between classes (in a precise sense discussed in the mathematics section below). The dimension of the output is necessarily less than the number of classes, so this is in general a rather strong dimensionality reduction, and only makes sense in a multiclass setting. This is implemented in the method. The desired dimensionality can be set using the parameter. This parameter has no influence on the and methods. Examples: Comparison of LDA and PCA 2D projection of Iris dataset : Comparison of LDA and PCA for dimensionality reduction of the Iris dataset ',\n",
       " 'sklearn_1_2_linear_and_quadratic_discriminant_analysis\\n1.2. Linear and Quadratic Discriminant Analysis\\nmodules/lda_qda.html\\n 1.2.2. Mathematical formulation of the LDA and QDA classifiers  Both LDA and QDA can be derived from simple probabilistic models which model the class conditional distribution of the data for each class . Predictions can then be obtained by using Bayes’ rule, for each training sample : \\\\[P(yk | x)  \\\\frac{P(x | yk) P(yk)}{P(x)}  \\\\frac{P(x | yk) P(y  k)}{ \\\\sum_{l} P(x | yl) \\\\cdot P(yl)}\\\\] and we select the class which maximizes this posterior probability. More specifically, for linear and quadratic discriminant analysis, is modeled as a multivariate Gaussian distribution with density: \\\\[P(x | yk)  \\\\frac{1}{(2\\\\pi)^{d/2} |\\\\Sigma_k|^{1/2}}\\\\exp\\\\left(-\\\\frac{1}{2} (x-\\\\mu_k)^t \\\\Sigma_k^{-1} (x-\\\\mu_k)\\\\right)\\\\] where is the number of features. 1.2.2.1. QDA  According to the model above, the log of the posterior is: \\\\[\\\\begin{split}\\\\log P(yk | x) & \\\\log P(x | yk) + \\\\log P(y  k) + Cst \\\\\\\\ & -\\\\frac{1}{2} \\\\log |\\\\Sigma_k| -\\\\frac{1}{2} (x-\\\\mu_k)^t \\\\Sigma_k^{-1} (x-\\\\mu_k) + \\\\log P(y  k) + Cst,\\\\end{split}\\\\] where the constant term corresponds to the denominator , in addition to other constant terms from the Gaussian. The predicted class is the one that maximises this log-posterior. Note Relation with Gaussian Naive Bayes If in the QDA model one assumes that the covariance matrices are diagonal, then the inputs are assumed to be conditionally independent in each class, and the resulting classifier is equivalent to the Gaussian Naive Bayes classifier naive_bayes.GaussianNB . 1.2.2.2. LDA  LDA is a special case of QDA, where the Gaussians for each class are assumed to share the same covariance matrix: for all . This reduces the log posterior to: \\\\[\\\\log P(yk | x)  -\\\\frac{1}{2} (x-\\\\mu_k)^t \\\\Sigma^{-1} (x-\\\\mu_k) + \\\\log P(y  k) + Cst.\\\\] The term corresponds to the Mahalanobis Distance between the sample and the mean . The Mahalanobis distance tells how close is from , while also accounting for the variance of each feature. We can thus interpret LDA as assigning to the class whose mean is the closest in terms of Mahalanobis distance, while also accounting for the class prior probabilities. The log-posterior of LDA can also be written 3 as: \\\\[\\\\log P(yk | x)  \\\\omega_k^t x + \\\\omega_{k0} + Cst.\\\\] where and . These quantities correspond to the and attributes, respectively. From the above formula, it is clear that LDA has a linear decision surface. In the case of QDA, there are no assumptions on the covariance matrices of the Gaussians, leading to quadratic decision surfaces. See 1 for more details. ',\n",
       " 'sklearn_1_2_linear_and_quadratic_discriminant_analysis\\n1.2. Linear and Quadratic Discriminant Analysis\\nmodules/lda_qda.html\\n 1.2.3. Mathematical formulation of LDA dimensionality reduction  First note that the K means are vectors in , and they lie in an affine subspace of dimension at least (2 points lie on a line, 3 points lie on a plane, etc). As mentioned above, we can interpret LDA as assigning to the class whose mean is the closest in terms of Mahalanobis distance, while also accounting for the class prior probabilities. Alternatively, LDA is equivalent to first sphering the data so that the covariance matrix is the identity, and then assigning to the closest mean in terms of Euclidean distance (still accounting for the class priors). Computing Euclidean distances in this d-dimensional space is equivalent to first projecting the data points into , and computing the distances there (since the other dimensions will contribute equally to each class in terms of distance). In other words, if is closest to in the original space, it will also be the case in . This shows that, implicit in the LDA classifier, there is a dimensionality reduction by linear projection onto a dimensional space. We can reduce the dimension even more, to a chosen , by projecting onto the linear subspace which maximizes the variance of the after projection (in effect, we are doing a form of PCA for the transformed class means ). This corresponds to the parameter used in the transform method. See 1 for more details. ',\n",
       " 'sklearn_1_2_linear_and_quadratic_discriminant_analysis\\n1.2. Linear and Quadratic Discriminant Analysis\\nmodules/lda_qda.html\\n 1.2.4. Shrinkage  Shrinkage is a form of regularization used to improve the estimation of covariance matrices in situations where the number of training samples is small compared to the number of features. In this scenario, the empirical sample covariance is a poor estimator, and shrinkage helps improving the generalization performance of the classifier. Shrinkage LDA can be used by setting the parameter of the LinearDiscriminantAnalysis class to ‘auto’. This automatically determines the optimal shrinkage parameter in an analytic way following the lemma introduced by Ledoit and Wolf 2 . Note that currently shrinkage only works when setting the parameter to ‘lsqr’ or ‘eigen’. The parameter can also be manually set between 0 and 1. In particular, a value of 0 corresponds to no shrinkage (which means the empirical covariance matrix will be used) and a value of 1 corresponds to complete shrinkage (which means that the diagonal matrix of variances will be used as an estimate for the covariance matrix). Setting this parameter to a value between these two extrema will estimate a shrunk version of the covariance matrix. ',\n",
       " 'sklearn_1_2_linear_and_quadratic_discriminant_analysis\\n1.2. Linear and Quadratic Discriminant Analysis\\nmodules/lda_qda.html\\n 1.2.5. Estimation algorithms  Using LDA and QDA requires computing the log-posterior which depends on the class priors , the class means , and the covariance matrices. The ‘svd’ solver is the default solver used for LinearDiscriminantAnalysis , and it is the only available solver for QuadraticDiscriminantAnalysis . It can perform both classification and transform (for LDA). As it does not rely on the calculation of the covariance matrix, the ‘svd’ solver may be preferable in situations where the number of features is large. The ‘svd’ solver cannot be used with shrinkage. For QDA, the use of the SVD solver relies on the fact that the covariance matrix is, by definition, equal to where comes from the SVD of the (centered) matrix: . It turns out that we can compute the log-posterior above without having to explictly compute : computing and via the SVD of is enough. For LDA, two SVDs are computed: the SVD of the centered input matrix and the SVD of the class-wise mean vectors. The ‘lsqr’ solver is an efficient algorithm that only works for classification. It needs to explicitly compute the covariance matrix , and supports shrinkage. This solver computes the coefficients by solving for , thus avoiding the explicit computation of the inverse . The ‘eigen’ solver is based on the optimization of the between class scatter to within class scatter ratio. It can be used for both classification and transform, and it supports shrinkage. However, the ‘eigen’ solver needs to compute the covariance matrix, so it might not be suitable for situations with a high number of features. Examples: Normal and Shrinkage Linear Discriminant Analysis for classification : Comparison of LDA classifiers with and without shrinkage. References: 1 ( 1 , 2 ) “The Elements of Statistical Learning”, Hastie T., Tibshirani R., Friedman J., Section 4.3, p.106-119, 2008. 2 Ledoit O, Wolf M. Honey, I Shrunk the Sample Covariance Matrix. The Journal of Portfolio Management 30(4), 110-119, 2004. 3 R. O. Duda, P. E. Hart, D. G. Stork. Pattern Classification (Second Edition), section 2.6.2. ',\n",
       " 'sklearn_1_4_support_vector_machines\\n1.4. Support Vector Machines\\nmodules/svm.html\\n 1.4.1. Classification  SVC , NuSVC and LinearSVC are classes capable of performing binary and multi-class classification on a dataset. SVC and NuSVC are similar methods, but accept slightly different sets of parameters and have different mathematical formulations (see section Mathematical formulation ). On the other hand, LinearSVC is another (faster) implementation of Support Vector Classification for the case of a linear kernel. Note that LinearSVC does not accept parameter , as this is assumed to be linear. It also lacks some of the attributes of SVC and NuSVC , like . As other classifiers, SVC , NuSVC and LinearSVC take as input two arrays: an array of shape holding the training samples, and an array of class labels (strings or integers), of shape : After being fitted, the model can then be used to predict new values: SVMs decision function (detailed in the Mathematical formulation ) depends on some subset of the training data, called the support vectors. Some properties of these support vectors can be found in attributes , and : Examples: SVM: Maximum margin separating hyperplane , Non-linear SVM SVM-Anova: SVM with univariate feature selection , 1.4.1.1. Multi-class classification  SVC and NuSVC implement the “one-versus-one” approach for multi-class classification. In total, classifiers are constructed and each one trains data from two classes. To provide a consistent interface with other classifiers, the option allows to monotonically transform the results of the “one-versus-one” classifiers to a “one-vs-rest” decision function of shape . On the other hand, LinearSVC implements “one-vs-the-rest” multi-class strategy, thus training models. See Mathematical formulation for a complete description of the decision function. Note that the LinearSVC also implements an alternative multi-class strategy, the so-called multi-class SVM formulated by Crammer and Singer 16 , by using the option . In practice, one-vs-rest classification is usually preferred, since the results are mostly similar, but the runtime is significantly less. For “one-vs-rest” LinearSVC the attributes and have the shape and respectively. Each row of the coefficients corresponds to one of the “one-vs-rest” classifiers and similar for the intercepts, in the order of the “one” class. In the case of “one-vs-one” SVC and NuSVC , the layout of the attributes is a little more involved. In the case of a linear kernel, the attributes and have the shape and respectively. This is similar to the layout for LinearSVC described above, with each row now corresponding to a binary classifier. The order for classes 0 to n is “0 vs 1”, “0 vs 2” , … “0 vs n”, “1 vs 2”, “1 vs 3”, “1 vs n”, . . . “n-1 vs n”. The shape of is with a somewhat hard to grasp layout. The columns correspond to the support vectors involved in any of the “one-vs-one” classifiers. Each of the support vectors is used in classifiers. The entries in each row correspond to the dual coefficients for these classifiers. This might be clearer with an example: consider a three class problem with class 0 having three support vectors and class 1 and 2 having two support vectors and respectively. For each support vector , there are two dual coefficients. Let’s call the coefficient of support vector in the classifier between classes and . Then looks like this: Coefficients for SVs of class 0 Coefficients for SVs of class 1 Coefficients for SVs of class 2 Examples: Plot different SVM classifiers in the iris dataset , 1.4.1.2. Scores and probabilities  The method of SVC and NuSVC gives per-class scores for each sample (or a single score per sample in the binary case). When the constructor option is set to , class membership probability estimates (from the methods and ) are enabled. In the binary case, the probabilities are calibrated using Platt scaling 9 : logistic regression on the SVM’s scores, fit by an additional cross-validation on the training data. In the multiclass case, this is extended as per 10 . Note The same probability calibration procedure is available for all estimators via the CalibratedClassifierCV (see Probability calibration ). In the case of SVC and NuSVC , this procedure is builtin in libsvm which is used under the hood, so it does not rely on scikit-learn’s CalibratedClassifierCV . The cross-validation involved in Platt scaling is an expensive operation for large datasets. In addition, the probability estimates may be inconsistent with the scores: the “argmax” of the scores may not be the argmax of the probabilities in binary classification, a sample may be labeled by as belonging to the positive class even if the output of is less than 0.5; and similarly, it could be labeled as negative even if the output of is more than 0.5. Platt’s method is also known to have theoretical issues. If confidence scores are required, but these do not have to be probabilities, then it is advisable to set and use instead of . Please note that when and , unlike , the method does not try to break ties by default. You can set for the output of to be the same as , otherwise the first class among the tied classes will always be returned; but have in mind that it comes with a computational cost. See SVM Tie Breaking Example for an example on tie breaking. 1.4.1.3. Unbalanced problems  In problems where it is desired to give more importance to certain classes or certain individual samples, the parameters and can be used. SVC (but not NuSVC ) implements the parameter in the method. It’s a dictionary of the form , where value is a floating point number > 0 that sets the parameter of class to . The figure below illustrates the decision boundary of an unbalanced problem, with and without weight correction. SVC , NuSVC , SVR , NuSVR , LinearSVC , LinearSVR and OneClassSVM implement also weights for individual samples in the method through the parameter. Similar to , this sets the parameter for the i-th example to , which will encourage the classifier to get these samples right. The figure below illustrates the effect of sample weighting on the decision boundary. The size of the circles is proportional to the sample weights: Examples: SVM: Separating hyperplane for unbalanced classes SVM: Weighted samples , ',\n",
       " 'sklearn_1_4_support_vector_machines\\n1.4. Support Vector Machines\\nmodules/svm.html\\n 1.4.2. Regression  The method of Support Vector Classification can be extended to solve regression problems. This method is called Support Vector Regression. The model produced by support vector classification (as described above) depends only on a subset of the training data, because the cost function for building the model does not care about training points that lie beyond the margin. Analogously, the model produced by Support Vector Regression depends only on a subset of the training data, because the cost function ignores samples whose prediction is close to their target. There are three different implementations of Support Vector Regression: SVR , NuSVR and LinearSVR . LinearSVR provides a faster implementation than SVR but only considers the linear kernel, while NuSVR implements a slightly different formulation than SVR and LinearSVR . See Implementation details for further details. As with classification classes, the fit method will take as argument vectors X, y, only that in this case y is expected to have floating point values instead of integer values: Examples: Support Vector Regression (SVR) using linear and non-linear kernels ',\n",
       " 'sklearn_1_4_support_vector_machines\\n1.4. Support Vector Machines\\nmodules/svm.html\\n 1.4.3. Density estimation, novelty detection  The class OneClassSVM implements a One-Class SVM which is used in outlier detection. See Novelty and Outlier Detection for the description and usage of OneClassSVM. ',\n",
       " 'sklearn_1_4_support_vector_machines\\n1.4. Support Vector Machines\\nmodules/svm.html\\n 1.4.4. Complexity  Support Vector Machines are powerful tools, but their compute and storage requirements increase rapidly with the number of training vectors. The core of an SVM is a quadratic programming problem (QP), separating support vectors from the rest of the training data. The QP solver used by the libsvm -based implementation scales between and depending on how efficiently the libsvm cache is used in practice (dataset dependent). If the data is very sparse should be replaced by the average number of non-zero features in a sample vector. For the linear case, the algorithm used in LinearSVC by the liblinear implementation is much more efficient than its libsvm -based SVC counterpart and can scale almost linearly to millions of samples and/or features. ',\n",
       " 'sklearn_1_4_support_vector_machines\\n1.4. Support Vector Machines\\nmodules/svm.html\\n 1.4.5. Tips on Practical Use  Avoiding data copy : For SVC , SVR , NuSVC and NuSVR , if the data passed to certain methods is not C-ordered contiguous and double precision, it will be copied before calling the underlying C implementation. You can check whether a given numpy array is C-contiguous by inspecting its attribute. For LinearSVC (and LogisticRegression ) any input passed as a numpy array will be copied and converted to the liblinear internal sparse data representation (double precision floats and int32 indices of non-zero components). If you want to fit a large-scale linear classifier without copying a dense numpy C-contiguous double precision array as input, we suggest to use the SGDClassifier class instead. The objective function can be configured to be almost the same as the LinearSVC model. Kernel cache size : For SVC , SVR , NuSVC and NuSVR , the size of the kernel cache has a strong impact on run times for larger problems. If you have enough RAM available, it is recommended to set to a higher value than the default of 200(MB), such as 500(MB) or 1000(MB). Setting C : is by default and it’s a reasonable default choice. If you have a lot of noisy observations you should decrease it: decreasing C corresponds to more regularization. LinearSVC and LinearSVR are less sensitive to when it becomes large, and prediction results stop improving after a certain threshold. Meanwhile, larger values will take more time to train, sometimes up to 10 times longer, as shown in 11 . Support Vector Machine algorithms are not scale invariant, so it is highly recommended to scale your data . For example, scale each attribute on the input vector X to [0,1] or [-1,+1], or standardize it to have mean 0 and variance 1. Note that the same scaling must be applied to the test vector to obtain meaningful results. This can be done easily by using a Pipeline : See section Preprocessing data for more details on scaling and normalization. Regarding the parameter, quoting 12 : We found that if the number of iterations is large, then shrinking can shorten the training time. However, if we loosely solve the optimization problem (e.g., by using a large stopping tolerance), the code without using shrinking may be much faster Parameter in NuSVC / OneClassSVM / NuSVR approximates the fraction of training errors and support vectors. In SVC , if the data is unbalanced (e.g. many positive and few negative), set and/or try different penalty parameters . Randomness of the underlying implementations : The underlying implementations of SVC and NuSVC use a random number generator only to shuffle the data for probability estimation (when is set to ). This randomness can be controlled with the parameter. If is set to these estimators are not random and has no effect on the results. The underlying OneClassSVM implementation is similar to the ones of SVC and NuSVC . As no probability estimation is provided for OneClassSVM , it is not random. The underlying LinearSVC implementation uses a random number generator to select features when fitting the model with a dual coordinate descent (i.e when is set to ). It is thus not uncommon to have slightly different results for the same input data. If that happens, try with a smaller parameter. This randomness can also be controlled with the parameter. When is set to the underlying implementation of LinearSVC is not random and has no effect on the results. Using L1 penalization as provided by yields a sparse solution, i.e. only a subset of feature weights is different from zero and contribute to the decision function. Increasing yields a more complex model (more features are selected). The value that yields a “null” model (all weights equal to zero) can be calculated using l1_min_c . ',\n",
       " 'sklearn_1_4_support_vector_machines\\n1.4. Support Vector Machines\\nmodules/svm.html\\n 1.4.6. Kernel functions  The kernel function can be any of the following: linear: . polynomial: , where is specified by parameter , by . rbf: , where is specified by parameter , must be greater than 0. sigmoid , where is specified by . Different kernels are specified by the parameter: 1.4.6.1. Parameters of the RBF Kernel  When training an SVM with the Radial Basis Function (RBF) kernel, two parameters must be considered: and . The parameter , common to all SVM kernels, trades off misclassification of training examples against simplicity of the decision surface. A low makes the decision surface smooth, while a high aims at classifying all training examples correctly. defines how much influence a single training example has. The larger is, the closer other examples must be to be affected. Proper choice of and is critical to the SVM’s performance. One is advised to use sklearn.model_selection.GridSearchCV with and spaced exponentially far apart to choose good values. Examples: RBF SVM parameters Non-linear SVM 1.4.6.2. Custom Kernels  You can define your own kernels by either giving the kernel as a python function or by precomputing the Gram matrix. Classifiers with custom kernels behave the same way as any other classifiers, except that: Field is now empty, only indices of support vectors are stored in A reference (and not a copy) of the first argument in the method is stored for future reference. If that array changes between the use of and you will have unexpected results. 1.4.6.2.1. Using Python functions as kernels  You can use your own defined kernels by passing a function to the parameter. Your kernel must take as arguments two matrices of shape , and return a kernel matrix of shape . The following code defines a linear kernel and creates a classifier instance that will use that kernel: Examples: SVM with custom kernel . 1.4.6.2.2. Using the Gram matrix  You can pass pre-computed kernels by using the option. You should then pass Gram matrix instead of X to the and methods. The kernel values between all training vectors and the test vectors must be provided: ',\n",
       " 'sklearn_1_4_support_vector_machines\\n1.4. Support Vector Machines\\nmodules/svm.html\\n 1.4.7. Mathematical formulation  A support vector machine constructs a hyper-plane or set of hyper-planes in a high or infinite dimensional space, which can be used for classification, regression or other tasks. Intuitively, a good separation is achieved by the hyper-plane that has the largest distance to the nearest training data points of any class (so-called functional margin), since in general the larger the margin the lower the generalization error of the classifier. The figure below shows the decision function for a linearly separable problem, with three samples on the margin boundaries, called “support vectors”: In general, when the problem isn’t linearly separable, the support vectors are the samples within the margin boundaries. We recommend 13 and 14 as good references for the theory and practicalities of SVMs. 1.4.7.1. SVC  Given training vectors , i1,…, n, in two classes, and a vector , our goal is to find and such that the prediction given by is correct for most samples. SVC solves the following primal problem: \\\\[ \\\\begin{align}\\\\begin{aligned}\\\\min_ {w, b, \\\\zeta} \\\\frac{1}{2} w^T w + C \\\\sum_{i1}^{n} \\\\zeta_i\\\\\\\\\\\\begin{split}\\\\textrm {subject to } & y_i (w^T \\\\phi (x_i) + b) \\\\geq 1 - \\\\zeta_i,\\\\\\\\ & \\\\zeta_i \\\\geq 0, i1, ..., n\\\\end{split}\\\\end{aligned}\\\\end{align} \\\\] Intuitively, we’re trying to maximize the margin (by minimizing ), while incurring a penalty when a sample is misclassified or within the margin boundary. Ideally, the value would be for all samples, which indicates a perfect prediction. But problems are usually not always perfectly separable with a hyperplane, so we allow some samples to be at a distance from their correct margin boundary. The penalty term controls the strengh of this penalty, and as a result, acts as an inverse regularization parameter (see note below). The dual problem to the primal is \\\\[ \\\\begin{align}\\\\begin{aligned}\\\\min_{\\\\alpha} \\\\frac{1}{2} \\\\alpha^T Q \\\\alpha - e^T \\\\alpha\\\\\\\\\\\\begin{split} \\\\textrm {subject to } & y^T \\\\alpha  0\\\\\\\\ & 0 \\\\leq \\\\alpha_i \\\\leq C, i1, ..., n\\\\end{split}\\\\end{aligned}\\\\end{align} \\\\] where is the vector of all ones, and is an by positive semidefinite matrix, , where is the kernel. The terms are called the dual coefficients, and they are upper-bounded by . This dual representation highlights the fact that training vectors are implicitly mapped into a higher (maybe infinite) dimensional space by the function : see kernel trick . Once the optimization problem is solved, the output of decision_function for a given sample becomes: \\\\[\\\\sum_{i\\\\in SV} y_i \\\\alpha_i K(x_i, x) + b,\\\\] and the predicted class correspond to its sign. We only need to sum over the support vectors (i.e. the samples that lie within the margin) because the dual coefficients are zero for the other samples. These parameters can be accessed through the attributes which holds the product , which holds the support vectors, and which holds the independent term Note While SVM models derived from libsvm and liblinear use as regularization parameter, most other estimators use . The exact equivalence between the amount of regularization of two models depends on the exact objective function optimized by the model. For example, when the estimator used is sklearn.linear_model.Ridge regression, the relation between them is given as . 1.4.7.2. LinearSVC  The primal problem can be equivalently formulated as \\\\[\\\\min_ {w, b} \\\\frac{1}{2} w^T w + C \\\\sum_{i1}\\\\max(0, y_i (w^T \\\\phi(x_i) + b)),\\\\] where we make use of the hinge loss . This is the form that is directly optimized by LinearSVC , but unlike the dual form, this one does not involve inner products between samples, so the famous kernel trick cannot be applied. This is why only the linear kernel is supported by LinearSVC ( is the identity function). 1.4.7.3. NuSVC  The -SVC formulation 15 is a reparameterization of the -SVC and therefore mathematically equivalent. We introduce a new parameter (instead of ) which controls the number of support vectors and margin errors : is an upper bound on the fraction of margin errors and a lower bound of the fraction of support vectors. A margin error corresponds to a sample that lies on the wrong side of its margin boundary: it is either misclassified, or it is correctly classified but does not lie beyond the margin. 1.4.7.4. SVR  Given training vectors , i1,…, n, and a vector -SVR solves the following primal problem: \\\\[ \\\\begin{align}\\\\begin{aligned}\\\\min_ {w, b, \\\\zeta, \\\\zeta^*} \\\\frac{1}{2} w^T w + C \\\\sum_{i1}^{n} (\\\\zeta_i + \\\\zeta_i^*)\\\\\\\\\\\\begin{split}\\\\textrm {subject to } & y_i - w^T \\\\phi (x_i) - b \\\\leq \\\\varepsilon + \\\\zeta_i,\\\\\\\\ & w^T \\\\phi (x_i) + b - y_i \\\\leq \\\\varepsilon + \\\\zeta_i^*,\\\\\\\\ & \\\\zeta_i, \\\\zeta_i^* \\\\geq 0, i1, ..., n\\\\end{split}\\\\end{aligned}\\\\end{align} \\\\] Here, we are penalizing samples whose prediction is at least away from their true target. These samples penalize the objective by or , depending on whether their predictions lie above or below the tube. The dual problem is \\\\[ \\\\begin{align}\\\\begin{aligned}\\\\min_{\\\\alpha, \\\\alpha^*} \\\\frac{1}{2} (\\\\alpha - \\\\alpha^*)^T Q (\\\\alpha - \\\\alpha^*) + \\\\varepsilon e^T (\\\\alpha + \\\\alpha^*) - y^T (\\\\alpha - \\\\alpha^*)\\\\\\\\\\\\begin{split} \\\\textrm {subject to } & e^T (\\\\alpha - \\\\alpha^*)  0\\\\\\\\ & 0 \\\\leq \\\\alpha_i, \\\\alpha_i^* \\\\leq C, i1, ..., n\\\\end{split}\\\\end{aligned}\\\\end{align} \\\\] where is the vector of all ones, is an by positive semidefinite matrix, is the kernel. Here training vectors are implicitly mapped into a higher (maybe infinite) dimensional space by the function . The prediction is: \\\\[\\\\sum_{i \\\\in SV}(\\\\alpha_i - \\\\alpha_i^*) K(x_i, x) + b\\\\] These parameters can be accessed through the attributes which holds the difference , which holds the support vectors, and which holds the independent term 1.4.7.5. LinearSVR  The primal problem can be equivalently formulated as \\\\[\\\\min_ {w, b} \\\\frac{1}{2} w^T w + C \\\\sum_{i1}\\\\max(0, |y_i - (w^T \\\\phi(x_i) + b)| - \\\\varepsilon),\\\\] where we make use of the epsilon-insensitive loss, i.e. errors of less than are ignored. This is the form that is directly optimized by LinearSVR . ',\n",
       " 'sklearn_1_4_support_vector_machines\\n1.4. Support Vector Machines\\nmodules/svm.html\\n 1.4.8. Implementation details  Internally, we use libsvm 12 and liblinear 11 to handle all computations. These libraries are wrapped using C and Cython. For a description of the implementation and details of the algorithms used, please refer to their respective papers. References: 9 Platt “Probabilistic outputs for SVMs and comparisons to regularized likelihood methods” . 10 Wu, Lin and Weng, “Probability estimates for multi-class classification by pairwise coupling” , JMLR 5:975-1005, 2004. 11 ( 1 , 2 ) Fan, Rong-En, et al., “LIBLINEAR: A library for large linear classification.” , Journal of machine learning research 9.Aug (2008): 1871-1874. 12 ( 1 , 2 ) Chang and Lin, LIBSVM: A Library for Support Vector Machines . 13 Bishop, Pattern recognition and machine learning , chapter 7 Sparse Kernel Machines 14 “A Tutorial on Support Vector Regression” , Alex J. Smola, Bernhard Schölkopf - Statistics and Computing archive Volume 14 Issue 3, August 2004, p. 199-222. 15 Schölkopf et. al New Support Vector Algorithms 16 Crammer and Singer On the Algorithmic Implementation ofMulticlass Kernel-based Vector Machines , JMLR 2001. ',\n",
       " 'sklearn_1_5_stochastic_gradient_descent\\n1.5. Stochastic Gradient Descent\\nmodules/sgd.html\\n 1.5.1. Classification  The class SGDClassifier implements a plain stochastic gradient descent learning routine which supports different loss functions and penalties for classification. Below is the decision boundary of a SGDClassifier trained with the hinge loss, equivalent to a linear SVM. As other classifiers, SGD has to be fitted with two arrays: an array of shape (n_samples, n_features) holding the training samples, and an array y of shape (n_samples,) holding the target values (class labels) for the training samples: After being fitted, the model can then be used to predict new values: SGD fits a linear model to the training data. The attribute holds the model parameters: The attribute holds the intercept (aka offset or bias): Whether or not the model should use an intercept, i.e. a biased hyperplane, is controlled by the parameter . The signed distance to the hyperplane (computed as the dot product between the coefficients and the input sample, plus the intercept) is given by SGDClassifier.decision_function : The concrete loss function can be set via the parameter. SGDClassifier supports the following loss functions: : (soft-margin) linear Support Vector Machine, : smoothed hinge loss, : logistic regression, and all regression losses below. In this case the target is encoded as -1 or 1, and the problem is treated as a regression problem. The predicted class then correspond to the sign of the predicted target. Please refer to the mathematical section below for formulas. The first two loss functions are lazy, they only update the model parameters if an example violates the margin constraint, which makes training very efficient and may result in sparser models (i.e. with more zero coefficents), even when L2 penalty is used. Using or enables the method, which gives a vector of probability estimates per sample : The concrete penalty can be set via the parameter. SGD supports the following penalties: : L2 norm penalty on . : L1 norm penalty on . : Convex combination of L2 and L1; . The default setting is . The L1 penalty leads to sparse solutions, driving most coefficients to zero. The Elastic Net 11 solves some deficiencies of the L1 penalty in the presence of highly correlated attributes. The parameter controls the convex combination of L1 and L2 penalty. SGDClassifier supports multi-class classification by combining multiple binary classifiers in a “one versus all” (OVA) scheme. For each of the classes, a binary classifier is learned that discriminates between that and all other classes. At testing time, we compute the confidence score (i.e. the signed distances to the hyperplane) for each classifier and choose the class with the highest confidence. The Figure below illustrates the OVA approach on the iris dataset. The dashed lines represent the three OVA classifiers; the background colors show the decision surface induced by the three classifiers. In the case of multi-class classification is a two-dimensional array of shape (n_classes, n_features) and is a one-dimensional array of shape (n_classes,). The i-th row of holds the weight vector of the OVA classifier for the i-th class; classes are indexed in ascending order (see attribute ). Note that, in principle, since they allow to create a probability model, and are more suitable for one-vs-all classification. SGDClassifier supports both weighted classes and weighted instances via the fit parameters and . See the examples below and the docstring of SGDClassifier.fit for further information. SGDClassifier supports averaged SGD (ASGD) 10 . Averaging can be enabled by setting . ASGD performs the same updates as the regular SGD (see Mathematical formulation ), but instead of using the last value of the coefficients as the attribute (i.e. the values of the last update), is set instead to the average value of the coefficients across all updates. The same is done for the attribute. When using ASGD the learning rate can be larger and even constant, leading on some datasets to a speed up in training time. For classification with a logistic loss, another variant of SGD with an averaging strategy is available with Stochastic Average Gradient (SAG) algorithm, available as a solver in LogisticRegression . Examples: SGD: Maximum margin separating hyperplane , Plot multi-class SGD on the iris dataset SGD: Weighted samples Comparing various online solvers SVM: Separating hyperplane for unbalanced classes (See the Note in the example) ',\n",
       " 'sklearn_1_5_stochastic_gradient_descent\\n1.5. Stochastic Gradient Descent\\nmodules/sgd.html\\n 1.5.2. Regression  The class SGDRegressor implements a plain stochastic gradient descent learning routine which supports different loss functions and penalties to fit linear regression models. SGDRegressor is well suited for regression problems with a large number of training samples (> 10.000), for other problems we recommend Ridge , Lasso , or ElasticNet . The concrete loss function can be set via the parameter. SGDRegressor supports the following loss functions: : Ordinary least squares, : Huber loss for robust regression, : linear Support Vector Regression. Please refer to the mathematical section below for formulas. The Huber and epsilon-insensitive loss functions can be used for robust regression. The width of the insensitive region has to be specified via the parameter . This parameter depends on the scale of the target variables. The parameter determines the regularization to be used (see description above in the classification section). SGDRegressor also supports averaged SGD 10 (here again, see description above in the classification section). For regression with a squared loss and a l2 penalty, another variant of SGD with an averaging strategy is available with Stochastic Average Gradient (SAG) algorithm, available as a solver in Ridge . ',\n",
       " 'sklearn_1_5_stochastic_gradient_descent\\n1.5. Stochastic Gradient Descent\\nmodules/sgd.html\\n 1.5.3. Stochastic Gradient Descent for sparse data  Note The sparse implementation produces slightly different results from the dense implementation, due to a shrunk learning rate for the intercept. See Implementation details . There is built-in support for sparse data given in any matrix in a format supported by scipy.sparse . For maximum efficiency, however, use the CSR matrix format as defined in scipy.sparse.csr_matrix . Examples: Classification of text documents using sparse features ',\n",
       " 'sklearn_1_5_stochastic_gradient_descent\\n1.5. Stochastic Gradient Descent\\nmodules/sgd.html\\n 1.5.4. Complexity  The major advantage of SGD is its efficiency, which is basically linear in the number of training examples. If X is a matrix of size (n, p) training has a cost of , where k is the number of iterations (epochs) and is the average number of non-zero attributes per sample. Recent theoretical results, however, show that the runtime to get some desired optimization accuracy does not increase as the training set size increases. ',\n",
       " 'sklearn_1_5_stochastic_gradient_descent\\n1.5. Stochastic Gradient Descent\\nmodules/sgd.html\\n 1.5.5. Stopping criterion  The classes SGDClassifier and SGDRegressor provide two criteria to stop the algorithm when a given level of convergence is reached: With , the input data is split into a training set and a validation set. The model is then fitted on the training set, and the stopping criterion is based on the prediction score (using the method) computed on the validation set. The size of the validation set can be changed with the parameter . With , the model is fitted on the entire input data and the stopping criterion is based on the objective function computed on the training data. In both cases, the criterion is evaluated once by epoch, and the algorithm stops when the criterion does not improve times in a row. The improvement is evaluated with absolute tolerance , and the algorithm stops in any case after a maximum number of iteration . ',\n",
       " 'sklearn_1_5_stochastic_gradient_descent\\n1.5. Stochastic Gradient Descent\\nmodules/sgd.html\\n 1.5.6. Tips on Practical Use  Stochastic Gradient Descent is sensitive to feature scaling, so it is highly recommended to scale your data. For example, scale each attribute on the input vector X to [0,1] or [-1,+1], or standardize it to have mean 0 and variance 1. Note that the same scaling must be applied to the test vector to obtain meaningful results. This can be easily done using StandardScaler : If your attributes have an intrinsic scale (e.g. word frequencies or indicator features) scaling is not needed. Finding a reasonable regularization term is best done using automatic hyper-parameter search, e.g. GridSearchCV or RandomizedSearchCV , usually in the range . Empirically, we found that SGD converges after observing approximately 10^6 training samples. Thus, a reasonable first guess for the number of iterations is , where is the size of the training set. If you apply SGD to features extracted using PCA we found that it is often wise to scale the feature values by some constant such that the average L2 norm of the training data equals one. We found that Averaged SGD works best with a larger number of features and a higher eta0 References: “Efficient BackProp” Y. LeCun, L. Bottou, G. Orr, K. Müller - In Neural Networks: Tricks of the Trade 1998. ',\n",
       " 'sklearn_1_5_stochastic_gradient_descent\\n1.5. Stochastic Gradient Descent\\nmodules/sgd.html\\n 1.5.7. Mathematical formulation  We describe here the mathematical details of the SGD procedure. A good overview with convergence rates can be found in 12 . Given a set of training examples where and ( for classification), our goal is to learn a linear scoring function with model parameters and intercept . In order to make predictions for binary classification, we simply look at the sign of . To find the model parameters, we minimize the regularized training error given by \\\\[E(w,b)  \\\\frac{1}{n}\\\\sum_{i1}^{n} L(y_i, f(x_i)) + \\\\alpha R(w)\\\\] where is a loss function that measures model (mis)fit and is a regularization term (aka penalty) that penalizes model complexity; is a non-negative hyperparameter that controls the regularization stength. Different choices for entail different classifiers or regressors: Hinge (soft-margin): equivalent to Support Vector Classification. . Perceptron: . Modified Huber: if , and otherwise. Log: equivalent to Logistic Regression. . Least-Squares: Linear regression (Ridge or Lasso depending on ). . Huber: less sensitive to outliers than least-squares. It is equivalent to least squares when , and otherwise. Epsilon-Insensitive: (soft-margin) equivalent to Support Vector Regression. . All of the above loss functions can be regarded as an upper bound on the misclassification error (Zero-one loss) as shown in the Figure below. Popular choices for the regularization term (the parameter) include: L2 norm: , L1 norm: , which leads to sparse solutions. Elastic Net: , a convex combination of L2 and L1, where is given by . The Figure below shows the contours of the different regularization terms in a 2-dimensional parameter space ( ) when . 1.5.7.1. SGD  Stochastic gradient descent is an optimization method for unconstrained optimization problems. In contrast to (batch) gradient descent, SGD approximates the true gradient of by considering a single training example at a time. The class SGDClassifier implements a first-order SGD learning routine. The algorithm iterates over the training examples and for each example updates the model parameters according to the update rule given by \\\\[w \\\\leftarrow w - \\\\eta \\\\left[\\\\alpha \\\\frac{\\\\partial R(w)}{\\\\partial w} + \\\\frac{\\\\partial L(w^T x_i + b, y_i)}{\\\\partial w}\\\\right]\\\\] where is the learning rate which controls the step-size in the parameter space. The intercept is updated similarly but without regularization (and with additional decay for sparse matrices, as detailed in Implementation details ). The learning rate can be either constant or gradually decaying. For classification, the default learning rate schedule ( ) is given by \\\\[\\\\eta^{(t)}  \\\\frac {1}{\\\\alpha (t_0 + t)}\\\\] where is the time step (there are a total of time steps), is determined based on a heuristic proposed by Léon Bottou such that the expected initial updates are comparable with the expected size of the weights (this assuming that the norm of the training samples is approx. 1). The exact definition can be found in in BaseSGD . For regression the default learning rate schedule is inverse scaling ( ), given by \\\\[\\\\eta^{(t)}  \\\\frac{eta_0}{t^{power\\\\_t}}\\\\] where and are hyperparameters chosen by the user via and , resp. For a constant learning rate use and use to specify the learning rate. For an adaptively decreasing learning rate, use and use to specify the starting learning rate. When the stopping criterion is reached, the learning rate is divided by 5, and the algorithm does not stop. The algorithm stops when the learning rate goes below 1e-6. The model parameters can be accessed through the and attributes: holds the weights and holds . When using Averaged SGD (with the parameter), is set to the average weight across all updates: , where is the total number of updates, found in the attribute. ',\n",
       " 'sklearn_1_5_stochastic_gradient_descent\\n1.5. Stochastic Gradient Descent\\nmodules/sgd.html\\n 1.5.8. Implementation details  The implementation of SGD is influenced by the of 7 . Similar to SvmSGD, the weight vector is represented as the product of a scalar and a vector which allows an efficient weight update in the case of L2 regularization. In the case of sparse input , the intercept is updated with a smaller learning rate (multiplied by 0.01) to account for the fact that it is updated more frequently. Training examples are picked up sequentially and the learning rate is lowered after each observed example. We adopted the learning rate schedule from 8 . For multi-class classification, a “one versus all” approach is used. We use the truncated gradient algorithm proposed in 9 for L1 regularization (and the Elastic Net). The code is written in Cython. References: 7 “Stochastic Gradient Descent” L. Bottou - Website, 2010. 8 “Pegasos: Primal estimated sub-gradient solver for svm” S. Shalev-Shwartz, Y. Singer, N. Srebro - In Proceedings of ICML ‘07. 9 “Stochastic gradient descent training for l1-regularized log-linear models with cumulative penalty” Y. Tsuruoka, J. Tsujii, S. Ananiadou - In Proceedings of the AFNLP/ACL ‘09. 10 ( 1 , 2 ) “Towards Optimal One Pass Large Scale Learning with Averaged Stochastic Gradient Descent” Xu, Wei 11 “Regularization and variable selection via the elastic net” H. Zou, T. Hastie - Journal of the Royal Statistical Society Series B, 67 (2), 301-320. 12 “Solving large scale linear prediction problems using stochastic gradient descent algorithms” T. Zhang - In Proceedings of ICML ‘04. ',\n",
       " 'sklearn_1_6_nearest_neighbors\\n1.6. Nearest Neighbors\\nmodules/neighbors.html\\n 1.6.1. Unsupervised Nearest Neighbors  NearestNeighbors implements unsupervised nearest neighbors learning. It acts as a uniform interface to three different nearest neighbors algorithms: BallTree , KDTree , and a brute-force algorithm based on routines in sklearn.metrics.pairwise . The choice of neighbors search algorithm is controlled through the keyword , which must be one of . When the default value is passed, the algorithm attempts to determine the best approach from the training data. For a discussion of the strengths and weaknesses of each option, see Nearest Neighbor Algorithms . Warning Regarding the Nearest Neighbors algorithms, if two neighbors and have identical distances but different labels, the result will depend on the ordering of the training data. 1.6.1.1. Finding the Nearest Neighbors  For the simple task of finding the nearest neighbors between two sets of data, the unsupervised algorithms within sklearn.neighbors can be used: Because the query set matches the training set, the nearest neighbor of each point is the point itself, at a distance of zero. It is also possible to efficiently produce a sparse graph showing the connections between neighboring points: The dataset is structured such that points nearby in index order are nearby in parameter space, leading to an approximately block-diagonal matrix of K-nearest neighbors. Such a sparse graph is useful in a variety of circumstances which make use of spatial relationships between points for unsupervised learning: in particular, see sklearn.manifold.Isomap , sklearn.manifold.LocallyLinearEmbedding , and sklearn.cluster.SpectralClustering . 1.6.1.2. KDTree and BallTree Classes  Alternatively, one can use the KDTree or BallTree classes directly to find nearest neighbors. This is the functionality wrapped by the NearestNeighbors class used above. The Ball Tree and KD Tree have the same interface; we’ll show an example of using the KD Tree here: Refer to the KDTree and BallTree class documentation for more information on the options available for nearest neighbors searches, including specification of query strategies, distance metrics, etc. For a list of available metrics, see the documentation of the DistanceMetric class. ',\n",
       " 'sklearn_1_6_nearest_neighbors\\n1.6. Nearest Neighbors\\nmodules/neighbors.html\\n 1.6.2. Nearest Neighbors Classification  Neighbors-based classification is a type of instance-based learning or non-generalizing learning : it does not attempt to construct a general internal model, but simply stores instances of the training data. Classification is computed from a simple majority vote of the nearest neighbors of each point: a query point is assigned the data class which has the most representatives within the nearest neighbors of the point. scikit-learn implements two different nearest neighbors classifiers: KNeighborsClassifier implements learning based on the nearest neighbors of each query point, where is an integer value specified by the user. RadiusNeighborsClassifier implements learning based on the number of neighbors within a fixed radius of each training point, where is a floating-point value specified by the user. The -neighbors classification in KNeighborsClassifier is the most commonly used technique. The optimal choice of the value is highly data-dependent: in general a larger suppresses the effects of noise, but makes the classification boundaries less distinct. In cases where the data is not uniformly sampled, radius-based neighbors classification in RadiusNeighborsClassifier can be a better choice. The user specifies a fixed radius , such that points in sparser neighborhoods use fewer nearest neighbors for the classification. For high-dimensional parameter spaces, this method becomes less effective due to the so-called “curse of dimensionality”. The basic nearest neighbors classification uses uniform weights: that is, the value assigned to a query point is computed from a simple majority vote of the nearest neighbors. Under some circumstances, it is better to weight the neighbors such that nearer neighbors contribute more to the fit. This can be accomplished through the keyword. The default value, , assigns uniform weights to each neighbor. assigns weights proportional to the inverse of the distance from the query point. Alternatively, a user-defined function of the distance can be supplied to compute the weights. Examples: Nearest Neighbors Classification : an example of classification using nearest neighbors. ',\n",
       " 'sklearn_1_6_nearest_neighbors\\n1.6. Nearest Neighbors\\nmodules/neighbors.html\\n 1.6.3. Nearest Neighbors Regression  Neighbors-based regression can be used in cases where the data labels are continuous rather than discrete variables. The label assigned to a query point is computed based on the mean of the labels of its nearest neighbors. scikit-learn implements two different neighbors regressors: KNeighborsRegressor implements learning based on the nearest neighbors of each query point, where is an integer value specified by the user. RadiusNeighborsRegressor implements learning based on the neighbors within a fixed radius of the query point, where is a floating-point value specified by the user. The basic nearest neighbors regression uses uniform weights: that is, each point in the local neighborhood contributes uniformly to the classification of a query point. Under some circumstances, it can be advantageous to weight points such that nearby points contribute more to the regression than faraway points. This can be accomplished through the keyword. The default value, , assigns equal weights to all points. assigns weights proportional to the inverse of the distance from the query point. Alternatively, a user-defined function of the distance can be supplied, which will be used to compute the weights. The use of multi-output nearest neighbors for regression is demonstrated in Face completion with a multi-output estimators . In this example, the inputs X are the pixels of the upper half of faces and the outputs Y are the pixels of the lower half of those faces. Examples: Nearest Neighbors regression : an example of regression using nearest neighbors. Face completion with a multi-output estimators : an example of multi-output regression using nearest neighbors. ',\n",
       " 'sklearn_1_6_nearest_neighbors\\n1.6. Nearest Neighbors\\nmodules/neighbors.html\\n 1.6.4. Nearest Neighbor Algorithms  1.6.4.1. Brute Force  Fast computation of nearest neighbors is an active area of research in machine learning. The most naive neighbor search implementation involves the brute-force computation of distances between all pairs of points in the dataset: for samples in dimensions, this approach scales as . Efficient brute-force neighbors searches can be very competitive for small data samples. However, as the number of samples grows, the brute-force approach quickly becomes infeasible. In the classes within sklearn.neighbors , brute-force neighbors searches are specified using the keyword , and are computed using the routines available in sklearn.metrics.pairwise . 1.6.4.2. K-D Tree  To address the computational inefficiencies of the brute-force approach, a variety of tree-based data structures have been invented. In general, these structures attempt to reduce the required number of distance calculations by efficiently encoding aggregate distance information for the sample. The basic idea is that if point is very distant from point , and point is very close to point , then we know that points and are very distant, without having to explicitly calculate their distance . In this way, the computational cost of a nearest neighbors search can be reduced to or better. This is a significant improvement over brute-force for large . An early approach to taking advantage of this aggregate information was the KD tree data structure (short for K-dimensional tree ), which generalizes two-dimensional Quad-trees and 3-dimensional Oct-trees to an arbitrary number of dimensions. The KD tree is a binary tree structure which recursively partitions the parameter space along the data axes, dividing it into nested orthotropic regions into which data points are filed. The construction of a KD tree is very fast: because partitioning is performed only along the data axes, no -dimensional distances need to be computed. Once constructed, the nearest neighbor of a query point can be determined with only distance computations. Though the KD tree approach is very fast for low-dimensional ( ) neighbors searches, it becomes inefficient as grows very large: this is one manifestation of the so-called “curse of dimensionality”. In scikit-learn, KD tree neighbors searches are specified using the keyword , and are computed using the class KDTree . References: “Multidimensional binary search trees used for associative searching” , Bentley, J.L., Communications of the ACM (1975) 1.6.4.3. Ball Tree  To address the inefficiencies of KD Trees in higher dimensions, the ball tree data structure was developed. Where KD trees partition data along Cartesian axes, ball trees partition data in a series of nesting hyper-spheres. This makes tree construction more costly than that of the KD tree, but results in a data structure which can be very efficient on highly structured data, even in very high dimensions. A ball tree recursively divides the data into nodes defined by a centroid and radius , such that each point in the node lies within the hyper-sphere defined by and . The number of candidate points for a neighbor search is reduced through use of the triangle inequality : \\\\[|x+y| \\\\leq |x| + |y|\\\\] With this setup, a single distance calculation between a test point and the centroid is sufficient to determine a lower and upper bound on the distance to all points within the node. Because of the spherical geometry of the ball tree nodes, it can out-perform a KD-tree in high dimensions, though the actual performance is highly dependent on the structure of the training data. In scikit-learn, ball-tree-based neighbors searches are specified using the keyword , and are computed using the class sklearn.neighbors.BallTree . Alternatively, the user can work with the BallTree class directly. References: “Five balltree construction algorithms” , Omohundro, S.M., International Computer Science Institute Technical Report (1989) 1.6.4.4. Choice of Nearest Neighbors Algorithm  The optimal algorithm for a given dataset is a complicated choice, and depends on a number of factors: number of samples (i.e. ) and dimensionality (i.e. ). Brute force query time grows as Ball tree query time grows as approximately KD tree query time changes with in a way that is difficult to precisely characterise. For small (less than 20 or so) the cost is approximately , and the KD tree query can be very efficient. For larger , the cost increases to nearly , and the overhead due to the tree structure can lead to queries which are slower than brute force. For small data sets ( less than 30 or so), is comparable to , and brute force algorithms can be more efficient than a tree-based approach. Both KDTree and BallTree address this through providing a leaf size parameter: this controls the number of samples at which a query switches to brute-force. This allows both algorithms to approach the efficiency of a brute-force computation for small . data structure: intrinsic dimensionality of the data and/or sparsity of the data. Intrinsic dimensionality refers to the dimension of a manifold on which the data lies, which can be linearly or non-linearly embedded in the parameter space. Sparsity refers to the degree to which the data fills the parameter space (this is to be distinguished from the concept as used in “sparse” matrices. The data matrix may have no zero entries, but the structure can still be “sparse” in this sense). Brute force query time is unchanged by data structure. Ball tree and KD tree query times can be greatly influenced by data structure. In general, sparser data with a smaller intrinsic dimensionality leads to faster query times. Because the KD tree internal representation is aligned with the parameter axes, it will not generally show as much improvement as ball tree for arbitrarily structured data. Datasets used in machine learning tend to be very structured, and are very well-suited for tree-based queries. number of neighbors requested for a query point. Brute force query time is largely unaffected by the value of Ball tree and KD tree query time will become slower as increases. This is due to two effects: first, a larger leads to the necessity to search a larger portion of the parameter space. Second, using requires internal queueing of results as the tree is traversed. As becomes large compared to , the ability to prune branches in a tree-based query is reduced. In this situation, Brute force queries can be more efficient. number of query points. Both the ball tree and the KD Tree require a construction phase. The cost of this construction becomes negligible when amortized over many queries. If only a small number of queries will be performed, however, the construction can make up a significant fraction of the total cost. If very few query points will be required, brute force is better than a tree-based method. Currently, selects if , the input data is sparse, or isn’t in the list for either or . Otherwise, it selects the first out of and that has in its list. This choice is based on the assumption that the number of query points is at least the same order as the number of training points, and that is close to its default value of . 1.6.4.5. Effect of  As noted above, for small sample sizes a brute force search can be more efficient than a tree-based query. This fact is accounted for in the ball tree and KD tree by internally switching to brute force searches within leaf nodes. The level of this switch can be specified with the parameter . This parameter choice has many effects: construction time A larger leads to a faster tree construction time, because fewer nodes need to be created query time Both a large or small can lead to suboptimal query cost. For approaching 1, the overhead involved in traversing nodes can significantly slow query times. For approaching the size of the training set, queries become essentially brute force. A good compromise between these is , the default value of the parameter. memory As increases, the memory required to store a tree structure decreases. This is especially important in the case of ball tree, which stores a -dimensional centroid for each node. The required storage space for BallTree is approximately times the size of the training set. is not referenced for brute force queries. ',\n",
       " 'sklearn_1_6_nearest_neighbors\\n1.6. Nearest Neighbors\\nmodules/neighbors.html\\n 1.6.5. Nearest Centroid Classifier  The NearestCentroid classifier is a simple algorithm that represents each class by the centroid of its members. In effect, this makes it similar to the label updating phase of the sklearn.cluster.KMeans algorithm. It also has no parameters to choose, making it a good baseline classifier. It does, however, suffer on non-convex classes, as well as when classes have drastically different variances, as equal variance in all dimensions is assumed. See Linear Discriminant Analysis ( sklearn.discriminant_analysis.LinearDiscriminantAnalysis ) and Quadratic Discriminant Analysis ( sklearn.discriminant_analysis.QuadraticDiscriminantAnalysis ) for more complex methods that do not make this assumption. Usage of the default NearestCentroid is simple: 1.6.5.1. Nearest Shrunken Centroid  The NearestCentroid classifier has a parameter, which implements the nearest shrunken centroid classifier. In effect, the value of each feature for each centroid is divided by the within-class variance of that feature. The feature values are then reduced by . Most notably, if a particular feature value crosses zero, it is set to zero. In effect, this removes the feature from affecting the classification. This is useful, for example, for removing noisy features. In the example below, using a small shrink threshold increases the accuracy of the model from 0.81 to 0.82. Examples: Nearest Centroid Classification : an example of classification using nearest centroid with different shrink thresholds. ',\n",
       " 'sklearn_1_6_nearest_neighbors\\n1.6. Nearest Neighbors\\nmodules/neighbors.html\\n 1.6.6. Nearest Neighbors Transformer  Many scikit-learn estimators rely on nearest neighbors: Several classifiers and regressors such as KNeighborsClassifier and KNeighborsRegressor , but also some clustering methods such as DBSCAN and SpectralClustering , and some manifold embeddings such as TSNE and Isomap . All these estimators can compute internally the nearest neighbors, but most of them also accept precomputed nearest neighbors sparse graph , as given by kneighbors_graph and radius_neighbors_graph . With mode , these functions return a binary adjacency sparse graph as required, for instance, in SpectralClustering . Whereas with , they return a distance sparse graph as required, for instance, in DBSCAN . To include these functions in a scikit-learn pipeline, one can also use the corresponding classes KNeighborsTransformer and RadiusNeighborsTransformer . The benefits of this sparse graph API are multiple. First, the precomputed graph can be re-used multiple times, for instance while varying a parameter of the estimator. This can be done manually by the user, or using the caching properties of the scikit-learn pipeline: Second, precomputing the graph can give finer control on the nearest neighbors estimation, for instance enabling multiprocessing though the parameter , which might not be available in all estimators. Finally, the precomputation can be performed by custom estimators to use different implementations, such as approximate nearest neighbors methods, or implementation with special data types. The precomputed neighbors sparse graph needs to be formatted as in radius_neighbors_graph output: a CSR matrix (although COO, CSC or LIL will be accepted). only explicitly store nearest neighborhoods of each sample with respect to the training data. This should include those at 0 distance from a query point, including the matrix diagonal when computing the nearest neighborhoods between the training data and itself. each row’s should store the distance in increasing order (optional. Unsorted data will be stable-sorted, adding a computational overhead). all values in data should be non-negative. there should be no duplicate in any row (see https://github.com/scipy/scipy/issues/5807 ). if the algorithm being passed the precomputed matrix uses k nearest neighbors (as opposed to radius neighborhood), at least k neighbors must be stored in each row (or k+1, as explained in the following note). Note When a specific number of neighbors is queried (using KNeighborsTransformer ), the definition of is ambiguous since it can either include each training point as its own neighbor, or exclude them. Neither choice is perfect, since including them leads to a different number of non-self neighbors during training and testing, while excluding them leads to a difference between and , which is against scikit-learn API. In KNeighborsTransformer we use the definition which includes each training point as its own neighbor in the count of . However, for compatibility reasons with other estimators which use the other definition, one extra neighbor will be computed when . To maximise compatibility with all estimators, a safe choice is to always include one extra neighbor in a custom nearest neighbors estimator, since unnecessary neighbors will be filtered by following estimators. Examples: Approximate nearest neighbors in TSNE : an example of pipelining KNeighborsTransformer and TSNE . Also proposes two custom nearest neighbors estimators based on external packages. Caching nearest neighbors : an example of pipelining KNeighborsTransformer and KNeighborsClassifier to enable caching of the neighbors graph during a hyper-parameter grid-search. ',\n",
       " 'sklearn_1_6_nearest_neighbors\\n1.6. Nearest Neighbors\\nmodules/neighbors.html\\n 1.6.7. Neighborhood Components Analysis  Neighborhood Components Analysis (NCA, NeighborhoodComponentsAnalysis ) is a distance metric learning algorithm which aims to improve the accuracy of nearest neighbors classification compared to the standard Euclidean distance. The algorithm directly maximizes a stochastic variant of the leave-one-out k-nearest neighbors (KNN) score on the training set. It can also learn a low-dimensional linear projection of data that can be used for data visualization and fast classification. In the above illustrating figure, we consider some points from a randomly generated dataset. We focus on the stochastic KNN classification of point no. 3. The thickness of a link between sample 3 and another point is proportional to their distance, and can be seen as the relative weight (or probability) that a stochastic nearest neighbor prediction rule would assign to this point. In the original space, sample 3 has many stochastic neighbors from various classes, so the right class is not very likely. However, in the projected space learned by NCA, the only stochastic neighbors with non-negligible weight are from the same class as sample 3, guaranteeing that the latter will be well classified. See the mathematical formulation for more details. 1.6.7.1. Classification  Combined with a nearest neighbors classifier ( KNeighborsClassifier ), NCA is attractive for classification because it can naturally handle multi-class problems without any increase in the model size, and does not introduce additional parameters that require fine-tuning by the user. NCA classification has been shown to work well in practice for data sets of varying size and difficulty. In contrast to related methods such as Linear Discriminant Analysis, NCA does not make any assumptions about the class distributions. The nearest neighbor classification can naturally produce highly irregular decision boundaries. To use this model for classification, one needs to combine a NeighborhoodComponentsAnalysis instance that learns the optimal transformation with a KNeighborsClassifier instance that performs the classification in the projected space. Here is an example using the two classes: The plot shows decision boundaries for Nearest Neighbor Classification and Neighborhood Components Analysis classification on the iris dataset, when training and scoring on only two features, for visualisation purposes. 1.6.7.2. Dimensionality reduction  NCA can be used to perform supervised dimensionality reduction. The input data are projected onto a linear subspace consisting of the directions which minimize the NCA objective. The desired dimensionality can be set using the parameter . For instance, the following figure shows a comparison of dimensionality reduction with Principal Component Analysis ( sklearn.decomposition.PCA ), Linear Discriminant Analysis ( sklearn.discriminant_analysis.LinearDiscriminantAnalysis ) and Neighborhood Component Analysis ( NeighborhoodComponentsAnalysis ) on the Digits dataset, a dataset with size and . The data set is split into a training and a test set of equal size, then standardized. For evaluation the 3-nearest neighbor classification accuracy is computed on the 2-dimensional projected points found by each method. Each data sample belongs to one of 10 classes. Examples: Comparing Nearest Neighbors with and without Neighborhood Components Analysis Dimensionality Reduction with Neighborhood Components Analysis Manifold learning on handwritten digits: Locally Linear Embedding, Isomap… 1.6.7.3. Mathematical formulation  The goal of NCA is to learn an optimal linear transformation matrix of size , which maximises the sum over all samples of the probability that is correctly classified, i.e.: \\\\[\\\\underset{L}{\\\\arg\\\\max} \\\\sum\\\\limits_{i0}^{N - 1} p_{i}\\\\] with  and the probability of sample being correctly classified according to a stochastic nearest neighbors rule in the learned embedded space: \\\\[p_{i}\\\\sum\\\\limits_{j \\\\in C_i}{p_{i j}}\\\\] where is the set of points in the same class as sample , and is the softmax over Euclidean distances in the embedded space: \\\\[p_{i j}  \\\\frac{\\\\exp(-||L x_i - L x_j||^2)}{\\\\sum\\\\limits_{k \\\\ne i} {\\\\exp{-(||L x_i - L x_k||^2)}}} , \\\\quad p_{i i}  0\\\\] 1.6.7.3.1. Mahalanobis distance  NCA can be seen as learning a (squared) Mahalanobis distance metric: \\\\[|| L(x_i - x_j)||^2  (x_i - x_j)^TM(x_i - x_j),\\\\] where is a symmetric positive semi-definite matrix of size . 1.6.7.4. Implementation  This implementation follows what is explained in the original paper 1 . For the optimisation method, it currently uses scipy’s L-BFGS-B with a full gradient computation at each iteration, to avoid to tune the learning rate and provide stable learning. See the examples below and the docstring of NeighborhoodComponentsAnalysis.fit for further information. 1.6.7.5. Complexity  1.6.7.5.1. Training  NCA stores a matrix of pairwise distances, taking memory. Time complexity depends on the number of iterations done by the optimisation algorithm. However, one can set the maximum number of iterations with the argument . For each iteration, time complexity is . 1.6.7.5.2. Transform  Here the operation returns , therefore its time complexity equals . There is no added space complexity in the operation. References: 1 “Neighbourhood Components Analysis” , J. Goldberger, S. Roweis, G. Hinton, R. Salakhutdinov, Advances in Neural Information Processing Systems, Vol. 17, May 2005, pp. 513-520. Wikipedia entry on Neighborhood Components Analysis ',\n",
       " 'sklearn_1_7_gaussian_processes\\n1.7. Gaussian Processes\\nmodules/gaussian_process.html\\n 1.7.1. Gaussian Process Regression (GPR)  The GaussianProcessRegressor implements Gaussian processes (GP) for regression purposes. For this, the prior of the GP needs to be specified. The prior mean is assumed to be constant and zero (for ) or the training data’s mean (for ). The prior’s covariance is specified by passing a kernel object. The hyperparameters of the kernel are optimized during fitting of GaussianProcessRegressor by maximizing the log-marginal-likelihood (LML) based on the passed . As the LML may have multiple local optima, the optimizer can be started repeatedly by specifying . The first run is always conducted starting from the initial hyperparameter values of the kernel; subsequent runs are conducted from hyperparameter values that have been chosen randomly from the range of allowed values. If the initial hyperparameters should be kept fixed, can be passed as optimizer. The noise level in the targets can be specified by passing it via the parameter , either globally as a scalar or per datapoint. Note that a moderate noise level can also be helpful for dealing with numeric issues during fitting as it is effectively implemented as Tikhonov regularization, i.e., by adding it to the diagonal of the kernel matrix. An alternative to specifying the noise level explicitly is to include a WhiteKernel component into the kernel, which can estimate the global noise level from the data (see example below). The implementation is based on Algorithm 2.1 of [RW2006] . In addition to the API of standard scikit-learn estimators, GaussianProcessRegressor: allows prediction without prior fitting (based on the GP prior) provides an additional method , which evaluates samples drawn from the GPR (prior or posterior) at given inputs exposes a method , which can be used externally for other ways of selecting hyperparameters, e.g., via Markov chain Monte Carlo. ',\n",
       " 'sklearn_1_7_gaussian_processes\\n1.7. Gaussian Processes\\nmodules/gaussian_process.html\\n 1.7.2. GPR examples  1.7.2.1. GPR with noise-level estimation  This example illustrates that GPR with a sum-kernel including a WhiteKernel can estimate the noise level of data. An illustration of the log-marginal-likelihood (LML) landscape shows that there exist two local maxima of LML. The first corresponds to a model with a high noise level and a large length scale, which explains all variations in the data by noise. The second one has a smaller noise level and shorter length scale, which explains most of the variation by the noise-free functional relationship. The second model has a higher likelihood; however, depending on the initial value for the hyperparameters, the gradient-based optimization might also converge to the high-noise solution. It is thus important to repeat the optimization several times for different initializations. 1.7.2.2. Comparison of GPR and Kernel Ridge Regression  Both kernel ridge regression (KRR) and GPR learn a target function by employing internally the “kernel trick”. KRR learns a linear function in the space induced by the respective kernel which corresponds to a non-linear function in the original space. The linear function in the kernel space is chosen based on the mean-squared error loss with ridge regularization. GPR uses the kernel to define the covariance of a prior distribution over the target functions and uses the observed training data to define a likelihood function. Based on Bayes theorem, a (Gaussian) posterior distribution over target functions is defined, whose mean is used for prediction. A major difference is that GPR can choose the kernel’s hyperparameters based on gradient-ascent on the marginal likelihood function while KRR needs to perform a grid search on a cross-validated loss function (mean-squared error loss). A further difference is that GPR learns a generative, probabilistic model of the target function and can thus provide meaningful confidence intervals and posterior samples along with the predictions while KRR only provides predictions. The following figure illustrates both methods on an artificial dataset, which consists of a sinusoidal target function and strong noise. The figure compares the learned model of KRR and GPR based on a ExpSineSquared kernel, which is suited for learning periodic functions. The kernel’s hyperparameters control the smoothness (length_scale) and periodicity of the kernel (periodicity). Moreover, the noise level of the data is learned explicitly by GPR by an additional WhiteKernel component in the kernel and by the regularization parameter alpha of KRR. The figure shows that both methods learn reasonable models of the target function. GPR correctly identifies the periodicity of the function to be roughly (6.28), while KRR chooses the doubled periodicity . Besides that, GPR provides reasonable confidence bounds on the prediction which are not available for KRR. A major difference between the two methods is the time required for fitting and predicting: while fitting KRR is fast in principle, the grid-search for hyperparameter optimization scales exponentially with the number of hyperparameters (“curse of dimensionality”). The gradient-based optimization of the parameters in GPR does not suffer from this exponential scaling and is thus considerable faster on this example with 3-dimensional hyperparameter space. The time for predicting is similar; however, generating the variance of the predictive distribution of GPR takes considerable longer than just predicting the mean. 1.7.2.3. GPR on Mauna Loa CO2 data  This example is based on Section 5.4.3 of [RW2006] . It illustrates an example of complex kernel engineering and hyperparameter optimization using gradient ascent on the log-marginal-likelihood. The data consists of the monthly average atmospheric CO2 concentrations (in parts per million by volume (ppmv)) collected at the Mauna Loa Observatory in Hawaii, between 1958 and 1997. The objective is to model the CO2 concentration as a function of the time t. The kernel is composed of several terms that are responsible for explaining different properties of the signal: a long term, smooth rising trend is to be explained by an RBF kernel. The RBF kernel with a large length-scale enforces this component to be smooth; it is not enforced that the trend is rising which leaves this choice to the GP. The specific length-scale and the amplitude are free hyperparameters. a seasonal component, which is to be explained by the periodic ExpSineSquared kernel with a fixed periodicity of 1 year. The length-scale of this periodic component, controlling its smoothness, is a free parameter. In order to allow decaying away from exact periodicity, the product with an RBF kernel is taken. The length-scale of this RBF component controls the decay time and is a further free parameter. smaller, medium term irregularities are to be explained by a RationalQuadratic kernel component, whose length-scale and alpha parameter, which determines the diffuseness of the length-scales, are to be determined. According to [RW2006] , these irregularities can better be explained by a RationalQuadratic than an RBF kernel component, probably because it can accommodate several length-scales. a “noise” term, consisting of an RBF kernel contribution, which shall explain the correlated noise components such as local weather phenomena, and a WhiteKernel contribution for the white noise. The relative amplitudes and the RBF’s length scale are further free parameters. Maximizing the log-marginal-likelihood after subtracting the target’s mean yields the following kernel with an LML of -83.214: Thus, most of the target signal (34.4ppm) is explained by a long-term rising trend (length-scale 41.8 years). The periodic component has an amplitude of 3.27ppm, a decay time of 180 years and a length-scale of 1.44. The long decay time indicates that we have a locally very close to periodic seasonal component. The correlated noise has an amplitude of 0.197ppm with a length scale of 0.138 years and a white-noise contribution of 0.197ppm. Thus, the overall noise level is very small, indicating that the data can be very well explained by the model. The figure shows also that the model makes very confident predictions until around 2015 ',\n",
       " 'sklearn_1_7_gaussian_processes\\n1.7. Gaussian Processes\\nmodules/gaussian_process.html\\n 1.7.3. Gaussian Process Classification (GPC)  The GaussianProcessClassifier implements Gaussian processes (GP) for classification purposes, more specifically for probabilistic classification, where test predictions take the form of class probabilities. GaussianProcessClassifier places a GP prior on a latent function , which is then squashed through a link function to obtain the probabilistic classification. The latent function is a so-called nuisance function, whose values are not observed and are not relevant by themselves. Its purpose is to allow a convenient formulation of the model, and is removed (integrated out) during prediction. GaussianProcessClassifier implements the logistic link function, for which the integral cannot be computed analytically but is easily approximated in the binary case. In contrast to the regression setting, the posterior of the latent function is not Gaussian even for a GP prior since a Gaussian likelihood is inappropriate for discrete class labels. Rather, a non-Gaussian likelihood corresponding to the logistic link function (logit) is used. GaussianProcessClassifier approximates the non-Gaussian posterior with a Gaussian based on the Laplace approximation. More details can be found in Chapter 3 of [RW2006] . The GP prior mean is assumed to be zero. The prior’s covariance is specified by passing a kernel object. The hyperparameters of the kernel are optimized during fitting of GaussianProcessRegressor by maximizing the log-marginal-likelihood (LML) based on the passed . As the LML may have multiple local optima, the optimizer can be started repeatedly by specifying . The first run is always conducted starting from the initial hyperparameter values of the kernel; subsequent runs are conducted from hyperparameter values that have been chosen randomly from the range of allowed values. If the initial hyperparameters should be kept fixed, can be passed as optimizer. GaussianProcessClassifier supports multi-class classification by performing either one-versus-rest or one-versus-one based training and prediction. In one-versus-rest, one binary Gaussian process classifier is fitted for each class, which is trained to separate this class from the rest. In “one_vs_one”, one binary Gaussian process classifier is fitted for each pair of classes, which is trained to separate these two classes. The predictions of these binary predictors are combined into multi-class predictions. See the section on multi-class classification for more details. In the case of Gaussian process classification, “one_vs_one” might be computationally cheaper since it has to solve many problems involving only a subset of the whole training set rather than fewer problems on the whole dataset. Since Gaussian process classification scales cubically with the size of the dataset, this might be considerably faster. However, note that “one_vs_one” does not support predicting probability estimates but only plain predictions. Moreover, note that GaussianProcessClassifier does not (yet) implement a true multi-class Laplace approximation internally, but as discussed above is based on solving several binary classification tasks internally, which are combined using one-versus-rest or one-versus-one. ',\n",
       " 'sklearn_1_7_gaussian_processes\\n1.7. Gaussian Processes\\nmodules/gaussian_process.html\\n 1.7.4. GPC examples  1.7.4.1. Probabilistic predictions with GPC  This example illustrates the predicted probability of GPC for an RBF kernel with different choices of the hyperparameters. The first figure shows the predicted probability of GPC with arbitrarily chosen hyperparameters and with the hyperparameters corresponding to the maximum log-marginal-likelihood (LML). While the hyperparameters chosen by optimizing LML have a considerable larger LML, they perform slightly worse according to the log-loss on test data. The figure shows that this is because they exhibit a steep change of the class probabilities at the class boundaries (which is good) but have predicted probabilities close to 0.5 far away from the class boundaries (which is bad) This undesirable effect is caused by the Laplace approximation used internally by GPC. The second figure shows the log-marginal-likelihood for different choices of the kernel’s hyperparameters, highlighting the two choices of the hyperparameters used in the first figure by black dots. 1.7.4.2. Illustration of GPC on the XOR dataset  This example illustrates GPC on XOR data. Compared are a stationary, isotropic kernel ( RBF ) and a non-stationary kernel ( DotProduct ). On this particular dataset, the DotProduct kernel obtains considerably better results because the class-boundaries are linear and coincide with the coordinate axes. In practice, however, stationary kernels such as RBF often obtain better results. 1.7.4.3. Gaussian process classification (GPC) on iris dataset  This example illustrates the predicted probability of GPC for an isotropic and anisotropic RBF kernel on a two-dimensional version for the iris-dataset. This illustrates the applicability of GPC to non-binary classification. The anisotropic RBF kernel obtains slightly higher log-marginal-likelihood by assigning different length-scales to the two feature dimensions. ',\n",
       " 'sklearn_1_7_gaussian_processes\\n1.7. Gaussian Processes\\nmodules/gaussian_process.html\\n 1.7.5. Kernels for Gaussian Processes  Kernels (also called “covariance functions” in the context of GPs) are a crucial ingredient of GPs which determine the shape of prior and posterior of the GP. They encode the assumptions on the function being learned by defining the “similarity” of two datapoints combined with the assumption that similar datapoints should have similar target values. Two categories of kernels can be distinguished: stationary kernels depend only on the distance of two datapoints and not on their absolute values and are thus invariant to translations in the input space, while non-stationary kernels depend also on the specific values of the datapoints. Stationary kernels can further be subdivided into isotropic and anisotropic kernels, where isotropic kernels are also invariant to rotations in the input space. For more details, we refer to Chapter 4 of [RW2006] . For guidance on how to best combine different kernels, we refer to [Duv2014] . 1.7.5.1. Gaussian Process Kernel API  The main usage of a Kernel is to compute the GP’s covariance between datapoints. For this, the method of the kernel can be called. This method can either be used to compute the “auto-covariance” of all pairs of datapoints in a 2d array X, or the “cross-covariance” of all combinations of datapoints of a 2d array X with datapoints in a 2d array Y. The following identity holds true for all kernels k (except for the WhiteKernel ): If only the diagonal of the auto-covariance is being used, the method of a kernel can be called, which is more computationally efficient than the equivalent call to : Kernels are parameterized by a vector of hyperparameters. These hyperparameters can for instance control length-scales or periodicity of a kernel (see below). All kernels support computing analytic gradients of the kernel’s auto-covariance with respect to via setting in the method. This gradient is used by the Gaussian process (both regressor and classifier) in computing the gradient of the log-marginal-likelihood, which in turn is used to determine the value of , which maximizes the log-marginal-likelihood, via gradient ascent. For each hyperparameter, the initial value and the bounds need to be specified when creating an instance of the kernel. The current value of can be get and set via the property of the kernel object. Moreover, the bounds of the hyperparameters can be accessed by the property of the kernel. Note that both properties (theta and bounds) return log-transformed values of the internally used values since those are typically more amenable to gradient-based optimization. The specification of each hyperparameter is stored in the form of an instance of Hyperparameter in the respective kernel. Note that a kernel using a hyperparameter with name “x” must have the attributes self.x and self.x_bounds. The abstract base class for all kernels is Kernel . Kernel implements a similar interface as Estimator , providing the methods , , and . This allows setting kernel values also via meta-estimators such as Pipeline or GridSearch . Note that due to the nested structure of kernels (by applying kernel operators, see below), the names of kernel parameters might become relatively complicated. In general, for a binary kernel operator, parameters of the left operand are prefixed with and parameters of the right operand with . An additional convenience method is , which returns a cloned version of the kernel but with the hyperparameters set to . An illustrative example: All Gaussian process kernels are interoperable with sklearn.metrics.pairwise and vice versa: instances of subclasses of Kernel can be passed as to from sklearn.metrics.pairwise . Moreover, kernel functions from pairwise can be used as GP kernels by using the wrapper class PairwiseKernel . The only caveat is that the gradient of the hyperparameters is not analytic but numeric and all those kernels support only isotropic distances. The parameter is considered to be a hyperparameter and may be optimized. The other kernel parameters are set directly at initialization and are kept fixed. 1.7.5.2. Basic kernels  The ConstantKernel kernel can be used as part of a Product kernel where it scales the magnitude of the other factor (kernel) or as part of a Sum kernel, where it modifies the mean of the Gaussian process. It depends on a parameter . It is defined as: \\\\[k(x_i, x_j)  constant\\\\_value \\\\;\\\\forall\\\\; x_1, x_2\\\\] The main use-case of the WhiteKernel kernel is as part of a sum-kernel where it explains the noise-component of the signal. Tuning its parameter corresponds to estimating the noise-level. It is defined as: \\\\[k(x_i, x_j)  noise\\\\_level \\\\text{ if } x_i  x_j \\\\text{ else } 0\\\\] 1.7.5.3. Kernel operators  Kernel operators take one or two base kernels and combine them into a new kernel. The Sum kernel takes two kernels and and combines them via . The Product kernel takes two kernels and and combines them via . The Exponentiation kernel takes one base kernel and a scalar parameter and combines them via . Note that magic methods , and are overridden on the Kernel objects, so one can use e.g. as a shortcut for . 1.7.5.4. Radial-basis function (RBF) kernel  The RBF kernel is a stationary kernel. It is also known as the “squared exponential” kernel. It is parameterized by a length-scale parameter , which can either be a scalar (isotropic variant of the kernel) or a vector with the same number of dimensions as the inputs (anisotropic variant of the kernel). The kernel is given by: \\\\[k(x_i, x_j)  \\\\text{exp}\\\\left(- \\\\frac{d(x_i, x_j)^2}{2l^2} \\\\right)\\\\] where is the Euclidean distance. This kernel is infinitely differentiable, which implies that GPs with this kernel as covariance function have mean square derivatives of all orders, and are thus very smooth. The prior and posterior of a GP resulting from an RBF kernel are shown in the following figure: 1.7.5.5. Matérn kernel  The Matern kernel is a stationary kernel and a generalization of the RBF kernel. It has an additional parameter which controls the smoothness of the resulting function. It is parameterized by a length-scale parameter , which can either be a scalar (isotropic variant of the kernel) or a vector with the same number of dimensions as the inputs (anisotropic variant of the kernel). The kernel is given by: \\\\[k(x_i, x_j)  \\\\frac{1}{\\\\Gamma(\\\\nu)2^{\\\\nu-1}}\\\\Bigg(\\\\frac{\\\\sqrt{2\\\\nu}}{l} d(x_i , x_j )\\\\Bigg)^\\\\nu K_\\\\nu\\\\Bigg(\\\\frac{\\\\sqrt{2\\\\nu}}{l} d(x_i , x_j )\\\\Bigg),\\\\] where is the Euclidean distance, is a modified Bessel function and is the gamma function. As , the Matérn kernel converges to the RBF kernel. When , the Matérn kernel becomes identical to the absolute exponential kernel, i.e., \\\\[k(x_i, x_j)  \\\\exp \\\\Bigg(- \\\\frac{1}{l} d(x_i , x_j ) \\\\Bigg) \\\\quad \\\\quad \\\\nu \\\\tfrac{1}{2}\\\\] In particular, : \\\\[k(x_i, x_j)  \\\\Bigg(1 + \\\\frac{\\\\sqrt{3}}{l} d(x_i , x_j )\\\\Bigg) \\\\exp \\\\Bigg(-\\\\frac{\\\\sqrt{3}}{l} d(x_i , x_j ) \\\\Bigg) \\\\quad \\\\quad \\\\nu \\\\tfrac{3}{2}\\\\] and : \\\\[k(x_i, x_j)  \\\\Bigg(1 + \\\\frac{\\\\sqrt{5}}{l} d(x_i , x_j ) +\\\\frac{5}{3l} d(x_i , x_j )^2 \\\\Bigg) \\\\exp \\\\Bigg(-\\\\frac{\\\\sqrt{5}}{l} d(x_i , x_j ) \\\\Bigg) \\\\quad \\\\quad \\\\nu \\\\tfrac{5}{2}\\\\] are popular choices for learning functions that are not infinitely differentiable (as assumed by the RBF kernel) but at least once ( ) or twice differentiable ( ). The flexibility of controlling the smoothness of the learned function via allows adapting to the properties of the true underlying functional relation. The prior and posterior of a GP resulting from a Matérn kernel are shown in the following figure: See [RW2006] , pp84 for further details regarding the different variants of the Matérn kernel. 1.7.5.6. Rational quadratic kernel  The RationalQuadratic kernel can be seen as a scale mixture (an infinite sum) of RBF kernels with different characteristic length-scales. It is parameterized by a length-scale parameter and a scale mixture parameter Only the isotropic variant where is a scalar is supported at the moment. The kernel is given by: \\\\[k(x_i, x_j)  \\\\left(1 + \\\\frac{d(x_i, x_j)^2}{2\\\\alpha l^2}\\\\right)^{-\\\\alpha}\\\\] The prior and posterior of a GP resulting from a RationalQuadratic kernel are shown in the following figure: 1.7.5.7. Exp-Sine-Squared kernel  The ExpSineSquared kernel allows modeling periodic functions. It is parameterized by a length-scale parameter and a periodicity parameter . Only the isotropic variant where is a scalar is supported at the moment. The kernel is given by: \\\\[k(x_i, x_j)  \\\\text{exp}\\\\left(- \\\\frac{ 2\\\\sin^2(\\\\pi d(x_i, x_j) / p) }{ l^ 2} \\\\right)\\\\] The prior and posterior of a GP resulting from an ExpSineSquared kernel are shown in the following figure: 1.7.5.8. Dot-Product kernel  The DotProduct kernel is non-stationary and can be obtained from linear regression by putting priors on the coefficients of and a prior of on the bias. The DotProduct kernel is invariant to a rotation of the coordinates about the origin, but not translations. It is parameterized by a parameter . For , the kernel is called the homogeneous linear kernel, otherwise it is inhomogeneous. The kernel is given by \\\\[k(x_i, x_j)  \\\\sigma_0 ^ 2 + x_i \\\\cdot x_j\\\\] The DotProduct kernel is commonly combined with exponentiation. An example with exponent 2 is shown in the following figure: 1.7.5.9. References  RW2006 ( 1 , 2 , 3 , 4 , 5 , 6 ) Carl Eduard Rasmussen and Christopher K.I. Williams, “Gaussian Processes for Machine Learning”, MIT Press 2006, Link to an official complete PDF version of the book here . Duv2014 David Duvenaud, “The Kernel Cookbook: Advice on Covariance functions”, 2014, Link . ',\n",
       " 'sklearn_1_9_naive_bayes\\n1.9. Naive Bayes\\nmodules/naive_bayes.html\\n 1.9.1. Gaussian Naive Bayes  GaussianNB implements the Gaussian Naive Bayes algorithm for classification. The likelihood of the features is assumed to be Gaussian: \\\\[P(x_i \\\\mid y)  \\\\frac{1}{\\\\sqrt{2\\\\pi\\\\sigma^2_y}} \\\\exp\\\\left(-\\\\frac{(x_i - \\\\mu_y)^2}{2\\\\sigma^2_y}\\\\right)\\\\] The parameters and are estimated using maximum likelihood. ',\n",
       " 'sklearn_1_9_naive_bayes\\n1.9. Naive Bayes\\nmodules/naive_bayes.html\\n 1.9.2. Multinomial Naive Bayes  MultinomialNB implements the naive Bayes algorithm for multinomially distributed data, and is one of the two classic naive Bayes variants used in text classification (where the data are typically represented as word vector counts, although tf-idf vectors are also known to work well in practice). The distribution is parametrized by vectors for each class , where is the number of features (in text classification, the size of the vocabulary) and is the probability of feature appearing in a sample belonging to class . The parameters is estimated by a smoothed version of maximum likelihood, i.e. relative frequency counting: \\\\[\\\\hat{\\\\theta}_{yi}  \\\\frac{ N_{yi} + \\\\alpha}{N_y + \\\\alpha n}\\\\] where is the number of times feature appears in a sample of class in the training set , and is the total count of all features for class . The smoothing priors accounts for features not present in the learning samples and prevents zero probabilities in further computations. Setting is called Laplace smoothing, while is called Lidstone smoothing. ',\n",
       " 'sklearn_1_9_naive_bayes\\n1.9. Naive Bayes\\nmodules/naive_bayes.html\\n 1.9.3. Complement Naive Bayes  ComplementNB implements the complement naive Bayes (CNB) algorithm. CNB is an adaptation of the standard multinomial naive Bayes (MNB) algorithm that is particularly suited for imbalanced data sets. Specifically, CNB uses statistics from the complement of each class to compute the model’s weights. The inventors of CNB show empirically that the parameter estimates for CNB are more stable than those for MNB. Further, CNB regularly outperforms MNB (often by a considerable margin) on text classification tasks. The procedure for calculating the weights is as follows: \\\\[ \\\\begin{align}\\\\begin{aligned}\\\\hat{\\\\theta}_{ci}  \\\\frac{\\\\alpha_i + \\\\sum_{j:y_j \\\\neq c} d_{ij}} {\\\\alpha + \\\\sum_{j:y_j \\\\neq c} \\\\sum_{k} d_{kj}}\\\\\\\\w_{ci}  \\\\log \\\\hat{\\\\theta}_{ci}\\\\\\\\w_{ci}  \\\\frac{w_{ci}}{\\\\sum_{j} |w_{cj}|}\\\\end{aligned}\\\\end{align} \\\\] where the summations are over all documents not in class , is either the count or tf-idf value of term in document , is a smoothing hyperparameter like that found in MNB, and . The second normalization addresses the tendency for longer documents to dominate parameter estimates in MNB. The classification rule is: \\\\[\\\\hat{c}  \\\\arg\\\\min_c \\\\sum_{i} t_i w_{ci}\\\\] i.e., a document is assigned to the class that is the poorest complement match. References: Rennie, J. D., Shih, L., Teevan, J., & Karger, D. R. (2003). Tackling the poor assumptions of naive bayes text classifiers. In ICML (Vol. 3, pp. 616-623). ',\n",
       " 'sklearn_1_9_naive_bayes\\n1.9. Naive Bayes\\nmodules/naive_bayes.html\\n 1.9.4. Bernoulli Naive Bayes  BernoulliNB implements the naive Bayes training and classification algorithms for data that is distributed according to multivariate Bernoulli distributions; i.e., there may be multiple features but each one is assumed to be a binary-valued (Bernoulli, boolean) variable. Therefore, this class requires samples to be represented as binary-valued feature vectors; if handed any other kind of data, a instance may binarize its input (depending on the parameter). The decision rule for Bernoulli naive Bayes is based on \\\\[P(x_i \\\\mid y)  P(i \\\\mid y) x_i + (1 - P(i \\\\mid y)) (1 - x_i)\\\\] which differs from multinomial NB’s rule in that it explicitly penalizes the non-occurrence of a feature that is an indicator for class , where the multinomial variant would simply ignore a non-occurring feature. In the case of text classification, word occurrence vectors (rather than word count vectors) may be used to train and use this classifier. might perform better on some datasets, especially those with shorter documents. It is advisable to evaluate both models, if time permits. References: C.D. Manning, P. Raghavan and H. Schütze (2008). Introduction to Information Retrieval. Cambridge University Press, pp. 234-265. A. McCallum and K. Nigam (1998). A comparison of event models for Naive Bayes text classification. Proc. AAAI/ICML-98 Workshop on Learning for Text Categorization, pp. 41-48. V. Metsis, I. Androutsopoulos and G. Paliouras (2006). Spam filtering with Naive Bayes – Which Naive Bayes? 3rd Conf. on Email and Anti-Spam (CEAS). ',\n",
       " 'sklearn_1_9_naive_bayes\\n1.9. Naive Bayes\\nmodules/naive_bayes.html\\n 1.9.5. Categorical Naive Bayes  CategoricalNB implements the categorical naive Bayes algorithm for categorically distributed data. It assumes that each feature, which is described by the index , has its own categorical distribution. For each feature in the training set , CategoricalNB estimates a categorical distribution for each feature i of X conditioned on the class y. The index set of the samples is defined as , with as the number of samples. The probability of category in feature given class is estimated as: \\\\[P(x_i  t \\\\mid y  c \\\\: ;\\\\, \\\\alpha)  \\\\frac{ N_{tic} + \\\\alpha}{N_{c} + \\\\alpha n_i},\\\\] where is the number of times category appears in the samples , which belong to class , is the number of samples with class c, is a smoothing parameter and is the number of available categories of feature . CategoricalNB assumes that the sample matrix is encoded (for instance with the help of OrdinalEncoder ) such that all categories for each feature are represented with numbers where is the number of available categories of feature . ',\n",
       " 'sklearn_1_9_naive_bayes\\n1.9. Naive Bayes\\nmodules/naive_bayes.html\\n 1.9.6. Out-of-core naive Bayes model fitting  Naive Bayes models can be used to tackle large scale classification problems for which the full training set might not fit in memory. To handle this case, MultinomialNB , BernoulliNB , and GaussianNB expose a method that can be used incrementally as done with other classifiers as demonstrated in Out-of-core classification of text documents . All naive Bayes classifiers support sample weighting. Contrary to the method, the first call to needs to be passed the list of all the expected class labels. For an overview of available strategies in scikit-learn, see also the out-of-core learning documentation. Note The method call of naive Bayes models introduces some computational overhead. It is recommended to use data chunk sizes that are as large as possible, that is as the available RAM allows. ',\n",
       " 'sklearn_2_1_gaussian_mixture_models\\n2.1. Gaussian mixture models\\nmodules/mixture.html\\n 2.1.1. Gaussian Mixture  The GaussianMixture object implements the expectation-maximization (EM) algorithm for fitting mixture-of-Gaussian models. It can also draw confidence ellipsoids for multivariate models, and compute the Bayesian Information Criterion to assess the number of clusters in the data. A GaussianMixture.fit method is provided that learns a Gaussian Mixture Model from train data. Given test data, it can assign to each sample the Gaussian it mostly probably belong to using the GaussianMixture.predict method. The GaussianMixture comes with different options to constrain the covariance of the difference classes estimated: spherical, diagonal, tied or full covariance. Examples: See GMM covariances for an example of using the Gaussian mixture as clustering on the iris dataset. See Density Estimation for a Gaussian mixture for an example on plotting the density estimation. 2.1.1.1. Pros and cons of class GaussianMixture  2.1.1.1.1. Pros  Speed It is the fastest algorithm for learning mixture models Agnostic As this algorithm maximizes only the likelihood, it will not bias the means towards zero, or bias the cluster sizes to have specific structures that might or might not apply. 2.1.1.1.2. Cons  Singularities When one has insufficiently many points per mixture, estimating the covariance matrices becomes difficult, and the algorithm is known to diverge and find solutions with infinite likelihood unless one regularizes the covariances artificially. Number of components This algorithm will always use all the components it has access to, needing held-out data or information theoretical criteria to decide how many components to use in the absence of external cues. 2.1.1.2. Selecting the number of components in a classical Gaussian Mixture Model  The BIC criterion can be used to select the number of components in a Gaussian Mixture in an efficient way. In theory, it recovers the true number of components only in the asymptotic regime (i.e. if much data is available and assuming that the data was actually generated i.i.d. from a mixture of Gaussian distribution). Note that using a Variational Bayesian Gaussian mixture avoids the specification of the number of components for a Gaussian mixture model. Examples: See Gaussian Mixture Model Selection for an example of model selection performed with classical Gaussian mixture. 2.1.1.3. Estimation algorithm Expectation-maximization  The main difficulty in learning Gaussian mixture models from unlabeled data is that it is one usually doesn’t know which points came from which latent component (if one has access to this information it gets very easy to fit a separate Gaussian distribution to each set of points). Expectation-maximization is a well-founded statistical algorithm to get around this problem by an iterative process. First one assumes random components (randomly centered on data points, learned from k-means, or even just normally distributed around the origin) and computes for each point a probability of being generated by each component of the model. Then, one tweaks the parameters to maximize the likelihood of the data given those assignments. Repeating this process is guaranteed to always converge to a local optimum. ',\n",
       " 'sklearn_2_1_gaussian_mixture_models\\n2.1. Gaussian mixture models\\nmodules/mixture.html\\n 2.1.2. Variational Bayesian Gaussian Mixture  The BayesianGaussianMixture object implements a variant of the Gaussian mixture model with variational inference algorithms. The API is similar as the one defined by GaussianMixture . 2.1.2.1. Estimation algorithm: variational inference  Variational inference is an extension of expectation-maximization that maximizes a lower bound on model evidence (including priors) instead of data likelihood. The principle behind variational methods is the same as expectation-maximization (that is both are iterative algorithms that alternate between finding the probabilities for each point to be generated by each mixture and fitting the mixture to these assigned points), but variational methods add regularization by integrating information from prior distributions. This avoids the singularities often found in expectation-maximization solutions but introduces some subtle biases to the model. Inference is often notably slower, but not usually as much so as to render usage unpractical. Due to its Bayesian nature, the variational algorithm needs more hyper- parameters than expectation-maximization, the most important of these being the concentration parameter . Specifying a low value for the concentration prior will make the model put most of the weight on few components set the remaining components weights very close to zero. High values of the concentration prior will allow a larger number of components to be active in the mixture. The parameters implementation of the BayesianGaussianMixture class proposes two types of prior for the weights distribution: a finite mixture model with Dirichlet distribution and an infinite mixture model with the Dirichlet Process. In practice Dirichlet Process inference algorithm is approximated and uses a truncated distribution with a fixed maximum number of components (called the Stick-breaking representation). The number of components actually used almost always depends on the data. The next figure compares the results obtained for the different type of the weight concentration prior (parameter ) for different values of . Here, we can see the value of the parameter has a strong impact on the effective number of active components obtained. We can also notice that large values for the concentration weight prior lead to more uniform weights when the type of prior is ‘dirichlet_distribution’ while this is not necessarily the case for the ‘dirichlet_process’ type (used by default). The examples below compare Gaussian mixture models with a fixed number of components, to the variational Gaussian mixture models with a Dirichlet process prior. Here, a classical Gaussian mixture is fitted with 5 components on a dataset composed of 2 clusters. We can see that the variational Gaussian mixture with a Dirichlet process prior is able to limit itself to only 2 components whereas the Gaussian mixture fits the data with a fixed number of components that has to be set a priori by the user. In this case the user has selected which does not match the true generative distribution of this toy dataset. Note that with very little observations, the variational Gaussian mixture models with a Dirichlet process prior can take a conservative stand, and fit only one component. On the following figure we are fitting a dataset not well-depicted by a Gaussian mixture. Adjusting the , parameter of the BayesianGaussianMixture controls the number of components used to fit this data. We also present on the last two plots a random sampling generated from the two resulting mixtures. Examples: See Gaussian Mixture Model Ellipsoids for an example on plotting the confidence ellipsoids for both GaussianMixture and BayesianGaussianMixture . Gaussian Mixture Model Sine Curve shows using GaussianMixture and BayesianGaussianMixture to fit a sine wave. See Concentration Prior Type Analysis of Variation Bayesian Gaussian Mixture for an example plotting the confidence ellipsoids for the BayesianGaussianMixture with different for different values of the parameter . 2.1.2.2. Pros and cons of variational inference with BayesianGaussianMixture  2.1.2.2.1. Pros  Automatic selection when is small enough and is larger than what is found necessary by the model, the Variational Bayesian mixture model has a natural tendency to set some mixture weights values close to zero. This makes it possible to let the model choose a suitable number of effective components automatically. Only an upper bound of this number needs to be provided. Note however that the “ideal” number of active components is very application specific and is typically ill-defined in a data exploration setting. Less sensitivity to the number of parameters unlike finite models, which will almost always use all components as much as they can, and hence will produce wildly different solutions for different numbers of components, the variational inference with a Dirichlet process prior ( ) won’t change much with changes to the parameters, leading to more stability and less tuning. Regularization due to the incorporation of prior information, variational solutions have less pathological special cases than expectation-maximization solutions. 2.1.2.2.2. Cons  Speed the extra parametrization necessary for variational inference make inference slower, although not by much. Hyperparameters this algorithm needs an extra hyperparameter that might need experimental tuning via cross-validation. Bias there are many implicit biases in the inference algorithms (and also in the Dirichlet process if used), and whenever there is a mismatch between these biases and the data it might be possible to fit better models using a finite mixture. 2.1.2.3. The Dirichlet Process  Here we describe variational inference algorithms on Dirichlet process mixture. The Dirichlet process is a prior probability distribution on clusterings with an infinite, unbounded, number of partitions . Variational techniques let us incorporate this prior structure on Gaussian mixture models at almost no penalty in inference time, comparing with a finite Gaussian mixture model. An important question is how can the Dirichlet process use an infinite, unbounded number of clusters and still be consistent. While a full explanation doesn’t fit this manual, one can think of its stick breaking process analogy to help understanding it. The stick breaking process is a generative story for the Dirichlet process. We start with a unit-length stick and in each step we break off a portion of the remaining stick. Each time, we associate the length of the piece of the stick to the proportion of points that falls into a group of the mixture. At the end, to represent the infinite mixture, we associate the last remaining piece of the stick to the proportion of points that don’t fall into all the other groups. The length of each piece is a random variable with probability proportional to the concentration parameter. Smaller value of the concentration will divide the unit-length into larger pieces of the stick (defining more concentrated distribution). Larger concentration values will create smaller pieces of the stick (increasing the number of components with non zero weights). Variational inference techniques for the Dirichlet process still work with a finite approximation to this infinite mixture model, but instead of having to specify a priori how many components one wants to use, one just specifies the concentration parameter and an upper bound on the number of mixture components (this upper bound, assuming it is higher than the “true” number of components, affects only algorithmic complexity, not the actual number of components used). ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.10. Tips on practical use  Make sure the same scale is used over all features. Because manifold learning methods are based on a nearest-neighbor search, the algorithm may perform poorly otherwise. See StandardScaler for convenient ways of scaling heterogeneous data. The reconstruction error computed by each routine can be used to choose the optimal output dimension. For a -dimensional manifold embedded in a -dimensional parameter space, the reconstruction error will decrease as is increased until . Note that noisy data can “short-circuit” the manifold, in essence acting as a bridge between parts of the manifold that would otherwise be well-separated. Manifold learning on noisy and/or incomplete data is an active area of research. Certain input configurations can lead to singular weight matrices, for example when more than two points in the dataset are identical, or when the data is split into disjointed groups. In this case, will fail to find the null space. The easiest way to address this is to use which will work on a singular matrix, though it may be very slow depending on the number of input points. Alternatively, one can attempt to understand the source of the singularity: if it is due to disjoint sets, increasing may help. If it is due to identical points in the dataset, removing these points may help. See also Totally Random Trees Embedding can also be useful to derive non-linear representations of feature space, also it does not perform dimensionality reduction. ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.1. Introduction  High-dimensional datasets can be very difficult to visualize. While data in two or three dimensions can be plotted to show the inherent structure of the data, equivalent high-dimensional plots are much less intuitive. To aid visualization of the structure of a dataset, the dimension must be reduced in some way. The simplest way to accomplish this dimensionality reduction is by taking a random projection of the data. Though this allows some degree of visualization of the data structure, the randomness of the choice leaves much to be desired. In a random projection, it is likely that the more interesting structure within the data will be lost. To address this concern, a number of supervised and unsupervised linear dimensionality reduction frameworks have been designed, such as Principal Component Analysis (PCA), Independent Component Analysis, Linear Discriminant Analysis, and others. These algorithms define specific rubrics to choose an “interesting” linear projection of the data. These methods can be powerful, but often miss important non-linear structure in the data. Manifold Learning can be thought of as an attempt to generalize linear frameworks like PCA to be sensitive to non-linear structure in data. Though supervised variants exist, the typical manifold learning problem is unsupervised: it learns the high-dimensional structure of the data from the data itself, without the use of predetermined classifications. Examples: See Manifold learning on handwritten digits: Locally Linear Embedding, Isomap… for an example of dimensionality reduction on handwritten digits. See Comparison of Manifold Learning methods for an example of dimensionality reduction on a toy “S-curve” dataset. The manifold learning implementations available in scikit-learn are summarized below ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.2. Isomap  One of the earliest approaches to manifold learning is the Isomap algorithm, short for Isometric Mapping. Isomap can be viewed as an extension of Multi-dimensional Scaling (MDS) or Kernel PCA. Isomap seeks a lower-dimensional embedding which maintains geodesic distances between all points. Isomap can be performed with the object Isomap . 2.2.2.1. Complexity  The Isomap algorithm comprises three stages: Nearest neighbor search. Isomap uses sklearn.neighbors.BallTree for efficient neighbor search. The cost is approximately , for nearest neighbors of points in dimensions. Shortest-path graph search. The most efficient known algorithms for this are Dijkstra’s Algorithm , which is approximately , or the Floyd-Warshall algorithm , which is . The algorithm can be selected by the user with the keyword of . If unspecified, the code attempts to choose the best algorithm for the input data. Partial eigenvalue decomposition. The embedding is encoded in the eigenvectors corresponding to the largest eigenvalues of the isomap kernel. For a dense solver, the cost is approximately . This cost can often be improved using the solver. The eigensolver can be specified by the user with the keyword of . If unspecified, the code attempts to choose the best algorithm for the input data. The overall complexity of Isomap is . : number of training data points : input dimension : number of nearest neighbors : output dimension References: “A global geometric framework for nonlinear dimensionality reduction” Tenenbaum, J.B.; De Silva, V.; & Langford, J.C. Science 290 (5500) ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.3. Locally Linear Embedding  Locally linear embedding (LLE) seeks a lower-dimensional projection of the data which preserves distances within local neighborhoods. It can be thought of as a series of local Principal Component Analyses which are globally compared to find the best non-linear embedding. Locally linear embedding can be performed with function locally_linear_embedding or its object-oriented counterpart LocallyLinearEmbedding . 2.2.3.1. Complexity  The standard LLE algorithm comprises three stages: Nearest Neighbors Search . See discussion under Isomap above. Weight Matrix Construction . . The construction of the LLE weight matrix involves the solution of a linear equation for each of the local neighborhoods Partial Eigenvalue Decomposition . See discussion under Isomap above. The overall complexity of standard LLE is . : number of training data points : input dimension : number of nearest neighbors : output dimension References: “Nonlinear dimensionality reduction by locally linear embedding” Roweis, S. & Saul, L. Science 290:2323 (2000) ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.4. Modified Locally Linear Embedding  One well-known issue with LLE is the regularization problem. When the number of neighbors is greater than the number of input dimensions, the matrix defining each local neighborhood is rank-deficient. To address this, standard LLE applies an arbitrary regularization parameter , which is chosen relative to the trace of the local weight matrix. Though it can be shown formally that as , the solution converges to the desired embedding, there is no guarantee that the optimal solution will be found for . This problem manifests itself in embeddings which distort the underlying geometry of the manifold. One method to address the regularization problem is to use multiple weight vectors in each neighborhood. This is the essence of modified locally linear embedding (MLLE). MLLE can be performed with function locally_linear_embedding or its object-oriented counterpart LocallyLinearEmbedding , with the keyword . It requires . 2.2.4.1. Complexity  The MLLE algorithm comprises three stages: Nearest Neighbors Search . Same as standard LLE Weight Matrix Construction . Approximately . The first term is exactly equivalent to that of standard LLE. The second term has to do with constructing the weight matrix from multiple weights. In practice, the added cost of constructing the MLLE weight matrix is relatively small compared to the cost of stages 1 and 3. Partial Eigenvalue Decomposition . Same as standard LLE The overall complexity of MLLE is . : number of training data points : input dimension : number of nearest neighbors : output dimension References: “MLLE: Modified Locally Linear Embedding Using Multiple Weights” Zhang, Z. & Wang, J. ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.5. Hessian Eigenmapping  Hessian Eigenmapping (also known as Hessian-based LLE: HLLE) is another method of solving the regularization problem of LLE. It revolves around a hessian-based quadratic form at each neighborhood which is used to recover the locally linear structure. Though other implementations note its poor scaling with data size, implements some algorithmic improvements which make its cost comparable to that of other LLE variants for small output dimension. HLLE can be performed with function locally_linear_embedding or its object-oriented counterpart LocallyLinearEmbedding , with the keyword . It requires . 2.2.5.1. Complexity  The HLLE algorithm comprises three stages: Nearest Neighbors Search . Same as standard LLE Weight Matrix Construction . Approximately . The first term reflects a similar cost to that of standard LLE. The second term comes from a QR decomposition of the local hessian estimator. Partial Eigenvalue Decomposition . Same as standard LLE The overall complexity of standard HLLE is . : number of training data points : input dimension : number of nearest neighbors : output dimension References: “Hessian Eigenmaps: Locally linear embedding techniques for high-dimensional data” Donoho, D. & Grimes, C. Proc Natl Acad Sci USA. 100:5591 (2003) ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.6. Spectral Embedding  Spectral Embedding is an approach to calculating a non-linear embedding. Scikit-learn implements Laplacian Eigenmaps, which finds a low dimensional representation of the data using a spectral decomposition of the graph Laplacian. The graph generated can be considered as a discrete approximation of the low dimensional manifold in the high dimensional space. Minimization of a cost function based on the graph ensures that points close to each other on the manifold are mapped close to each other in the low dimensional space, preserving local distances. Spectral embedding can be performed with the function spectral_embedding or its object-oriented counterpart SpectralEmbedding . 2.2.6.1. Complexity  The Spectral Embedding (Laplacian Eigenmaps) algorithm comprises three stages: Weighted Graph Construction . Transform the raw input data into graph representation using affinity (adjacency) matrix representation. Graph Laplacian Construction . unnormalized Graph Laplacian is constructed as for and normalized one as . Partial Eigenvalue Decomposition . Eigenvalue decomposition is done on graph Laplacian The overall complexity of spectral embedding is . : number of training data points : input dimension : number of nearest neighbors : output dimension References: “Laplacian Eigenmaps for Dimensionality Reduction and Data Representation” M. Belkin, P. Niyogi, Neural Computation, June 2003; 15 (6):1373-1396 ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.7. Local Tangent Space Alignment  Though not technically a variant of LLE, Local tangent space alignment (LTSA) is algorithmically similar enough to LLE that it can be put in this category. Rather than focusing on preserving neighborhood distances as in LLE, LTSA seeks to characterize the local geometry at each neighborhood via its tangent space, and performs a global optimization to align these local tangent spaces to learn the embedding. LTSA can be performed with function locally_linear_embedding or its object-oriented counterpart LocallyLinearEmbedding , with the keyword . 2.2.7.1. Complexity  The LTSA algorithm comprises three stages: Nearest Neighbors Search . Same as standard LLE Weight Matrix Construction . Approximately . The first term reflects a similar cost to that of standard LLE. Partial Eigenvalue Decomposition . Same as standard LLE The overall complexity of standard LTSA is . : number of training data points : input dimension : number of nearest neighbors : output dimension References: “Principal manifolds and nonlinear dimensionality reduction via tangent space alignment” Zhang, Z. & Zha, H. Journal of Shanghai Univ. 8:406 (2004) ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.8. Multi-dimensional Scaling (MDS)  Multidimensional scaling ( MDS ) seeks a low-dimensional representation of the data in which the distances respect well the distances in the original high-dimensional space. In general, MDS is a technique used for analyzing similarity or dissimilarity data. It attempts to model similarity or dissimilarity data as distances in a geometric spaces. The data can be ratings of similarity between objects, interaction frequencies of molecules, or trade indices between countries. There exists two types of MDS algorithm: metric and non metric. In the scikit-learn, the class MDS implements both. In Metric MDS, the input similarity matrix arises from a metric (and thus respects the triangular inequality), the distances between output two points are then set to be as close as possible to the similarity or dissimilarity data. In the non-metric version, the algorithms will try to preserve the order of the distances, and hence seek for a monotonic relationship between the distances in the embedded space and the similarities/dissimilarities. Let be the similarity matrix, and the coordinates of the input points. Disparities are transformation of the similarities chosen in some optimal ways. The objective, called the stress, is then defined by 2.2.8.1. Metric MDS  The simplest metric MDS model, called absolute MDS , disparities are defined by . With absolute MDS, the value should then correspond exactly to the distance between point and in the embedding point. Most commonly, disparities are set to . 2.2.8.2. Nonmetric MDS  Non metric MDS focuses on the ordination of the data. If , then the embedding should enforce . A simple algorithm to enforce that is to use a monotonic regression of on , yielding disparities in the same order as . A trivial solution to this problem is to set all the points on the origin. In order to avoid that, the disparities are normalized. References: “Modern Multidimensional Scaling - Theory and Applications” Borg, I.; Groenen P. Springer Series in Statistics (1997) “Nonmetric multidimensional scaling: a numerical method” Kruskal, J. Psychometrika, 29 (1964) “Multidimensional scaling by optimizing goodness of fit to a nonmetric hypothesis” Kruskal, J. Psychometrika, 29, (1964) ',\n",
       " 'sklearn_2_2_manifold_learning\\n2.2. Manifold learning\\nmodules/manifold.html\\n 2.2.9. t-distributed Stochastic Neighbor Embedding (t-SNE)  t-SNE ( TSNE ) converts affinities of data points to probabilities. The affinities in the original space are represented by Gaussian joint probabilities and the affinities in the embedded space are represented by Student’s t-distributions. This allows t-SNE to be particularly sensitive to local structure and has a few other advantages over existing techniques: Revealing the structure at many scales on a single map Revealing data that lie in multiple, different, manifolds or clusters Reducing the tendency to crowd points together at the center While Isomap, LLE and variants are best suited to unfold a single continuous low dimensional manifold, t-SNE will focus on the local structure of the data and will tend to extract clustered local groups of samples as highlighted on the S-curve example. This ability to group samples based on the local structure might be beneficial to visually disentangle a dataset that comprises several manifolds at once as is the case in the digits dataset. The Kullback-Leibler (KL) divergence of the joint probabilities in the original space and the embedded space will be minimized by gradient descent. Note that the KL divergence is not convex, i.e. multiple restarts with different initializations will end up in local minima of the KL divergence. Hence, it is sometimes useful to try different seeds and select the embedding with the lowest KL divergence. The disadvantages to using t-SNE are roughly: t-SNE is computationally expensive, and can take several hours on million-sample datasets where PCA will finish in seconds or minutes The Barnes-Hut t-SNE method is limited to two or three dimensional embeddings. The algorithm is stochastic and multiple restarts with different seeds can yield different embeddings. However, it is perfectly legitimate to pick the embedding with the least error. Global structure is not explicitly preserved. This problem is mitigated by initializing points with PCA (using ). 2.2.9.1. Optimizing t-SNE  The main purpose of t-SNE is visualization of high-dimensional data. Hence, it works best when the data will be embedded on two or three dimensions. Optimizing the KL divergence can be a little bit tricky sometimes. There are five parameters that control the optimization of t-SNE and therefore possibly the quality of the resulting embedding: perplexity early exaggeration factor learning rate maximum number of iterations angle (not used in the exact method) The perplexity is defined as where is the Shannon entropy of the conditional probability distribution. The perplexity of a -sided die is , so that is effectively the number of nearest neighbors t-SNE considers when generating the conditional probabilities. Larger perplexities lead to more nearest neighbors and less sensitive to small structure. Conversely a lower perplexity considers a smaller number of neighbors, and thus ignores more global information in favour of the local neighborhood. As dataset sizes get larger more points will be required to get a reasonable sample of the local neighborhood, and hence larger perplexities may be required. Similarly noisier datasets will require larger perplexity values to encompass enough local neighbors to see beyond the background noise. The maximum number of iterations is usually high enough and does not need any tuning. The optimization consists of two phases: the early exaggeration phase and the final optimization. During early exaggeration the joint probabilities in the original space will be artificially increased by multiplication with a given factor. Larger factors result in larger gaps between natural clusters in the data. If the factor is too high, the KL divergence could increase during this phase. Usually it does not have to be tuned. A critical parameter is the learning rate. If it is too low gradient descent will get stuck in a bad local minimum. If it is too high the KL divergence will increase during optimization. More tips can be found in Laurens van der Maaten’s FAQ (see references). The last parameter, angle, is a tradeoff between performance and accuracy. Larger angles imply that we can approximate larger regions by a single point, leading to better speed but less accurate results. “How to Use t-SNE Effectively” provides a good discussion of the effects of the various parameters, as well as interactive plots to explore the effects of different parameters. 2.2.9.2. Barnes-Hut t-SNE  The Barnes-Hut t-SNE that has been implemented here is usually much slower than other manifold learning algorithms. The optimization is quite difficult and the computation of the gradient is , where is the number of output dimensions and is the number of samples. The Barnes-Hut method improves on the exact method where t-SNE complexity is , but has several other notable differences: The Barnes-Hut implementation only works when the target dimensionality is 3 or less. The 2D case is typical when building visualizations. Barnes-Hut only works with dense input data. Sparse data matrices can only be embedded with the exact method or can be approximated by a dense low rank projection for instance using sklearn.decomposition.TruncatedSVD Barnes-Hut is an approximation of the exact method. The approximation is parameterized with the angle parameter, therefore the angle parameter is unused when method”exact” Barnes-Hut is significantly more scalable. Barnes-Hut can be used to embed hundred of thousands of data points while the exact method can handle thousands of samples before becoming computationally intractable For visualization purpose (which is the main use case of t-SNE), using the Barnes-Hut method is strongly recommended. The exact t-SNE method is useful for checking the theoretically properties of the embedding possibly in higher dimensional space but limit to small datasets due to computational constraints. Also note that the digits labels roughly match the natural grouping found by t-SNE while the linear 2D projection of the PCA model yields a representation where label regions largely overlap. This is a strong clue that this data can be well separated by non linear methods that focus on the local structure (e.g. an SVM with a Gaussian RBF kernel). However, failing to visualize well separated homogeneously labeled groups with t-SNE in 2D does not necessarily imply that the data cannot be correctly classified by a supervised model. It might be the case that 2 dimensions are not low enough to accurately represents the internal structure of the data. References: “Visualizing High-Dimensional Data Using t-SNE” van der Maaten, L.J.P.; Hinton, G. Journal of Machine Learning Research (2008) “t-Distributed Stochastic Neighbor Embedding” van der Maaten, L.J.P. “Accelerating t-SNE using Tree-Based Algorithms.” L.J.P. van der Maaten. Journal of Machine Learning Research 15(Oct):3221-3245, 2014. ',\n",
       " \"sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.10. Clustering performance evaluation  Evaluating the performance of a clustering algorithm is not as trivial as counting the number of errors or the precision and recall of a supervised classification algorithm. In particular any evaluation metric should not take the absolute values of the cluster labels into account but rather if this clustering define separations of the data similar to some ground truth set of classes or satisfying some assumption such that members belong to the same class are more similar than members of different classes according to some similarity metric. 2.3.10.1. Adjusted Rand index  Given the knowledge of the ground truth class assignments and our clustering algorithm assignments of the same samples , the adjusted Rand index is a function that measures the similarity of the two assignments, ignoring permutations and with chance normalization : One can permute 0 and 1 in the predicted labels, rename 2 to 3, and get the same score: Furthermore, adjusted_rand_score is symmetric : swapping the argument does not change the score. It can thus be used as a consensus measure : Perfect labeling is scored 1.0: Bad (e.g. independent labelings) have negative or close to 0.0 scores: 2.3.10.1.1. Advantages  Random (uniform) label assignments have a ARI score close to 0.0 for any value of and (which is not the case for raw Rand index or the V-measure for instance). Bounded range [-1, 1] : negative values are bad (independent labelings), similar clusterings have a positive ARI, 1.0 is the perfect match score. No assumption is made on the cluster structure : can be used to compare clustering algorithms such as k-means which assumes isotropic blob shapes with results of spectral clustering algorithms which can find cluster with “folded” shapes. 2.3.10.1.2. Drawbacks  Contrary to inertia, ARI requires knowledge of the ground truth classes while is almost never available in practice or requires manual assignment by human annotators (as in the supervised learning setting). However ARI can also be useful in a purely unsupervised setting as a building block for a Consensus Index that can be used for clustering model selection (TODO). Examples: Adjustment for chance in clustering performance evaluation : Analysis of the impact of the dataset size on the value of clustering measures for random assignments. 2.3.10.1.3. Mathematical formulation  If C is a ground truth class assignment and K the clustering, let us define and as: , the number of pairs of elements that are in the same set in C and in the same set in K , the number of pairs of elements that are in different sets in C and in different sets in K The raw (unadjusted) Rand index is then given by: \\\\[\\\\text{RI}  \\\\frac{a + b}{C_2^{n_{samples}}}\\\\] Where is the total number of possible pairs in the dataset (without ordering). However the RI score does not guarantee that random label assignments will get a value close to zero (esp. if the number of clusters is in the same order of magnitude as the number of samples). To counter this effect we can discount the expected RI of random labelings by defining the adjusted Rand index as follows: \\\\[\\\\text{ARI}  \\\\frac{\\\\text{RI} - E[\\\\text{RI}]}{\\\\max(\\\\text{RI}) - E[\\\\text{RI}]}\\\\] References Comparing Partitions L. Hubert and P. Arabie, Journal of Classification 1985 Wikipedia entry for the adjusted Rand index 2.3.10.2. Mutual Information based scores  Given the knowledge of the ground truth class assignments and our clustering algorithm assignments of the same samples , the Mutual Information is a function that measures the agreement of the two assignments, ignoring permutations. Two different normalized versions of this measure are available, Normalized Mutual Information (NMI) and Adjusted Mutual Information (AMI) . NMI is often used in the literature, while AMI was proposed more recently and is normalized against chance : One can permute 0 and 1 in the predicted labels, rename 2 to 3 and get the same score: All, mutual_info_score , adjusted_mutual_info_score and normalized_mutual_info_score are symmetric: swapping the argument does not change the score. Thus they can be used as a consensus measure : Perfect labeling is scored 1.0: This is not true for , which is therefore harder to judge: Bad (e.g. independent labelings) have non-positive scores: 2.3.10.2.1. Advantages  Random (uniform) label assignments have a AMI score close to 0.0 for any value of and (which is not the case for raw Mutual Information or the V-measure for instance). Upper bound of 1 : Values close to zero indicate two label assignments that are largely independent, while values close to one indicate significant agreement. Further, an AMI of exactly 1 indicates that the two label assignments are equal (with or without permutation). 2.3.10.2.2. Drawbacks  Contrary to inertia, MI-based measures require the knowledge of the ground truth classes while almost never available in practice or requires manual assignment by human annotators (as in the supervised learning setting). However MI-based measures can also be useful in purely unsupervised setting as a building block for a Consensus Index that can be used for clustering model selection. NMI and MI are not adjusted against chance. Examples: Adjustment for chance in clustering performance evaluation : Analysis of the impact of the dataset size on the value of clustering measures for random assignments. This example also includes the Adjusted Rand Index. 2.3.10.2.3. Mathematical formulation  Assume two label assignments (of the same N objects), and . Their entropy is the amount of uncertainty for a partition set, defined by: \\\\[H(U)  - \\\\sum_{i1}^{|U|}P(i)\\\\log(P(i))\\\\] where is the probability that an object picked at random from falls into class . Likewise for : \\\\[H(V)  - \\\\sum_{j1}^{|V|}P'(j)\\\\log(P'(j))\\\\] With . The mutual information (MI) between and is calculated by: \\\\[\\\\text{MI}(U, V)  \\\\sum_{i1}^{|U|}\\\\sum_{j1}^{|V|}P(i, j)\\\\log\\\\left(\\\\frac{P(i,j)}{P(i)P'(j)}\\\\right)\\\\] where is the probability that an object picked at random falls into both classes and . It also can be expressed in set cardinality formulation: \\\\[\\\\text{MI}(U, V)  \\\\sum_{i1}^{|U|} \\\\sum_{j1}^{|V|} \\\\frac{|U_i \\\\cap V_j|}{N}\\\\log\\\\left(\\\\frac{N|U_i \\\\cap V_j|}{|U_i||V_j|}\\\\right)\\\\] The normalized mutual information is defined as \\\\[\\\\text{NMI}(U, V)  \\\\frac{\\\\text{MI}(U, V)}{\\\\text{mean}(H(U), H(V))}\\\\] This value of the mutual information and also the normalized variant is not adjusted for chance and will tend to increase as the number of different labels (clusters) increases, regardless of the actual amount of “mutual information” between the label assignments. The expected value for the mutual information can be calculated using the following equation [VEB2009] . In this equation, (the number of elements in ) and (the number of elements in ). \\\\[E[\\\\text{MI}(U,V)]\\\\sum_{i1}^{|U|} \\\\sum_{j1}^{|V|} \\\\sum_{n_{ij}(a_i+b_j-N)^+ }^{\\\\min(a_i, b_j)} \\\\frac{n_{ij}}{N}\\\\log \\\\left( \\\\frac{ N.n_{ij}}{a_i b_j}\\\\right) \\\\frac{a_i!b_j!(N-a_i)!(N-b_j)!}{N!n_{ij}!(a_i-n_{ij})!(b_j-n_{ij})! (N-a_i-b_j+n_{ij})!}\\\\] Using the expected value, the adjusted mutual information can then be calculated using a similar form to that of the adjusted Rand index: \\\\[\\\\text{AMI}  \\\\frac{\\\\text{MI} - E[\\\\text{MI}]}{\\\\text{mean}(H(U), H(V)) - E[\\\\text{MI}]}\\\\] For normalized mutual information and adjusted mutual information, the normalizing value is typically some generalized mean of the entropies of each clustering. Various generalized means exist, and no firm rules exist for preferring one over the others. The decision is largely a field-by-field basis; for instance, in community detection, the arithmetic mean is most common. Each normalizing method provides “qualitatively similar behaviours” [YAT2016] . In our implementation, this is controlled by the parameter. Vinh et al. (2010) named variants of NMI and AMI by their averaging method [VEB2010] . Their ‘sqrt’ and ‘sum’ averages are the geometric and arithmetic means; we use these more broadly common names. References Strehl, Alexander, and Joydeep Ghosh (2002). “Cluster ensembles – a knowledge reuse framework for combining multiple partitions”. Journal of Machine Learning Research 3: 583–617. doi:10.1162/153244303321897735 . Wikipedia entry for the (normalized) Mutual Information Wikipedia entry for the Adjusted Mutual Information VEB2009 Vinh, Epps, and Bailey, (2009). “Information theoretic measures for clusterings comparison”. Proceedings of the 26th Annual International Conference on Machine Learning - ICML ‘09. doi:10.1145/1553374.1553511 . ISBN 9781605585161. VEB2010 Vinh, Epps, and Bailey, (2010). “Information Theoretic Measures for Clusterings Comparison: Variants, Properties, Normalization and Correction for Chance”. JMLR < http://jmlr.csail.mit.edu/papers/volume11/vinh10a/vinh10a.pdf > YAT2016 Yang, Algesheimer, and Tessone, (2016). “A comparative analysis of community detection algorithms on artificial networks”. Scientific Reports 6: 30750. doi:10.1038/srep30750 . 2.3.10.3. Homogeneity, completeness and V-measure  Given the knowledge of the ground truth class assignments of the samples, it is possible to define some intuitive metric using conditional entropy analysis. In particular Rosenberg and Hirschberg (2007) define the following two desirable objectives for any cluster assignment: homogeneity : each cluster contains only members of a single class. completeness : all members of a given class are assigned to the same cluster. We can turn those concept as scores homogeneity_score and completeness_score . Both are bounded below by 0.0 and above by 1.0 (higher is better): Their harmonic mean called V-measure is computed by v_measure_score : This function’s formula is as follows: \\\\[v  \\\\frac{(1 + \\\\beta) \\\\times \\\\text{homogeneity} \\\\times \\\\text{completeness}}{(\\\\beta \\\\times \\\\text{homogeneity} + \\\\text{completeness})}\\\\] defaults to a value of 1.0, but for using a value less than 1 for beta: more weight will be attributed to homogeneity, and using a value greater than 1: more weight will be attributed to completeness. The V-measure is actually equivalent to the mutual information (NMI) discussed above, with the aggregation function being the arithmetic mean [B2011] . Homogeneity, completeness and V-measure can be computed at once using homogeneity_completeness_v_measure as follows: The following clustering assignment is slightly better, since it is homogeneous but not complete: Note v_measure_score is symmetric : it can be used to evaluate the agreement of two independent assignments on the same dataset. This is not the case for completeness_score and homogeneity_score : both are bound by the relationship: 2.3.10.3.1. Advantages  Bounded scores : 0.0 is as bad as it can be, 1.0 is a perfect score. Intuitive interpretation: clustering with bad V-measure can be qualitatively analyzed in terms of homogeneity and completeness to better feel what ‘kind’ of mistakes is done by the assignment. No assumption is made on the cluster structure : can be used to compare clustering algorithms such as k-means which assumes isotropic blob shapes with results of spectral clustering algorithms which can find cluster with “folded” shapes. 2.3.10.3.2. Drawbacks  The previously introduced metrics are not normalized with regards to random labeling : this means that depending on the number of samples, clusters and ground truth classes, a completely random labeling will not always yield the same values for homogeneity, completeness and hence v-measure. In particular random labeling won’t yield zero scores especially when the number of clusters is large . This problem can safely be ignored when the number of samples is more than a thousand and the number of clusters is less than 10. For smaller sample sizes or larger number of clusters it is safer to use an adjusted index such as the Adjusted Rand Index (ARI) . These metrics require the knowledge of the ground truth classes while almost never available in practice or requires manual assignment by human annotators (as in the supervised learning setting). Examples: Adjustment for chance in clustering performance evaluation : Analysis of the impact of the dataset size on the value of clustering measures for random assignments. 2.3.10.3.3. Mathematical formulation  Homogeneity and completeness scores are formally given by: \\\\[h  1 - \\\\frac{H(C|K)}{H(C)}\\\\] \\\\[c  1 - \\\\frac{H(K|C)}{H(K)}\\\\] where is the conditional entropy of the classes given the cluster assignments and is given by: \\\\[H(C|K)  - \\\\sum_{c1}^{|C|} \\\\sum_{k1}^{|K|} \\\\frac{n_{c,k}}{n} \\\\cdot \\\\log\\\\left(\\\\frac{n_{c,k}}{n_k}\\\\right)\\\\] and is the entropy of the classes and is given by: \\\\[H(C)  - \\\\sum_{c1}^{|C|} \\\\frac{n_c}{n} \\\\cdot \\\\log\\\\left(\\\\frac{n_c}{n}\\\\right)\\\\] with the total number of samples, and the number of samples respectively belonging to class and cluster , and finally the number of samples from class assigned to cluster . The conditional entropy of clusters given class and the entropy of clusters are defined in a symmetric manner. Rosenberg and Hirschberg further define V-measure as the harmonic mean of homogeneity and completeness : \\\\[v  2 \\\\cdot \\\\frac{h \\\\cdot c}{h + c}\\\\] References V-Measure: A conditional entropy-based external cluster evaluation measure Andrew Rosenberg and Julia Hirschberg, 2007 B2011 Identication and Characterization of Events in Social Media , Hila Becker, PhD Thesis. 2.3.10.4. Fowlkes-Mallows scores  The Fowlkes-Mallows index ( sklearn.metrics.fowlkes_mallows_score ) can be used when the ground truth class assignments of the samples is known. The Fowlkes-Mallows score FMI is defined as the geometric mean of the pairwise precision and recall: \\\\[\\\\text{FMI}  \\\\frac{\\\\text{TP}}{\\\\sqrt{(\\\\text{TP} + \\\\text{FP}) (\\\\text{TP} + \\\\text{FN})}}\\\\] Where is the number of True Positive (i.e. the number of pair of points that belong to the same clusters in both the true labels and the predicted labels), is the number of False Positive (i.e. the number of pair of points that belong to the same clusters in the true labels and not in the predicted labels) and is the number of False Negative (i.e the number of pair of points that belongs in the same clusters in the predicted labels and not in the true labels). The score ranges from 0 to 1. A high value indicates a good similarity between two clusters. One can permute 0 and 1 in the predicted labels, rename 2 to 3 and get the same score: Perfect labeling is scored 1.0: Bad (e.g. independent labelings) have zero scores: 2.3.10.4.1. Advantages  Random (uniform) label assignments have a FMI score close to 0.0 for any value of and (which is not the case for raw Mutual Information or the V-measure for instance). Upper-bounded at 1 : Values close to zero indicate two label assignments that are largely independent, while values close to one indicate significant agreement. Further, values of exactly 0 indicate purely independent label assignments and a FMI of exactly 1 indicates that the two label assignments are equal (with or without permutation). No assumption is made on the cluster structure : can be used to compare clustering algorithms such as k-means which assumes isotropic blob shapes with results of spectral clustering algorithms which can find cluster with “folded” shapes. 2.3.10.4.2. Drawbacks  Contrary to inertia, FMI-based measures require the knowledge of the ground truth classes while almost never available in practice or requires manual assignment by human annotators (as in the supervised learning setting). References E. B. Fowkles and C. L. Mallows, 1983. “A method for comparing two hierarchical clusterings”. Journal of the American Statistical Association. http://wildfire.stat.ucla.edu/pdflibrary/fowlkes.pdf Wikipedia entry for the Fowlkes-Mallows Index 2.3.10.5. Silhouette Coefficient  If the ground truth labels are not known, evaluation must be performed using the model itself. The Silhouette Coefficient ( sklearn.metrics.silhouette_score ) is an example of such an evaluation, where a higher Silhouette Coefficient score relates to a model with better defined clusters. The Silhouette Coefficient is defined for each sample and is composed of two scores: a : The mean distance between a sample and all other points in the same class. b : The mean distance between a sample and all other points in the next nearest cluster . The Silhouette Coefficient s for a single sample is then given as: \\\\[s  \\\\frac{b - a}{max(a, b)}\\\\] The Silhouette Coefficient for a set of samples is given as the mean of the Silhouette Coefficient for each sample. In normal usage, the Silhouette Coefficient is applied to the results of a cluster analysis. References Peter J. Rousseeuw (1987). “Silhouettes: a Graphical Aid to the Interpretation and Validation of Cluster Analysis”. Computational and Applied Mathematics 20: 53–65. doi:10.1016/0377-0427(87)90125-7 . 2.3.10.5.1. Advantages  The score is bounded between -1 for incorrect clustering and +1 for highly dense clustering. Scores around zero indicate overlapping clusters. The score is higher when clusters are dense and well separated, which relates to a standard concept of a cluster. 2.3.10.5.2. Drawbacks  The Silhouette Coefficient is generally higher for convex clusters than other concepts of clusters, such as density based clusters like those obtained through DBSCAN. Examples: Selecting the number of clusters with silhouette analysis on KMeans clustering : In this example the silhouette analysis is used to choose an optimal value for n_clusters. 2.3.10.6. Calinski-Harabasz Index  If the ground truth labels are not known, the Calinski-Harabasz index ( sklearn.metrics.calinski_harabasz_score ) - also known as the Variance Ratio Criterion - can be used to evaluate the model, where a higher Calinski-Harabasz score relates to a model with better defined clusters. The index is the ratio of the sum of between-clusters dispersion and of inter-cluster dispersion for all clusters (where dispersion is defined as the sum of distances squared): In normal usage, the Calinski-Harabasz index is applied to the results of a cluster analysis: 2.3.10.6.1. Advantages  The score is higher when clusters are dense and well separated, which relates to a standard concept of a cluster. The score is fast to compute. 2.3.10.6.2. Drawbacks  The Calinski-Harabasz index is generally higher for convex clusters than other concepts of clusters, such as density based clusters like those obtained through DBSCAN. 2.3.10.6.3. Mathematical formulation  For a set of data of size which has been clustered into clusters, the Calinski-Harabasz score is defined as the ratio of the between-clusters dispersion mean and the within-cluster dispersion: \\\\[s  \\\\frac{\\\\mathrm{tr}(B_k)}{\\\\mathrm{tr}(W_k)} \\\\times \\\\frac{n_E - k}{k - 1}\\\\] where is trace of the between group dispersion matrix and is the trace of the within-cluster dispersion matrix defined by: \\\\[W_k  \\\\sum_{q1}^k \\\\sum_{x \\\\in C_q} (x - c_q) (x - c_q)^T\\\\] \\\\[B_k  \\\\sum_{q1}^k n_q (c_q - c_E) (c_q - c_E)^T\\\\] with the set of points in cluster , the center of cluster , the center of , and the number of points in cluster . References Caliński, T., & Harabasz, J. (1974). “A Dendrite Method for Cluster Analysis” . Communications in Statistics-theory and Methods 3: 1-27. doi:10.1080/03610927408827101 . 2.3.10.7. Davies-Bouldin Index  If the ground truth labels are not known, the Davies-Bouldin index ( sklearn.metrics.davies_bouldin_score ) can be used to evaluate the model, where a lower Davies-Bouldin index relates to a model with better separation between the clusters. This index signifies the average ‘similarity’ between clusters, where the similarity is a measure that compares the distance between clusters with the size of the clusters themselves. Zero is the lowest possible score. Values closer to zero indicate a better partition. In normal usage, the Davies-Bouldin index is applied to the results of a cluster analysis as follows: 2.3.10.7.1. Advantages  The computation of Davies-Bouldin is simpler than that of Silhouette scores. The index is computed only quantities and features inherent to the dataset. 2.3.10.7.2. Drawbacks  The Davies-Boulding index is generally higher for convex clusters than other concepts of clusters, such as density based clusters like those obtained from DBSCAN. The usage of centroid distance limits the distance metric to Euclidean space. 2.3.10.7.3. Mathematical formulation  The index is defined as the average similarity between each cluster for and its most similar one . In the context of this index, similarity is defined as a measure that trades off: , the average distance between each point of cluster and the centroid of that cluster – also know as cluster diameter. , the distance between cluster centroids and . A simple choice to construct so that it is nonnegative and symmetric is: \\\\[R_{ij}  \\\\frac{s_i + s_j}{d_{ij}}\\\\] Then the Davies-Bouldin index is defined as: \\\\[DB  \\\\frac{1}{k} \\\\sum_{i1}^k \\\\max_{i \\\\neq j} R_{ij}\\\\] References Davies, David L.; Bouldin, Donald W. (1979). “A Cluster Separation Measure” IEEE Transactions on Pattern Analysis and Machine Intelligence. PAMI-1 (2): 224-227. doi:10.1109/TPAMI.1979.4766909 . Halkidi, Maria; Batistakis, Yannis; Vazirgiannis, Michalis (2001). “On Clustering Validation Techniques” Journal of Intelligent Information Systems, 17(2-3), 107-145. doi:10.1023/A:1012801612483 . Wikipedia entry for Davies-Bouldin index . 2.3.10.8. Contingency Matrix  Contingency matrix ( sklearn.metrics.cluster.contingency_matrix ) reports the intersection cardinality for every true/predicted cluster pair. The contingency matrix provides sufficient statistics for all clustering metrics where the samples are independent and identically distributed and one doesn’t need to account for some instances not being clustered. Here is an example: The first row of output array indicates that there are three samples whose true cluster is “a”. Of them, two are in predicted cluster 0, one is in 1, and none is in 2. And the second row indicates that there are three samples whose true cluster is “b”. Of them, none is in predicted cluster 0, one is in 1 and two are in 2. A confusion matrix for classification is a square contingency matrix where the order of rows and columns correspond to a list of classes. 2.3.10.8.1. Advantages  Allows to examine the spread of each true cluster across predicted clusters and vice versa. The contingency table calculated is typically utilized in the calculation of a similarity statistic (like the others listed in this document) between the two clusterings. 2.3.10.8.2. Drawbacks  Contingency matrix is easy to interpret for a small number of clusters, but becomes very hard to interpret for a large number of clusters. It doesn’t give a single metric to use as an objective for clustering optimisation. References Wikipedia entry for contingency matrix \",\n",
       " 'sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.1. Overview of clustering methods  A comparison of the clustering algorithms in scikit-learn  Method name Parameters Scalability Usecase Geometry (metric used) K-Means number of clusters Very large , medium with MiniBatch code General-purpose, even cluster size, flat geometry, not too many clusters Distances between points Affinity propagation damping, sample preference Not scalable with n_samples Many clusters, uneven cluster size, non-flat geometry Graph distance (e.g. nearest-neighbor graph) Mean-shift bandwidth Not scalable with Many clusters, uneven cluster size, non-flat geometry Distances between points Spectral clustering number of clusters Medium , small Few clusters, even cluster size, non-flat geometry Graph distance (e.g. nearest-neighbor graph) Ward hierarchical clustering number of clusters or distance threshold Large and Many clusters, possibly connectivity constraints Distances between points Agglomerative clustering number of clusters or distance threshold, linkage type, distance Large and Many clusters, possibly connectivity constraints, non Euclidean distances Any pairwise distance DBSCAN neighborhood size Very large , medium Non-flat geometry, uneven cluster sizes Distances between nearest points OPTICS minimum cluster membership Very large , large Non-flat geometry, uneven cluster sizes, variable cluster density Distances between points Gaussian mixtures many Not scalable Flat geometry, good for density estimation Mahalanobis distances to centers Birch branching factor, threshold, optional global clusterer. Large and Large dataset, outlier removal, data reduction. Euclidean distance between points Non-flat geometry clustering is useful when the clusters have a specific shape, i.e. a non-flat manifold, and the standard euclidean distance is not the right metric. This case arises in the two top rows of the figure above. Gaussian mixture models, useful for clustering, are described in another chapter of the documentation dedicated to mixture models. KMeans can be seen as a special case of Gaussian mixture model with equal covariance per component. ',\n",
       " 'sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.2. K-means  The KMeans algorithm clusters data by trying to separate samples in n groups of equal variance, minimizing a criterion known as the inertia or within-cluster sum-of-squares (see below). This algorithm requires the number of clusters to be specified. It scales well to large number of samples and has been used across a large range of application areas in many different fields. The k-means algorithm divides a set of samples into disjoint clusters , each described by the mean of the samples in the cluster. The means are commonly called the cluster “centroids”; note that they are not, in general, points from , although they live in the same space. The K-means algorithm aims to choose centroids that minimise the inertia , or within-cluster sum-of-squares criterion : \\\\[\\\\sum_{i0}^{n}\\\\min_{\\\\mu_j \\\\in C}(||x_i - \\\\mu_j||^2)\\\\] Inertia can be recognized as a measure of how internally coherent clusters are. It suffers from various drawbacks: Inertia makes the assumption that clusters are convex and isotropic, which is not always the case. It responds poorly to elongated clusters, or manifolds with irregular shapes. Inertia is not a normalized metric: we just know that lower values are better and zero is optimal. But in very high-dimensional spaces, Euclidean distances tend to become inflated (this is an instance of the so-called “curse of dimensionality”). Running a dimensionality reduction algorithm such as Principal component analysis (PCA) prior to k-means clustering can alleviate this problem and speed up the computations. K-means is often referred to as Lloyd’s algorithm. In basic terms, the algorithm has three steps. The first step chooses the initial centroids, with the most basic method being to choose samples from the dataset . After initialization, K-means consists of looping between the two other steps. The first step assigns each sample to its nearest centroid. The second step creates new centroids by taking the mean value of all of the samples assigned to each previous centroid. The difference between the old and the new centroids are computed and the algorithm repeats these last two steps until this value is less than a threshold. In other words, it repeats until the centroids do not move significantly. K-means is equivalent to the expectation-maximization algorithm with a small, all-equal, diagonal covariance matrix. The algorithm can also be understood through the concept of Voronoi diagrams . First the Voronoi diagram of the points is calculated using the current centroids. Each segment in the Voronoi diagram becomes a separate cluster. Secondly, the centroids are updated to the mean of each segment. The algorithm then repeats this until a stopping criterion is fulfilled. Usually, the algorithm stops when the relative decrease in the objective function between iterations is less than the given tolerance value. This is not the case in this implementation: iteration stops when centroids move less than the tolerance. Given enough time, K-means will always converge, however this may be to a local minimum. This is highly dependent on the initialization of the centroids. As a result, the computation is often done several times, with different initializations of the centroids. One method to help address this issue is the k-means++ initialization scheme, which has been implemented in scikit-learn (use the parameter). This initializes the centroids to be (generally) distant from each other, leading to provably better results than random initialization, as shown in the reference. The algorithm supports sample weights, which can be given by a parameter . This allows to assign more weight to some samples when computing cluster centers and values of inertia. For example, assigning a weight of 2 to a sample is equivalent to adding a duplicate of that sample to the dataset . K-means can be used for vector quantization. This is achieved using the transform method of a trained model of KMeans . 2.3.2.1. Low-level parallelism  KMeans benefits from OpenMP based parallelism through Cython. Small chunks of data (256 samples) are processed in parallel, which in addition yields a low memory footprint. For more details on how to control the number of threads, please refer to our Parallelism notes. Examples: Demonstration of k-means assumptions : Demonstrating when k-means performs intuitively and when it does not A demo of K-Means clustering on the handwritten digits data : Clustering handwritten digits References: “k-means++: The advantages of careful seeding” Arthur, David, and Sergei Vassilvitskii, Proceedings of the eighteenth annual ACM-SIAM symposium on Discrete algorithms , Society for Industrial and Applied Mathematics (2007) 2.3.2.2. Mini Batch K-Means  The MiniBatchKMeans is a variant of the KMeans algorithm which uses mini-batches to reduce the computation time, while still attempting to optimise the same objective function. Mini-batches are subsets of the input data, randomly sampled in each training iteration. These mini-batches drastically reduce the amount of computation required to converge to a local solution. In contrast to other algorithms that reduce the convergence time of k-means, mini-batch k-means produces results that are generally only slightly worse than the standard algorithm. The algorithm iterates between two major steps, similar to vanilla k-means. In the first step, samples are drawn randomly from the dataset, to form a mini-batch. These are then assigned to the nearest centroid. In the second step, the centroids are updated. In contrast to k-means, this is done on a per-sample basis. For each sample in the mini-batch, the assigned centroid is updated by taking the streaming average of the sample and all previous samples assigned to that centroid. This has the effect of decreasing the rate of change for a centroid over time. These steps are performed until convergence or a predetermined number of iterations is reached. MiniBatchKMeans converges faster than KMeans , but the quality of the results is reduced. In practice this difference in quality can be quite small, as shown in the example and cited reference. Examples: Comparison of the K-Means and MiniBatchKMeans clustering algorithms : Comparison of KMeans and MiniBatchKMeans Clustering text documents using k-means : Document clustering using sparse MiniBatchKMeans Online learning of a dictionary of parts of faces References: “Web Scale K-Means clustering” D. Sculley, Proceedings of the 19th international conference on World wide web (2010) ',\n",
       " \"sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.3. Affinity Propagation  AffinityPropagation creates clusters by sending messages between pairs of samples until convergence. A dataset is then described using a small number of exemplars, which are identified as those most representative of other samples. The messages sent between pairs represent the suitability for one sample to be the exemplar of the other, which is updated in response to the values from other pairs. This updating happens iteratively until convergence, at which point the final exemplars are chosen, and hence the final clustering is given. Affinity Propagation can be interesting as it chooses the number of clusters based on the data provided. For this purpose, the two important parameters are the preference , which controls how many exemplars are used, and the damping factor which damps the responsibility and availability messages to avoid numerical oscillations when updating these messages. The main drawback of Affinity Propagation is its complexity. The algorithm has a time complexity of the order , where is the number of samples and is the number of iterations until convergence. Further, the memory complexity is of the order if a dense similarity matrix is used, but reducible if a sparse similarity matrix is used. This makes Affinity Propagation most appropriate for small to medium sized datasets. Examples: Demo of affinity propagation clustering algorithm : Affinity Propagation on a synthetic 2D datasets with 3 classes. Visualizing the stock market structure Affinity Propagation on Financial time series to find groups of companies Algorithm description: The messages sent between points belong to one of two categories. The first is the responsibility , which is the accumulated evidence that sample should be the exemplar for sample . The second is the availability which is the accumulated evidence that sample should choose sample to be its exemplar, and considers the values for all other samples that should be an exemplar. In this way, exemplars are chosen by samples if they are (1) similar enough to many samples and (2) chosen by many samples to be representative of themselves. More formally, the responsibility of a sample to be the exemplar of sample is given by: \\\\[r(i, k) \\\\leftarrow s(i, k) - max [ a(i, k') + s(i, k') \\\\forall k' \\\\neq k ]\\\\] Where is the similarity between samples and . The availability of sample to be the exemplar of sample is given by: \\\\[a(i, k) \\\\leftarrow min [0, r(k, k) + \\\\sum_{i'~s.t.~i' \\\\notin \\\\{i, k\\\\}}{r(i', k)}]\\\\] To begin with, all values for and are set to zero, and the calculation of each iterates until convergence. As discussed above, in order to avoid numerical oscillations when updating the messages, the damping factor is introduced to iteration process: \\\\[r_{t+1}(i, k)  \\\\lambda\\\\cdot r_{t}(i, k) + (1-\\\\lambda)\\\\cdot r_{t+1}(i, k)\\\\] \\\\[a_{t+1}(i, k)  \\\\lambda\\\\cdot a_{t}(i, k) + (1-\\\\lambda)\\\\cdot a_{t+1}(i, k)\\\\] where indicates the iteration times. \",\n",
       " 'sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.4. Mean Shift  MeanShift clustering aims to discover blobs in a smooth density of samples. It is a centroid based algorithm, which works by updating candidates for centroids to be the mean of the points within a given region. These candidates are then filtered in a post-processing stage to eliminate near-duplicates to form the final set of centroids. Given a candidate centroid for iteration , the candidate is updated according to the following equation: \\\\[x_i^{t+1}  m(x_i^t)\\\\] Where is the neighborhood of samples within a given distance around and is the mean shift vector that is computed for each centroid that points towards a region of the maximum increase in the density of points. This is computed using the following equation, effectively updating a centroid to be the mean of the samples within its neighborhood: \\\\[m(x_i)  \\\\frac{\\\\sum_{x_j \\\\in N(x_i)}K(x_j - x_i)x_j}{\\\\sum_{x_j \\\\in N(x_i)}K(x_j - x_i)}\\\\] The algorithm automatically sets the number of clusters, instead of relying on a parameter , which dictates the size of the region to search through. This parameter can be set manually, but can be estimated using the provided function, which is called if the bandwidth is not set. The algorithm is not highly scalable, as it requires multiple nearest neighbor searches during the execution of the algorithm. The algorithm is guaranteed to converge, however the algorithm will stop iterating when the change in centroids is small. Labelling a new sample is performed by finding the nearest centroid for a given sample. Examples: A demo of the mean-shift clustering algorithm : Mean Shift clustering on a synthetic 2D datasets with 3 classes. References: “Mean shift: A robust approach toward feature space analysis.” D. Comaniciu and P. Meer, IEEE Transactions on Pattern Analysis and Machine Intelligence (2002) ',\n",
       " 'sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.5. Spectral clustering  SpectralClustering performs a low-dimension embedding of the affinity matrix between samples, followed by clustering, e.g., by KMeans, of the components of the eigenvectors in the low dimensional space. It is especially computationally efficient if the affinity matrix is sparse and the solver is used for the eigenvalue problem (Note, the solver requires that the pyamg module is installed.) The present version of SpectralClustering requires the number of clusters to be specified in advance. It works well for a small number of clusters, but is not advised for many clusters. For two clusters, SpectralClustering solves a convex relaxation of the normalised cuts problem on the similarity graph: cutting the graph in two so that the weight of the edges cut is small compared to the weights of the edges inside each cluster. This criteria is especially interesting when working on images, where graph vertices are pixels, and weights of the edges of the similarity graph are computed using a function of a gradient of the image. Warning Transforming distance to well-behaved similarities Note that if the values of your similarity matrix are not well distributed, e.g. with negative values or with a distance matrix rather than a similarity, the spectral problem will be singular and the problem not solvable. In which case it is advised to apply a transformation to the entries of the matrix. For instance, in the case of a signed distance matrix, is common to apply a heat kernel: See the examples for such an application. Examples: Spectral clustering for image segmentation : Segmenting objects from a noisy background using spectral clustering. Segmenting the picture of greek coins in regions : Spectral clustering to split the image of coins in regions. 2.3.5.1. Different label assignment strategies  Different label assignment strategies can be used, corresponding to the parameter of SpectralClustering . strategy can match finer details, but can be unstable. In particular, unless you control the , it may not be reproducible from run-to-run, as it depends on random initialization. The alternative strategy is 100% reproducible, but tends to create parcels of fairly even and geometrical shape. 2.3.5.2. Spectral Clustering Graphs  Spectral Clustering can also be used to partition graphs via their spectral embeddings. In this case, the affinity matrix is the adjacency matrix of the graph, and SpectralClustering is initialized with : References: “A Tutorial on Spectral Clustering” Ulrike von Luxburg, 2007 “Normalized cuts and image segmentation” Jianbo Shi, Jitendra Malik, 2000 “A Random Walks View of Spectral Segmentation” Marina Meila, Jianbo Shi, 2001 “On Spectral Clustering: Analysis and an algorithm” Andrew Y. Ng, Michael I. Jordan, Yair Weiss, 2001 “Preconditioned Spectral Clustering for Stochastic Block Partition Streaming Graph Challenge” David Zhuzhunashvili, Andrew Knyazev ',\n",
       " 'sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.6. Hierarchical clustering  Hierarchical clustering is a general family of clustering algorithms that build nested clusters by merging or splitting them successively. This hierarchy of clusters is represented as a tree (or dendrogram). The root of the tree is the unique cluster that gathers all the samples, the leaves being the clusters with only one sample. See the Wikipedia page for more details. The AgglomerativeClustering object performs a hierarchical clustering using a bottom up approach: each observation starts in its own cluster, and clusters are successively merged together. The linkage criteria determines the metric used for the merge strategy: Ward minimizes the sum of squared differences within all clusters. It is a variance-minimizing approach and in this sense is similar to the k-means objective function but tackled with an agglomerative hierarchical approach. Maximum or complete linkage minimizes the maximum distance between observations of pairs of clusters. Average linkage minimizes the average of the distances between all observations of pairs of clusters. Single linkage minimizes the distance between the closest observations of pairs of clusters. AgglomerativeClustering can also scale to large number of samples when it is used jointly with a connectivity matrix, but is computationally expensive when no connectivity constraints are added between samples: it considers at each step all the possible merges. FeatureAgglomeration The FeatureAgglomeration uses agglomerative clustering to group together features that look very similar, thus decreasing the number of features. It is a dimensionality reduction tool, see Unsupervised dimensionality reduction . 2.3.6.1. Different linkage type: Ward, complete, average, and single linkage  AgglomerativeClustering supports Ward, single, average, and complete linkage strategies. Agglomerative cluster has a “rich get richer” behavior that leads to uneven cluster sizes. In this regard, single linkage is the worst strategy, and Ward gives the most regular sizes. However, the affinity (or distance used in clustering) cannot be varied with Ward, thus for non Euclidean metrics, average linkage is a good alternative. Single linkage, while not robust to noisy data, can be computed very efficiently and can therefore be useful to provide hierarchical clustering of larger datasets. Single linkage can also perform well on non-globular data. Examples: Various Agglomerative Clustering on a 2D embedding of digits : exploration of the different linkage strategies in a real dataset. 2.3.6.2. Visualization of cluster hierarchy  It’s possible to visualize the tree representing the hierarchical merging of clusters as a dendrogram. Visual inspection can often be useful for understanding the structure of the data, though more so in the case of small sample sizes. 2.3.6.3. Adding connectivity constraints  An interesting aspect of AgglomerativeClustering is that connectivity constraints can be added to this algorithm (only adjacent clusters can be merged together), through a connectivity matrix that defines for each sample the neighboring samples following a given structure of the data. For instance, in the swiss-roll example below, the connectivity constraints forbid the merging of points that are not adjacent on the swiss roll, and thus avoid forming clusters that extend across overlapping folds of the roll. These constraint are useful to impose a certain local structure, but they also make the algorithm faster, especially when the number of the samples is high. The connectivity constraints are imposed via an connectivity matrix: a scipy sparse matrix that has elements only at the intersection of a row and a column with indices of the dataset that should be connected. This matrix can be constructed from a-priori information: for instance, you may wish to cluster web pages by only merging pages with a link pointing from one to another. It can also be learned from the data, for instance using sklearn.neighbors.kneighbors_graph to restrict merging to nearest neighbors as in this example , or using sklearn.feature_extraction.image.grid_to_graph to enable only merging of neighboring pixels on an image, as in the coin example. Examples: A demo of structured Ward hierarchical clustering on an image of coins : Ward clustering to split the image of coins in regions. Hierarchical clustering: structured vs unstructured ward : Example of Ward algorithm on a swiss-roll, comparison of structured approaches versus unstructured approaches. Feature agglomeration vs. univariate selection : Example of dimensionality reduction with feature agglomeration based on Ward hierarchical clustering. Agglomerative clustering with and without structure Warning Connectivity constraints with single, average and complete linkage Connectivity constraints and single, complete or average linkage can enhance the ‘rich getting richer’ aspect of agglomerative clustering, particularly so if they are built with sklearn.neighbors.kneighbors_graph . In the limit of a small number of clusters, they tend to give a few macroscopically occupied clusters and almost empty ones. (see the discussion in Agglomerative clustering with and without structure ). Single linkage is the most brittle linkage option with regard to this issue. 2.3.6.4. Varying the metric  Single, average and complete linkage can be used with a variety of distances (or affinities), in particular Euclidean distance ( l2 ), Manhattan distance (or Cityblock, or l1 ), cosine distance, or any precomputed affinity matrix. l1 distance is often good for sparse features, or sparse noise: i.e. many of the features are zero, as in text mining using occurrences of rare words. cosine distance is interesting because it is invariant to global scalings of the signal. The guidelines for choosing a metric is to use one that maximizes the distance between samples in different classes, and minimizes that within each class. Examples: Agglomerative clustering with different metrics ',\n",
       " 'sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.7. DBSCAN  The DBSCAN algorithm views clusters as areas of high density separated by areas of low density. Due to this rather generic view, clusters found by DBSCAN can be any shape, as opposed to k-means which assumes that clusters are convex shaped. The central component to the DBSCAN is the concept of core samples , which are samples that are in areas of high density. A cluster is therefore a set of core samples, each close to each other (measured by some distance measure) and a set of non-core samples that are close to a core sample (but are not themselves core samples). There are two parameters to the algorithm, and , which define formally what we mean when we say dense . Higher or lower indicate higher density necessary to form a cluster. More formally, we define a core sample as being a sample in the dataset such that there exist other samples within a distance of , which are defined as neighbors of the core sample. This tells us that the core sample is in a dense area of the vector space. A cluster is a set of core samples that can be built by recursively taking a core sample, finding all of its neighbors that are core samples, finding all of their neighbors that are core samples, and so on. A cluster also has a set of non-core samples, which are samples that are neighbors of a core sample in the cluster but are not themselves core samples. Intuitively, these samples are on the fringes of a cluster. Any core sample is part of a cluster, by definition. Any sample that is not a core sample, and is at least in distance from any core sample, is considered an outlier by the algorithm. While the parameter primarily controls how tolerant the algorithm is towards noise (on noisy and large data sets it may be desirable to increase this parameter), the parameter is crucial to choose appropriately for the data set and distance function and usually cannot be left at the default value. It controls the local neighborhood of the points. When chosen too small, most data will not be clustered at all (and labeled as for “noise”). When chosen too large, it causes close clusters to be merged into one cluster, and eventually the entire data set to be returned as a single cluster. Some heuristics for choosing this parameter have been discussed in the literature, for example based on a knee in the nearest neighbor distances plot (as discussed in the references below). In the figure below, the color indicates cluster membership, with large circles indicating core samples found by the algorithm. Smaller circles are non-core samples that are still part of a cluster. Moreover, the outliers are indicated by black points below. Examples: Demo of DBSCAN clustering algorithm Implementation The DBSCAN algorithm is deterministic, always generating the same clusters when given the same data in the same order. However, the results can differ when data is provided in a different order. First, even though the core samples will always be assigned to the same clusters, the labels of those clusters will depend on the order in which those samples are encountered in the data. Second and more importantly, the clusters to which non-core samples are assigned can differ depending on the data order. This would happen when a non-core sample has a distance lower than to two core samples in different clusters. By the triangular inequality, those two core samples must be more distant than from each other, or they would be in the same cluster. The non-core sample is assigned to whichever cluster is generated first in a pass through the data, and so the results will depend on the data ordering. The current implementation uses ball trees and kd-trees to determine the neighborhood of points, which avoids calculating the full distance matrix (as was done in scikit-learn versions before 0.14). The possibility to use custom metrics is retained; for details, see NearestNeighbors . Memory consumption for large sample sizes This implementation is by default not memory efficient because it constructs a full pairwise similarity matrix in the case where kd-trees or ball-trees cannot be used (e.g., with sparse matrices). This matrix will consume n^2 floats. A couple of mechanisms for getting around this are: Use OPTICS clustering in conjunction with the method. OPTICS clustering also calculates the full pairwise matrix, but only keeps one row in memory at a time (memory complexity n). A sparse radius neighborhood graph (where missing entries are presumed to be out of eps) can be precomputed in a memory-efficient way and dbscan can be run over this with . See sklearn.neighbors.NearestNeighbors.radius_neighbors_graph . The dataset can be compressed, either by removing exact duplicates if these occur in your data, or by using BIRCH. Then you only have a relatively small number of representatives for a large number of points. You can then provide a when fitting DBSCAN. References: “A Density-Based Algorithm for Discovering Clusters in Large Spatial Databases with Noise” Ester, M., H. P. Kriegel, J. Sander, and X. Xu, In Proceedings of the 2nd International Conference on Knowledge Discovery and Data Mining, Portland, OR, AAAI Press, pp. 226–231. 1996 “DBSCAN revisited, revisited: why and how you should (still) use DBSCAN. Schubert, E., Sander, J., Ester, M., Kriegel, H. P., & Xu, X. (2017). In ACM Transactions on Database Systems (TODS), 42(3), 19. ',\n",
       " 'sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.8. OPTICS  The OPTICS algorithm shares many similarities with the DBSCAN algorithm, and can be considered a generalization of DBSCAN that relaxes the requirement from a single value to a value range. The key difference between DBSCAN and OPTICS is that the OPTICS algorithm builds a reachability graph, which assigns each sample both a distance, and a spot within the cluster attribute; these two attributes are assigned when the model is fitted, and are used to determine cluster membership. If OPTICS is run with the default value of inf set for , then DBSCAN style cluster extraction can be performed repeatedly in linear time for any given value using the method. Setting to a lower value will result in shorter run times, and can be thought of as the maximum neighborhood radius from each point to find other potential reachable points. The reachability distances generated by OPTICS allow for variable density extraction of clusters within a single data set. As shown in the above plot, combining reachability distances and data set produces a reachability plot , where point density is represented on the Y-axis, and points are ordered such that nearby points are adjacent. ‘Cutting’ the reachability plot at a single value produces DBSCAN like results; all points above the ‘cut’ are classified as noise, and each time that there is a break when reading from left to right signifies a new cluster. The default cluster extraction with OPTICS looks at the steep slopes within the graph to find clusters, and the user can define what counts as a steep slope using the parameter . There are also other possibilities for analysis on the graph itself, such as generating hierarchical representations of the data through reachability-plot dendrograms, and the hierarchy of clusters detected by the algorithm can be accessed through the parameter. The plot above has been color-coded so that cluster colors in planar space match the linear segment clusters of the reachability plot. Note that the blue and red clusters are adjacent in the reachability plot, and can be hierarchically represented as children of a larger parent cluster. Examples: Demo of OPTICS clustering algorithm Comparison with DBSCAN The results from OPTICS method and DBSCAN are very similar, but not always identical; specifically, labeling of periphery and noise points. This is in part because the first samples of each dense area processed by OPTICS have a large reachability value while being close to other points in their area, and will thus sometimes be marked as noise rather than periphery. This affects adjacent points when they are considered as candidates for being marked as either periphery or noise. Note that for any single value of , DBSCAN will tend to have a shorter run time than OPTICS; however, for repeated runs at varying values, a single run of OPTICS may require less cumulative runtime than DBSCAN. It is also important to note that OPTICS’ output is close to DBSCAN’s only if and are close. Computational Complexity Spatial indexing trees are used to avoid calculating the full distance matrix, and allow for efficient memory usage on large sets of samples. Different distance metrics can be supplied via the keyword. For large datasets, similar (but not identical) results can be obtained via HDBSCAN . The HDBSCAN implementation is multithreaded, and has better algorithmic runtime complexity than OPTICS, at the cost of worse memory scaling. For extremely large datasets that exhaust system memory using HDBSCAN, OPTICS will maintain n (as opposed to n^2 ) memory scaling; however, tuning of the parameter will likely need to be used to give a solution in a reasonable amount of wall time. References: “OPTICS: ordering points to identify the clustering structure.” Ankerst, Mihael, Markus M. Breunig, Hans-Peter Kriegel, and Jörg Sander. In ACM Sigmod Record, vol. 28, no. 2, pp. 49-60. ACM, 1999. ',\n",
       " 'sklearn_2_3_clustering\\n2.3. Clustering\\nmodules/clustering.html\\n 2.3.9. Birch  The Birch builds a tree called the Clustering Feature Tree (CFT) for the given data. The data is essentially lossy compressed to a set of Clustering Feature nodes (CF Nodes). The CF Nodes have a number of subclusters called Clustering Feature subclusters (CF Subclusters) and these CF Subclusters located in the non-terminal CF Nodes can have CF Nodes as children. The CF Subclusters hold the necessary information for clustering which prevents the need to hold the entire input data in memory. This information includes: Number of samples in a subcluster. Linear Sum - A n-dimensional vector holding the sum of all samples Squared Sum - Sum of the squared L2 norm of all samples. Centroids - To avoid recalculation linear sum / n_samples. Squared norm of the centroids. The Birch algorithm has two parameters, the threshold and the branching factor. The branching factor limits the number of subclusters in a node and the threshold limits the distance between the entering sample and the existing subclusters. This algorithm can be viewed as an instance or data reduction method, since it reduces the input data to a set of subclusters which are obtained directly from the leaves of the CFT. This reduced data can be further processed by feeding it into a global clusterer. This global clusterer can be set by . If is set to None, the subclusters from the leaves are directly read off, otherwise a global clustering step labels these subclusters into global clusters (labels) and the samples are mapped to the global label of the nearest subcluster. Algorithm description: A new sample is inserted into the root of the CF Tree which is a CF Node. It is then merged with the subcluster of the root, that has the smallest radius after merging, constrained by the threshold and branching factor conditions. If the subcluster has any child node, then this is done repeatedly till it reaches a leaf. After finding the nearest subcluster in the leaf, the properties of this subcluster and the parent subclusters are recursively updated. If the radius of the subcluster obtained by merging the new sample and the nearest subcluster is greater than the square of the threshold and if the number of subclusters is greater than the branching factor, then a space is temporarily allocated to this new sample. The two farthest subclusters are taken and the subclusters are divided into two groups on the basis of the distance between these subclusters. If this split node has a parent subcluster and there is room for a new subcluster, then the parent is split into two. If there is no room, then this node is again split into two and the process is continued recursively, till it reaches the root. Birch or MiniBatchKMeans? Birch does not scale very well to high dimensional data. As a rule of thumb if is greater than twenty, it is generally better to use MiniBatchKMeans. If the number of instances of data needs to be reduced, or if one wants a large number of subclusters either as a preprocessing step or otherwise, Birch is more useful than MiniBatchKMeans. How to use partial_fit? To avoid the computation of global clustering, for every call of the user is advised To set initially Train all data by multiple calls to partial_fit. Set to a required value using . Call finally with no arguments, i.e. which performs the global clustering. References: Tian Zhang, Raghu Ramakrishnan, Maron Livny BIRCH: An efficient data clustering method for large databases. https://www.cs.sfu.ca/CourseCentral/459/han/papers/zhang96.pdf Roberto Perdisci JBirch - Java implementation of BIRCH clustering algorithm https://code.google.com/archive/p/jbirch ',\n",
       " 'sklearn_2_4_biclustering\\n2.4. Biclustering\\nmodules/biclustering.html\\n 2.4.1. Spectral Co-Clustering  The SpectralCoclustering algorithm finds biclusters with values higher than those in the corresponding other rows and columns. Each row and each column belongs to exactly one bicluster, so rearranging the rows and columns to make partitions contiguous reveals these high values along the diagonal: Note The algorithm treats the input data matrix as a bipartite graph: the rows and columns of the matrix correspond to the two sets of vertices, and each entry corresponds to an edge between a row and a column. The algorithm approximates the normalized cut of this graph to find heavy subgraphs. 2.4.1.1. Mathematical formulation  An approximate solution to the optimal normalized cut may be found via the generalized eigenvalue decomposition of the Laplacian of the graph. Usually this would mean working directly with the Laplacian matrix. If the original data matrix has shape , the Laplacian matrix for the corresponding bipartite graph has shape . However, in this case it is possible to work directly with , which is smaller and more efficient. The input matrix is preprocessed as follows: \\\\[A_n  R^{-1/2} A C^{-1/2}\\\\] Where is the diagonal matrix with entry equal to and is the diagonal matrix with entry equal to . The singular value decomposition, , provides the partitions of the rows and columns of . A subset of the left singular vectors gives the row partitions, and a subset of the right singular vectors gives the column partitions. The singular vectors, starting from the second, provide the desired partitioning information. They are used to form the matrix : \\\\[\\\\begin{split}Z  \\\\begin{bmatrix} R^{-1/2} U \\\\\\\\\\\\\\\\ C^{-1/2} V \\\\end{bmatrix}\\\\end{split}\\\\] where the columns of are , and similarly for . Then the rows of are clustered using k-means . The first labels provide the row partitioning, and the remaining labels provide the column partitioning. Examples: A demo of the Spectral Co-Clustering algorithm : A simple example showing how to generate a data matrix with biclusters and apply this method to it. Biclustering documents with the Spectral Co-clustering algorithm : An example of finding biclusters in the twenty newsgroup dataset. References: Dhillon, Inderjit S, 2001. Co-clustering documents and words using bipartite spectral graph partitioning . ',\n",
       " 'sklearn_2_4_biclustering\\n2.4. Biclustering\\nmodules/biclustering.html\\n 2.4.2. Spectral Biclustering  The SpectralBiclustering algorithm assumes that the input data matrix has a hidden checkerboard structure. The rows and columns of a matrix with this structure may be partitioned so that the entries of any bicluster in the Cartesian product of row clusters and column clusters are approximately constant. For instance, if there are two row partitions and three column partitions, each row will belong to three biclusters, and each column will belong to two biclusters. The algorithm partitions the rows and columns of a matrix so that a corresponding blockwise-constant checkerboard matrix provides a good approximation to the original matrix. 2.4.2.1. Mathematical formulation  The input matrix is first normalized to make the checkerboard pattern more obvious. There are three possible methods: Independent row and column normalization , as in Spectral Co-Clustering. This method makes the rows sum to a constant and the columns sum to a different constant. Bistochastization : repeated row and column normalization until convergence. This method makes both rows and columns sum to the same constant. Log normalization : the log of the data matrix is computed: . Then the column mean , row mean , and overall mean of are computed. The final matrix is computed according to the formula \\\\[K_{ij}  L_{ij} - \\\\overline{L_{i \\\\cdot}} - \\\\overline{L_{\\\\cdot j}} + \\\\overline{L_{\\\\cdot \\\\cdot}}\\\\] After normalizing, the first few singular vectors are computed, just as in the Spectral Co-Clustering algorithm. If log normalization was used, all the singular vectors are meaningful. However, if independent normalization or bistochastization were used, the first singular vectors, and . are discarded. From now on, the “first” singular vectors refers to and except in the case of log normalization. Given these singular vectors, they are ranked according to which can be best approximated by a piecewise-constant vector. The approximations for each vector are found using one-dimensional k-means and scored using the Euclidean distance. Some subset of the best left and right singular vector are selected. Next, the data is projected to this best subset of singular vectors and clustered. For instance, if singular vectors were calculated, the best are found as described, where . Let be the matrix with columns the best left singular vectors, and similarly for the right. To partition the rows, the rows of are projected to a dimensional space: . Treating the rows of this matrix as samples and clustering using k-means yields the row labels. Similarly, projecting the columns to and clustering this matrix yields the column labels. Examples: A demo of the Spectral Biclustering algorithm : a simple example showing how to generate a checkerboard matrix and bicluster it. References: Kluger, Yuval, et. al., 2003. Spectral biclustering of microarray data: coclustering genes and conditions . ',\n",
       " 'sklearn_2_4_biclustering\\n2.4. Biclustering\\nmodules/biclustering.html\\n 2.4.3. Biclustering evaluation  There are two ways of evaluating a biclustering result: internal and external. Internal measures, such as cluster stability, rely only on the data and the result themselves. Currently there are no internal bicluster measures in scikit-learn. External measures refer to an external source of information, such as the true solution. When working with real data the true solution is usually unknown, but biclustering artificial data may be useful for evaluating algorithms precisely because the true solution is known. To compare a set of found biclusters to the set of true biclusters, two similarity measures are needed: a similarity measure for individual biclusters, and a way to combine these individual similarities into an overall score. To compare individual biclusters, several measures have been used. For now, only the Jaccard index is implemented: \\\\[J(A, B)  \\\\frac{|A \\\\cap B|}{|A| + |B| - |A \\\\cap B|}\\\\] where and are biclusters, is the number of elements in their intersection. The Jaccard index achieves its minimum of 0 when the biclusters to not overlap at all and its maximum of 1 when they are identical. Several methods have been developed to compare two sets of biclusters. For now, only consensus_score (Hochreiter et. al., 2010) is available: Compute bicluster similarities for pairs of biclusters, one in each set, using the Jaccard index or a similar measure. Assign biclusters from one set to another in a one-to-one fashion to maximize the sum of their similarities. This step is performed using the Hungarian algorithm. The final sum of similarities is divided by the size of the larger set. The minimum consensus score, 0, occurs when all pairs of biclusters are totally dissimilar. The maximum score, 1, occurs when both sets are identical. References: Hochreiter, Bodenhofer, et. al., 2010. FABIA: factor analysis for bicluster acquisition . ',\n",
       " 'sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems\\n2.5. Decomposing signals in components (matrix factorization problems)\\nmodules/decomposition.html\\n 2.5.1. Principal component analysis (PCA)  2.5.1.1. Exact PCA and probabilistic interpretation  PCA is used to decompose a multivariate dataset in a set of successive orthogonal components that explain a maximum amount of the variance. In scikit-learn, PCA is implemented as a transformer object that learns components in its method, and can be used on new data to project it on these components. PCA centers but does not scale the input data for each feature before applying the SVD. The optional parameter makes it possible to project the data onto the singular space while scaling each component to unit variance. This is often useful if the models down-stream make strong assumptions on the isotropy of the signal: this is for example the case for Support Vector Machines with the RBF kernel and the K-Means clustering algorithm. Below is an example of the iris dataset, which is comprised of 4 features, projected on the 2 dimensions that explain most variance: The PCA object also provides a probabilistic interpretation of the PCA that can give a likelihood of data based on the amount of variance it explains. As such it implements a score method that can be used in cross-validation: Examples: Comparison of LDA and PCA 2D projection of Iris dataset Model selection with Probabilistic PCA and Factor Analysis (FA) 2.5.1.2. Incremental PCA  The PCA object is very useful, but has certain limitations for large datasets. The biggest limitation is that PCA only supports batch processing, which means all of the data to be processed must fit in main memory. The IncrementalPCA object uses a different form of processing and allows for partial computations which almost exactly match the results of PCA while processing the data in a minibatch fashion. IncrementalPCA makes it possible to implement out-of-core Principal Component Analysis either by: Using its method on chunks of data fetched sequentially from the local hard drive or a network database. Calling its fit method on a sparse matrix or a memory mapped file using . IncrementalPCA only stores estimates of component and noise variances, in order update incrementally. This is why memory usage depends on the number of samples per batch, rather than the number of samples to be processed in the dataset. As in PCA , IncrementalPCA centers but does not scale the input data for each feature before applying the SVD. Examples: Incremental PCA 2.5.1.3. PCA using randomized SVD  It is often interesting to project data to a lower-dimensional space that preserves most of the variance, by dropping the singular vector of components associated with lower singular values. For instance, if we work with 64x64 pixel gray-level pictures for face recognition, the dimensionality of the data is 4096 and it is slow to train an RBF support vector machine on such wide data. Furthermore we know that the intrinsic dimensionality of the data is much lower than 4096 since all pictures of human faces look somewhat alike. The samples lie on a manifold of much lower dimension (say around 200 for instance). The PCA algorithm can be used to linearly transform the data while both reducing the dimensionality and preserve most of the explained variance at the same time. The class PCA used with the optional parameter is very useful in that case: since we are going to drop most of the singular vectors it is much more efficient to limit the computation to an approximated estimate of the singular vectors we will keep to actually perform the transform. For instance, the following shows 16 sample portraits (centered around 0.0) from the Olivetti dataset. On the right hand side are the first 16 singular vectors reshaped as portraits. Since we only require the top 16 singular vectors of a dataset with size and , the computation time is less than 1s: If we note and , the time complexity of the randomized PCA is instead of for the exact method implemented in PCA . The memory footprint of randomized PCA is also proportional to instead of for the exact method. Note: the implementation of in PCA with is not the exact inverse transform of even when (default). Examples: Faces recognition example using eigenfaces and SVMs Faces dataset decompositions References: “Finding structure with randomness: Stochastic algorithms for constructing approximate matrix decompositions” Halko, et al., 2009 2.5.1.4. Kernel PCA  KernelPCA is an extension of PCA which achieves non-linear dimensionality reduction through the use of kernels (see Pairwise metrics, Affinities and Kernels ). It has many applications including denoising, compression and structured prediction (kernel dependency estimation). KernelPCA supports both and . Examples: Kernel PCA 2.5.1.5. Sparse principal components analysis (SparsePCA and MiniBatchSparsePCA)  SparsePCA is a variant of PCA, with the goal of extracting the set of sparse components that best reconstruct the data. Mini-batch sparse PCA ( MiniBatchSparsePCA ) is a variant of SparsePCA that is faster but less accurate. The increased speed is reached by iterating over small chunks of the set of features, for a given number of iterations. Principal component analysis ( PCA ) has the disadvantage that the components extracted by this method have exclusively dense expressions, i.e. they have non-zero coefficients when expressed as linear combinations of the original variables. This can make interpretation difficult. In many cases, the real underlying components can be more naturally imagined as sparse vectors; for example in face recognition, components might naturally map to parts of faces. Sparse principal components yields a more parsimonious, interpretable representation, clearly emphasizing which of the original features contribute to the differences between samples. The following example illustrates 16 components extracted using sparse PCA from the Olivetti faces dataset. It can be seen how the regularization term induces many zeros. Furthermore, the natural structure of the data causes the non-zero coefficients to be vertically adjacent. The model does not enforce this mathematically: each component is a vector , and there is no notion of vertical adjacency except during the human-friendly visualization as 64x64 pixel images. The fact that the components shown below appear local is the effect of the inherent structure of the data, which makes such local patterns minimize reconstruction error. There exist sparsity-inducing norms that take into account adjacency and different kinds of structure; see [Jen09] for a review of such methods. For more details on how to use Sparse PCA, see the Examples section, below. Note that there are many different formulations for the Sparse PCA problem. The one implemented here is based on [Mrl09] . The optimization problem solved is a PCA problem (dictionary learning) with an penalty on the components: \\\\[\\\\begin{split}(U^*, V^*)  \\\\underset{U, V}{\\\\operatorname{arg\\\\,min\\\\,}} & \\\\frac{1}{2} ||X-UV||_2^2+\\\\alpha||V||_1 \\\\\\\\ \\\\text{subject to } & ||U_k||_2  1 \\\\text{ for all } 0 \\\\leq k < n_{components}\\\\end{split}\\\\] The sparsity-inducing norm also prevents learning components from noise when few training samples are available. The degree of penalization (and thus sparsity) can be adjusted through the hyperparameter . Small values lead to a gently regularized factorization, while larger values shrink many coefficients to zero. Note While in the spirit of an online algorithm, the class MiniBatchSparsePCA does not implement because the algorithm is online along the features direction, not the samples direction. Examples: Faces dataset decompositions References: Mrl09 “Online Dictionary Learning for Sparse Coding” J. Mairal, F. Bach, J. Ponce, G. Sapiro, 2009 Jen09 “Structured Sparse Principal Component Analysis” R. Jenatton, G. Obozinski, F. Bach, 2009 ',\n",
       " \"sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems\\n2.5. Decomposing signals in components (matrix factorization problems)\\nmodules/decomposition.html\\n 2.5.2. Truncated singular value decomposition and latent semantic analysis  TruncatedSVD implements a variant of singular value decomposition (SVD) that only computes the largest singular values, where is a user-specified parameter. When truncated SVD is applied to term-document matrices (as returned by CountVectorizer or TfidfVectorizer ), this transformation is known as latent semantic analysis (LSA), because it transforms such matrices to a “semantic” space of low dimensionality. In particular, LSA is known to combat the effects of synonymy and polysemy (both of which roughly mean there are multiple meanings per word), which cause term-document matrices to be overly sparse and exhibit poor similarity under measures such as cosine similarity. Note LSA is also known as latent semantic indexing, LSI, though strictly that refers to its use in persistent indexes for information retrieval purposes. Mathematically, truncated SVD applied to training samples produces a low-rank approximation : \\\\[X \\\\approx X_k  U_k \\\\Sigma_k V_k^\\\\top\\\\] After this operation, is the transformed training set with features (called in the API). To also transform a test set , we multiply it with : \\\\[X'  X V_k\\\\] Note Most treatments of LSA in the natural language processing (NLP) and information retrieval (IR) literature swap the axes of the matrix so that it has shape × . We present LSA in a different way that matches the scikit-learn API better, but the singular values found are the same. TruncatedSVD is very similar to PCA , but differs in that the matrix does not need to be centered. When the columnwise (per-feature) means of are subtracted from the feature values, truncated SVD on the resulting matrix is equivalent to PCA. In practical terms, this means that the TruncatedSVD transformer accepts matrices without the need to densify them, as densifying may fill up memory even for medium-sized document collections. While the TruncatedSVD transformer works with any feature matrix, using it on tf–idf matrices is recommended over raw frequency counts in an LSA/document processing setting. In particular, sublinear scaling and inverse document frequency should be turned on ( ) to bring the feature values closer to a Gaussian distribution, compensating for LSA’s erroneous assumptions about textual data. Examples: Clustering text documents using k-means References: Christopher D. Manning, Prabhakar Raghavan and Hinrich Schütze (2008), Introduction to Information Retrieval , Cambridge University Press, chapter 18: Matrix decompositions & latent semantic indexing \",\n",
       " 'sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems\\n2.5. Decomposing signals in components (matrix factorization problems)\\nmodules/decomposition.html\\n 2.5.3. Dictionary Learning  2.5.3.1. Sparse coding with a precomputed dictionary  The SparseCoder object is an estimator that can be used to transform signals into sparse linear combination of atoms from a fixed, precomputed dictionary such as a discrete wavelet basis. This object therefore does not implement a method. The transformation amounts to a sparse coding problem: finding a representation of the data as a linear combination of as few dictionary atoms as possible. All variations of dictionary learning implement the following transform methods, controllable via the initialization parameter: Orthogonal matching pursuit ( Orthogonal Matching Pursuit (OMP) ) Least-angle regression ( Least Angle Regression ) Lasso computed by least-angle regression Lasso using coordinate descent ( Lasso ) Thresholding Thresholding is very fast but it does not yield accurate reconstructions. They have been shown useful in literature for classification tasks. For image reconstruction tasks, orthogonal matching pursuit yields the most accurate, unbiased reconstruction. The dictionary learning objects offer, via the parameter, the possibility to separate the positive and negative values in the results of sparse coding. This is useful when dictionary learning is used for extracting features that will be used for supervised learning, because it allows the learning algorithm to assign different weights to negative loadings of a particular atom, from to the corresponding positive loading. The split code for a single sample has length and is constructed using the following rule: First, the regular code of length is computed. Then, the first entries of the are filled with the positive part of the regular code vector. The second half of the split code is filled with the negative part of the code vector, only with a positive sign. Therefore, the split_code is non-negative. Examples: Sparse coding with a precomputed dictionary 2.5.3.2. Generic dictionary learning  Dictionary learning ( DictionaryLearning ) is a matrix factorization problem that amounts to finding a (usually overcomplete) dictionary that will perform well at sparsely encoding the fitted data. Representing data as sparse combinations of atoms from an overcomplete dictionary is suggested to be the way the mammalian primary visual cortex works. Consequently, dictionary learning applied on image patches has been shown to give good results in image processing tasks such as image completion, inpainting and denoising, as well as for supervised recognition tasks. Dictionary learning is an optimization problem solved by alternatively updating the sparse code, as a solution to multiple Lasso problems, considering the dictionary fixed, and then updating the dictionary to best fit the sparse code. \\\\[\\\\begin{split}(U^*, V^*)  \\\\underset{U, V}{\\\\operatorname{arg\\\\,min\\\\,}} & \\\\frac{1}{2} ||X-UV||_2^2+\\\\alpha||U||_1 \\\\\\\\ \\\\text{subject to } & ||V_k||_2  1 \\\\text{ for all } 0 \\\\leq k < n_{\\\\mathrm{atoms}}\\\\end{split}\\\\] After using such a procedure to fit the dictionary, the transform is simply a sparse coding step that shares the same implementation with all dictionary learning objects (see Sparse coding with a precomputed dictionary ). It is also possible to constrain the dictionary and/or code to be positive to match constraints that may be present in the data. Below are the faces with different positivity constraints applied. Red indicates negative values, blue indicates positive values, and white represents zeros. The following image shows how a dictionary learned from 4x4 pixel image patches extracted from part of the image of a raccoon face looks like. Examples: Image denoising using dictionary learning References: “Online dictionary learning for sparse coding” J. Mairal, F. Bach, J. Ponce, G. Sapiro, 2009 2.5.3.3. Mini-batch dictionary learning  MiniBatchDictionaryLearning implements a faster, but less accurate version of the dictionary learning algorithm that is better suited for large datasets. By default, MiniBatchDictionaryLearning divides the data into mini-batches and optimizes in an online manner by cycling over the mini-batches for the specified number of iterations. However, at the moment it does not implement a stopping condition. The estimator also implements , which updates the dictionary by iterating only once over a mini-batch. This can be used for online learning when the data is not readily available from the start, or for when the data does not fit into the memory. Clustering for dictionary learning Note that when using dictionary learning to extract a representation (e.g. for sparse coding) clustering can be a good proxy to learn the dictionary. For instance the MiniBatchKMeans estimator is computationally efficient and implements on-line learning with a method. Example: Online learning of a dictionary of parts of faces ',\n",
       " 'sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems\\n2.5. Decomposing signals in components (matrix factorization problems)\\nmodules/decomposition.html\\n 2.5.4. Factor Analysis  In unsupervised learning we only have a dataset . How can this dataset be described mathematically? A very simple model for is \\\\[x_i  W h_i + \\\\mu + \\\\epsilon\\\\] The vector is called “latent” because it is unobserved. is considered a noise term distributed according to a Gaussian with mean 0 and covariance (i.e. ), is some arbitrary offset vector. Such a model is called “generative” as it describes how is generated from . If we use all the ’s as columns to form a matrix and all the ’s as columns of a matrix then we can write (with suitably defined and ): \\\\[\\\\mathbf{X}  W \\\\mathbf{H} + \\\\mathbf{M} + \\\\mathbf{E}\\\\] In other words, we decomposed matrix . If is given, the above equation automatically implies the following probabilistic interpretation: \\\\[p(x_i|h_i)  \\\\mathcal{N}(Wh_i + \\\\mu, \\\\Psi)\\\\] For a complete probabilistic model we also need a prior distribution for the latent variable . The most straightforward assumption (based on the nice properties of the Gaussian distribution) is . This yields a Gaussian as the marginal distribution of : \\\\[p(x)  \\\\mathcal{N}(\\\\mu, WW^T + \\\\Psi)\\\\] Now, without any further assumptions the idea of having a latent variable would be superfluous – can be completely modelled with a mean and a covariance. We need to impose some more specific structure on one of these two parameters. A simple additional assumption regards the structure of the error covariance : : This assumption leads to the probabilistic model of PCA . : This model is called FactorAnalysis , a classical statistical model. The matrix W is sometimes called the “factor loading matrix”. Both models essentially estimate a Gaussian with a low-rank covariance matrix. Because both models are probabilistic they can be integrated in more complex models, e.g. Mixture of Factor Analysers. One gets very different models (e.g. FastICA ) if non-Gaussian priors on the latent variables are assumed. Factor analysis can produce similar components (the columns of its loading matrix) to PCA . However, one can not make any general statements about these components (e.g. whether they are orthogonal): The main advantage for Factor Analysis over PCA is that it can model the variance in every direction of the input space independently (heteroscedastic noise): This allows better model selection than probabilistic PCA in the presence of heteroscedastic noise: Examples: Model selection with Probabilistic PCA and Factor Analysis (FA) ',\n",
       " 'sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems\\n2.5. Decomposing signals in components (matrix factorization problems)\\nmodules/decomposition.html\\n 2.5.5. Independent component analysis (ICA)  Independent component analysis separates a multivariate signal into additive subcomponents that are maximally independent. It is implemented in scikit-learn using the Fast ICA algorithm. Typically, ICA is not used for reducing dimensionality but for separating superimposed signals. Since the ICA model does not include a noise term, for the model to be correct, whitening must be applied. This can be done internally using the whiten argument or manually using one of the PCA variants. It is classically used to separate mixed signals (a problem known as blind source separation ), as in the example below: ICA can also be used as yet another non linear decomposition that finds components with some sparsity: Examples: Blind source separation using FastICA FastICA on 2D point clouds Faces dataset decompositions ',\n",
       " 'sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems\\n2.5. Decomposing signals in components (matrix factorization problems)\\nmodules/decomposition.html\\n 2.5.6. Non-negative matrix factorization (NMF or NNMF)  2.5.6.1. NMF with the Frobenius norm  NMF 1 is an alternative approach to decomposition that assumes that the data and the components are non-negative. NMF can be plugged in instead of PCA or its variants, in the cases where the data matrix does not contain negative values. It finds a decomposition of samples into two matrices and of non-negative elements, by optimizing the distance between and the matrix product . The most widely used distance function is the squared Frobenius norm, which is an obvious extension of the Euclidean norm to matrices: \\\\[d_{\\\\mathrm{Fro}}(X, Y)  \\\\frac{1}{2} ||X - Y||_{\\\\mathrm{Fro}}^2  \\\\frac{1}{2} \\\\sum_{i,j} (X_{ij} - {Y}_{ij})^2\\\\] Unlike PCA , the representation of a vector is obtained in an additive fashion, by superimposing the components, without subtracting. Such additive models are efficient for representing images and text. It has been observed in [Hoyer, 2004] 2 that, when carefully constrained, NMF can produce a parts-based representation of the dataset, resulting in interpretable models. The following example displays 16 sparse components found by NMF from the images in the Olivetti faces dataset, in comparison with the PCA eigenfaces. The init attribute determines the initialization method applied, which has a great impact on the performance of the method. NMF implements the method Nonnegative Double Singular Value Decomposition. NNDSVD 4 is based on two SVD processes, one approximating the data matrix, the other approximating positive sections of the resulting partial SVD factors utilizing an algebraic property of unit rank matrices. The basic NNDSVD algorithm is better fit for sparse factorization. Its variants NNDSVDa (in which all zeros are set equal to the mean of all elements of the data), and NNDSVDar (in which the zeros are set to random perturbations less than the mean of the data divided by 100) are recommended in the dense case. Note that the Multiplicative Update (‘mu’) solver cannot update zeros present in the initialization, so it leads to poorer results when used jointly with the basic NNDSVD algorithm which introduces a lot of zeros; in this case, NNDSVDa or NNDSVDar should be preferred. NMF can also be initialized with correctly scaled random non-negative matrices by setting init\"random\" . An integer seed or a can also be passed to random_state to control reproducibility. In NMF , L1 and L2 priors can be added to the loss function in order to regularize the model. The L2 prior uses the Frobenius norm, while the L1 prior uses an elementwise L1 norm. As in ElasticNet , we control the combination of L1 and L2 with the l1_ratio ( ) parameter, and the intensity of the regularization with the alpha ( ) parameter. Then the priors terms are: \\\\[\\\\alpha \\\\rho ||W||_1 + \\\\alpha \\\\rho ||H||_1 + \\\\frac{\\\\alpha(1-\\\\rho)}{2} ||W||_{\\\\mathrm{Fro}} ^ 2 + \\\\frac{\\\\alpha(1-\\\\rho)}{2} ||H||_{\\\\mathrm{Fro}} ^ 2\\\\] and the regularized objective function is: \\\\[d_{\\\\mathrm{Fro}}(X, WH) + \\\\alpha \\\\rho ||W||_1 + \\\\alpha \\\\rho ||H||_1 + \\\\frac{\\\\alpha(1-\\\\rho)}{2} ||W||_{\\\\mathrm{Fro}} ^ 2 + \\\\frac{\\\\alpha(1-\\\\rho)}{2} ||H||_{\\\\mathrm{Fro}} ^ 2\\\\] NMF regularizes both W and H. The public function non_negative_factorization allows a finer control through the regularization attribute, and may regularize only W, only H, or both. 2.5.6.2. NMF with a beta-divergence  As described previously, the most widely used distance function is the squared Frobenius norm, which is an obvious extension of the Euclidean norm to matrices: \\\\[d_{\\\\mathrm{Fro}}(X, Y)  \\\\frac{1}{2} ||X - Y||_{Fro}^2  \\\\frac{1}{2} \\\\sum_{i,j} (X_{ij} - {Y}_{ij})^2\\\\] Other distance functions can be used in NMF as, for example, the (generalized) Kullback-Leibler (KL) divergence, also referred as I-divergence: \\\\[d_{KL}(X, Y)  \\\\sum_{i,j} (X_{ij} \\\\log(\\\\frac{X_{ij}}{Y_{ij}}) - X_{ij} + Y_{ij})\\\\] Or, the Itakura-Saito (IS) divergence: \\\\[d_{IS}(X, Y)  \\\\sum_{i,j} (\\\\frac{X_{ij}}{Y_{ij}} - \\\\log(\\\\frac{X_{ij}}{Y_{ij}}) - 1)\\\\] These three distances are special cases of the beta-divergence family, with respectively 6 . The beta-divergence are defined by : \\\\[d_{\\\\beta}(X, Y)  \\\\sum_{i,j} \\\\frac{1}{\\\\beta(\\\\beta - 1)}(X_{ij}^\\\\beta + (\\\\beta-1)Y_{ij}^\\\\beta - \\\\beta X_{ij} Y_{ij}^{\\\\beta - 1})\\\\] Note that this definition is not valid if , yet it can be continuously extended to the definitions of and respectively. NMF implements two solvers, using Coordinate Descent (‘cd’) 5 , and Multiplicative Update (‘mu’) 6 . The ‘mu’ solver can optimize every beta-divergence, including of course the Frobenius norm ( ), the (generalized) Kullback-Leibler divergence ( ) and the Itakura-Saito divergence ( ). Note that for , the ‘mu’ solver is significantly faster than for other values of . Note also that with a negative (or 0, i.e. ‘itakura-saito’) , the input matrix cannot contain zero values. The ‘cd’ solver can only optimize the Frobenius norm. Due to the underlying non-convexity of NMF, the different solvers may converge to different minima, even when optimizing the same distance function. NMF is best used with the method, which returns the matrix W. The matrix H is stored into the fitted model in the attribute; the method will decompose a new matrix X_new based on these stored components: Examples: Faces dataset decompositions Topic extraction with Non-negative Matrix Factorization and Latent Dirichlet Allocation Beta-divergence loss functions References: 1 “Learning the parts of objects by non-negative matrix factorization” D. Lee, S. Seung, 1999 2 “Non-negative Matrix Factorization with Sparseness Constraints” P. Hoyer, 2004 4 “SVD based initialization: A head start for nonnegative matrix factorization” C. Boutsidis, E. Gallopoulos, 2008 5 “Fast local algorithms for large scale nonnegative matrix and tensor factorizations.” A. Cichocki, A. Phan, 2009 6 ( 1 , 2 ) “Algorithms for nonnegative matrix factorization with the beta-divergence” C. Fevotte, J. Idier, 2011 ',\n",
       " 'sklearn_2_5_decomposing_signals_in_components_matrix_factorization_problems\\n2.5. Decomposing signals in components (matrix factorization problems)\\nmodules/decomposition.html\\n 2.5.7. Latent Dirichlet Allocation (LDA)  Latent Dirichlet Allocation is a generative probabilistic model for collections of discrete dataset such as text corpora. It is also a topic model that is used for discovering abstract topics from a collection of documents. The graphical model of LDA is a three-level generative model: Note on notations presented in the graphical model above, which can be found in Hoffman et al. (2013): The corpus is a collection of documents. A document is a sequence of words. There are topics in the corpus. The boxes represent repeated sampling. In the graphical model, each node is a random variable and has a role in the generative process. A shaded node indicates an observed variable and an unshaded node indicates a hidden (latent) variable. In this case, words in the corpus are the only data that we observe. The latent variables determine the random mixture of topics in the corpus and the distribution of words in the documents. The goal of LDA is to use the observed words to infer the hidden topic structure. When modeling text corpora, the model assumes the following generative process for a corpus with documents and topics, with corresponding to n_components in the API: For each topic , draw . This provides a distribution over the words, i.e. the probability of a word appearing in topic . corresponds to topic_word_prior . For each document , draw the topic proportions . corresponds to doc_topic_prior . For each word in document : Draw the topic assignment Draw the observed word For parameter estimation, the posterior distribution is: \\\\[p(z, \\\\theta, \\\\beta |w, \\\\alpha, \\\\eta)  \\\\frac{p(z, \\\\theta, \\\\beta|\\\\alpha, \\\\eta)}{p(w|\\\\alpha, \\\\eta)}\\\\] Since the posterior is intractable, variational Bayesian method uses a simpler distribution to approximate it, and those variational parameters , , are optimized to maximize the Evidence Lower Bound (ELBO): \\\\[\\\\log\\\\: P(w | \\\\alpha, \\\\eta) \\\\geq L(w,\\\\phi,\\\\gamma,\\\\lambda) \\\\overset{\\\\triangle}{} E_{q}[\\\\log\\\\:p(w,z,\\\\theta,\\\\beta|\\\\alpha,\\\\eta)] - E_{q}[\\\\log\\\\:q(z, \\\\theta, \\\\beta)]\\\\] Maximizing ELBO is equivalent to minimizing the Kullback-Leibler(KL) divergence between and the true posterior . LatentDirichletAllocation implements the online variational Bayes algorithm and supports both online and batch update methods. While the batch method updates variational variables after each full pass through the data, the online method updates variational variables from mini-batch data points. Note Although the online method is guaranteed to converge to a local optimum point, the quality of the optimum point and the speed of convergence may depend on mini-batch size and attributes related to learning rate setting. When LatentDirichletAllocation is applied on a “document-term” matrix, the matrix will be decomposed into a “topic-term” matrix and a “document-topic” matrix. While “topic-term” matrix is stored as components_ in the model, “document-topic” matrix can be calculated from method. LatentDirichletAllocation also implements method. This is used when data can be fetched sequentially. Examples: Topic extraction with Non-negative Matrix Factorization and Latent Dirichlet Allocation References: “Latent Dirichlet Allocation” D. Blei, A. Ng, M. Jordan, 2003 “Online Learning for Latent Dirichlet Allocation” M. Hoffman, D. Blei, F. Bach, 2010 “Stochastic Variational Inference” M. Hoffman, D. Blei, C. Wang, J. Paisley, 2013 See also Dimensionality reduction for dimensionality reduction with Neighborhood Components Analysis. ',\n",
       " 'sklearn_2_6_covariance_estimation\\n2.6. Covariance estimation\\nmodules/covariance.html\\n 2.6.1. Empirical covariance  The covariance matrix of a data set is known to be well approximated by the classical maximum likelihood estimator (or “empirical covariance”), provided the number of observations is large enough compared to the number of features (the variables describing the observations). More precisely, the Maximum Likelihood Estimator of a sample is an unbiased estimator of the corresponding population’s covariance matrix. The empirical covariance matrix of a sample can be computed using the empirical_covariance function of the package, or by fitting an EmpiricalCovariance object to the data sample with the EmpiricalCovariance.fit method. Be careful that results depend on whether the data are centered, so one may want to use the parameter accurately. More precisely, if , then the test set is supposed to have the same mean vector as the training set. If not, both should be centered by the user, and should be used. Examples: See Shrinkage covariance estimation: LedoitWolf vs OAS and max-likelihood for an example on how to fit an EmpiricalCovariance object to data. ',\n",
       " 'sklearn_2_6_covariance_estimation\\n2.6. Covariance estimation\\nmodules/covariance.html\\n 2.6.2. Shrunk Covariance  2.6.2.1. Basic shrinkage  Despite being an unbiased estimator of the covariance matrix, the Maximum Likelihood Estimator is not a good estimator of the eigenvalues of the covariance matrix, so the precision matrix obtained from its inversion is not accurate. Sometimes, it even occurs that the empirical covariance matrix cannot be inverted for numerical reasons. To avoid such an inversion problem, a transformation of the empirical covariance matrix has been introduced: the . In scikit-learn, this transformation (with a user-defined shrinkage coefficient) can be directly applied to a pre-computed covariance with the shrunk_covariance method. Also, a shrunk estimator of the covariance can be fitted to data with a ShrunkCovariance object and its ShrunkCovariance.fit method. Again, results depend on whether the data are centered, so one may want to use the parameter accurately. Mathematically, this shrinkage consists in reducing the ratio between the smallest and the largest eigenvalues of the empirical covariance matrix. It can be done by simply shifting every eigenvalue according to a given offset, which is equivalent of finding the l2-penalized Maximum Likelihood Estimator of the covariance matrix. In practice, shrinkage boils down to a simple a convex transformation : . Choosing the amount of shrinkage, amounts to setting a bias/variance trade-off, and is discussed below. Examples: See Shrinkage covariance estimation: LedoitWolf vs OAS and max-likelihood for an example on how to fit a ShrunkCovariance object to data. 2.6.2.2. Ledoit-Wolf shrinkage  In their 2004 paper 1 , O. Ledoit and M. Wolf propose a formula to compute the optimal shrinkage coefficient that minimizes the Mean Squared Error between the estimated and the real covariance matrix. The Ledoit-Wolf estimator of the covariance matrix can be computed on a sample with the ledoit_wolf function of the sklearn.covariance package, or it can be otherwise obtained by fitting a LedoitWolf object to the same sample. Note Case when population covariance matrix is isotropic It is important to note that when the number of samples is much larger than the number of features, one would expect that no shrinkage would be necessary. The intuition behind this is that if the population covariance is full rank, when the number of sample grows, the sample covariance will also become positive definite. As a result, no shrinkage would necessary and the method should automatically do this. This, however, is not the case in the Ledoit-Wolf procedure when the population covariance happens to be a multiple of the identity matrix. In this case, the Ledoit-Wolf shrinkage estimate approaches 1 as the number of samples increases. This indicates that the optimal estimate of the covariance matrix in the Ledoit-Wolf sense is multiple of the identity. Since the population covariance is already a multiple of the identity matrix, the Ledoit-Wolf solution is indeed a reasonable estimate. Examples: See Shrinkage covariance estimation: LedoitWolf vs OAS and max-likelihood for an example on how to fit a LedoitWolf object to data and for visualizing the performances of the Ledoit-Wolf estimator in terms of likelihood. References: 1 O. Ledoit and M. Wolf, “A Well-Conditioned Estimator for Large-Dimensional Covariance Matrices”, Journal of Multivariate Analysis, Volume 88, Issue 2, February 2004, pages 365-411. 2.6.2.3. Oracle Approximating Shrinkage  Under the assumption that the data are Gaussian distributed, Chen et al. 2 derived a formula aimed at choosing a shrinkage coefficient that yields a smaller Mean Squared Error than the one given by Ledoit and Wolf’s formula. The resulting estimator is known as the Oracle Shrinkage Approximating estimator of the covariance. The OAS estimator of the covariance matrix can be computed on a sample with the oas function of the sklearn.covariance package, or it can be otherwise obtained by fitting an OAS object to the same sample. Bias-variance trade-off when setting the shrinkage: comparing the choices of Ledoit-Wolf and OAS estimators  References: 2 Chen et al., “Shrinkage Algorithms for MMSE Covariance Estimation”, IEEE Trans. on Sign. Proc., Volume 58, Issue 10, October 2010. Examples: See Shrinkage covariance estimation: LedoitWolf vs OAS and max-likelihood for an example on how to fit an OAS object to data. See Ledoit-Wolf vs OAS estimation to visualize the Mean Squared Error difference between a LedoitWolf and an OAS estimator of the covariance. ',\n",
       " 'sklearn_2_6_covariance_estimation\\n2.6. Covariance estimation\\nmodules/covariance.html\\n 2.6.3. Sparse inverse covariance  The matrix inverse of the covariance matrix, often called the precision matrix, is proportional to the partial correlation matrix. It gives the partial independence relationship. In other words, if two features are independent conditionally on the others, the corresponding coefficient in the precision matrix will be zero. This is why it makes sense to estimate a sparse precision matrix: the estimation of the covariance matrix is better conditioned by learning independence relations from the data. This is known as covariance selection . In the small-samples situation, in which is on the order of or smaller, sparse inverse covariance estimators tend to work better than shrunk covariance estimators. However, in the opposite situation, or for very correlated data, they can be numerically unstable. In addition, unlike shrinkage estimators, sparse estimators are able to recover off-diagonal structure. The GraphicalLasso estimator uses an l1 penalty to enforce sparsity on the precision matrix: the higher its parameter, the more sparse the precision matrix. The corresponding GraphicalLassoCV object uses cross-validation to automatically set the parameter. A comparison of maximum likelihood, shrinkage and sparse estimates of the covariance and precision matrix in the very small samples settings.  Note Structure recovery Recovering a graphical structure from correlations in the data is a challenging thing. If you are interested in such recovery keep in mind that: Recovery is easier from a correlation matrix than a covariance matrix: standardize your observations before running GraphicalLasso If the underlying graph has nodes with much more connections than the average node, the algorithm will miss some of these connections. If your number of observations is not large compared to the number of edges in your underlying graph, you will not recover it. Even if you are in favorable recovery conditions, the alpha parameter chosen by cross-validation (e.g. using the GraphicalLassoCV object) will lead to selecting too many edges. However, the relevant edges will have heavier weights than the irrelevant ones. The mathematical formulation is the following: \\\\[\\\\hat{K}  \\\\mathrm{argmin}_K \\\\big( \\\\mathrm{tr} S K - \\\\mathrm{log} \\\\mathrm{det} K + \\\\alpha \\\\|K\\\\|_1 \\\\big)\\\\] Where is the precision matrix to be estimated, and is the sample covariance matrix. is the sum of the absolute values of off-diagonal coefficients of . The algorithm employed to solve this problem is the GLasso algorithm, from the Friedman 2008 Biostatistics paper. It is the same algorithm as in the R package. Examples: Sparse inverse covariance estimation : example on synthetic data showing some recovery of a structure, and comparing to other covariance estimators. Visualizing the stock market structure : example on real stock market data, finding which symbols are most linked. References: Friedman et al, “Sparse inverse covariance estimation with the graphical lasso” , Biostatistics 9, pp 432, 2008 ',\n",
       " 'sklearn_2_6_covariance_estimation\\n2.6. Covariance estimation\\nmodules/covariance.html\\n 2.6.4. Robust Covariance Estimation  Real data sets are often subject to measurement or recording errors. Regular but uncommon observations may also appear for a variety of reasons. Observations which are very uncommon are called outliers. The empirical covariance estimator and the shrunk covariance estimators presented above are very sensitive to the presence of outliers in the data. Therefore, one should use robust covariance estimators to estimate the covariance of its real data sets. Alternatively, robust covariance estimators can be used to perform outlier detection and discard/downweight some observations according to further processing of the data. The package implements a robust estimator of covariance, the Minimum Covariance Determinant 3 . 2.6.4.1. Minimum Covariance Determinant  The Minimum Covariance Determinant estimator is a robust estimator of a data set’s covariance introduced by P.J. Rousseeuw in 3 . The idea is to find a given proportion (h) of “good” observations which are not outliers and compute their empirical covariance matrix. This empirical covariance matrix is then rescaled to compensate the performed selection of observations (“consistency step”). Having computed the Minimum Covariance Determinant estimator, one can give weights to observations according to their Mahalanobis distance, leading to a reweighted estimate of the covariance matrix of the data set (“reweighting step”). Rousseeuw and Van Driessen 4 developed the FastMCD algorithm in order to compute the Minimum Covariance Determinant. This algorithm is used in scikit-learn when fitting an MCD object to data. The FastMCD algorithm also computes a robust estimate of the data set location at the same time. Raw estimates can be accessed as and attributes of a MinCovDet robust covariance estimator object. References: 3 ( 1 , 2 ) P. J. Rousseeuw. Least median of squares regression. J. Am Stat Ass, 79:871, 1984. 4 A Fast Algorithm for the Minimum Covariance Determinant Estimator, 1999, American Statistical Association and the American Society for Quality, TECHNOMETRICS. Examples: See Robust vs Empirical covariance estimate for an example on how to fit a MinCovDet object to data and see how the estimate remains accurate despite the presence of outliers. See Robust covariance estimation and Mahalanobis distances relevance to visualize the difference between EmpiricalCovariance and MinCovDet covariance estimators in terms of Mahalanobis distance (so we get a better estimate of the precision matrix too). Influence of outliers on location and covariance estimates Separating inliers from outliers using a Mahalanobis distance ',\n",
       " 'sklearn_2_7_novelty_and_outlier_detection\\n2.7. Novelty and Outlier Detection\\nmodules/outlier_detection.html\\n 2.7.1. Overview of outlier detection methods  A comparison of the outlier detection algorithms in scikit-learn. Local Outlier Factor (LOF) does not show a decision boundary in black as it has no predict method to be applied on new data when it is used for outlier detection. ensemble.IsolationForest and neighbors.LocalOutlierFactor perform reasonably well on the data sets considered here. The svm.OneClassSVM is known to be sensitive to outliers and thus does not perform very well for outlier detection. Finally, covariance.EllipticEnvelope assumes the data is Gaussian and learns an ellipse. For more details on the different estimators refer to the example Comparing anomaly detection algorithms for outlier detection on toy datasets and the sections hereunder. Examples: See Comparing anomaly detection algorithms for outlier detection on toy datasets for a comparison of the svm.OneClassSVM , the ensemble.IsolationForest , the neighbors.LocalOutlierFactor and covariance.EllipticEnvelope . ',\n",
       " 'sklearn_2_7_novelty_and_outlier_detection\\n2.7. Novelty and Outlier Detection\\nmodules/outlier_detection.html\\n 2.7.2. Novelty Detection  Consider a data set of observations from the same distribution described by features. Consider now that we add one more observation to that data set. Is the new observation so different from the others that we can doubt it is regular? (i.e. does it come from the same distribution?) Or on the contrary, is it so similar to the other that we cannot distinguish it from the original observations? This is the question addressed by the novelty detection tools and methods. In general, it is about to learn a rough, close frontier delimiting the contour of the initial observations distribution, plotted in embedding -dimensional space. Then, if further observations lay within the frontier-delimited subspace, they are considered as coming from the same population than the initial observations. Otherwise, if they lay outside the frontier, we can say that they are abnormal with a given confidence in our assessment. The One-Class SVM has been introduced by Schölkopf et al. for that purpose and implemented in the Support Vector Machines module in the svm.OneClassSVM object. It requires the choice of a kernel and a scalar parameter to define a frontier. The RBF kernel is usually chosen although there exists no exact formula or algorithm to set its bandwidth parameter. This is the default in the scikit-learn implementation. The parameter, also known as the margin of the One-Class SVM, corresponds to the probability of finding a new, but regular, observation outside the frontier. References: Estimating the support of a high-dimensional distribution Schölkopf, Bernhard, et al. Neural computation 13.7 (2001): 1443-1471. Examples: See One-class SVM with non-linear kernel (RBF) for visualizing the frontier learned around some data by a svm.OneClassSVM object. Species distribution modeling ',\n",
       " 'sklearn_2_7_novelty_and_outlier_detection\\n2.7. Novelty and Outlier Detection\\nmodules/outlier_detection.html\\n 2.7.3. Outlier Detection  Outlier detection is similar to novelty detection in the sense that the goal is to separate a core of regular observations from some polluting ones, called outliers . Yet, in the case of outlier detection, we don’t have a clean data set representing the population of regular observations that can be used to train any tool. 2.7.3.1. Fitting an elliptic envelope  One common way of performing outlier detection is to assume that the regular data come from a known distribution (e.g. data are Gaussian distributed). From this assumption, we generally try to define the “shape” of the data, and can define outlying observations as observations which stand far enough from the fit shape. The scikit-learn provides an object covariance.EllipticEnvelope that fits a robust covariance estimate to the data, and thus fits an ellipse to the central data points, ignoring points outside the central mode. For instance, assuming that the inlier data are Gaussian distributed, it will estimate the inlier location and covariance in a robust way (i.e. without being influenced by outliers). The Mahalanobis distances obtained from this estimate is used to derive a measure of outlyingness. This strategy is illustrated below. Examples: See Robust covariance estimation and Mahalanobis distances relevance for an illustration of the difference between using a standard ( covariance.EmpiricalCovariance ) or a robust estimate ( covariance.MinCovDet ) of location and covariance to assess the degree of outlyingness of an observation. References: Rousseeuw, P.J., Van Driessen, K. “A fast algorithm for the minimum covariance determinant estimator” Technometrics 41(3), 212 (1999) 2.7.3.2. Isolation Forest  One efficient way of performing outlier detection in high-dimensional datasets is to use random forests. The ensemble.IsolationForest ‘isolates’ observations by randomly selecting a feature and then randomly selecting a split value between the maximum and minimum values of the selected feature. Since recursive partitioning can be represented by a tree structure, the number of splittings required to isolate a sample is equivalent to the path length from the root node to the terminating node. This path length, averaged over a forest of such random trees, is a measure of normality and our decision function. Random partitioning produces noticeably shorter paths for anomalies. Hence, when a forest of random trees collectively produce shorter path lengths for particular samples, they are highly likely to be anomalies. The implementation of ensemble.IsolationForest is based on an ensemble of tree.ExtraTreeRegressor . Following Isolation Forest original paper, the maximum depth of each tree is set to where is the number of samples used to build the tree (see (Liu et al., 2008) for more details). This algorithm is illustrated below. The ensemble.IsolationForest supports which allows you to add more trees to an already fitted model: Examples: See IsolationForest example for an illustration of the use of IsolationForest. See Comparing anomaly detection algorithms for outlier detection on toy datasets for a comparison of ensemble.IsolationForest with neighbors.LocalOutlierFactor , svm.OneClassSVM (tuned to perform like an outlier detection method) and a covariance-based outlier detection with covariance.EllipticEnvelope . References: Liu, Fei Tony, Ting, Kai Ming and Zhou, Zhi-Hua. “Isolation forest.” Data Mining, 2008. ICDM’08. Eighth IEEE International Conference on. 2.7.3.3. Local Outlier Factor  Another efficient way to perform outlier detection on moderately high dimensional datasets is to use the Local Outlier Factor (LOF) algorithm. The neighbors.LocalOutlierFactor (LOF) algorithm computes a score (called local outlier factor) reflecting the degree of abnormality of the observations. It measures the local density deviation of a given data point with respect to its neighbors. The idea is to detect the samples that have a substantially lower density than their neighbors. In practice the local density is obtained from the k-nearest neighbors. The LOF score of an observation is equal to the ratio of the average local density of his k-nearest neighbors, and its own local density: a normal instance is expected to have a local density similar to that of its neighbors, while abnormal data are expected to have much smaller local density. The number k of neighbors considered, (alias parameter n_neighbors) is typically chosen 1) greater than the minimum number of objects a cluster has to contain, so that other objects can be local outliers relative to this cluster, and 2) smaller than the maximum number of close by objects that can potentially be local outliers. In practice, such informations are generally not available, and taking n_neighbors20 appears to work well in general. When the proportion of outliers is high (i.e. greater than 10 %, as in the example below), n_neighbors should be greater (n_neighbors35 in the example below). The strength of the LOF algorithm is that it takes both local and global properties of datasets into consideration: it can perform well even in datasets where abnormal samples have different underlying densities. The question is not, how isolated the sample is, but how isolated it is with respect to the surrounding neighborhood. When applying LOF for outlier detection, there are no , and methods but only a method. The scores of abnormality of the training samples are accessible through the attribute. Note that , and can be used on new unseen data when LOF is applied for novelty detection, i.e. when the parameter is set to . See Novelty detection with Local Outlier Factor . This strategy is illustrated below. Examples: See Outlier detection with Local Outlier Factor (LOF) for an illustration of the use of neighbors.LocalOutlierFactor . See Comparing anomaly detection algorithms for outlier detection on toy datasets for a comparison with other anomaly detection methods. References: Breunig, Kriegel, Ng, and Sander (2000) LOF: identifying density-based local outliers. Proc. ACM SIGMOD ',\n",
       " 'sklearn_2_7_novelty_and_outlier_detection\\n2.7. Novelty and Outlier Detection\\nmodules/outlier_detection.html\\n 2.7.4. Novelty detection with Local Outlier Factor  To use neighbors.LocalOutlierFactor for novelty detection, i.e. predict labels or compute the score of abnormality of new unseen data, you need to instantiate the estimator with the parameter set to before fitting the estimator: Note that is not available in this case. Warning Novelty detection with Local Outlier Factor` When is set to be aware that you must only use , and on new unseen data and not on the training samples as this would lead to wrong results. The scores of abnormality of the training samples are always accessible through the attribute. Novelty detection with Local Outlier Factor is illustrated below. ',\n",
       " 'sklearn_2_8_density_estimation\\n2.8. Density Estimation\\nmodules/density.html\\n 2.8.1. Density Estimation: Histograms  A histogram is a simple visualization of data where bins are defined, and the number of data points within each bin is tallied. An example of a histogram can be seen in the upper-left panel of the following figure: A major problem with histograms, however, is that the choice of binning can have a disproportionate effect on the resulting visualization. Consider the upper-right panel of the above figure. It shows a histogram over the same data, with the bins shifted right. The results of the two visualizations look entirely different, and might lead to different interpretations of the data. Intuitively, one can also think of a histogram as a stack of blocks, one block per point. By stacking the blocks in the appropriate grid space, we recover the histogram. But what if, instead of stacking the blocks on a regular grid, we center each block on the point it represents, and sum the total height at each location? This idea leads to the lower-left visualization. It is perhaps not as clean as a histogram, but the fact that the data drive the block locations mean that it is a much better representation of the underlying data. This visualization is an example of a kernel density estimation , in this case with a top-hat kernel (i.e. a square block at each point). We can recover a smoother distribution by using a smoother kernel. The bottom-right plot shows a Gaussian kernel density estimate, in which each point contributes a Gaussian curve to the total. The result is a smooth density estimate which is derived from the data, and functions as a powerful non-parametric model of the distribution of points. ',\n",
       " 'sklearn_2_8_density_estimation\\n2.8. Density Estimation\\nmodules/density.html\\n 2.8.2. Kernel Density Estimation  Kernel density estimation in scikit-learn is implemented in the sklearn.neighbors.KernelDensity estimator, which uses the Ball Tree or KD Tree for efficient queries (see Nearest Neighbors for a discussion of these). Though the above example uses a 1D data set for simplicity, kernel density estimation can be performed in any number of dimensions, though in practice the curse of dimensionality causes its performance to degrade in high dimensions. In the following figure, 100 points are drawn from a bimodal distribution, and the kernel density estimates are shown for three choices of kernels: It’s clear how the kernel shape affects the smoothness of the resulting distribution. The scikit-learn kernel density estimator can be used as follows: Here we have used , as seen above. Mathematically, a kernel is a positive function which is controlled by the bandwidth parameter . Given this kernel form, the density estimate at a point within a group of points is given by: \\\\[\\\\rho_K(y)  \\\\sum_{i1}^{N} K(y - x_i; h)\\\\] The bandwidth here acts as a smoothing parameter, controlling the tradeoff between bias and variance in the result. A large bandwidth leads to a very smooth (i.e. high-bias) density distribution. A small bandwidth leads to an unsmooth (i.e. high-variance) density distribution. sklearn.neighbors.KernelDensity implements several common kernel forms, which are shown in the following figure: The form of these kernels is as follows: Gaussian kernel ( ) Tophat kernel ( ) if Epanechnikov kernel ( ) Exponential kernel ( ) Linear kernel ( ) if Cosine kernel ( ) if The kernel density estimator can be used with any of the valid distance metrics (see sklearn.neighbors.DistanceMetric for a list of available metrics), though the results are properly normalized only for the Euclidean metric. One particularly useful metric is the Haversine distance which measures the angular distance between points on a sphere. Here is an example of using a kernel density estimate for a visualization of geospatial data, in this case the distribution of observations of two different species on the South American continent: One other useful application of kernel density estimation is to learn a non-parametric generative model of a dataset in order to efficiently draw new samples from this generative model. Here is an example of using this process to create a new set of hand-written digits, using a Gaussian kernel learned on a PCA projection of the data: The “new” data consists of linear combinations of the input data, with weights probabilistically drawn given the KDE model. Examples: Simple 1D Kernel Density Estimation : computation of simple kernel density estimates in one dimension. Kernel Density Estimation : an example of using Kernel Density estimation to learn a generative model of the hand-written digits data, and drawing new samples from this model. Kernel Density Estimate of Species Distributions : an example of Kernel Density estimation using the Haversine distance metric to visualize geospatial data ',\n",
       " 'sklearn_2_9_neural_network_models_unsupervised\\n2.9. Neural network models (unsupervised)\\nmodules/neural_networks_unsupervised.html\\n 2.9.1. Restricted Boltzmann machines  Restricted Boltzmann machines (RBM) are unsupervised nonlinear feature learners based on a probabilistic model. The features extracted by an RBM or a hierarchy of RBMs often give good results when fed into a linear classifier such as a linear SVM or a perceptron. The model makes assumptions regarding the distribution of inputs. At the moment, scikit-learn only provides BernoulliRBM , which assumes the inputs are either binary values or values between 0 and 1, each encoding the probability that the specific feature would be turned on. The RBM tries to maximize the likelihood of the data using a particular graphical model. The parameter learning algorithm used ( Stochastic Maximum Likelihood ) prevents the representations from straying far from the input data, which makes them capture interesting regularities, but makes the model less useful for small datasets, and usually not useful for density estimation. The method gained popularity for initializing deep neural networks with the weights of independent RBMs. This method is known as unsupervised pre-training. Examples: Restricted Boltzmann Machine features for digit classification 2.9.1.1. Graphical model and parametrization  The graphical model of an RBM is a fully-connected bipartite graph. The nodes are random variables whose states depend on the state of the other nodes they are connected to. The model is therefore parameterized by the weights of the connections, as well as one intercept (bias) term for each visible and hidden unit, omitted from the image for simplicity. The energy function measures the quality of a joint assignment: \\\\[E(\\\\mathbf{v}, \\\\mathbf{h})  -\\\\sum_i \\\\sum_j w_{ij}v_ih_j - \\\\sum_i b_iv_i - \\\\sum_j c_jh_j\\\\] In the formula above, and are the intercept vectors for the visible and hidden layers, respectively. The joint probability of the model is defined in terms of the energy: \\\\[P(\\\\mathbf{v}, \\\\mathbf{h})  \\\\frac{e^{-E(\\\\mathbf{v}, \\\\mathbf{h})}}{Z}\\\\] The word restricted refers to the bipartite structure of the model, which prohibits direct interaction between hidden units, or between visible units. This means that the following conditional independencies are assumed: \\\\[\\\\begin{split}h_i \\\\bot h_j | \\\\mathbf{v} \\\\\\\\ v_i \\\\bot v_j | \\\\mathbf{h}\\\\end{split}\\\\] The bipartite structure allows for the use of efficient block Gibbs sampling for inference. 2.9.1.2. Bernoulli Restricted Boltzmann machines  In the BernoulliRBM , all units are binary stochastic units. This means that the input data should either be binary, or real-valued between 0 and 1 signifying the probability that the visible unit would turn on or off. This is a good model for character recognition, where the interest is on which pixels are active and which aren’t. For images of natural scenes it no longer fits because of background, depth and the tendency of neighbouring pixels to take the same values. The conditional probability distribution of each unit is given by the logistic sigmoid activation function of the input it receives: \\\\[\\\\begin{split}P(v_i1|\\\\mathbf{h})  \\\\sigma(\\\\sum_j w_{ij}h_j + b_i) \\\\\\\\ P(h_i1|\\\\mathbf{v})  \\\\sigma(\\\\sum_i w_{ij}v_i + c_j)\\\\end{split}\\\\] where is the logistic sigmoid function: \\\\[\\\\sigma(x)  \\\\frac{1}{1 + e^{-x}}\\\\] 2.9.1.3. Stochastic Maximum Likelihood learning  The training algorithm implemented in BernoulliRBM is known as Stochastic Maximum Likelihood (SML) or Persistent Contrastive Divergence (PCD). Optimizing maximum likelihood directly is infeasible because of the form of the data likelihood: \\\\[\\\\log P(v)  \\\\log \\\\sum_h e^{-E(v, h)} - \\\\log \\\\sum_{x, y} e^{-E(x, y)}\\\\] For simplicity the equation above is written for a single training example. The gradient with respect to the weights is formed of two terms corresponding to the ones above. They are usually known as the positive gradient and the negative gradient, because of their respective signs. In this implementation, the gradients are estimated over mini-batches of samples. In maximizing the log-likelihood, the positive gradient makes the model prefer hidden states that are compatible with the observed training data. Because of the bipartite structure of RBMs, it can be computed efficiently. The negative gradient, however, is intractable. Its goal is to lower the energy of joint states that the model prefers, therefore making it stay true to the data. It can be approximated by Markov chain Monte Carlo using block Gibbs sampling by iteratively sampling each of and given the other, until the chain mixes. Samples generated in this way are sometimes referred as fantasy particles. This is inefficient and it is difficult to determine whether the Markov chain mixes. The Contrastive Divergence method suggests to stop the chain after a small number of iterations, , usually even 1. This method is fast and has low variance, but the samples are far from the model distribution. Persistent Contrastive Divergence addresses this. Instead of starting a new chain each time the gradient is needed, and performing only one Gibbs sampling step, in PCD we keep a number of chains (fantasy particles) that are updated Gibbs steps after each weight update. This allows the particles to explore the space more thoroughly. References: “A fast learning algorithm for deep belief nets” G. Hinton, S. Osindero, Y.-W. Teh, 2006 “Training Restricted Boltzmann Machines using Approximations to the Likelihood Gradient” T. Tieleman, 2008 ',\n",
       " 'sklearn_3_1_cross-validation_evaluating_estimator_performance\\n3.1. Cross-validation: evaluating estimator performance\\nmodules/cross_validation.html\\n 3.1.1. Computing cross-validated metrics  The simplest way to use cross-validation is to call the cross_val_score helper function on the estimator and the dataset. The following example demonstrates how to estimate the accuracy of a linear kernel support vector machine on the iris dataset by splitting the data, fitting a model and computing the score 5 consecutive times (with different splits each time): The mean score and the 95% confidence interval of the score estimate are hence given by: By default, the score computed at each CV iteration is the method of the estimator. It is possible to change this by using the scoring parameter: See The scoring parameter: defining model evaluation rules for details. In the case of the Iris dataset, the samples are balanced across target classes hence the accuracy and the F1-score are almost equal. When the argument is an integer, cross_val_score uses the KFold or StratifiedKFold strategies by default, the latter being used if the estimator derives from ClassifierMixin . It is also possible to use other cross validation strategies by passing a cross validation iterator instead, for instance: Another option is to use an iterable yielding (train, test) splits as arrays of indices, for example: Data transformation with held out data Just as it is important to test a predictor on data held-out from training, preprocessing (such as standardization, feature selection, etc.) and similar data transformations similarly should be learnt from a training set and applied to held-out data for prediction: A Pipeline makes it easier to compose estimators, providing this behavior under cross-validation: See Pipelines and composite estimators . 3.1.1.1. The cross_validate function and multiple metric evaluation  The cross_validate function differs from cross_val_score in two ways: It allows specifying multiple metrics for evaluation. It returns a dict containing fit-times, score-times (and optionally training scores as well as fitted estimators) in addition to the test score. For single metric evaluation, where the scoring parameter is a string, callable or None, the keys will be - And for multiple metric evaluation, the return value is a dict with the following keys - is set to by default to save computation time. To evaluate the scores on the training set as well you need to be set to . You may also retain the estimator fitted on each training set by setting . The multiple metrics can be specified either as a list, tuple or set of predefined scorer names: Or as a dict mapping scorer name to a predefined or custom scoring function: Here is an example of using a single metric: 3.1.1.2. Obtaining predictions by cross-validation  The function cross_val_predict has a similar interface to cross_val_score , but returns, for each element in the input, the prediction that was obtained for that element when it was in the test set. Only cross-validation strategies that assign all elements to a test set exactly once can be used (otherwise, an exception is raised). Warning Note on inappropriate usage of cross_val_predict The result of cross_val_predict may be different from those obtained using cross_val_score as the elements are grouped in different ways. The function cross_val_score takes an average over cross-validation folds, whereas cross_val_predict simply returns the labels (or probabilities) from several distinct models undistinguished. Thus, cross_val_predict is not an appropriate measure of generalisation error. The function cross_val_predict is appropriate for: Visualization of predictions obtained from different models. Model blending: When predictions of one supervised estimator are used to train another estimator in ensemble methods. The available cross validation iterators are introduced in the following section. Examples Receiver Operating Characteristic (ROC) with cross validation , Recursive feature elimination with cross-validation , Parameter estimation using grid search with cross-validation , Sample pipeline for text feature extraction and evaluation , Plotting Cross-Validated Predictions , Nested versus non-nested cross-validation . ',\n",
       " 'sklearn_3_1_cross-validation_evaluating_estimator_performance\\n3.1. Cross-validation: evaluating estimator performance\\nmodules/cross_validation.html\\n 3.1.2. Cross validation iterators  The following sections list utilities to generate indices that can be used to generate dataset splits according to different cross validation strategies. 3.1.2.1. Cross-validation iterators for i.i.d. data  Assuming that some data is Independent and Identically Distributed (i.i.d.) is making the assumption that all samples stem from the same generative process and that the generative process is assumed to have no memory of past generated samples. The following cross-validators can be used in such cases. NOTE While i.i.d. data is a common assumption in machine learning theory, it rarely holds in practice. If one knows that the samples have been generated using a time-dependent process, it is safer to use a time-series aware cross-validation scheme . Similarly, if we know that the generative process has a group structure (samples collected from different subjects, experiments, measurement devices), it is safer to use group-wise cross-validation . 3.1.2.1.1. K-fold  KFold divides all the samples in groups of samples, called folds (if , this is equivalent to the Leave One Out strategy), of equal sizes (if possible). The prediction function is learned using folds, and the fold left out is used for test. Example of 2-fold cross-validation on a dataset with 4 samples: Here is a visualization of the cross-validation behavior. Note that KFold is not affected by classes or groups. Each fold is constituted by two arrays: the first one is related to the training set , and the second one to the test set . Thus, one can create the training/test sets using numpy indexing: 3.1.2.1.2. Repeated K-Fold  RepeatedKFold repeats K-Fold n times. It can be used when one requires to run KFold n times, producing different splits in each repetition. Example of 2-fold K-Fold repeated 2 times: Similarly, RepeatedStratifiedKFold repeats Stratified K-Fold n times with different randomization in each repetition. 3.1.2.1.3. Leave One Out (LOO)  LeaveOneOut (or LOO) is a simple cross-validation. Each learning set is created by taking all the samples except one, the test set being the sample left out. Thus, for samples, we have different training sets and different tests set. This cross-validation procedure does not waste much data as only one sample is removed from the training set: Potential users of LOO for model selection should weigh a few known caveats. When compared with -fold cross validation, one builds models from samples instead of models, where . Moreover, each is trained on samples rather than . In both ways, assuming is not too large and , LOO is more computationally expensive than -fold cross validation. In terms of accuracy, LOO often results in high variance as an estimator for the test error. Intuitively, since of the samples are used to build each model, models constructed from folds are virtually identical to each other and to the model built from the entire training set. However, if the learning curve is steep for the training size in question, then 5- or 10- fold cross validation can overestimate the generalization error. As a general rule, most authors, and empirical evidence, suggest that 5- or 10- fold cross validation should be preferred to LOO. References: http://www.faqs.org/faqs/ai-faq/neural-nets/part3/section-12.html ; T. Hastie, R. Tibshirani, J. Friedman, The Elements of Statistical Learning , Springer 2009 L. Breiman, P. Spector Submodel selection and evaluation in regression: The X-random case , International Statistical Review 1992; R. Kohavi, A Study of Cross-Validation and Bootstrap for Accuracy Estimation and Model Selection , Intl. Jnt. Conf. AI R. Bharat Rao, G. Fung, R. Rosales, On the Dangers of Cross-Validation. An Experimental Evaluation , SIAM 2008; G. James, D. Witten, T. Hastie, R Tibshirani, An Introduction to Statistical Learning , Springer 2013. 3.1.2.1.4. Leave P Out (LPO)  LeavePOut is very similar to LeaveOneOut as it creates all the possible training/test sets by removing samples from the complete set. For samples, this produces train-test pairs. Unlike LeaveOneOut and KFold , the test sets will overlap for . Example of Leave-2-Out on a dataset with 4 samples: 3.1.2.1.5. Random permutations cross-validation a.k.a. Shuffle & Split  ShuffleSplit The ShuffleSplit iterator will generate a user defined number of independent train / test dataset splits. Samples are first shuffled and then split into a pair of train and test sets. It is possible to control the randomness for reproducibility of the results by explicitly seeding the pseudo random number generator. Here is a usage example: Here is a visualization of the cross-validation behavior. Note that ShuffleSplit is not affected by classes or groups. ShuffleSplit is thus a good alternative to KFold cross validation that allows a finer control on the number of iterations and the proportion of samples on each side of the train / test split. 3.1.2.2. Cross-validation iterators with stratification based on class labels.  Some classification problems can exhibit a large imbalance in the distribution of the target classes: for instance there could be several times more negative samples than positive samples. In such cases it is recommended to use stratified sampling as implemented in StratifiedKFold and StratifiedShuffleSplit to ensure that relative class frequencies is approximately preserved in each train and validation fold. 3.1.2.2.1. Stratified k-fold  StratifiedKFold is a variation of k-fold which returns stratified folds: each set contains approximately the same percentage of samples of each target class as the complete set. Here is an example of stratified 3-fold cross-validation on a dataset with 50 samples from two unbalanced classes. We show the number of samples in each class and compare with KFold . We can see that StratifiedKFold preserves the class ratios (approximately 1 / 10) in both train and test dataset. Here is a visualization of the cross-validation behavior. RepeatedStratifiedKFold can be used to repeat Stratified K-Fold n times with different randomization in each repetition. 3.1.2.2.2. Stratified Shuffle Split  StratifiedShuffleSplit is a variation of ShuffleSplit , which returns stratified splits, i.e which creates splits by preserving the same percentage for each target class as in the complete set. Here is a visualization of the cross-validation behavior. 3.1.2.3. Cross-validation iterators for grouped data.  The i.i.d. assumption is broken if the underlying generative process yield groups of dependent samples. Such a grouping of data is domain specific. An example would be when there is medical data collected from multiple patients, with multiple samples taken from each patient. And such data is likely to be dependent on the individual group. In our example, the patient id for each sample will be its group identifier. In this case we would like to know if a model trained on a particular set of groups generalizes well to the unseen groups. To measure this, we need to ensure that all the samples in the validation fold come from groups that are not represented at all in the paired training fold. The following cross-validation splitters can be used to do that. The grouping identifier for the samples is specified via the parameter. 3.1.2.3.1. Group k-fold  GroupKFold is a variation of k-fold which ensures that the same group is not represented in both testing and training sets. For example if the data is obtained from different subjects with several samples per-subject and if the model is flexible enough to learn from highly person specific features it could fail to generalize to new subjects. GroupKFold makes it possible to detect this kind of overfitting situations. Imagine you have three subjects, each with an associated number from 1 to 3: Each subject is in a different testing fold, and the same subject is never in both testing and training. Notice that the folds do not have exactly the same size due to the imbalance in the data. Here is a visualization of the cross-validation behavior. 3.1.2.3.2. Leave One Group Out  LeaveOneGroupOut is a cross-validation scheme which holds out the samples according to a third-party provided array of integer groups. This group information can be used to encode arbitrary domain specific pre-defined cross-validation folds. Each training set is thus constituted by all the samples except the ones related to a specific group. For example, in the cases of multiple experiments, LeaveOneGroupOut can be used to create a cross-validation based on the different experiments: we create a training set using the samples of all the experiments except one: Another common application is to use time information: for instance the groups could be the year of collection of the samples and thus allow for cross-validation against time-based splits. 3.1.2.3.3. Leave P Groups Out  LeavePGroupsOut is similar as LeaveOneGroupOut , but removes samples related to groups for each training/test set. Example of Leave-2-Group Out: 3.1.2.3.4. Group Shuffle Split  The GroupShuffleSplit iterator behaves as a combination of ShuffleSplit and LeavePGroupsOut , and generates a sequence of randomized partitions in which a subset of groups are held out for each split. Here is a usage example: Here is a visualization of the cross-validation behavior. This class is useful when the behavior of LeavePGroupsOut is desired, but the number of groups is large enough that generating all possible partitions with groups withheld would be prohibitively expensive. In such a scenario, GroupShuffleSplit provides a random sample (with replacement) of the train / test splits generated by LeavePGroupsOut . 3.1.2.4. Predefined Fold-Splits / Validation-Sets  For some datasets, a pre-defined split of the data into training- and validation fold or into several cross-validation folds already exists. Using PredefinedSplit it is possible to use these folds e.g. when searching for hyperparameters. For example, when using a validation set, set the to 0 for all samples that are part of the validation set, and to -1 for all other samples. 3.1.2.5. Cross validation of time series data  Time series data is characterised by the correlation between observations that are near in time ( autocorrelation ). However, classical cross-validation techniques such as KFold and ShuffleSplit assume the samples are independent and identically distributed, and would result in unreasonable correlation between training and testing instances (yielding poor estimates of generalisation error) on time series data. Therefore, it is very important to evaluate our model for time series data on the “future” observations least like those that are used to train the model. To achieve this, one solution is provided by TimeSeriesSplit . 3.1.2.5.1. Time Series Split  TimeSeriesSplit is a variation of k-fold which returns first folds as train set and the th fold as test set. Note that unlike standard cross-validation methods, successive training sets are supersets of those that come before them. Also, it adds all surplus data to the first training partition, which is always used to train the model. This class can be used to cross-validate time series data samples that are observed at fixed time intervals. Example of 3-split time series cross-validation on a dataset with 6 samples: Here is a visualization of the cross-validation behavior. ',\n",
       " 'sklearn_3_1_cross-validation_evaluating_estimator_performance\\n3.1. Cross-validation: evaluating estimator performance\\nmodules/cross_validation.html\\n 3.1.3. A note on shuffling  If the data ordering is not arbitrary (e.g. samples with the same class label are contiguous), shuffling it first may be essential to get a meaningful cross- validation result. However, the opposite may be true if the samples are not independently and identically distributed. For example, if samples correspond to news articles, and are ordered by their time of publication, then shuffling the data will likely lead to a model that is overfit and an inflated validation score: it will be tested on samples that are artificially similar (close in time) to training samples. Some cross validation iterators, such as KFold , have an inbuilt option to shuffle the data indices before splitting them. Note that: This consumes less memory than shuffling the data directly. By default no shuffling occurs, including for the (stratified) K fold cross- validation performed by specifying to cross_val_score , grid search, etc. Keep in mind that train_test_split still returns a random split. The parameter defaults to , meaning that the shuffling will be different every time is iterated. However, will use the same shuffling for each set of parameters validated by a single call to its method. To get identical results for each split, set to an integer. ',\n",
       " 'sklearn_3_1_cross-validation_evaluating_estimator_performance\\n3.1. Cross-validation: evaluating estimator performance\\nmodules/cross_validation.html\\n 3.1.4. Cross validation and model selection  Cross validation iterators can also be used to directly perform model selection using Grid Search for the optimal hyperparameters of the model. This is the topic of the next section: Tuning the hyper-parameters of an estimator . ',\n",
       " 'sklearn_3_2_tuning_the_hyper-parameters_of_an_estimator\\n3.2. Tuning the hyper-parameters of an estimator\\nmodules/grid_search.html\\n 3.2.1. Exhaustive Grid Search  The grid search provided by GridSearchCV exhaustively generates candidates from a grid of parameter values specified with the parameter. For instance, the following : specifies that two grids should be explored: one with a linear kernel and C values in [1, 10, 100, 1000], and the second one with an RBF kernel, and the cross-product of C values ranging in [1, 10, 100, 1000] and gamma values in [0.001, 0.0001]. The GridSearchCV instance implements the usual estimator API: when “fitting” it on a dataset all the possible combinations of parameter values are evaluated and the best combination is retained. Examples: See Parameter estimation using grid search with cross-validation for an example of Grid Search computation on the digits dataset. See Sample pipeline for text feature extraction and evaluation for an example of Grid Search coupling parameters from a text documents feature extractor (n-gram count vectorizer and TF-IDF transformer) with a classifier (here a linear SVM trained with SGD with either elastic net or L2 penalty) using a pipeline.Pipeline instance. See Nested versus non-nested cross-validation for an example of Grid Search within a cross validation loop on the iris dataset. This is the best practice for evaluating the performance of a model with grid search. See Demonstration of multi-metric evaluation on cross_val_score and GridSearchCV for an example of GridSearchCV being used to evaluate multiple metrics simultaneously. See Balance model complexity and cross-validated score for an example of using interface in GridSearchCV . The example shows how this interface adds certain amount of flexibility in identifying the “best” estimator. This interface can also be used in multiple metrics evaluation. ',\n",
       " 'sklearn_3_2_tuning_the_hyper-parameters_of_an_estimator\\n3.2. Tuning the hyper-parameters of an estimator\\nmodules/grid_search.html\\n 3.2.2. Randomized Parameter Optimization  While using a grid of parameter settings is currently the most widely used method for parameter optimization, other search methods have more favourable properties. RandomizedSearchCV implements a randomized search over parameters, where each setting is sampled from a distribution over possible parameter values. This has two main benefits over an exhaustive search: A budget can be chosen independent of the number of parameters and possible values. Adding parameters that do not influence the performance does not decrease efficiency. Specifying how parameters should be sampled is done using a dictionary, very similar to specifying parameters for GridSearchCV . Additionally, a computation budget, being the number of sampled candidates or sampling iterations, is specified using the parameter. For each parameter, either a distribution over possible values or a list of discrete choices (which will be sampled uniformly) can be specified: This example uses the module, which contains many useful distributions for sampling parameters, such as , , or . In principle, any function can be passed that provides a (random variate sample) method to sample a value. A call to the function should provide independent random samples from possible parameter values on consecutive calls. Warning The distributions in prior to version scipy 0.16 do not allow specifying a random state. Instead, they use the global numpy random state, that can be seeded via or set using . However, beginning scikit-learn 0.18, the sklearn.model_selection module sets the random state provided by the user if scipy > 0.16 is also available. For continuous parameters, such as above, it is important to specify a continuous distribution to take full advantage of the randomization. This way, increasing will always lead to a finer search. A continuous log-uniform random variable is available through loguniform . This is a continuous version of log-spaced parameters. For example to specify above, can be used instead of or . This is an alias to SciPy’s stats.reciprocal . Mirroring the example above in grid search, we can specify a continuous random variable that is log-uniformly distributed between and : Examples: Comparing randomized search and grid search for hyperparameter estimation compares the usage and efficiency of randomized search and grid search. References: Bergstra, J. and Bengio, Y., Random search for hyper-parameter optimization, The Journal of Machine Learning Research (2012) ',\n",
       " 'sklearn_3_2_tuning_the_hyper-parameters_of_an_estimator\\n3.2. Tuning the hyper-parameters of an estimator\\nmodules/grid_search.html\\n 3.2.3. Tips for parameter search  3.2.3.1. Specifying an objective metric  By default, parameter search uses the function of the estimator to evaluate a parameter setting. These are the sklearn.metrics.accuracy_score for classification and sklearn.metrics.r2_score for regression. For some applications, other scoring functions are better suited (for example in unbalanced classification, the accuracy score is often uninformative). An alternative scoring function can be specified via the parameter to GridSearchCV , RandomizedSearchCV and many of the specialized cross-validation tools described below. See The scoring parameter: defining model evaluation rules for more details. 3.2.3.2. Specifying multiple metrics for evaluation  and allow specifying multiple metrics for the parameter. Multimetric scoring can either be specified as a list of strings of predefined scores names or a dict mapping the scorer name to the scorer function and/or the predefined scorer name(s). See Using multiple metric evaluation for more details. When specifying multiple metrics, the parameter must be set to the metric (string) for which the will be found and used to build the on the whole dataset. If the search should not be refit, set . Leaving refit to the default value will result in an error when using multiple metrics. See Demonstration of multi-metric evaluation on cross_val_score and GridSearchCV for an example usage. 3.2.3.3. Composite estimators and parameter spaces  GridSearchCV and RandomizedSearchCV allow searching over parameters of composite or nested estimators such as Pipeline , ColumnTransformer , VotingClassifier or CalibratedClassifierCV using a dedicated syntax: Here, is the parameter name of the nested estimator, in this case . If the meta-estimator is constructed as a collection of estimators as in , then refers to the name of the estimator, see Nested parameters . In practice, there can be several levels of nesting: 3.2.3.4. Model selection: development and evaluation  Model selection by evaluating various parameter settings can be seen as a way to use the labeled data to “train” the parameters of the grid. When evaluating the resulting model it is important to do it on held-out samples that were not seen during the grid search process: it is recommended to split the data into a development set (to be fed to the instance) and an evaluation set to compute performance metrics. This can be done by using the train_test_split utility function. 3.2.3.5. Parallelism  GridSearchCV and RandomizedSearchCV evaluate each parameter setting independently. Computations can be run in parallel if your OS supports it, by using the keyword . See function signature for more details. 3.2.3.6. Robustness to failure  Some parameter settings may result in a failure to one or more folds of the data. By default, this will cause the entire search to fail, even if some parameter settings could be fully evaluated. Setting (or ) will make the procedure robust to such failure, issuing a warning and setting the score for that fold to 0 (or ), but completing the search. ',\n",
       " 'sklearn_3_2_tuning_the_hyper-parameters_of_an_estimator\\n3.2. Tuning the hyper-parameters of an estimator\\nmodules/grid_search.html\\n 3.2.4. Alternatives to brute force parameter search  3.2.4.1. Model specific cross-validation  Some models can fit data for a range of values of some parameter almost as efficiently as fitting the estimator for a single value of the parameter. This feature can be leveraged to perform a more efficient cross-validation used for model selection of this parameter. The most common parameter amenable to this strategy is the parameter encoding the strength of the regularizer. In this case we say that we compute the regularization path of the estimator. Here is the list of such models: linear_model.ElasticNetCV (*[,\\xa0l1_ratio,\\xa0…]) Elastic Net model with iterative fitting along a regularization path. linear_model.LarsCV (*[,\\xa0fit_intercept,\\xa0…]) Cross-validated Least Angle Regression model. linear_model.LassoCV (*[,\\xa0eps,\\xa0n_alphas,\\xa0…]) Lasso linear model with iterative fitting along a regularization path. linear_model.LassoLarsCV (*[,\\xa0fit_intercept,\\xa0…]) Cross-validated Lasso, using the LARS algorithm. linear_model.LogisticRegressionCV (*[,\\xa0Cs,\\xa0…]) Logistic Regression CV (aka logit, MaxEnt) classifier. linear_model.MultiTaskElasticNetCV (*[,\\xa0…]) Multi-task L1/L2 ElasticNet with built-in cross-validation. linear_model.MultiTaskLassoCV (*[,\\xa0eps,\\xa0…]) Multi-task Lasso model trained with L1/L2 mixed-norm as regularizer. linear_model.OrthogonalMatchingPursuitCV (*) Cross-validated Orthogonal Matching Pursuit model (OMP). linear_model.RidgeCV ([alphas,\\xa0…]) Ridge regression with built-in cross-validation. linear_model.RidgeClassifierCV ([alphas,\\xa0…]) Ridge classifier with built-in cross-validation. 3.2.4.2. Information Criterion  Some models can offer an information-theoretic closed-form formula of the optimal estimate of the regularization parameter by computing a single regularization path (instead of several when using cross-validation). Here is the list of models benefiting from the Akaike Information Criterion (AIC) or the Bayesian Information Criterion (BIC) for automated model selection: linear_model.LassoLarsIC ([criterion,\\xa0…]) Lasso model fit with Lars using BIC or AIC for model selection 3.2.4.3. Out of Bag Estimates  When using ensemble methods base upon bagging, i.e. generating new training sets using sampling with replacement, part of the training set remains unused. For each classifier in the ensemble, a different part of the training set is left out. This left out portion can be used to estimate the generalization error without having to rely on a separate validation set. This estimate comes “for free” as no additional data is needed and can be used for model selection. This is currently implemented in the following classes: ensemble.RandomForestClassifier ([…]) A random forest classifier. ensemble.RandomForestRegressor ([…]) A random forest regressor. ensemble.ExtraTreesClassifier ([…]) An extra-trees classifier. ensemble.ExtraTreesRegressor ([n_estimators,\\xa0…]) An extra-trees regressor. ensemble.GradientBoostingClassifier (*[,\\xa0…]) Gradient Boosting for classification. ensemble.GradientBoostingRegressor (*[,\\xa0…]) Gradient Boosting for regression. ',\n",
       " 'sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions\\n3.3. Metrics and scoring: quantifying the quality of predictions\\nmodules/model_evaluation.html\\n 3.3.1. The parameter: defining model evaluation rules  Model selection and evaluation using tools, such as model_selection.GridSearchCV and model_selection.cross_val_score , take a parameter that controls what metric they apply to the estimators evaluated. 3.3.1.1. Common cases: predefined values  For the most common use cases, you can designate a scorer object with the parameter; the table below shows all possible values. All scorer objects follow the convention that higher return values are better than lower return values . Thus metrics which measure the distance between the model and the data, like metrics.mean_squared_error , are available as neg_mean_squared_error which return the negated value of the metric. Scoring Function Comment Classification ‘accuracy’ metrics.accuracy_score ‘balanced_accuracy’ metrics.balanced_accuracy_score ‘average_precision’ metrics.average_precision_score ‘neg_brier_score’ metrics.brier_score_loss ‘f1’ metrics.f1_score for binary targets ‘f1_micro’ metrics.f1_score micro-averaged ‘f1_macro’ metrics.f1_score macro-averaged ‘f1_weighted’ metrics.f1_score weighted average ‘f1_samples’ metrics.f1_score by multilabel sample ‘neg_log_loss’ metrics.log_loss requires support ‘precision’ etc. metrics.precision_score suffixes apply as with ‘f1’ ‘recall’ etc. metrics.recall_score suffixes apply as with ‘f1’ ‘jaccard’ etc. metrics.jaccard_score suffixes apply as with ‘f1’ ‘roc_auc’ metrics.roc_auc_score ‘roc_auc_ovr’ metrics.roc_auc_score ‘roc_auc_ovo’ metrics.roc_auc_score ‘roc_auc_ovr_weighted’ metrics.roc_auc_score ‘roc_auc_ovo_weighted’ metrics.roc_auc_score Clustering ‘adjusted_mutual_info_score’ metrics.adjusted_mutual_info_score ‘adjusted_rand_score’ metrics.adjusted_rand_score ‘completeness_score’ metrics.completeness_score ‘fowlkes_mallows_score’ metrics.fowlkes_mallows_score ‘homogeneity_score’ metrics.homogeneity_score ‘mutual_info_score’ metrics.mutual_info_score ‘normalized_mutual_info_score’ metrics.normalized_mutual_info_score ‘v_measure_score’ metrics.v_measure_score Regression ‘explained_variance’ metrics.explained_variance_score ‘max_error’ metrics.max_error ‘neg_mean_absolute_error’ metrics.mean_absolute_error ‘neg_mean_squared_error’ metrics.mean_squared_error ‘neg_root_mean_squared_error’ metrics.mean_squared_error ‘neg_mean_squared_log_error’ metrics.mean_squared_log_error ‘neg_median_absolute_error’ metrics.median_absolute_error ‘r2’ metrics.r2_score ‘neg_mean_poisson_deviance’ metrics.mean_poisson_deviance ‘neg_mean_gamma_deviance’ metrics.mean_gamma_deviance Usage examples: Note The values listed by the ValueError exception correspond to the functions measuring prediction accuracy described in the following sections. The scorer objects for those functions are stored in the dictionary . 3.3.1.2. Defining your scoring strategy from metric functions  The module sklearn.metrics also exposes a set of simple functions measuring a prediction error given ground truth and prediction: functions ending with return a value to maximize, the higher the better. functions ending with or return a value to minimize, the lower the better. When converting into a scorer object using make_scorer , set the parameter to False (True by default; see the parameter description below). Metrics available for various machine learning tasks are detailed in sections below. Many metrics are not given names to be used as values, sometimes because they require additional parameters, such as fbeta_score . In such cases, you need to generate an appropriate scoring object. The simplest way to generate a callable object for scoring is by using make_scorer . That function converts metrics into callables that can be used for model evaluation. One typical use case is to wrap an existing metric function from the library with non-default values for its parameters, such as the parameter for the fbeta_score function: The second use case is to build a completely custom scorer object from a simple python function using make_scorer , which can take several parameters: the python function you want to use ( in the example below) whether the python function returns a score ( , the default) or a loss ( ). If a loss, the output of the python function is negated by the scorer object, conforming to the cross validation convention that scorers return higher values for better models. for classification metrics only: whether the python function you provided requires continuous decision certainties ( ). The default value is False. any additional parameters, such as or in f1_score . Here is an example of building custom scorers, and of using the parameter: 3.3.1.3. Implementing your own scoring object  You can generate even more flexible model scorers by constructing your own scoring object from scratch, without using the make_scorer factory. For a callable to be a scorer, it needs to meet the protocol specified by the following two rules: It can be called with parameters , where is the model that should be evaluated, is validation data, and is the ground truth target for (in the supervised case) or (in the unsupervised case). It returns a floating point number that quantifies the prediction quality on , with reference to . Again, by convention higher numbers are better, so if your scorer returns loss, that value should be negated. Note Using custom scorers in functions where n_jobs > 1 While defining the custom scoring function alongside the calling function should work out of the box with the default joblib backend (loky), importing it from another module will be a more robust approach and work independently of the joblib backend. For example, to use greater than 1 in the example below, function is saved in a user-created module ( ) and imported: 3.3.1.4. Using multiple metric evaluation  Scikit-learn also permits evaluation of multiple metrics in , and . There are two ways to specify multiple scoring metrics for the parameter: As an iterable of string metrics:: As a mapping the scorer name to the scoring function:: Note that the dict values can either be scorer functions or one of the predefined metric strings. Currently only those scorer functions that return a single score can be passed inside the dict. Scorer functions that return multiple values are not permitted and will require a wrapper to return a single metric: ',\n",
       " 'sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions\\n3.3. Metrics and scoring: quantifying the quality of predictions\\nmodules/model_evaluation.html\\n 3.3.2. Classification metrics  The sklearn.metrics module implements several loss, score, and utility functions to measure classification performance. Some metrics might require probability estimates of the positive class, confidence values, or binary decisions values. Most implementations allow each sample to provide a weighted contribution to the overall score, through the parameter. Some of these are restricted to the binary classification case: precision_recall_curve (y_true,\\xa0probas_pred,\\xa0*) Compute precision-recall pairs for different probability thresholds roc_curve (y_true,\\xa0y_score,\\xa0*[,\\xa0pos_label,\\xa0…]) Compute Receiver operating characteristic (ROC) Others also work in the multiclass case: balanced_accuracy_score (y_true,\\xa0y_pred,\\xa0*[,\\xa0…]) Compute the balanced accuracy cohen_kappa_score (y1,\\xa0y2,\\xa0*[,\\xa0labels,\\xa0…]) Cohen’s kappa: a statistic that measures inter-annotator agreement. confusion_matrix (y_true,\\xa0y_pred,\\xa0*[,\\xa0…]) Compute confusion matrix to evaluate the accuracy of a classification. hinge_loss (y_true,\\xa0pred_decision,\\xa0*[,\\xa0…]) Average hinge loss (non-regularized) matthews_corrcoef (y_true,\\xa0y_pred,\\xa0*[,\\xa0…]) Compute the Matthews correlation coefficient (MCC) roc_auc_score (y_true,\\xa0y_score,\\xa0*[,\\xa0average,\\xa0…]) Compute Area Under the Receiver Operating Characteristic Curve (ROC AUC) from prediction scores. Some also work in the multilabel case: accuracy_score (y_true,\\xa0y_pred,\\xa0*[,\\xa0…]) Accuracy classification score. classification_report (y_true,\\xa0y_pred,\\xa0*[,\\xa0…]) Build a text report showing the main classification metrics. f1_score (y_true,\\xa0y_pred,\\xa0*[,\\xa0labels,\\xa0…]) Compute the F1 score, also known as balanced F-score or F-measure fbeta_score (y_true,\\xa0y_pred,\\xa0*,\\xa0beta[,\\xa0…]) Compute the F-beta score hamming_loss (y_true,\\xa0y_pred,\\xa0*[,\\xa0sample_weight]) Compute the average Hamming loss. jaccard_score (y_true,\\xa0y_pred,\\xa0*[,\\xa0labels,\\xa0…]) Jaccard similarity coefficient score log_loss (y_true,\\xa0y_pred,\\xa0*[,\\xa0eps,\\xa0…]) Log loss, aka logistic loss or cross-entropy loss. multilabel_confusion_matrix (y_true,\\xa0y_pred,\\xa0*) Compute a confusion matrix for each class or sample precision_recall_fscore_support (y_true,\\xa0…) Compute precision, recall, F-measure and support for each class precision_score (y_true,\\xa0y_pred,\\xa0*[,\\xa0labels,\\xa0…]) Compute the precision recall_score (y_true,\\xa0y_pred,\\xa0*[,\\xa0labels,\\xa0…]) Compute the recall roc_auc_score (y_true,\\xa0y_score,\\xa0*[,\\xa0average,\\xa0…]) Compute Area Under the Receiver Operating Characteristic Curve (ROC AUC) from prediction scores. zero_one_loss (y_true,\\xa0y_pred,\\xa0*[,\\xa0…]) Zero-one classification loss. And some work with binary and multilabel (but not multiclass) problems: average_precision_score (y_true,\\xa0y_score,\\xa0*) Compute average precision (AP) from prediction scores In the following sub-sections, we will describe each of those functions, preceded by some notes on common API and metric definition. 3.3.2.1. From binary to multiclass and multilabel  Some metrics are essentially defined for binary classification tasks (e.g. f1_score , roc_auc_score ). In these cases, by default only the positive label is evaluated, assuming by default that the positive class is labelled (though this may be configurable through the parameter). In extending a binary metric to multiclass or multilabel problems, the data is treated as a collection of binary problems, one for each class. There are then a number of ways to average binary metric calculations across the set of classes, each of which may be useful in some scenario. Where available, you should select among these using the parameter. simply calculates the mean of the binary metrics, giving equal weight to each class. In problems where infrequent classes are nonetheless important, macro-averaging may be a means of highlighting their performance. On the other hand, the assumption that all classes are equally important is often untrue, such that macro-averaging will over-emphasize the typically low performance on an infrequent class. accounts for class imbalance by computing the average of binary metrics in which each class’s score is weighted by its presence in the true data sample. gives each sample-class pair an equal contribution to the overall metric (except as a result of sample-weight). Rather than summing the metric per class, this sums the dividends and divisors that make up the per-class metrics to calculate an overall quotient. Micro-averaging may be preferred in multilabel settings, including multiclass classification where a majority class is to be ignored. applies only to multilabel problems. It does not calculate a per-class measure, instead calculating the metric over the true and predicted classes for each sample in the evaluation data, and returning their ( -weighted) average. Selecting will return an array with the score for each class. While multiclass data is provided to the metric, like binary targets, as an array of class labels, multilabel data is specified as an indicator matrix, in which cell has value 1 if sample has label and value 0 otherwise. 3.3.2.2. Accuracy score  The accuracy_score function computes the accuracy , either the fraction (default) or the count (normalizeFalse) of correct predictions. In multilabel classification, the function returns the subset accuracy. If the entire set of predicted labels for a sample strictly match with the true set of labels, then the subset accuracy is 1.0; otherwise it is 0.0. If is the predicted value of the -th sample and is the corresponding true value, then the fraction of correct predictions over is defined as \\\\[\\\\texttt{accuracy}(y, \\\\hat{y})  \\\\frac{1}{n_\\\\text{samples}} \\\\sum_{i0}^{n_\\\\text{samples}-1} 1(\\\\hat{y}_i  y_i)\\\\] where is the indicator function . In the multilabel case with binary label indicators: Example: See Test with permutations the significance of a classification score for an example of accuracy score usage using permutations of the dataset. 3.3.2.3. Balanced accuracy score  The balanced_accuracy_score function computes the balanced accuracy , which avoids inflated performance estimates on imbalanced datasets. It is the macro-average of recall scores per class or, equivalently, raw accuracy where each sample is weighted according to the inverse prevalence of its true class. Thus for balanced datasets, the score is equal to accuracy. In the binary case, balanced accuracy is equal to the arithmetic mean of sensitivity (true positive rate) and specificity (true negative rate), or the area under the ROC curve with binary predictions rather than scores: \\\\[\\\\texttt{balanced-accuracy}  \\\\frac{1}{2}\\\\left( \\\\frac{TP}{TP + FN} + \\\\frac{TN}{TN + FP}\\\\right )\\\\] If the classifier performs equally well on either class, this term reduces to the conventional accuracy (i.e., the number of correct predictions divided by the total number of predictions). In contrast, if the conventional accuracy is above chance only because the classifier takes advantage of an imbalanced test set, then the balanced accuracy, as appropriate, will drop to . The score ranges from 0 to 1, or when is used, it rescaled to the range to 1, inclusive, with performance at random scoring 0. If is the true value of the -th sample, and is the corresponding sample weight, then we adjust the sample weight to: \\\\[\\\\hat{w}_i  \\\\frac{w_i}{\\\\sum_j{1(y_j  y_i) w_j}}\\\\] where is the indicator function . Given predicted for sample , balanced accuracy is defined as: \\\\[\\\\texttt{balanced-accuracy}(y, \\\\hat{y}, w)  \\\\frac{1}{\\\\sum{\\\\hat{w}_i}} \\\\sum_i 1(\\\\hat{y}_i  y_i) \\\\hat{w}_i\\\\] With , balanced accuracy reports the relative increase from . In the binary case, this is also known as *Youden’s J statistic* , or informedness . Note The multiclass definition here seems the most reasonable extension of the metric used in binary classification, though there is no certain consensus in the literature: Our definition: [Mosley2013] , [Kelleher2015] and [Guyon2015] , where [Guyon2015] adopt the adjusted version to ensure that random predictions have a score of and perfect predictions have a score of .. Class balanced accuracy as described in [Mosley2013] : the minimum between the precision and the recall for each class is computed. Those values are then averaged over the total number of classes to get the balanced accuracy. Balanced Accuracy as described in [Urbanowicz2015] : the average of sensitivity and specificity is computed for each class and then averaged over total number of classes. References: Guyon2015 ( 1 , 2 ) I. Guyon, K. Bennett, G. Cawley, H.J. Escalante, S. Escalera, T.K. Ho, N. Macià, B. Ray, M. Saeed, A.R. Statnikov, E. Viegas, Design of the 2015 ChaLearn AutoML Challenge , IJCNN 2015. Mosley2013 ( 1 , 2 ) L. Mosley, A balanced approach to the multi-class imbalance problem , IJCV 2010. Kelleher2015 John. D. Kelleher, Brian Mac Namee, Aoife D’Arcy, Fundamentals of Machine Learning for Predictive Data Analytics: Algorithms, Worked Examples, and Case Studies , 2015. Urbanowicz2015 Urbanowicz R.J., Moore, J.H. ExSTraCS 2.0: description and evaluation of a scalable learning classifier system , Evol. Intel. (2015) 8: 89. 3.3.2.4. Cohen’s kappa  The function cohen_kappa_score computes Cohen’s kappa statistic. This measure is intended to compare labelings by different human annotators, not a classifier versus a ground truth. The kappa score (see docstring) is a number between -1 and 1. Scores above .8 are generally considered good agreement; zero or lower means no agreement (practically random labels). Kappa scores can be computed for binary or multiclass problems, but not for multilabel problems (except by manually computing a per-label score) and not for more than two annotators. 3.3.2.5. Confusion matrix  The confusion_matrix function evaluates classification accuracy by computing the confusion matrix with each row corresponding to the true class (Wikipedia and other references may use different convention for axes). By definition, entry in a confusion matrix is the number of observations actually in group , but predicted to be in group . Here is an example: plot_confusion_matrix can be used to visually represent a confusion matrix as shown in the Confusion matrix example, which creates the following figure: The parameter allows to report ratios instead of counts. The confusion matrix can be normalized in 3 different ways: , , and which will divide the counts by the sum of each columns, rows, or the entire matrix, respectively. For binary problems, we can get counts of true negatives, false positives, false negatives and true positives as follows: Example: See Confusion matrix for an example of using a confusion matrix to evaluate classifier output quality. See Recognizing hand-written digits for an example of using a confusion matrix to classify hand-written digits. See Classification of text documents using sparse features for an example of using a confusion matrix to classify text documents. 3.3.2.6. Classification report  The classification_report function builds a text report showing the main classification metrics. Here is a small example with custom and inferred labels: Example: See Recognizing hand-written digits for an example of classification report usage for hand-written digits. See Classification of text documents using sparse features for an example of classification report usage for text documents. See Parameter estimation using grid search with cross-validation for an example of classification report usage for grid search with nested cross-validation. 3.3.2.7. Hamming loss  The hamming_loss computes the average Hamming loss or Hamming distance between two sets of samples. If is the predicted value for the -th label of a given sample, is the corresponding true value, and is the number of classes or labels, then the Hamming loss between two samples is defined as: \\\\[L_{Hamming}(y, \\\\hat{y})  \\\\frac{1}{n_\\\\text{labels}} \\\\sum_{j0}^{n_\\\\text{labels} - 1} 1(\\\\hat{y}_j \\\\not y_j)\\\\] where is the indicator function . In the multilabel case with binary label indicators: Note In multiclass classification, the Hamming loss corresponds to the Hamming distance between and which is similar to the Zero one loss function. However, while zero-one loss penalizes prediction sets that do not strictly match true sets, the Hamming loss penalizes individual labels. Thus the Hamming loss, upper bounded by the zero-one loss, is always between zero and one, inclusive; and predicting a proper subset or superset of the true labels will give a Hamming loss between zero and one, exclusive. 3.3.2.8. Precision, recall and F-measures  Intuitively, precision is the ability of the classifier not to label as positive a sample that is negative, and recall is the ability of the classifier to find all the positive samples. The F-measure ( and measures) can be interpreted as a weighted harmonic mean of the precision and recall. A measure reaches its best value at 1 and its worst score at 0. With , and are equivalent, and the recall and the precision are equally important. The precision_recall_curve computes a precision-recall curve from the ground truth label and a score given by the classifier by varying a decision threshold. The average_precision_score function computes the average precision (AP) from prediction scores. The value is between 0 and 1 and higher is better. AP is defined as \\\\[\\\\text{AP}  \\\\sum_n (R_n - R_{n-1}) P_n\\\\] where and are the precision and recall at the nth threshold. With random predictions, the AP is the fraction of positive samples. References [Manning2008] and [Everingham2010] present alternative variants of AP that interpolate the precision-recall curve. Currently, average_precision_score does not implement any interpolated variant. References [Davis2006] and [Flach2015] describe why a linear interpolation of points on the precision-recall curve provides an overly-optimistic measure of classifier performance. This linear interpolation is used when computing area under the curve with the trapezoidal rule in auc . Several functions allow you to analyze the precision, recall and F-measures score: average_precision_score (y_true,\\xa0y_score,\\xa0*) Compute average precision (AP) from prediction scores f1_score (y_true,\\xa0y_pred,\\xa0*[,\\xa0labels,\\xa0…]) Compute the F1 score, also known as balanced F-score or F-measure fbeta_score (y_true,\\xa0y_pred,\\xa0*,\\xa0beta[,\\xa0…]) Compute the F-beta score precision_recall_curve (y_true,\\xa0probas_pred,\\xa0*) Compute precision-recall pairs for different probability thresholds precision_recall_fscore_support (y_true,\\xa0…) Compute precision, recall, F-measure and support for each class precision_score (y_true,\\xa0y_pred,\\xa0*[,\\xa0labels,\\xa0…]) Compute the precision recall_score (y_true,\\xa0y_pred,\\xa0*[,\\xa0labels,\\xa0…]) Compute the recall Note that the precision_recall_curve function is restricted to the binary case. The average_precision_score function works only in binary classification and multilabel indicator format. The plot_precision_recall_curve function plots the precision recall as follows. Examples: See Classification of text documents using sparse features for an example of f1_score usage to classify text documents. See Parameter estimation using grid search with cross-validation for an example of precision_score and recall_score usage to estimate parameters using grid search with nested cross-validation. See Precision-Recall for an example of precision_recall_curve usage to evaluate classifier output quality. References: Manning2008 C.D. Manning, P. Raghavan, H. Schütze, Introduction to Information Retrieval , 2008. Everingham2010 M. Everingham, L. Van Gool, C.K.I. Williams, J. Winn, A. Zisserman, The Pascal Visual Object Classes (VOC) Challenge , IJCV 2010. Davis2006 J. Davis, M. Goadrich, The Relationship Between Precision-Recall and ROC Curves , ICML 2006. Flach2015 P.A. Flach, M. Kull, Precision-Recall-Gain Curves: PR Analysis Done Right , NIPS 2015. 3.3.2.8.1. Binary classification  In a binary classification task, the terms ‘’positive’’ and ‘’negative’’ refer to the classifier’s prediction, and the terms ‘’true’’ and ‘’false’’ refer to whether that prediction corresponds to the external judgment (sometimes known as the ‘’observation’’). Given these definitions, we can formulate the following table: Actual class (observation) Predicted class (expectation) tp (true positive) Correct result fp (false positive) Unexpected result fn (false negative) Missing result tn (true negative) Correct absence of result In this context, we can define the notions of precision, recall and F-measure: \\\\[\\\\text{precision}  \\\\frac{tp}{tp + fp},\\\\] \\\\[\\\\text{recall}  \\\\frac{tp}{tp + fn},\\\\] \\\\[F_\\\\beta  (1 + \\\\beta^2) \\\\frac{\\\\text{precision} \\\\times \\\\text{recall}}{\\\\beta^2 \\\\text{precision} + \\\\text{recall}}.\\\\] Here are some small examples in binary classification: 3.3.2.8.2. Multiclass and multilabel classification  In multiclass and multilabel classification task, the notions of precision, recall, and F-measures can be applied to each label independently. There are a few ways to combine results across labels, specified by the argument to the average_precision_score (multilabel only), f1_score , fbeta_score , precision_recall_fscore_support , precision_score and recall_score functions, as described above . Note that if all labels are included, “micro”-averaging in a multiclass setting will produce precision, recall and that are all identical to accuracy. Also note that “weighted” averaging may produce an F-score that is not between precision and recall. To make this more explicit, consider the following notation: the set of predicted pairs the set of true pairs the set of labels the set of samples the subset of with sample , i.e. the subset of with label similarly, and are subsets of for some sets and (Conventions vary on handling ; this implementation uses , and similar for .) Then the metrics are defined as: Precision Recall F_beta For multiclass classification with a “negative class”, it is possible to exclude some labels: Similarly, labels not present in the data sample may be accounted for in macro-averaging. 3.3.2.9. Jaccard similarity coefficient score  The jaccard_score function computes the average of Jaccard similarity coefficients , also called the Jaccard index, between pairs of label sets. The Jaccard similarity coefficient of the -th samples, with a ground truth label set and predicted label set , is defined as \\\\[J(y_i, \\\\hat{y}_i)  \\\\frac{|y_i \\\\cap \\\\hat{y}_i|}{|y_i \\\\cup \\\\hat{y}_i|}.\\\\] jaccard_score works like precision_recall_fscore_support as a naively set-wise measure applying natively to binary targets, and extended to apply to multilabel and multiclass through the use of (see above ). In the binary case: In the multilabel case with binary label indicators: Multiclass problems are binarized and treated like the corresponding multilabel problem: 3.3.2.10. Hinge loss  The hinge_loss function computes the average distance between the model and the data using hinge loss , a one-sided metric that considers only prediction errors. (Hinge loss is used in maximal margin classifiers such as support vector machines.) If the labels are encoded with +1 and -1, : is the true value, and is the predicted decisions as output by , then the hinge loss is defined as: \\\\[L_\\\\text{Hinge}(y, w)  \\\\max\\\\left\\\\{1 - wy, 0\\\\right\\\\}  \\\\left|1 - wy\\\\right|_+\\\\] If there are more than two labels, hinge_loss uses a multiclass variant due to Crammer & Singer. Here is the paper describing it. If is the predicted decision for true label and is the maximum of the predicted decisions for all other labels, where predicted decisions are output by decision function, then multiclass hinge loss is defined by: \\\\[L_\\\\text{Hinge}(y_w, y_t)  \\\\max\\\\left\\\\{1 + y_t - y_w, 0\\\\right\\\\}\\\\] Here a small example demonstrating the use of the hinge_loss function with a svm classifier in a binary class problem: Here is an example demonstrating the use of the hinge_loss function with a svm classifier in a multiclass problem: 3.3.2.11. Log loss  Log loss, also called logistic regression loss or cross-entropy loss, is defined on probability estimates. It is commonly used in (multinomial) logistic regression and neural networks, as well as in some variants of expectation-maximization, and can be used to evaluate the probability outputs ( ) of a classifier instead of its discrete predictions. For binary classification with a true label and a probability estimate , the log loss per sample is the negative log-likelihood of the classifier given the true label: \\\\[L_{\\\\log}(y, p)  -\\\\log \\\\operatorname{Pr}(y|p)  -(y \\\\log (p) + (1 - y) \\\\log (1 - p))\\\\] This extends to the multiclass case as follows. Let the true labels for a set of samples be encoded as a 1-of-K binary indicator matrix , i.e., if sample has label taken from a set of labels. Let be a matrix of probability estimates, with . Then the log loss of the whole set is \\\\[L_{\\\\log}(Y, P)  -\\\\log \\\\operatorname{Pr}(Y|P)  - \\\\frac{1}{N} \\\\sum_{i0}^{N-1} \\\\sum_{k0}^{K-1} y_{i,k} \\\\log p_{i,k}\\\\] To see how this generalizes the binary log loss given above, note that in the binary case, and , so expanding the inner sum over gives the binary log loss. The log_loss function computes log loss given a list of ground-truth labels and a probability matrix, as returned by an estimator’s method. The first in denotes 90% probability that the first sample has label 0. The log loss is non-negative. 3.3.2.12. Matthews correlation coefficient  The matthews_corrcoef function computes the Matthew’s correlation coefficient (MCC) for binary classes. Quoting Wikipedia: “The Matthews correlation coefficient is used in machine learning as a measure of the quality of binary (two-class) classifications. It takes into account true and false positives and negatives and is generally regarded as a balanced measure which can be used even if the classes are of very different sizes. The MCC is in essence a correlation coefficient value between -1 and +1. A coefficient of +1 represents a perfect prediction, 0 an average random prediction and -1 an inverse prediction. The statistic is also known as the phi coefficient.” In the binary (two-class) case, , , and are respectively the number of true positives, true negatives, false positives and false negatives, the MCC is defined as \\\\[MCC  \\\\frac{tp \\\\times tn - fp \\\\times fn}{\\\\sqrt{(tp + fp)(tp + fn)(tn + fp)(tn + fn)}}.\\\\] In the multiclass case, the Matthews correlation coefficient can be defined in terms of a confusion_matrix for classes. To simplify the definition consider the following intermediate variables: the number of times class truly occurred, the number of times class was predicted, the total number of samples correctly predicted, the total number of samples. Then the multiclass MCC is defined as: \\\\[MCC  \\\\frac{ c \\\\times s - \\\\sum_{k}^{K} p_k \\\\times t_k }{\\\\sqrt{ (s^2 - \\\\sum_{k}^{K} p_k^2) \\\\times (s^2 - \\\\sum_{k}^{K} t_k^2) }}\\\\] When there are more than two labels, the value of the MCC will no longer range between -1 and +1. Instead the minimum value will be somewhere between -1 and 0 depending on the number and distribution of ground true labels. The maximum value is always +1. Here is a small example illustrating the usage of the matthews_corrcoef function: 3.3.2.13. Multi-label confusion matrix  The multilabel_confusion_matrix function computes class-wise (default) or sample-wise (samplewiseTrue) multilabel confusion matrix to evaluate the accuracy of a classification. multilabel_confusion_matrix also treats multiclass data as if it were multilabel, as this is a transformation commonly applied to evaluate multiclass problems with binary classification metrics (such as precision, recall, etc.). When calculating class-wise multilabel confusion matrix , the count of true negatives for class is , false negatives is , true positives is and false positives is . Here is an example demonstrating the use of the multilabel_confusion_matrix function with multilabel indicator matrix input: Or a confusion matrix can be constructed for each sample’s labels: Here is an example demonstrating the use of the multilabel_confusion_matrix function with multiclass input: Here are some examples demonstrating the use of the multilabel_confusion_matrix function to calculate recall (or sensitivity), specificity, fall out and miss rate for each class in a problem with multilabel indicator matrix input. Calculating recall (also called the true positive rate or the sensitivity) for each class: Calculating specificity (also called the true negative rate) for each class: Calculating fall out (also called the false positive rate) for each class: Calculating miss rate (also called the false negative rate) for each class: 3.3.2.14. Receiver operating characteristic (ROC)  The function roc_curve computes the receiver operating characteristic curve, or ROC curve . Quoting Wikipedia : “A receiver operating characteristic (ROC), or simply ROC curve, is a graphical plot which illustrates the performance of a binary classifier system as its discrimination threshold is varied. It is created by plotting the fraction of true positives out of the positives (TPR  true positive rate) vs. the fraction of false positives out of the negatives (FPR  false positive rate), at various threshold settings. TPR is also known as sensitivity, and FPR is one minus the specificity or true negative rate.” This function requires the true binary value and the target scores, which can either be probability estimates of the positive class, confidence values, or binary decisions. Here is a small example of how to use the roc_curve function: This figure shows an example of such an ROC curve: The roc_auc_score function computes the area under the receiver operating characteristic (ROC) curve, which is also denoted by AUC or AUROC. By computing the area under the roc curve, the curve information is summarized in one number. For more information see the Wikipedia article on AUC . In multi-label classification, the roc_auc_score function is extended by averaging over the labels as above . Compared to metrics such as the subset accuracy, the Hamming loss, or the F1 score, ROC doesn’t require optimizing a threshold for each label. The roc_auc_score function can also be used in multi-class classification. Two averaging strategies are currently supported: the one-vs-one algorithm computes the average of the pairwise ROC AUC scores, and the one-vs-rest algorithm computes the average of the ROC AUC scores for each class against all other classes. In both cases, the predicted labels are provided in an array with values from 0 to , and the scores correspond to the probability estimates that a sample belongs to a particular class. The OvO and OvR algorithms support weighting uniformly ( ) and by prevalence ( ). One-vs-one Algorithm : Computes the average AUC of all possible pairwise combinations of classes. [HT2001] defines a multiclass AUC metric weighted uniformly: \\\\[\\\\frac{2}{c(c-1)}\\\\sum_{j1}^{c}\\\\sum_{k > j}^c (\\\\text{AUC}(j | k) + \\\\text{AUC}(k | j))\\\\] where is the number of classes and is the AUC with class as the positive class and class as the negative class. In general, in the multiclass case. This algorithm is used by setting the keyword argument to and to . The [HT2001] multiclass AUC metric can be extended to be weighted by the prevalence: \\\\[\\\\frac{2}{c(c-1)}\\\\sum_{j1}^{c}\\\\sum_{k > j}^c p(j \\\\cup k)( \\\\text{AUC}(j | k) + \\\\text{AUC}(k | j))\\\\] where is the number of classes. This algorithm is used by setting the keyword argument to and to . The option returns a prevalence-weighted average as described in [FC2009] . One-vs-rest Algorithm : Computes the AUC of each class against the rest [PD2000] . The algorithm is functionally the same as the multilabel case. To enable this algorithm set the keyword argument to . Like OvO, OvR supports two types of averaging: [F2006] and [F2001] . In applications where a high false positive rate is not tolerable the parameter of roc_auc_score can be used to summarize the ROC curve up to the given limit. Examples: See Receiver Operating Characteristic (ROC) for an example of using ROC to evaluate the quality of the output of a classifier. See Receiver Operating Characteristic (ROC) with cross validation for an example of using ROC to evaluate classifier output quality, using cross-validation. See Species distribution modeling for an example of using ROC to model species distribution. References: HT2001 ( 1 , 2 ) Hand, D.J. and Till, R.J., (2001). A simple generalisation of the area under the ROC curve for multiple class classification problems. Machine learning, 45(2), pp.171-186. FC2009 Ferri, Cèsar & Hernandez-Orallo, Jose & Modroiu, R. (2009). An Experimental Comparison of Performance Measures for Classification. Pattern Recognition Letters. 30. 27-38. PD2000 Provost, F., Domingos, P. (2000). Well-trained PETs: Improving probability estimation trees (Section 6.2), CeDER Working Paper #IS-00-04, Stern School of Business, New York University. F2006 Fawcett, T., 2006. An introduction to ROC analysis. Pattern Recognition Letters, 27(8), pp. 861-874. F2001 Fawcett, T., 2001. Using rule sets to maximize ROC performance In Data Mining, 2001. Proceedings IEEE International Conference, pp. 131-138. 3.3.2.15. Zero one loss  The zero_one_loss function computes the sum or the average of the 0-1 classification loss ( ) over . By default, the function normalizes over the sample. To get the sum of the , set to . In multilabel classification, the zero_one_loss scores a subset as one if its labels strictly match the predictions, and as a zero if there are any errors. By default, the function returns the percentage of imperfectly predicted subsets. To get the count of such subsets instead, set to If is the predicted value of the -th sample and is the corresponding true value, then the 0-1 loss is defined as: \\\\[L_{0-1}(y_i, \\\\hat{y}_i)  1(\\\\hat{y}_i \\\\not y_i)\\\\] where is the indicator function . In the multilabel case with binary label indicators, where the first label set [0,1] has an error: Example: See Recursive feature elimination with cross-validation for an example of zero one loss usage to perform recursive feature elimination with cross-validation. 3.3.2.16. Brier score loss  The brier_score_loss function computes the Brier score for binary classes. Quoting Wikipedia: “The Brier score is a proper score function that measures the accuracy of probabilistic predictions. It is applicable to tasks in which predictions must assign probabilities to a set of mutually exclusive discrete outcomes.” This function returns a score of the mean square difference between the actual outcome and the predicted probability of the possible outcome. The actual outcome has to be 1 or 0 (true or false), while the predicted probability of the actual outcome can be a value between 0 and 1. The brier score loss is also between 0 to 1 and the lower the score (the mean square difference is smaller), the more accurate the prediction is. It can be thought of as a measure of the “calibration” of a set of probabilistic predictions. \\\\[BS  \\\\frac{1}{N} \\\\sum_{t1}^{N}(f_t - o_t)^2\\\\] where : is the total number of predictions, is the predicted probability of the actual outcome . Here is a small example of usage of this function:: Example: See Probability calibration of classifiers for an example of Brier score loss usage to perform probability calibration of classifiers. References: G. Brier, Verification of forecasts expressed in terms of probability , Monthly weather review 78.1 (1950) ',\n",
       " 'sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions\\n3.3. Metrics and scoring: quantifying the quality of predictions\\nmodules/model_evaluation.html\\n 3.3.3. Multilabel ranking metrics  In multilabel learning, each sample can have any number of ground truth labels associated with it. The goal is to give high scores and better rank to the ground truth labels. 3.3.3.1. Coverage error  The coverage_error function computes the average number of labels that have to be included in the final prediction such that all true labels are predicted. This is useful if you want to know how many top-scored-labels you have to predict in average without missing any true one. The best value of this metrics is thus the average number of true labels. Note Our implementation’s score is 1 greater than the one given in Tsoumakas et al., 2010. This extends it to handle the degenerate case in which an instance has 0 true labels. Formally, given a binary indicator matrix of the ground truth labels and the score associated with each label , the coverage is defined as \\\\[coverage(y, \\\\hat{f})  \\\\frac{1}{n_{\\\\text{samples}}} \\\\sum_{i0}^{n_{\\\\text{samples}} - 1} \\\\max_{j:y_{ij}  1} \\\\text{rank}_{ij}\\\\] with . Given the rank definition, ties in are broken by giving the maximal rank that would have been assigned to all tied values. Here is a small example of usage of this function: 3.3.3.2. Label ranking average precision  The label_ranking_average_precision_score function implements label ranking average precision (LRAP). This metric is linked to the average_precision_score function, but is based on the notion of label ranking instead of precision and recall. Label ranking average precision (LRAP) averages over the samples the answer to the following question: for each ground truth label, what fraction of higher-ranked labels were true labels? This performance measure will be higher if you are able to give better rank to the labels associated with each sample. The obtained score is always strictly greater than 0, and the best value is 1. If there is exactly one relevant label per sample, label ranking average precision is equivalent to the mean reciprocal rank . Formally, given a binary indicator matrix of the ground truth labels and the score associated with each label , the average precision is defined as \\\\[LRAP(y, \\\\hat{f})  \\\\frac{1}{n_{\\\\text{samples}}} \\\\sum_{i0}^{n_{\\\\text{samples}} - 1} \\\\frac{1}{||y_i||_0} \\\\sum_{j:y_{ij}  1} \\\\frac{|\\\\mathcal{L}_{ij}|}{\\\\text{rank}_{ij}}\\\\] where , , computes the cardinality of the set (i.e., the number of elements in the set), and is the “norm” (which computes the number of nonzero elements in a vector). Here is a small example of usage of this function: 3.3.3.3. Ranking loss  The label_ranking_loss function computes the ranking loss which averages over the samples the number of label pairs that are incorrectly ordered, i.e. true labels have a lower score than false labels, weighted by the inverse of the number of ordered pairs of false and true labels. The lowest achievable ranking loss is zero. Formally, given a binary indicator matrix of the ground truth labels and the score associated with each label , the ranking loss is defined as \\\\[ranking\\\\_loss(y, \\\\hat{f})  \\\\frac{1}{n_{\\\\text{samples}}} \\\\sum_{i0}^{n_{\\\\text{samples}} - 1} \\\\frac{1}{||y_i||_0(n_\\\\text{labels} - ||y_i||_0)} \\\\left|\\\\left\\\\{(k, l): \\\\hat{f}_{ik} \\\\leq \\\\hat{f}_{il}, y_{ik}  1, y_{il}  0\\xa0\\\\right\\\\}\\\\right|\\\\] where computes the cardinality of the set (i.e., the number of elements in the set) and is the “norm” (which computes the number of nonzero elements in a vector). Here is a small example of usage of this function: References: Tsoumakas, G., Katakis, I., & Vlahavas, I. (2010). Mining multi-label data. In Data mining and knowledge discovery handbook (pp. 667-685). Springer US. 3.3.3.4. Normalized Discounted Cumulative Gain  Discounted Cumulative Gain (DCG) and Normalized Discounted Cumulative Gain (NDCG) are ranking metrics; they compare a predicted order to ground-truth scores, such as the relevance of answers to a query. From the Wikipedia page for Discounted Cumulative Gain: “Discounted cumulative gain (DCG) is a measure of ranking quality. In information retrieval, it is often used to measure effectiveness of web search engine algorithms or related applications. Using a graded relevance scale of documents in a search-engine result set, DCG measures the usefulness, or gain, of a document based on its position in the result list. The gain is accumulated from the top of the result list to the bottom, with the gain of each result discounted at lower ranks” DCG orders the true targets (e.g. relevance of query answers) in the predicted order, then multiplies them by a logarithmic decay and sums the result. The sum can be truncated after the first results, in which case we call it DCG@K. NDCG, or NDCG@K is DCG divided by the DCG obtained by a perfect prediction, so that it is always between 0 and 1. Usually, NDCG is preferred to DCG. Compared with the ranking loss, NDCG can take into account relevance scores, rather than a ground-truth ranking. So if the ground-truth consists only of an ordering, the ranking loss should be preferred; if the ground-truth consists of actual usefulness scores (e.g. 0 for irrelevant, 1 for relevant, 2 for very relevant), NDCG can be used. For one sample, given the vector of continuous ground-truth values for each target , where is the number of outputs, and the prediction , which induces the ranking function , the DCG score is \\\\[\\\\sum_{r1}^{\\\\min(K, M)}\\\\frac{y_{f(r)}}{\\\\log(1 + r)}\\\\] and the NDCG score is the DCG score divided by the DCG score obtained for . References: Wikipedia entry for Discounted Cumulative Gain Jarvelin, K., & Kekalainen, J. (2002). Cumulated gain-based evaluation of IR techniques. ACM Transactions on Information Systems (TOIS), 20(4), 422-446. Wang, Y., Wang, L., Li, Y., He, D., Chen, W., & Liu, T. Y. (2013, May). A theoretical analysis of NDCG ranking measures. In Proceedings of the 26th Annual Conference on Learning Theory (COLT 2013) McSherry, F., & Najork, M. (2008, March). Computing information retrieval performance measures efficiently in the presence of tied scores. In European conference on information retrieval (pp. 414-421). Springer, Berlin, Heidelberg. ',\n",
       " 'sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions\\n3.3. Metrics and scoring: quantifying the quality of predictions\\nmodules/model_evaluation.html\\n 3.3.4. Regression metrics  The sklearn.metrics module implements several loss, score, and utility functions to measure regression performance. Some of those have been enhanced to handle the multioutput case: mean_squared_error , mean_absolute_error , explained_variance_score and r2_score . These functions have an keyword argument which specifies the way the scores or losses for each individual target should be averaged. The default is , which specifies a uniformly weighted mean over outputs. If an of shape is passed, then its entries are interpreted as weights and an according weighted average is returned. If is is specified, then all unaltered individual scores or losses will be returned in an array of shape . The r2_score and explained_variance_score accept an additional value for the parameter. This option leads to a weighting of each individual score by the variance of the corresponding target variable. This setting quantifies the globally captured unscaled variance. If the target variables are of different scale, then this score puts more importance on well explaining the higher variance variables. is the default value for r2_score for backward compatibility. This will be changed to in the future. 3.3.4.1. Explained variance score  The explained_variance_score computes the explained variance regression score . If is the estimated target output, the corresponding (correct) target output, and is Variance , the square of the standard deviation, then the explained variance is estimated as follow: \\\\[explained\\\\_{}variance(y, \\\\hat{y})  1 - \\\\frac{Var\\\\{ y - \\\\hat{y}\\\\}}{Var\\\\{y\\\\}}\\\\] The best possible score is 1.0, lower values are worse. Here is a small example of usage of the explained_variance_score function: 3.3.4.2. Max error  The max_error function computes the maximum residual error , a metric that captures the worst case error between the predicted value and the true value. In a perfectly fitted single output regression model, would be on the training set and though this would be highly unlikely in the real world, this metric shows the extent of error that the model had when it was fitted. If is the predicted value of the -th sample, and is the corresponding true value, then the max error is defined as \\\\[\\\\text{Max Error}(y, \\\\hat{y})  max(| y_i - \\\\hat{y}_i |)\\\\] Here is a small example of usage of the max_error function: The max_error does not support multioutput. 3.3.4.3. Mean absolute error  The mean_absolute_error function computes mean absolute error , a risk metric corresponding to the expected value of the absolute error loss or -norm loss. If is the predicted value of the -th sample, and is the corresponding true value, then the mean absolute error (MAE) estimated over is defined as \\\\[\\\\text{MAE}(y, \\\\hat{y})  \\\\frac{1}{n_{\\\\text{samples}}} \\\\sum_{i0}^{n_{\\\\text{samples}}-1} \\\\left| y_i - \\\\hat{y}_i \\\\right|.\\\\] Here is a small example of usage of the mean_absolute_error function: 3.3.4.4. Mean squared error  The mean_squared_error function computes mean square error , a risk metric corresponding to the expected value of the squared (quadratic) error or loss. If is the predicted value of the -th sample, and is the corresponding true value, then the mean squared error (MSE) estimated over is defined as \\\\[\\\\text{MSE}(y, \\\\hat{y})  \\\\frac{1}{n_\\\\text{samples}} \\\\sum_{i0}^{n_\\\\text{samples} - 1} (y_i - \\\\hat{y}_i)^2.\\\\] Here is a small example of usage of the mean_squared_error function: Examples: See Gradient Boosting regression for an example of mean squared error usage to evaluate gradient boosting regression. 3.3.4.5. Mean squared logarithmic error  The mean_squared_log_error function computes a risk metric corresponding to the expected value of the squared logarithmic (quadratic) error or loss. If is the predicted value of the -th sample, and is the corresponding true value, then the mean squared logarithmic error (MSLE) estimated over is defined as \\\\[\\\\text{MSLE}(y, \\\\hat{y})  \\\\frac{1}{n_\\\\text{samples}} \\\\sum_{i0}^{n_\\\\text{samples} - 1} (\\\\log_e (1 + y_i) - \\\\log_e (1 + \\\\hat{y}_i) )^2.\\\\] Where means the natural logarithm of . This metric is best to use when targets having exponential growth, such as population counts, average sales of a commodity over a span of years etc. Note that this metric penalizes an under-predicted estimate greater than an over-predicted estimate. Here is a small example of usage of the mean_squared_log_error function: 3.3.4.6. Median absolute error  The median_absolute_error is particularly interesting because it is robust to outliers. The loss is calculated by taking the median of all absolute differences between the target and the prediction. If is the predicted value of the -th sample and is the corresponding true value, then the median absolute error (MedAE) estimated over is defined as \\\\[\\\\text{MedAE}(y, \\\\hat{y})  \\\\text{median}(\\\\mid y_1 - \\\\hat{y}_1 \\\\mid, \\\\ldots, \\\\mid y_n - \\\\hat{y}_n \\\\mid).\\\\] The median_absolute_error does not support multioutput. Here is a small example of usage of the median_absolute_error function: 3.3.4.7. R² score, the coefficient of determination  The r2_score function computes the coefficient of determination , usually denoted as R². It represents the proportion of variance (of y) that has been explained by the independent variables in the model. It provides an indication of goodness of fit and therefore a measure of how well unseen samples are likely to be predicted by the model, through the proportion of explained variance. As such variance is dataset dependent, R² may not be meaningfully comparable across different datasets. Best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse). A constant model that always predicts the expected value of y, disregarding the input features, would get a R² score of 0.0. If is the predicted value of the -th sample and is the corresponding true value for total samples, the estimated R² is defined as: \\\\[R^2(y, \\\\hat{y})  1 - \\\\frac{\\\\sum_{i1}^{n} (y_i - \\\\hat{y}_i)^2}{\\\\sum_{i1}^{n} (y_i - \\\\bar{y})^2}\\\\] where and . Note that r2_score calculates unadjusted R² without correcting for bias in sample variance of y. Here is a small example of usage of the r2_score function: Example: See Lasso and Elastic Net for Sparse Signals for an example of R² score usage to evaluate Lasso and Elastic Net on sparse signals. 3.3.4.8. Mean Poisson, Gamma, and Tweedie deviances  The mean_tweedie_deviance function computes the mean Tweedie deviance error with a parameter ( ). This is a metric that elicits predicted expectation values of regression targets. Following special cases exist, when it is equivalent to mean_squared_error . when it is equivalent to mean_poisson_deviance . when it is equivalent to mean_gamma_deviance . If is the predicted value of the -th sample, and is the corresponding true value, then the mean Tweedie deviance error (D) for power , estimated over is defined as \\\\[\\\\begin{split}\\\\text{D}(y, \\\\hat{y})  \\\\frac{1}{n_\\\\text{samples}} \\\\sum_{i0}^{n_\\\\text{samples} - 1} \\\\begin{cases} (y_i-\\\\hat{y}_i)^2, & \\\\text{for }p0\\\\text{ (Normal)}\\\\\\\\ 2(y_i \\\\log(y/\\\\hat{y}_i) + \\\\hat{y}_i - y_i), & \\\\text{for}p1\\\\text{ (Poisson)}\\\\\\\\ 2(\\\\log(\\\\hat{y}_i/y_i) + y_i/\\\\hat{y}_i - 1), & \\\\text{for}p2\\\\text{ (Gamma)}\\\\\\\\ 2\\\\left(\\\\frac{\\\\max(y_i,0)^{2-p}}{(1-p)(2-p)}- \\\\frac{y\\\\,\\\\hat{y}^{1-p}_i}{1-p}+\\\\frac{\\\\hat{y}^{2-p}_i}{2-p}\\\\right), & \\\\text{otherwise} \\\\end{cases}\\\\end{split}\\\\] Tweedie deviance is a homogeneous function of degree . Thus, Gamma distribution with means that simultaneously scaling and has no effect on the deviance. For Poisson distribution the deviance scales linearly, and for Normal distribution ( ), quadratically. In general, the higher the less weight is given to extreme deviations between true and predicted targets. For instance, let’s compare the two predictions 1.0 and 100 that are both 50% of their corresponding true value. The mean squared error ( ) is very sensitive to the prediction difference of the second point,: If we increase to 1,: the difference in errors decreases. Finally, by setting, : we would get identical errors. The deviance when is thus only sensitive to relative errors. ',\n",
       " 'sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions\\n3.3. Metrics and scoring: quantifying the quality of predictions\\nmodules/model_evaluation.html\\n 3.3.5. Clustering metrics  The sklearn.metrics module implements several loss, score, and utility functions. For more information see the Clustering performance evaluation section for instance clustering, and Biclustering evaluation for biclustering. ',\n",
       " 'sklearn_3_3_metrics_and_scoring_quantifying_the_quality_of_predictions\\n3.3. Metrics and scoring: quantifying the quality of predictions\\nmodules/model_evaluation.html\\n 3.3.6. Dummy estimators  When doing supervised learning, a simple sanity check consists of comparing one’s estimator against simple rules of thumb. DummyClassifier implements several such simple strategies for classification: generates random predictions by respecting the training set class distribution. always predicts the most frequent label in the training set. always predicts the class that maximizes the class prior (like ) and returns the class prior. generates predictions uniformly at random. always predicts a constant label that is provided by the user. A major motivation of this method is F1-scoring, when the positive class is in the minority. Note that with all these strategies, the method completely ignores the input data! To illustrate DummyClassifier , first let’s create an imbalanced dataset: Next, let’s compare the accuracy of and : We see that doesn’t do much better than a dummy classifier. Now, let’s change the kernel: We see that the accuracy was boosted to almost 100%. A cross validation strategy is recommended for a better estimate of the accuracy, if it is not too CPU costly. For more information see the Cross-validation: evaluating estimator performance section. Moreover if you want to optimize over the parameter space, it is highly recommended to use an appropriate methodology; see the Tuning the hyper-parameters of an estimator section for details. More generally, when the accuracy of a classifier is too close to random, it probably means that something went wrong: features are not helpful, a hyperparameter is not correctly tuned, the classifier is suffering from class imbalance, etc… DummyRegressor also implements four simple rules of thumb for regression: always predicts the mean of the training targets. always predicts the median of the training targets. always predicts a user provided quantile of the training targets. always predicts a constant value that is provided by the user. In all these strategies, the method completely ignores the input data. ',\n",
       " 'sklearn_3_4_model_persistence\\n3.4. Model persistence\\nmodules/model_persistence.html\\n 3.4.1. Persistence example  It is possible to save a model in scikit-learn by using Python’s built-in persistence model, namely pickle : In the specific case of scikit-learn, it may be better to use joblib’s replacement of pickle ( & ), which is more efficient on objects that carry large numpy arrays internally as is often the case for fitted scikit-learn estimators, but can only pickle to the disk and not to a string: Later you can load back the pickled model (possibly in another Python process) with: Note and functions also accept file-like object instead of filenames. More information on data persistence with Joblib is available here . ',\n",
       " 'sklearn_3_4_model_persistence\\n3.4. Model persistence\\nmodules/model_persistence.html\\n 3.4.2. Security & maintainability limitations  pickle (and joblib by extension), has some issues regarding maintainability and security. Because of this, Never unpickle untrusted data as it could lead to malicious code being executed upon loading. While models saved using one version of scikit-learn might load in other versions, this is entirely unsupported and inadvisable. It should also be kept in mind that operations performed on such data could give different and unexpected results. In order to rebuild a similar model with future versions of scikit-learn, additional metadata should be saved along the pickled model: The training data, e.g. a reference to an immutable snapshot The python source code used to generate the model The versions of scikit-learn and its dependencies The cross validation score obtained on the training data This should make it possible to check that the cross-validation score is in the same range as before. Since a model internal representation may be different on two different architectures, dumping a model on one architecture and loading it on another architecture is not supported. If you want to know more about these issues and explore other possible serialization methods, please refer to this talk by Alex Gaynor . ',\n",
       " 'sklearn_3_5_validation_curves_plotting_scores_to_evaluate_models\\n3.5. Validation curves: plotting scores to evaluate models\\nmodules/learning_curve.html\\n 3.5.1. Validation curve  To validate a model we need a scoring function (see Metrics and scoring: quantifying the quality of predictions ), for example accuracy for classifiers. The proper way of choosing multiple hyperparameters of an estimator are of course grid search or similar methods (see Tuning the hyper-parameters of an estimator ) that select the hyperparameter with the maximum score on a validation set or multiple validation sets. Note that if we optimized the hyperparameters based on a validation score the validation score is biased and not a good estimate of the generalization any longer. To get a proper estimate of the generalization we have to compute the score on another test set. However, it is sometimes helpful to plot the influence of a single hyperparameter on the training score and the validation score to find out whether the estimator is overfitting or underfitting for some hyperparameter values. The function validation_curve can help in this case: If the training score and the validation score are both low, the estimator will be underfitting. If the training score is high and the validation score is low, the estimator is overfitting and otherwise it is working very well. A low training score and a high validation score is usually not possible. All three cases can be found in the plot below where we vary the parameter of an SVM on the digits dataset. ',\n",
       " 'sklearn_3_5_validation_curves_plotting_scores_to_evaluate_models\\n3.5. Validation curves: plotting scores to evaluate models\\nmodules/learning_curve.html\\n 3.5.2. Learning curve  A learning curve shows the validation and training score of an estimator for varying numbers of training samples. It is a tool to find out how much we benefit from adding more training data and whether the estimator suffers more from a variance error or a bias error. Consider the following example where we plot the learning curve of a naive Bayes classifier and an SVM. For the naive Bayes, both the validation score and the training score converge to a value that is quite low with increasing size of the training set. Thus, we will probably not benefit much from more training data. In contrast, for small amounts of data, the training score of the SVM is much greater than the validation score. Adding more training samples will most likely increase generalization. We can use the function learning_curve to generate the values that are required to plot such a learning curve (number of samples that have been used, the average scores on the training sets and the average scores on the validation sets): ',\n",
       " 'sklearn_4_1_partial_dependence_plots\\n4.1. Partial dependence plots\\nmodules/partial_dependence.html\\n 4.1.1. Mathematical Definition  Let be the set of target features (i.e. the parameter) and let be its complement. The partial dependence of the response at a point is defined as: \\\\[\\\\begin{split}pd_{X_S}(x_S) &\\\\overset{def}{} \\\\mathbb{E}_{X_C}\\\\left[ f(x_S, X_C) \\\\right]\\\\\\\\ & \\\\int f(x_S, x_C) p(x_C) dx_C,\\\\end{split}\\\\] where is the response function ( predict , predict_proba or decision_function ) for a given sample whose values are defined by for the features in , and by for the features in . Note that and may be tuples. Computing this integral for various values of produces a plot as above. ',\n",
       " 'sklearn_4_1_partial_dependence_plots\\n4.1. Partial dependence plots\\nmodules/partial_dependence.html\\n 4.1.2. Computation methods  There are two main methods to approximate the integral above, namely the ‘brute’ and ‘recursion’ methods. The parameter controls which method to use. The ‘brute’ method is a generic method that works with any estimator. It approximates the above integral by computing an average over the data : \\\\[pd_{X_S}(x_S) \\\\approx \\\\frac{1}{n_\\\\text{samples}} \\\\sum_{i1}^n f(x_S, x_C^{(i)}),\\\\] where is the value of the i-th sample for the features in . For each value of , this method requires a full pass over the dataset which is computationally intensive. The ‘recursion’ method is faster than the ‘brute’ method, but it is only supported by some tree-based estimators. It is computed as follows. For a given point , a weighted tree traversal is performed: if a split node involves a ‘target’ feature, the corresponding left or right branch is followed; otherwise both branches are followed, each branch being weighted by the fraction of training samples that entered that branch. Finally, the partial dependence is given by a weighted average of all the visited leaves values. With the ‘brute’ method, the parameter is used both for generating the grid of values and the complement feature values . However with the ‘recursion’ method, is only used for the grid values: implicitly, the values are those of the training data. By default, the ‘recursion’ method is used on tree-based estimators that support it, and ‘brute’ is used for the rest. Note While both methods should be close in general, they might differ in some specific settings. The ‘brute’ method assumes the existence of the data points . When the features are correlated, such artificial samples may have a very low probability mass. The ‘brute’ and ‘recursion’ methods will likely disagree regarding the value of the partial dependence, because they will treat these unlikely samples differently. Remember, however, that the primary assumption for interpreting PDPs is that the features should be independent. Footnotes 1 For classification, the target response may be the probability of a class (the positive class for binary classification), or the decision function. Examples: Partial Dependence Plots References T. Hastie, R. Tibshirani and J. Friedman, The Elements of Statistical Learning , Second Edition, Section 10.13.2, Springer, 2009. C. Molnar, Interpretable Machine Learning , Section 5.1, 2019. ',\n",
       " 'sklearn_4_2_permutation_feature_importance\\n4.2. Permutation feature importance\\nmodules/permutation_importance.html\\n 4.2.1. Outline of the permutation importance algorithm  Inputs: fitted predictive model , tabular dataset (training or validation) . Compute the reference score of the model on data (for instance the accuracy for a classifier or the for a regressor). For each feature (column of ): For each repetition in : Randomly shuffle column of dataset to generate a corrupted version of the data named . Compute the score of model on corrupted data . Compute importance for feature defined as: \\\\[i_j  s - \\\\frac{1}{K} \\\\sum_{k1}^{K} s_{k,j}\\\\] ',\n",
       " 'sklearn_4_2_permutation_feature_importance\\n4.2. Permutation feature importance\\nmodules/permutation_importance.html\\n 4.2.2. Relation to impurity-based importance in trees  Tree-based models provide an alternative measure of feature importances based on the mean decrease in impurity (MDI). Impurity is quantified by the splitting criterion of the decision trees (Gini, Entropy or Mean Squared Error). However, this method can give high importance to features that may not be predictive on unseen data when the model is overfitting. Permutation-based feature importance, on the other hand, avoids this issue, since it can be computed on unseen data. Furthermore, impurity-based feature importance for trees are strongly biased and favor high cardinality features (typically numerical features) over low cardinality features such as binary features or categorical variables with a small number of possible categories. Permutation-based feature importances do not exhibit such a bias. Additionally, the permutation feature importance may be computed performance metric on the model predictions predictions and can be used to analyze any model class (not just tree-based models). The following example highlights the limitations of impurity-based feature importance in contrast to permutation-based feature importance: Permutation Importance vs Random Forest Feature Importance (MDI) . ',\n",
       " 'sklearn_4_2_permutation_feature_importance\\n4.2. Permutation feature importance\\nmodules/permutation_importance.html\\n 4.2.3. Misleading values on strongly correlated features  When two features are correlated and one of the features is permuted, the model will still have access to the feature through its correlated feature. This will result in a lower importance value for both features, where they might actually be important. One way to handle this is to cluster features that are correlated and only keep one feature from each cluster. This strategy is explored in the following example: Permutation Importance with Multicollinear or Correlated Features . Examples: Permutation Importance vs Random Forest Feature Importance (MDI) Permutation Importance with Multicollinear or Correlated Features References: 1 L. Breiman, “Random Forests”, Machine Learning, 45(1), 5-32, 2001. https://doi.org/10.1023/A:1010933404324 ',\n",
       " 'sklearn_5_1_available_plotting_utilities\\n5.1. Available Plotting Utilities\\nvisualizations.html#available-plotting-utilities\\n 5.1. Available Plotting Utilities  5.1.1. Functions  inspection.plot_partial_dependence (…[,\\xa0…]) Partial dependence plots. metrics.plot_confusion_matrix (estimator,\\xa0X,\\xa0…) Plot Confusion Matrix. metrics.plot_precision_recall_curve (…[,\\xa0…]) Plot Precision Recall Curve for binary classifiers. metrics.plot_roc_curve (estimator,\\xa0X,\\xa0y,\\xa0*[,\\xa0…]) Plot Receiver operating characteristic (ROC) curve. 5.1.2. Display Objects  inspection.PartialDependenceDisplay (…) Partial Dependence Plot (PDP) visualization. metrics.ConfusionMatrixDisplay (…[,\\xa0…]) Confusion Matrix visualization. metrics.PrecisionRecallDisplay (precision,\\xa0…) Precision Recall visualization. metrics.RocCurveDisplay (*,\\xa0fpr,\\xa0tpr[,\\xa0…]) ROC Curve visualization. ',\n",
       " 'sklearn_6_1_pipelines_and_composite_estimators\\n6.1. Pipelines and composite estimators\\nmodules/compose.html\\n 6.1.1. Pipeline: chaining estimators  Pipeline can be used to chain multiple estimators into one. This is useful as there is often a fixed sequence of steps in processing the data, for example feature selection, normalization and classification. Pipeline serves multiple purposes here: Convenience and encapsulation You only have to call fit and predict once on your data to fit a whole sequence of estimators. Joint parameter selection You can grid search over parameters of all estimators in the pipeline at once. Safety Pipelines help avoid leaking statistics from your test data into the trained model in cross-validation, by ensuring that the same samples are used to train the transformers and predictors. All estimators in a pipeline, except the last one, must be transformers (i.e. must have a transform method). The last estimator may be any type (transformer, classifier, etc.). 6.1.1.1. Usage  6.1.1.1.1. Construction  The Pipeline is built using a list of pairs, where the is a string containing the name you want to give this step and is an estimator object: The utility function make_pipeline is a shorthand for constructing pipelines; it takes a variable number of estimators and returns a pipeline, filling in the names automatically: 6.1.1.1.2. Accessing steps  The estimators of a pipeline are stored as a list in the attribute, but can be accessed by index or name by indexing (with ) the Pipeline: Pipeline’s attribute allows accessing steps by name with tab completion in interactive environments: A sub-pipeline can also be extracted using the slicing notation commonly used for Python Sequences such as lists or strings (although only a step of 1 is permitted). This is convenient for performing only some of the transformations (or their inverse): 6.1.1.1.3. Nested parameters  Parameters of the estimators in the pipeline can be accessed using the syntax: This is particularly important for doing grid searches: Individual steps may also be replaced as parameters, and non-final steps may be ignored by setting them to : The estimators of the pipeline can be retrieved by index: or by name: Examples: Pipeline Anova SVM Sample pipeline for text feature extraction and evaluation Pipelining: chaining a PCA and a logistic regression Explicit feature map approximation for RBF kernels SVM-Anova: SVM with univariate feature selection Selecting dimensionality reduction with Pipeline and GridSearchCV See also: Composite estimators and parameter spaces 6.1.1.2. Notes  Calling on the pipeline is the same as calling on each estimator in turn, the input and pass it on to the next step. The pipeline has all the methods that the last estimator in the pipeline has, i.e. if the last estimator is a classifier, the Pipeline can be used as a classifier. If the last estimator is a transformer, again, so is the pipeline. 6.1.1.3. Caching transformers: avoid repeated computation  Fitting transformers may be computationally expensive. With its parameter set, Pipeline will cache each transformer after calling . This feature is used to avoid computing the fit transformers within a pipeline if the parameters and input data are identical. A typical example is the case of a grid search in which the transformers can be fitted only once and reused for each configuration. The parameter is needed in order to cache the transformers. can be either a string containing the directory where to cache the transformers or a joblib.Memory object: Warning Side effect of caching transformers Using a Pipeline without cache enabled, it is possible to inspect the original instance such as: Enabling caching triggers a clone of the transformers before fitting. Therefore, the transformer instance given to the pipeline cannot be inspected directly. In following example, accessing the PCA instance will raise an since will be an unfitted transformer. Instead, use the attribute to inspect estimators within the pipeline: Examples: Selecting dimensionality reduction with Pipeline and GridSearchCV ',\n",
       " 'sklearn_6_1_pipelines_and_composite_estimators\\n6.1. Pipelines and composite estimators\\nmodules/compose.html\\n 6.1.2. Transforming target in regression  TransformedTargetRegressor transforms the targets before fitting a regression model. The predictions are mapped back to the original space via an inverse transform. It takes as an argument the regressor that will be used for prediction, and the transformer that will be applied to the target variable: For simple transformations, instead of a Transformer object, a pair of functions can be passed, defining the transformation and its inverse mapping: Subsequently, the object is created as: By default, the provided functions are checked at each fit to be the inverse of each other. However, it is possible to bypass this checking by setting to : Note The transformation can be triggered by setting either or the pair of functions and . However, setting both options will raise an error. Examples: Effect of transforming the targets in regression model ',\n",
       " 'sklearn_6_1_pipelines_and_composite_estimators\\n6.1. Pipelines and composite estimators\\nmodules/compose.html\\n 6.1.3. FeatureUnion: composite feature spaces  FeatureUnion combines several transformer objects into a new transformer that combines their output. A FeatureUnion takes a list of transformer objects. During fitting, each of these is fit to the data independently. The transformers are applied in parallel, and the feature matrices they output are concatenated side-by-side into a larger matrix. When you want to apply different transformations to each field of the data, see the related class sklearn.compose.ColumnTransformer (see user guide ). FeatureUnion serves the same purposes as Pipeline - convenience and joint parameter estimation and validation. FeatureUnion and Pipeline can be combined to create complex models. (A FeatureUnion has no way of checking whether two transformers might produce identical features. It only produces a union when the feature sets are disjoint, and making sure they are the caller’s responsibility.) 6.1.3.1. Usage  A FeatureUnion is built using a list of pairs, where the is the name you want to give to a given transformation (an arbitrary string; it only serves as an identifier) and is an estimator object: Like pipelines, feature unions have a shorthand constructor called make_union that does not require explicit naming of the components. Like , individual steps may be replaced using , and ignored by setting to : Examples: Concatenating multiple feature extraction methods ',\n",
       " 'sklearn_6_1_pipelines_and_composite_estimators\\n6.1. Pipelines and composite estimators\\nmodules/compose.html\\n 6.1.4. ColumnTransformer for heterogeneous data  Many datasets contain features of different types, say text, floats, and dates, where each type of feature requires separate preprocessing or feature extraction steps. Often it is easiest to preprocess data before applying scikit-learn methods, for example using pandas . Processing your data before passing it to scikit-learn might be problematic for one of the following reasons: Incorporating statistics from test data into the preprocessors makes cross-validation scores unreliable (known as data leakage ), for example in the case of scalers or imputing missing values. You may want to include the parameters of the preprocessors in a parameter search . The ColumnTransformer helps performing different transformations for different columns of the data, within a Pipeline that is safe from data leakage and that can be parametrized. ColumnTransformer works on arrays, sparse matrices, and pandas DataFrames . To each column, a different transformation can be applied, such as preprocessing or a specific feature extraction method: For this data, we might want to encode the column as a categorical variable using preprocessing.OneHotEncoder but apply a feature_extraction.text.CountVectorizer to the column. As we might use multiple feature extraction methods on the same column, we give each transformer a unique name, say and . By default, the remaining rating columns are ignored ( ): In the above example, the CountVectorizer expects a 1D array as input and therefore the columns were specified as a string ( ). However, preprocessing.OneHotEncoder as most of other transformers expects 2D data, therefore in that case you need to specify the column as a list of strings ( ). Apart from a scalar or a single item list, the column selection can be specified as a list of multiple items, an integer array, a slice, a boolean mask, or with a make_column_selector . The make_column_selector is used to select columns based on data type or column name: Strings can reference columns if the input is a DataFrame, integers are always interpreted as the positional columns. We can keep the remaining rating columns by setting . The values are appended to the end of the transformation: The parameter can be set to an estimator to transform the remaining rating columns. The transformed values are appended to the end of the transformation: The make_column_transformer function is available to more easily create a ColumnTransformer object. Specifically, the names will be given automatically. The equivalent for the above example would be: ',\n",
       " 'sklearn_6_1_pipelines_and_composite_estimators\\n6.1. Pipelines and composite estimators\\nmodules/compose.html\\n 6.1.5. Visualizing Composite Estimators  Estimators can be displayed with a HTML representation when shown in a jupyter notebook. This can be useful to diagnose or visualize a Pipeline with many estimators. This visualization is activated by setting the option in sklearn.set_config : An example of the HTML output can be seen in the HTML representation of Pipeline section of Column Transformer with Mixed Types . As an alternative, the HTML can be written to a file using estimator_html_repr : Examples: Column Transformer with Heterogeneous Data Sources Column Transformer with Mixed Types ',\n",
       " 'sklearn_6_2_feature_extraction\\n6.2. Feature extraction\\nmodules/feature_extraction.html\\n 6.2.1. Loading features from dicts  The class DictVectorizer can be used to convert feature arrays represented as lists of standard Python objects to the NumPy/SciPy representation used by scikit-learn estimators. While not particularly fast to process, Python’s has the advantages of being convenient to use, being sparse (absent features need not be stored) and storing feature names in addition to values. DictVectorizer implements what is called one-of-K or “one-hot” coding for categorical (aka nominal, discrete) features. Categorical features are “attribute-value” pairs where the value is restricted to a list of discrete of possibilities without ordering (e.g. topic identifiers, types of objects, tags, names…). In the following, “city” is a categorical attribute while “temperature” is a traditional numerical feature: DictVectorizer is also a useful representation transformation for training sequence classifiers in Natural Language Processing models that typically work by extracting feature windows around a particular word of interest. For example, suppose that we have a first algorithm that extracts Part of Speech (PoS) tags that we want to use as complementary tags for training a sequence classifier (e.g. a chunker). The following dict could be such a window of features extracted around the word ‘sat’ in the sentence ‘The cat sat on the mat.’: This description can be vectorized into a sparse two-dimensional matrix suitable for feeding into a classifier (maybe after being piped into a text.TfidfTransformer for normalization): As you can imagine, if one extracts such a context around each individual word of a corpus of documents the resulting matrix will be very wide (many one-hot-features) with most of them being valued to zero most of the time. So as to make the resulting data structure able to fit in memory the class uses a matrix by default instead of a . ',\n",
       " 'sklearn_6_2_feature_extraction\\n6.2. Feature extraction\\nmodules/feature_extraction.html\\n 6.2.2. Feature hashing  The class FeatureHasher is a high-speed, low-memory vectorizer that uses a technique known as feature hashing , or the “hashing trick”. Instead of building a hash table of the features encountered in training, as the vectorizers do, instances of FeatureHasher apply a hash function to the features to determine their column index in sample matrices directly. The result is increased speed and reduced memory usage, at the expense of inspectability; the hasher does not remember what the input features looked like and has no method. Since the hash function might cause collisions between (unrelated) features, a signed hash function is used and the sign of the hash value determines the sign of the value stored in the output matrix for a feature. This way, collisions are likely to cancel out rather than accumulate error, and the expected mean of any output feature’s value is zero. This mechanism is enabled by default with and is particularly useful for small hash table sizes ( ). For large hash table sizes, it can be disabled, to allow the output to be passed to estimators like sklearn.naive_bayes.MultinomialNB or sklearn.feature_selection.chi2 feature selectors that expect non-negative inputs. FeatureHasher accepts either mappings (like Python’s and its variants in the module), pairs, or strings, depending on the constructor parameter . Mapping are treated as lists of pairs, while single strings have an implicit value of 1, so is interpreted as . If a single feature occurs multiple times in a sample, the associated values will be summed (so and become ). The output from FeatureHasher is always a matrix in the CSR format. Feature hashing can be employed in document classification, but unlike text.CountVectorizer , FeatureHasher does not do word splitting or any other preprocessing except Unicode-to-UTF-8 encoding; see Vectorizing a large text corpus with the hashing trick , below, for a combined tokenizer/hasher. As an example, consider a word-level natural language processing task that needs features extracted from pairs. One could use a Python generator function to extract features: Then, the to be fed to can be constructed using: and fed to a hasher with: to get a matrix . Note the use of a generator comprehension, which introduces laziness into the feature extraction: tokens are only processed on demand from the hasher. 6.2.2.1. Implementation details  FeatureHasher uses the signed 32-bit variant of MurmurHash3. As a result (and because of limitations in ), the maximum number of features supported is currently . The original formulation of the hashing trick by Weinberger et al. used two separate hash functions and to determine the column index and sign of a feature, respectively. The present implementation works under the assumption that the sign bit of MurmurHash3 is independent of its other bits. Since a simple modulo is used to transform the hash function to a column index, it is advisable to use a power of two as the parameter; otherwise the features will not be mapped evenly to the columns. References: Kilian Weinberger, Anirban Dasgupta, John Langford, Alex Smola and Josh Attenberg (2009). Feature hashing for large scale multitask learning . Proc. ICML. MurmurHash3 . ',\n",
       " 'sklearn_6_2_feature_extraction\\n6.2. Feature extraction\\nmodules/feature_extraction.html\\n 6.2.3. Text feature extraction  6.2.3.1. The Bag of Words representation  Text Analysis is a major application field for machine learning algorithms. However the raw data, a sequence of symbols cannot be fed directly to the algorithms themselves as most of them expect numerical feature vectors with a fixed size rather than the raw text documents with variable length. In order to address this, scikit-learn provides utilities for the most common ways to extract numerical features from text content, namely: tokenizing strings and giving an integer id for each possible token, for instance by using white-spaces and punctuation as token separators. counting the occurrences of tokens in each document. normalizing and weighting with diminishing importance tokens that occur in the majority of samples / documents. In this scheme, features and samples are defined as follows: each individual token occurrence frequency (normalized or not) is treated as a feature . the vector of all the token frequencies for a given document is considered a multivariate sample . A corpus of documents can thus be represented by a matrix with one row per document and one column per token (e.g. word) occurring in the corpus. We call vectorization the general process of turning a collection of text documents into numerical feature vectors. This specific strategy (tokenization, counting and normalization) is called the Bag of Words or “Bag of n-grams” representation. Documents are described by word occurrences while completely ignoring the relative position information of the words in the document. 6.2.3.2. Sparsity  As most documents will typically use a very small subset of the words used in the corpus, the resulting matrix will have many feature values that are zeros (typically more than 99% of them). For instance a collection of 10,000 short text documents (such as emails) will use a vocabulary with a size in the order of 100,000 unique words in total while each document will use 100 to 1000 unique words individually. In order to be able to store such a matrix in memory but also to speed up algebraic operations matrix / vector, implementations will typically use a sparse representation such as the implementations available in the package. 6.2.3.3. Common Vectorizer usage  CountVectorizer implements both tokenization and occurrence counting in a single class: This model has many parameters, however the default values are quite reasonable (please see the reference documentation for the details): Let’s use it to tokenize and count the word occurrences of a minimalistic corpus of text documents: The default configuration tokenizes the string by extracting words of at least 2 letters. The specific function that does this step can be requested explicitly: Each term found by the analyzer during the fit is assigned a unique integer index corresponding to a column in the resulting matrix. This interpretation of the columns can be retrieved as follows: The converse mapping from feature name to column index is stored in the attribute of the vectorizer: Hence words that were not seen in the training corpus will be completely ignored in future calls to the transform method: Note that in the previous corpus, the first and the last documents have exactly the same words hence are encoded in equal vectors. In particular we lose the information that the last document is an interrogative form. To preserve some of the local ordering information we can extract 2-grams of words in addition to the 1-grams (individual words): The vocabulary extracted by this vectorizer is hence much bigger and can now resolve ambiguities encoded in local positioning patterns: In particular the interrogative form “Is this” is only present in the last document: 6.2.3.3.1. Using stop words  Stop words are words like “and”, “the”, “him”, which are presumed to be uninformative in representing the content of a text, and which may be removed to avoid them being construed as signal for prediction. Sometimes, however, similar words are useful for prediction, such as in classifying writing style or personality. There are several known issues in our provided ‘english’ stop word list. It does not aim to be a general, ‘one-size-fits-all’ solution as some tasks may require a more custom solution. See [NQY18] for more details. Please take care in choosing a stop word list. Popular stop word lists may include words that are highly informative to some tasks, such as computer . You should also make sure that the stop word list has had the same preprocessing and tokenization applied as the one used in the vectorizer. The word we’ve is split into we and ve by CountVectorizer’s default tokenizer, so if we’ve is in , but ve is not, ve will be retained from we’ve in transformed text. Our vectorizers will try to identify and warn about some kinds of inconsistencies. References NQY18 J. Nothman, H. Qin and R. Yurchak (2018). “Stop Word Lists in Free Open-source Software Packages” . In Proc. Workshop for NLP Open Source Software . 6.2.3.4. Tf–idf term weighting  In a large text corpus, some words will be very present (e.g. “the”, “a”, “is” in English) hence carrying very little meaningful information about the actual contents of the document. If we were to feed the direct count data directly to a classifier those very frequent terms would shadow the frequencies of rarer yet more interesting terms. In order to re-weight the count features into floating point values suitable for usage by a classifier it is very common to use the tf–idf transform. Tf means term-frequency while tf–idf means term-frequency times inverse document-frequency : . Using the ’s default settings, the term frequency, the number of times a term occurs in a given document, is multiplied with idf component, which is computed as , where is the total number of documents in the document set, and is the number of documents in the document set that contain term . The resulting tf-idf vectors are then normalized by the Euclidean norm: . This was originally a term weighting scheme developed for information retrieval (as a ranking function for search engines results) that has also found good use in document classification and clustering. The following sections contain further explanations and examples that illustrate how the tf-idfs are computed exactly and how the tf-idfs computed in scikit-learn’s TfidfTransformer and TfidfVectorizer differ slightly from the standard textbook notation that defines the idf as In the TfidfTransformer and TfidfVectorizer with , the “1” count is added to the idf instead of the idf’s denominator: This normalization is implemented by the TfidfTransformer class: Again please see the reference documentation for the details on all the parameters. Let’s take an example with the following counts. The first term is present 100% of the time hence not very interesting. The two other features only in less than 50% of the time hence probably more representative of the content of the documents: Each row is normalized to have unit Euclidean norm: For example, we can compute the tf-idf of the first term in the first document in the array as follows: Now, if we repeat this computation for the remaining 2 terms in the document, we get and the vector of raw tf-idfs: Then, applying the Euclidean (L2) norm, we obtain the following tf-idfs for document 1: Furthermore, the default parameter adds “1” to the numerator and denominator as if an extra document was seen containing every term in the collection exactly once, which prevents zero divisions: Using this modification, the tf-idf of the third term in document 1 changes to 1.8473: And the L2-normalized tf-idf changes to : The weights of each feature computed by the method call are stored in a model attribute: As tf–idf is very often used for text features, there is also another class called TfidfVectorizer that combines all the options of CountVectorizer and TfidfTransformer in a single model: While the tf–idf normalization is often very useful, there might be cases where the binary occurrence markers might offer better features. This can be achieved by using the parameter of CountVectorizer . In particular, some estimators such as Bernoulli Naive Bayes explicitly model discrete boolean random variables. Also, very short texts are likely to have noisy tf–idf values while the binary occurrence info is more stable. As usual the best way to adjust the feature extraction parameters is to use a cross-validated grid search, for instance by pipelining the feature extractor with a classifier: Sample pipeline for text feature extraction and evaluation 6.2.3.5. Decoding text files  Text is made of characters, but files are made of bytes. These bytes represent characters according to some encoding . To work with text files in Python, their bytes must be decoded to a character set called Unicode. Common encodings are ASCII, Latin-1 (Western Europe), KOI8-R (Russian) and the universal encodings UTF-8 and UTF-16. Many others exist. Note An encoding can also be called a ‘character set’, but this term is less accurate: several encodings can exist for a single character set. The text feature extractors in scikit-learn know how to decode text files, but only if you tell them what encoding the files are in. The CountVectorizer takes an parameter for this purpose. For modern text files, the correct encoding is probably UTF-8, which is therefore the default ( ). If the text you are loading is not actually encoded with UTF-8, however, you will get a . The vectorizers can be told to be silent about decoding errors by setting the parameter to either or . See the documentation for the Python function for more details (type at the Python prompt). If you are having trouble decoding text, here are some things to try: Find out what the actual encoding of the text is. The file might come with a header or README that tells you the encoding, or there might be some standard encoding you can assume based on where the text comes from. You may be able to find out what kind of encoding it is in general using the UNIX command . The Python module comes with a script called that will guess the specific encoding, though you cannot rely on its guess being correct. You could try UTF-8 and disregard the errors. You can decode byte strings with to replace all decoding errors with a meaningless character, or set in the vectorizer. This may damage the usefulness of your features. Real text may come from a variety of sources that may have used different encodings, or even be sloppily decoded in a different encoding than the one it was encoded with. This is common in text retrieved from the Web. The Python package ftfy can automatically sort out some classes of decoding errors, so you could try decoding the unknown text as and then using to fix errors. If the text is in a mish-mash of encodings that is simply too hard to sort out (which is the case for the 20 Newsgroups dataset), you can fall back on a simple single-byte encoding such as . Some text may display incorrectly, but at least the same sequence of bytes will always represent the same feature. For example, the following snippet uses (not shipped with scikit-learn, must be installed separately) to figure out the encoding of three texts. It then vectorizes the texts and prints the learned vocabulary. The output is not shown here. (Depending on the version of , it might get the first one wrong.) For an introduction to Unicode and character encodings in general, see Joel Spolsky’s Absolute Minimum Every Software Developer Must Know About Unicode . 6.2.3.6. Applications and examples  The bag of words representation is quite simplistic but surprisingly useful in practice. In particular in a supervised setting it can be successfully combined with fast and scalable linear models to train document classifiers , for instance: Classification of text documents using sparse features In an unsupervised setting it can be used to group similar documents together by applying clustering algorithms such as K-means : Clustering text documents using k-means Finally it is possible to discover the main topics of a corpus by relaxing the hard assignment constraint of clustering, for instance by using Non-negative matrix factorization (NMF or NNMF) : Topic extraction with Non-negative Matrix Factorization and Latent Dirichlet Allocation 6.2.3.7. Limitations of the Bag of Words representation  A collection of unigrams (what bag of words is) cannot capture phrases and multi-word expressions, effectively disregarding any word order dependence. Additionally, the bag of words model doesn’t account for potential misspellings or word derivations. N-grams to the rescue! Instead of building a simple collection of unigrams (n1), one might prefer a collection of bigrams (n2), where occurrences of pairs of consecutive words are counted. One might alternatively consider a collection of character n-grams, a representation resilient against misspellings and derivations. For example, let’s say we’re dealing with a corpus of two documents: . The second document contains a misspelling of the word ‘words’. A simple bag of words representation would consider these two as very distinct documents, differing in both of the two possible features. A character 2-gram representation, however, would find the documents matching in 4 out of 8 features, which may help the preferred classifier decide better: In the above example, analyzer is used, which creates n-grams only from characters inside word boundaries (padded with space on each side). The analyzer, alternatively, creates n-grams that span across words: The word boundaries-aware variant is especially interesting for languages that use white-spaces for word separation as it generates significantly less noisy features than the raw variant in that case. For such languages it can increase both the predictive accuracy and convergence speed of classifiers trained using such features while retaining the robustness with regards to misspellings and word derivations. While some local positioning information can be preserved by extracting n-grams instead of individual words, bag of words and bag of n-grams destroy most of the inner structure of the document and hence most of the meaning carried by that internal structure. In order to address the wider task of Natural Language Understanding, the local structure of sentences and paragraphs should thus be taken into account. Many such models will thus be casted as “Structured output” problems which are currently outside of the scope of scikit-learn. 6.2.3.8. Vectorizing a large text corpus with the hashing trick  The above vectorization scheme is simple but the fact that it holds an in- memory mapping from the string tokens to the integer feature indices (the attribute) causes several problems when dealing with large datasets : the larger the corpus, the larger the vocabulary will grow and hence the memory use too, fitting requires the allocation of intermediate data structures of size proportional to that of the original dataset. building the word-mapping requires a full pass over the dataset hence it is not possible to fit text classifiers in a strictly online manner. pickling and un-pickling vectorizers with a large can be very slow (typically much slower than pickling / un-pickling flat data structures such as a NumPy array of the same size), it is not easily possible to split the vectorization work into concurrent sub tasks as the attribute would have to be a shared state with a fine grained synchronization barrier: the mapping from token string to feature index is dependent on ordering of the first occurrence of each token hence would have to be shared, potentially harming the concurrent workers’ performance to the point of making them slower than the sequential variant. It is possible to overcome those limitations by combining the “hashing trick” ( Feature hashing ) implemented by the sklearn.feature_extraction.FeatureHasher class and the text preprocessing and tokenization features of the CountVectorizer . This combination is implementing in HashingVectorizer , a transformer class that is mostly API compatible with CountVectorizer . HashingVectorizer is stateless, meaning that you don’t have to call on it: You can see that 16 non-zero feature tokens were extracted in the vector output: this is less than the 19 non-zeros extracted previously by the CountVectorizer on the same toy corpus. The discrepancy comes from hash function collisions because of the low value of the parameter. In a real world setting, the parameter can be left to its default value of (roughly one million possible features). If memory or downstream models size is an issue selecting a lower value such as might help without introducing too many additional collisions on typical text classification tasks. Note that the dimensionality does not affect the CPU training time of algorithms which operate on CSR matrices ( , , , ) but it does for algorithms that work with CSC matrices ( , , etc). Let’s try again with the default setting: We no longer get the collisions, but this comes at the expense of a much larger dimensionality of the output space. Of course, other terms than the 19 used here might still collide with each other. The HashingVectorizer also comes with the following limitations: it is not possible to invert the model (no method), nor to access the original string representation of the features, because of the one-way nature of the hash function that performs the mapping. it does not provide IDF weighting as that would introduce statefulness in the model. A TfidfTransformer can be appended to it in a pipeline if required. 6.2.3.9. Performing out-of-core scaling with HashingVectorizer  An interesting development of using a HashingVectorizer is the ability to perform out-of-core scaling. This means that we can learn from data that does not fit into the computer’s main memory. A strategy to implement out-of-core scaling is to stream data to the estimator in mini-batches. Each mini-batch is vectorized using HashingVectorizer so as to guarantee that the input space of the estimator has always the same dimensionality. The amount of memory used at any time is thus bounded by the size of a mini-batch. Although there is no limit to the amount of data that can be ingested using such an approach, from a practical point of view the learning time is often limited by the CPU time one wants to spend on the task. For a full-fledged example of out-of-core scaling in a text classification task see Out-of-core classification of text documents . 6.2.3.10. Customizing the vectorizer classes  It is possible to customize the behavior by passing a callable to the vectorizer constructor: In particular we name: : a callable that takes an entire document as input (as a single string), and returns a possibly transformed version of the document, still as an entire string. This can be used to remove HTML tags, lowercase the entire document, etc. : a callable that takes the output from the preprocessor and splits it into tokens, then returns a list of these. : a callable that replaces the preprocessor and tokenizer. The default analyzers all call the preprocessor and tokenizer, but custom analyzers will skip this. N-gram extraction and stop word filtering take place at the analyzer level, so a custom analyzer may have to reproduce these steps. (Lucene users might recognize these names, but be aware that scikit-learn concepts may not map one-to-one onto Lucene concepts.) To make the preprocessor, tokenizer and analyzers aware of the model parameters it is possible to derive from the class and override the , and factory methods instead of passing custom functions. Some tips and tricks: If documents are pre-tokenized by an external package, then store them in files (or strings) with the tokens separated by whitespace and pass Fancy token-level analysis such as stemming, lemmatizing, compound splitting, filtering based on part-of-speech, etc. are not included in the scikit-learn codebase, but can be added by customizing either the tokenizer or the analyzer. Here’s a with a tokenizer and lemmatizer using NLTK : (Note that this will not filter out punctuation.) The following example will, for instance, transform some British spelling to American spelling: for other styles of preprocessing; examples include stemming, lemmatization, or normalizing numerical tokens, with the latter illustrated in: Biclustering documents with the Spectral Co-clustering algorithm Customizing the vectorizer can also be useful when handling Asian languages that do not use an explicit word separator such as whitespace. ',\n",
       " 'sklearn_6_2_feature_extraction\\n6.2. Feature extraction\\nmodules/feature_extraction.html\\n 6.2.4. Image feature extraction  6.2.4.1. Patch extraction  The extract_patches_2d function extracts patches from an image stored as a two-dimensional array, or three-dimensional with color information along the third axis. For rebuilding an image from all its patches, use reconstruct_from_patches_2d . For example let use generate a 4x4 pixel picture with 3 color channels (e.g. in RGB format): Let us now try to reconstruct the original image from the patches by averaging on overlapping areas: The PatchExtractor class works in the same way as extract_patches_2d , only it supports multiple images as input. It is implemented as an estimator, so it can be used in pipelines. See: 6.2.4.2. Connectivity graph of an image  Several estimators in the scikit-learn can use connectivity information between features or samples. For instance Ward clustering ( Hierarchical clustering ) can cluster together only neighboring pixels of an image, thus forming contiguous patches: For this purpose, the estimators use a ‘connectivity’ matrix, giving which samples are connected. The function img_to_graph returns such a matrix from a 2D or 3D image. Similarly, grid_to_graph build a connectivity matrix for images given the shape of these image. These matrices can be used to impose connectivity in estimators that use connectivity information, such as Ward clustering ( Hierarchical clustering ), but also to build precomputed kernels, or similarity matrices. Note Examples A demo of structured Ward hierarchical clustering on an image of coins Spectral clustering for image segmentation Feature agglomeration vs. univariate selection ',\n",
       " 'sklearn_6_3_preprocessing_data\\n6.3. Preprocessing data\\nmodules/preprocessing.html\\n 6.3.1. Standardization, or mean removal and variance scaling  Standardization of datasets is a common requirement for many machine learning estimators implemented in scikit-learn; they might behave badly if the individual features do not more or less look like standard normally distributed data: Gaussian with zero mean and unit variance . In practice we often ignore the shape of the distribution and just transform the data to center it by removing the mean value of each feature, then scale it by dividing non-constant features by their standard deviation. For instance, many elements used in the objective function of a learning algorithm (such as the RBF kernel of Support Vector Machines or the l1 and l2 regularizers of linear models) assume that all features are centered around zero and have variance in the same order. If a feature has a variance that is orders of magnitude larger than others, it might dominate the objective function and make the estimator unable to learn from other features correctly as expected. The function scale provides a quick and easy way to perform this operation on a single array-like dataset: Scaled data has zero mean and unit variance: The module further provides a utility class StandardScaler that implements the API to compute the mean and standard deviation on a training set so as to be able to later reapply the same transformation on the testing set. This class is hence suitable for use in the early steps of a sklearn.pipeline.Pipeline : The scaler instance can then be used on new data to transform it the same way it did on the training set: It is possible to disable either centering or scaling by either passing or to the constructor of StandardScaler . 6.3.1.1. Scaling features to a range  An alternative standardization is scaling features to lie between a given minimum and maximum value, often between zero and one, or so that the maximum absolute value of each feature is scaled to unit size. This can be achieved using MinMaxScaler or MaxAbsScaler , respectively. The motivation to use this scaling include robustness to very small standard deviations of features and preserving zero entries in sparse data. Here is an example to scale a toy data matrix to the range: The same instance of the transformer can then be applied to some new test data unseen during the fit call: the same scaling and shifting operations will be applied to be consistent with the transformation performed on the train data: It is possible to introspect the scaler attributes to find about the exact nature of the transformation learned on the training data: If MinMaxScaler is given an explicit the full formula is: MaxAbsScaler works in a very similar fashion, but scales in a way that the training data lies within the range by dividing through the largest maximum value in each feature. It is meant for data that is already centered at zero or sparse data. Here is how to use the toy data from the previous example with this scaler: As with scale , the module further provides convenience functions minmax_scale and maxabs_scale if you don’t want to create an object. 6.3.1.2. Scaling sparse data  Centering sparse data would destroy the sparseness structure in the data, and thus rarely is a sensible thing to do. However, it can make sense to scale sparse inputs, especially if features are on different scales. MaxAbsScaler and maxabs_scale were specifically designed for scaling sparse data, and are the recommended way to go about this. However, scale and StandardScaler can accept matrices as input, as long as is explicitly passed to the constructor. Otherwise a will be raised as silently centering would break the sparsity and would often crash the execution by allocating excessive amounts of memory unintentionally. RobustScaler cannot be fitted to sparse inputs, but you can use the method on sparse inputs. Note that the scalers accept both Compressed Sparse Rows and Compressed Sparse Columns format (see and ). Any other sparse input will be converted to the Compressed Sparse Rows representation . To avoid unnecessary memory copies, it is recommended to choose the CSR or CSC representation upstream. Finally, if the centered data is expected to be small enough, explicitly converting the input to an array using the method of sparse matrices is another option. 6.3.1.3. Scaling data with outliers  If your data contains many outliers, scaling using the mean and variance of the data is likely to not work very well. In these cases, you can use robust_scale and RobustScaler as drop-in replacements instead. They use more robust estimates for the center and range of your data. References: Further discussion on the importance of centering and scaling data is available on this FAQ: Should I normalize/standardize/rescale the data? Scaling vs Whitening It is sometimes not enough to center and scale the features independently, since a downstream model can further make some assumption on the linear independence of the features. To address this issue you can use sklearn.decomposition.PCA with to further remove the linear correlation across features. Scaling a 1D array All above functions (i.e. scale , minmax_scale , maxabs_scale , and robust_scale ) accept 1D array which can be useful in some specific case. 6.3.1.4. Centering kernel matrices  If you have a kernel matrix of a kernel that computes a dot product in a feature space defined by function , a KernelCenterer can transform the kernel matrix so that it contains inner products in the feature space defined by followed by removal of the mean in that space. ',\n",
       " 'sklearn_6_3_preprocessing_data\\n6.3. Preprocessing data\\nmodules/preprocessing.html\\n 6.3.2. Non-linear transformation  Two types of transformations are available: quantile transforms and power transforms. Both quantile and power transforms are based on monotonic transformations of the features and thus preserve the rank of the values along each feature. Quantile transforms put all features into the same desired distribution based on the formula where is the cumulative distribution function of the feature and the quantile function of the desired output distribution . This formula is using the two following facts: (i) if is a random variable with a continuous cumulative distribution function then is uniformly distributed on ; (ii) if is a random variable with uniform distribution on then has distribution . By performing a rank transformation, a quantile transform smooths out unusual distributions and is less influenced by outliers than scaling methods. It does, however, distort correlations and distances within and across features. Power transforms are a family of parametric transformations that aim to map data from any distribution to as close to a Gaussian distribution. 6.3.2.1. Mapping to a Uniform distribution  QuantileTransformer and quantile_transform provide a non-parametric transformation to map the data to a uniform distribution with values between 0 and 1: This feature corresponds to the sepal length in cm. Once the quantile transformation applied, those landmarks approach closely the percentiles previously defined: This can be confirmed on a independent testing set with similar remarks: 6.3.2.2. Mapping to a Gaussian distribution  In many modeling scenarios, normality of the features in a dataset is desirable. Power transforms are a family of parametric, monotonic transformations that aim to map data from any distribution to as close to a Gaussian distribution as possible in order to stabilize variance and minimize skewness. PowerTransformer currently provides two such power transformations, the Yeo-Johnson transform and the Box-Cox transform. The Yeo-Johnson transform is given by: \\\\[\\\\begin{split}x_i^{(\\\\lambda)}  \\\\begin{cases} [(x_i + 1)^\\\\lambda - 1] / \\\\lambda & \\\\text{if } \\\\lambda \\\\neq 0, x_i \\\\geq 0, \\\\\\\\[8pt] \\\\ln{(x_i + 1)} & \\\\text{if } \\\\lambda  0, x_i \\\\geq 0 \\\\\\\\[8pt] -[(-x_i + 1)^{2 - \\\\lambda} - 1] / (2 - \\\\lambda) & \\\\text{if } \\\\lambda \\\\neq 2, x_i < 0, \\\\\\\\[8pt] - \\\\ln (- x_i + 1) & \\\\text{if } \\\\lambda  2, x_i < 0 \\\\end{cases}\\\\end{split}\\\\] while the Box-Cox transform is given by: \\\\[\\\\begin{split}x_i^{(\\\\lambda)}  \\\\begin{cases} \\\\dfrac{x_i^\\\\lambda - 1}{\\\\lambda} & \\\\text{if } \\\\lambda \\\\neq 0, \\\\\\\\[8pt] \\\\ln{(x_i)} & \\\\text{if } \\\\lambda  0, \\\\end{cases}\\\\end{split}\\\\] Box-Cox can only be applied to strictly positive data. In both methods, the transformation is parameterized by , which is determined through maximum likelihood estimation. Here is an example of using Box-Cox to map samples drawn from a lognormal distribution to a normal distribution: While the above example sets the option to , PowerTransformer will apply zero-mean, unit-variance normalization to the transformed output by default. Below are examples of Box-Cox and Yeo-Johnson applied to various probability distributions. Note that when applied to certain distributions, the power transforms achieve very Gaussian-like results, but with others, they are ineffective. This highlights the importance of visualizing the data before and after transformation. It is also possible to map data to a normal distribution using QuantileTransformer by setting . Using the earlier example with the iris dataset: Thus the median of the input becomes the mean of the output, centered at 0. The normal output is clipped so that the input’s minimum and maximum — corresponding to the 1e-7 and 1 - 1e-7 quantiles respectively — do not become infinite under the transformation. ',\n",
       " 'sklearn_6_3_preprocessing_data\\n6.3. Preprocessing data\\nmodules/preprocessing.html\\n 6.3.3. Normalization  Normalization is the process of scaling individual samples to have unit norm . This process can be useful if you plan to use a quadratic form such as the dot-product or any other kernel to quantify the similarity of any pair of samples. This assumption is the base of the Vector Space Model often used in text classification and clustering contexts. The function normalize provides a quick and easy way to perform this operation on a single array-like dataset, either using the or norms: The module further provides a utility class Normalizer that implements the same operation using the API (even though the method is useless in this case: the class is stateless as this operation treats samples independently). This class is hence suitable for use in the early steps of a sklearn.pipeline.Pipeline : The normalizer instance can then be used on sample vectors as any transformer: Note: L2 normalization is also known as spatial sign preprocessing. Sparse input normalize and Normalizer accept both dense array-like and sparse matrices from scipy.sparse as input . For sparse input the data is converted to the Compressed Sparse Rows representation (see ) before being fed to efficient Cython routines. To avoid unnecessary memory copies, it is recommended to choose the CSR representation upstream. ',\n",
       " 'sklearn_6_3_preprocessing_data\\n6.3. Preprocessing data\\nmodules/preprocessing.html\\n 6.3.4. Encoding categorical features  Often features are not given as continuous values but categorical. For example a person could have features , , . Such features can be efficiently coded as integers, for instance could be expressed as while would be . To convert categorical features to such integer codes, we can use the OrdinalEncoder . This estimator transforms each categorical feature to one new feature of integers (0 to n_categories - 1): Such integer representation can, however, not be used directly with all scikit-learn estimators, as these expect continuous input, and would interpret the categories as being ordered, which is often not desired (i.e. the set of browsers was ordered arbitrarily). Another possibility to convert categorical features to features that can be used with scikit-learn estimators is to use a one-of-K, also known as one-hot or dummy encoding. This type of encoding can be obtained with the OneHotEncoder , which transforms each categorical feature with possible values into binary features, with one of them 1, and all others 0. Continuing the example above: By default, the values each feature can take is inferred automatically from the dataset and can be found in the attribute: It is possible to specify this explicitly using the parameter . There are two genders, four possible continents and four web browsers in our dataset: If there is a possibility that the training data might have missing categorical features, it can often be better to specify instead of setting the manually as above. When is specified and unknown categories are encountered during transform, no error will be raised but the resulting one-hot encoded columns for this feature will be all zeros ( is only supported for one-hot encoding): It is also possible to encode each column into columns instead of columns by using the parameter. This parameter allows the user to specify a category for each feature to be dropped. This is useful to avoid co-linearity in the input matrix in some classifiers. Such functionality is useful, for example, when using non-regularized regression ( LinearRegression ), since co-linearity would cause the covariance matrix to be non-invertible. When this parameter is not None, must be set to : One might want to drop one of the two columns only for features with 2 categories. In this case, you can set the parameter . In the transformed , the first column is the encoding of the feature with categories “male”/”female”, while the remaining 6 columns is the encoding of the 2 features with respectively 3 categories each. See Loading features from dicts for categorical features that are represented as a dict, not as scalars. ',\n",
       " 'sklearn_6_3_preprocessing_data\\n6.3. Preprocessing data\\nmodules/preprocessing.html\\n 6.3.5. Discretization  Discretization (otherwise known as quantization or binning) provides a way to partition continuous features into discrete values. Certain datasets with continuous features may benefit from discretization, because discretization can transform the dataset of continuous attributes to one with only nominal attributes. One-hot encoded discretized features can make a model more expressive, while maintaining interpretability. For instance, pre-processing with a discretizer can introduce nonlinearity to linear models. 6.3.5.1. K-bins discretization  KBinsDiscretizer discretizes features into bins: By default the output is one-hot encoded into a sparse matrix (See Encoding categorical features ) and this can be configured with the parameter. For each feature, the bin edges are computed during and together with the number of bins, they will define the intervals. Therefore, for the current example, these intervals are defined as: feature 1: feature 2: feature 3: Based on these bin intervals, is transformed as follows: The resulting dataset contains ordinal attributes which can be further used in a sklearn.pipeline.Pipeline . Discretization is similar to constructing histograms for continuous data. However, histograms focus on counting features which fall into particular bins, whereas discretization focuses on assigning feature values to these bins. KBinsDiscretizer implements different binning strategies, which can be selected with the parameter. The ‘uniform’ strategy uses constant-width bins. The ‘quantile’ strategy uses the quantiles values to have equally populated bins in each feature. The ‘kmeans’ strategy defines bins based on a k-means clustering procedure performed on each feature independently. Examples: Using KBinsDiscretizer to discretize continuous features Feature discretization Demonstrating the different strategies of KBinsDiscretizer 6.3.5.2. Feature binarization  Feature binarization is the process of thresholding numerical features to get boolean values . This can be useful for downstream probabilistic estimators that make assumption that the input data is distributed according to a multi-variate Bernoulli distribution . For instance, this is the case for the sklearn.neural_network.BernoulliRBM . It is also common among the text processing community to use binary feature values (probably to simplify the probabilistic reasoning) even if normalized counts (a.k.a. term frequencies) or TF-IDF valued features often perform slightly better in practice. As for the Normalizer , the utility class Binarizer is meant to be used in the early stages of sklearn.pipeline.Pipeline . The method does nothing as each sample is treated independently of others: It is possible to adjust the threshold of the binarizer: As for the StandardScaler and Normalizer classes, the preprocessing module provides a companion function binarize to be used when the transformer API is not necessary. Note that the Binarizer is similar to the KBinsDiscretizer when , and when the bin edge is at the value . Sparse input binarize and Binarizer accept both dense array-like and sparse matrices from scipy.sparse as input . For sparse input the data is converted to the Compressed Sparse Rows representation (see ). To avoid unnecessary memory copies, it is recommended to choose the CSR representation upstream. ',\n",
       " 'sklearn_6_3_preprocessing_data\\n6.3. Preprocessing data\\nmodules/preprocessing.html\\n 6.3.6. Imputation of missing values  Tools for imputing missing values are discussed at Imputation of missing values . ',\n",
       " 'sklearn_6_3_preprocessing_data\\n6.3. Preprocessing data\\nmodules/preprocessing.html\\n 6.3.7. Generating polynomial features  Often it’s useful to add complexity to the model by considering nonlinear features of the input data. A simple and common method to use is polynomial features, which can get features’ high-order and interaction terms. It is implemented in PolynomialFeatures : The features of X have been transformed from to . In some cases, only interaction terms among features are required, and it can be gotten with the setting : The features of X have been transformed from to . Note that polynomial features are used implicitly in kernel methods (e.g., sklearn.svm.SVC , sklearn.decomposition.KernelPCA ) when using polynomial Kernel functions . See Polynomial interpolation for Ridge regression using created polynomial features. ',\n",
       " 'sklearn_6_3_preprocessing_data\\n6.3. Preprocessing data\\nmodules/preprocessing.html\\n 6.3.8. Custom transformers  Often, you will want to convert an existing Python function into a transformer to assist in data cleaning or processing. You can implement a transformer from an arbitrary function with FunctionTransformer . For example, to build a transformer that applies a log transformation in a pipeline, do: You can ensure that and are the inverse of each other by setting and calling before . Please note that a warning is raised and can be turned into an error with a : For a full code example that demonstrates using a FunctionTransformer to do custom feature selection, see Using FunctionTransformer to select columns ',\n",
       " 'sklearn_6_4_imputation_of_missing_values\\n6.4. Imputation of missing values\\nmodules/impute.html\\n 6.4.1. Univariate vs. Multivariate Imputation  One type of imputation algorithm is univariate, which imputes values in the i-th feature dimension using only non-missing values in that feature dimension (e.g. impute.SimpleImputer ). By contrast, multivariate imputation algorithms use the entire set of available feature dimensions to estimate the missing values (e.g. impute.IterativeImputer ). ',\n",
       " 'sklearn_6_4_imputation_of_missing_values\\n6.4. Imputation of missing values\\nmodules/impute.html\\n 6.4.2. Univariate feature imputation  The SimpleImputer class provides basic strategies for imputing missing values. Missing values can be imputed with a provided constant value, or using the statistics (mean, median or most frequent) of each column in which the missing values are located. This class also allows for different missing values encodings. The following snippet demonstrates how to replace missing values, encoded as , using the mean value of the columns (axis 0) that contain the missing values: The SimpleImputer class also supports sparse matrices: Note that this format is not meant to be used to implicitly store missing values in the matrix because it would densify it at transform time. Missing values encoded by 0 must be used with dense input. The SimpleImputer class also supports categorical data represented as string values or pandas categoricals when using the or strategy: ',\n",
       " 'sklearn_6_4_imputation_of_missing_values\\n6.4. Imputation of missing values\\nmodules/impute.html\\n 6.4.3. Multivariate feature imputation  A more sophisticated approach is to use the IterativeImputer class, which models each feature with missing values as a function of other features, and uses that estimate for imputation. It does so in an iterated round-robin fashion: at each step, a feature column is designated as output and the other feature columns are treated as inputs . A regressor is fit on for known . Then, the regressor is used to predict the missing values of . This is done for each feature in an iterative fashion, and then is repeated for imputation rounds. The results of the final imputation round are returned. Note This estimator is still experimental for now: default parameters or details of behaviour might change without any deprecation cycle. Resolving the following issues would help stabilize IterativeImputer : convergence criteria ( #14338 ), default estimators ( #13286 ), and use of random state ( #15611 ). To use it, you need to explicitly import . Both SimpleImputer and IterativeImputer can be used in a Pipeline as a way to build a composite estimator that supports imputation. See Imputing missing values before building an estimator . 6.4.3.1. Flexibility of IterativeImputer  There are many well-established imputation packages in the R data science ecosystem: Amelia, mi, mice, missForest, etc. missForest is popular, and turns out to be a particular instance of different sequential imputation algorithms that can all be implemented with IterativeImputer by passing in different regressors to be used for predicting missing feature values. In the case of missForest, this regressor is a Random Forest. See Imputing missing values with variants of IterativeImputer . 6.4.3.2. Multiple vs. Single Imputation  In the statistics community, it is common practice to perform multiple imputations, generating, for example, separate imputations for a single feature matrix. Each of these imputations is then put through the subsequent analysis pipeline (e.g. feature engineering, clustering, regression, classification). The final analysis results (e.g. held-out validation errors) allow the data scientist to obtain understanding of how analytic results may differ as a consequence of the inherent uncertainty caused by the missing values. The above practice is called multiple imputation. Our implementation of IterativeImputer was inspired by the R MICE package (Multivariate Imputation by Chained Equations) 1 , but differs from it by returning a single imputation instead of multiple imputations. However, IterativeImputer can also be used for multiple imputations by applying it repeatedly to the same dataset with different random seeds when . See 2 , chapter 4 for more discussion on multiple vs. single imputations. It is still an open problem as to how useful single vs. multiple imputation is in the context of prediction and classification when the user is not interested in measuring uncertainty due to missing values. Note that a call to the method of IterativeImputer is not allowed to change the number of samples. Therefore multiple imputations cannot be achieved by a single call to . ',\n",
       " 'sklearn_6_4_imputation_of_missing_values\\n6.4. Imputation of missing values\\nmodules/impute.html\\n 6.4.4. References  1 Stef van Buuren, Karin Groothuis-Oudshoorn (2011). “mice: Multivariate Imputation by Chained Equations in R”. Journal of Statistical Software 45: 1-67. 2 Roderick J A Little and Donald B Rubin (1986). “Statistical Analysis with Missing Data”. John Wiley & Sons, Inc., New York, NY, USA. ',\n",
       " 'sklearn_6_4_imputation_of_missing_values\\n6.4. Imputation of missing values\\nmodules/impute.html\\n 6.4.5. Nearest neighbors imputation  The KNNImputer class provides imputation for filling in missing values using the k-Nearest Neighbors approach. By default, a euclidean distance metric that supports missing values, nan_euclidean_distances , is used to find the nearest neighbors. Each missing feature is imputed using values from nearest neighbors that have a value for the feature. The feature of the neighbors are averaged uniformly or weighted by distance to each neighbor. If a sample has more than one feature missing, then the neighbors for that sample can be different depending on the particular feature being imputed. When the number of available neighbors is less than and there are no defined distances to the training set, the training set average for that feature is used during imputation. If there is at least one neighbor with a defined distance, the weighted or unweighted average of the remaining neighbors will be used during imputation. If a feature is always missing in training, it is removed during . For more information on the methodology, see ref. [OL2001] . The following snippet demonstrates how to replace missing values, encoded as , using the mean feature value of the two nearest neighbors of samples with missing values: OL2001 Olga Troyanskaya, Michael Cantor, Gavin Sherlock, Pat Brown, Trevor Hastie, Robert Tibshirani, David Botstein and Russ B. Altman, Missing value estimation methods for DNA microarrays, BIOINFORMATICS Vol. 17 no. 6, 2001 Pages 520-525. ',\n",
       " 'sklearn_6_4_imputation_of_missing_values\\n6.4. Imputation of missing values\\nmodules/impute.html\\n 6.4.6. Marking imputed values  The MissingIndicator transformer is useful to transform a dataset into corresponding binary matrix indicating the presence of missing values in the dataset. This transformation is useful in conjunction with imputation. When using imputation, preserving the information about which values had been missing can be informative. Note that both the SimpleImputer and IterativeImputer have the boolean parameter ( by default) which when set to provides a convenient way of stacking the output of the MissingIndicator transformer with the output of the imputer. is usually used as the placeholder for missing values. However, it enforces the data type to be float. The parameter allows to specify other placeholder such as integer. In the following example, we will use as missing values: The parameter is used to choose the features for which the mask is constructed. By default, it is which returns the imputer mask of the features containing missing values at time: The parameter can be set to to return all features whether or not they contain missing values: When using the MissingIndicator in a Pipeline , be sure to use the FeatureUnion or ColumnTransformer to add the indicator features to the regular features. First we obtain the dataset, and add some missing values to it. Now we create a FeatureUnion . All features will be imputed using SimpleImputer , in order to enable classifiers to work with this data. Additionally, it adds the the indicator variables from MissingIndicator . Of course, we cannot use the transformer to make any predictions. We should wrap this in a Pipeline with a classifier (e.g., a DecisionTreeClassifier ) to be able to make predictions. ',\n",
       " 'sklearn_6_5_unsupervised_dimensionality_reduction\\n6.5. Unsupervised dimensionality reduction\\nmodules/unsupervised_reduction.html\\n 6.5.1. PCA: principal component analysis  decomposition.PCA looks for a combination of features that capture well the variance of the original features. See Decomposing signals in components (matrix factorization problems) . Examples Faces recognition example using eigenfaces and SVMs ',\n",
       " 'sklearn_6_5_unsupervised_dimensionality_reduction\\n6.5. Unsupervised dimensionality reduction\\nmodules/unsupervised_reduction.html\\n 6.5.2. Random projections  The module: random_projection provides several tools for data reduction by random projections. See the relevant section of the documentation: Random Projection . Examples The Johnson-Lindenstrauss bound for embedding with random projections ',\n",
       " 'sklearn_6_5_unsupervised_dimensionality_reduction\\n6.5. Unsupervised dimensionality reduction\\nmodules/unsupervised_reduction.html\\n 6.5.3. Feature agglomeration  cluster.FeatureAgglomeration applies Hierarchical clustering to group together features that behave similarly. Examples Feature agglomeration vs. univariate selection Feature agglomeration Feature scaling Note that if features have very different scaling or statistical properties, cluster.FeatureAgglomeration may not be able to capture the links between related features. Using a preprocessing.StandardScaler can be useful in these settings. ',\n",
       " 'sklearn_6_6_random_projection\\n6.6. Random Projection\\nmodules/random_projection.html\\n 6.6.1. The Johnson-Lindenstrauss lemma  The main theoretical result behind the efficiency of random projection is the Johnson-Lindenstrauss lemma (quoting Wikipedia) : In mathematics, the Johnson-Lindenstrauss lemma is a result concerning low-distortion embeddings of points from high-dimensional into low-dimensional Euclidean space. The lemma states that a small set of points in a high-dimensional space can be embedded into a space of much lower dimension in such a way that distances between the points are nearly preserved. The map used for the embedding is at least Lipschitz, and can even be taken to be an orthogonal projection. Knowing only the number of samples, the sklearn.random_projection.johnson_lindenstrauss_min_dim estimates conservatively the minimal size of the random subspace to guarantee a bounded distortion introduced by the random projection: Example: See The Johnson-Lindenstrauss bound for embedding with random projections for a theoretical explication on the Johnson-Lindenstrauss lemma and an empirical validation using sparse random matrices. References: Sanjoy Dasgupta and Anupam Gupta, 1999. An elementary proof of the Johnson-Lindenstrauss Lemma. ',\n",
       " 'sklearn_6_6_random_projection\\n6.6. Random Projection\\nmodules/random_projection.html\\n 6.6.2. Gaussian random projection  The sklearn.random_projection.GaussianRandomProjection reduces the dimensionality by projecting the original input space on a randomly generated matrix where components are drawn from the following distribution . Here a small excerpt which illustrates how to use the Gaussian random projection transformer: ',\n",
       " 'sklearn_6_6_random_projection\\n6.6. Random Projection\\nmodules/random_projection.html\\n 6.6.3. Sparse random projection  The sklearn.random_projection.SparseRandomProjection reduces the dimensionality by projecting the original input space using a sparse random matrix. Sparse random matrices are an alternative to dense Gaussian random projection matrix that guarantees similar embedding quality while being much more memory efficient and allowing faster computation of the projected data. If we define , the elements of the random matrix are drawn from \\\\[\\\\begin{split}\\\\left\\\\{ \\\\begin{array}{c c l} -\\\\sqrt{\\\\frac{s}{n_{\\\\text{components}}}} & & 1 / 2s\\\\\\\\ 0 &\\\\text{with probability} & 1 - 1 / s \\\\\\\\ +\\\\sqrt{\\\\frac{s}{n_{\\\\text{components}}}} & & 1 / 2s\\\\\\\\ \\\\end{array} \\\\right.\\\\end{split}\\\\] where is the size of the projected subspace. By default the density of non zero elements is set to the minimum density as recommended by Ping Li et al.: . Here a small excerpt which illustrates how to use the sparse random projection transformer: References: D. Achlioptas. 2003. Database-friendly random projections: Johnson-Lindenstrauss with binary coins . Journal of Computer and System Sciences 66 (2003) 671–687 Ping Li, Trevor J. Hastie, and Kenneth W. Church. 2006. Very sparse random projections. In Proceedings of the 12th ACM SIGKDD international conference on Knowledge discovery and data mining (KDD ‘06). ACM, New York, NY, USA, 287-296. ',\n",
       " 'sklearn_6_7_kernel_approximation\\n6.7. Kernel Approximation\\nmodules/kernel_approximation.html\\n 6.7.1. Nystroem Method for Kernel Approximation  The Nystroem method, as implemented in Nystroem is a general method for low-rank approximations of kernels. It achieves this by essentially subsampling the data on which the kernel is evaluated. By default Nystroem uses the kernel, but it can use any kernel function or a precomputed kernel matrix. The number of samples used - which is also the dimensionality of the features computed - is given by the parameter . ',\n",
       " 'sklearn_6_7_kernel_approximation\\n6.7. Kernel Approximation\\nmodules/kernel_approximation.html\\n 6.7.2. Radial Basis Function Kernel  The RBFSampler constructs an approximate mapping for the radial basis function kernel, also known as Random Kitchen Sinks [RR2007] . This transformation can be used to explicitly model a kernel map, prior to applying a linear algorithm, for example a linear SVM: The mapping relies on a Monte Carlo approximation to the kernel values. The function performs the Monte Carlo sampling, whereas the method performs the mapping of the data. Because of the inherent randomness of the process, results may vary between different calls to the function. The function takes two arguments: , which is the target dimensionality of the feature transform, and , the parameter of the RBF-kernel. A higher will result in a better approximation of the kernel and will yield results more similar to those produced by a kernel SVM. Note that “fitting” the feature function does not actually depend on the data given to the function. Only the dimensionality of the data is used. Details on the method can be found in [RR2007] . For a given value of RBFSampler is often less accurate as Nystroem . RBFSampler is cheaper to compute, though, making use of larger feature spaces more efficient. Comparing an exact RBF kernel (left) with the approximation (right)  Examples: Explicit feature map approximation for RBF kernels ',\n",
       " 'sklearn_6_7_kernel_approximation\\n6.7. Kernel Approximation\\nmodules/kernel_approximation.html\\n 6.7.3. Additive Chi Squared Kernel  The additive chi squared kernel is a kernel on histograms, often used in computer vision. The additive chi squared kernel as used here is given by \\\\[k(x, y)  \\\\sum_i \\\\frac{2x_iy_i}{x_i+y_i}\\\\] This is not exactly the same as sklearn.metrics.additive_chi2_kernel . The authors of [VZ2010] prefer the version above as it is always positive definite. Since the kernel is additive, it is possible to treat all components separately for embedding. This makes it possible to sample the Fourier transform in regular intervals, instead of approximating using Monte Carlo sampling. The class AdditiveChi2Sampler implements this component wise deterministic sampling. Each component is sampled times, yielding dimensions per input dimension (the multiple of two stems from the real and complex part of the Fourier transform). In the literature, is usually chosen to be 1 or 2, transforming the dataset to size (in the case of ). The approximate feature map provided by AdditiveChi2Sampler can be combined with the approximate feature map provided by RBFSampler to yield an approximate feature map for the exponentiated chi squared kernel. See the [VZ2010] for details and [VVZ2010] for combination with the RBFSampler . ',\n",
       " 'sklearn_6_7_kernel_approximation\\n6.7. Kernel Approximation\\nmodules/kernel_approximation.html\\n 6.7.4. Skewed Chi Squared Kernel  The skewed chi squared kernel is given by: \\\\[k(x,y)  \\\\prod_i \\\\frac{2\\\\sqrt{x_i+c}\\\\sqrt{y_i+c}}{x_i + y_i + 2c}\\\\] It has properties that are similar to the exponentiated chi squared kernel often used in computer vision, but allows for a simple Monte Carlo approximation of the feature map. The usage of the SkewedChi2Sampler is the same as the usage described above for the RBFSampler . The only difference is in the free parameter, that is called . For a motivation for this mapping and the mathematical details see [LS2010] . ',\n",
       " 'sklearn_6_7_kernel_approximation\\n6.7. Kernel Approximation\\nmodules/kernel_approximation.html\\n 6.7.5. Mathematical Details  Kernel methods like support vector machines or kernelized PCA rely on a property of reproducing kernel Hilbert spaces. For any positive definite kernel function (a so called Mercer kernel), it is guaranteed that there exists a mapping into a Hilbert space , such that \\\\[k(x,y)  \\\\langle \\\\phi(x), \\\\phi(y) \\\\rangle\\\\] Where denotes the inner product in the Hilbert space. If an algorithm, such as a linear support vector machine or PCA, relies only on the scalar product of data points , one may use the value of , which corresponds to applying the algorithm to the mapped data points . The advantage of using is that the mapping never has to be calculated explicitly, allowing for arbitrary large features (even infinite). One drawback of kernel methods is, that it might be necessary to store many kernel values during optimization. If a kernelized classifier is applied to new data , needs to be computed to make predictions, possibly for many different in the training set. The classes in this submodule allow to approximate the embedding , thereby working explicitly with the representations , which obviates the need to apply the kernel or store training examples. References: RR2007 ( 1 , 2 ) “Random features for large-scale kernel machines” Rahimi, A. and Recht, B. - Advances in neural information processing 2007, LS2010 “Random Fourier approximations for skewed multiplicative histogram kernels” Random Fourier approximations for skewed multiplicative histogram kernels - Lecture Notes for Computer Sciencd (DAGM) VZ2010 ( 1 , 2 ) “Efficient additive kernels via explicit feature maps” Vedaldi, A. and Zisserman, A. - Computer Vision and Pattern Recognition 2010 VVZ2010 “Generalized RBF feature maps for Efficient Detection” Vempati, S. and Vedaldi, A. and Zisserman, A. and Jawahar, CV - 2010 ',\n",
       " 'sklearn_6_8_pairwise_metrics_affinities_and_kernels\\n6.8. Pairwise metrics, Affinities and Kernels\\nmodules/metrics.html\\n 6.8.1. Cosine similarity  cosine_similarity computes the L2-normalized dot product of vectors. That is, if and are row vectors, their cosine similarity is defined as: \\\\[k(x, y)  \\\\frac{x y^\\\\top}{\\\\|x\\\\| \\\\|y\\\\|}\\\\] This is called cosine similarity, because Euclidean (L2) normalization projects the vectors onto the unit sphere, and their dot product is then the cosine of the angle between the points denoted by the vectors. This kernel is a popular choice for computing the similarity of documents represented as tf-idf vectors. cosine_similarity accepts matrices. (Note that the tf-idf functionality in can produce normalized vectors, in which case cosine_similarity is equivalent to linear_kernel , only slower.) References: C.D. Manning, P. Raghavan and H. Schütze (2008). Introduction to Information Retrieval. Cambridge University Press. https://nlp.stanford.edu/IR-book/html/htmledition/the-vector-space-model-for-scoring-1.html ',\n",
       " 'sklearn_6_8_pairwise_metrics_affinities_and_kernels\\n6.8. Pairwise metrics, Affinities and Kernels\\nmodules/metrics.html\\n 6.8.2. Linear kernel  The function linear_kernel computes the linear kernel, that is, a special case of polynomial_kernel with and (homogeneous). If and are column vectors, their linear kernel is: \\\\[k(x, y)  x^\\\\top y\\\\] ',\n",
       " 'sklearn_6_8_pairwise_metrics_affinities_and_kernels\\n6.8. Pairwise metrics, Affinities and Kernels\\nmodules/metrics.html\\n 6.8.3. Polynomial kernel  The function polynomial_kernel computes the degree-d polynomial kernel between two vectors. The polynomial kernel represents the similarity between two vectors. Conceptually, the polynomial kernels considers not only the similarity between vectors under the same dimension, but also across dimensions. When used in machine learning algorithms, this allows to account for feature interaction. The polynomial kernel is defined as: \\\\[k(x, y)  (\\\\gamma x^\\\\top y +c_0)^d\\\\] where: , are the input vectors is the kernel degree If the kernel is said to be homogeneous. ',\n",
       " 'sklearn_6_8_pairwise_metrics_affinities_and_kernels\\n6.8. Pairwise metrics, Affinities and Kernels\\nmodules/metrics.html\\n 6.8.4. Sigmoid kernel  The function sigmoid_kernel computes the sigmoid kernel between two vectors. The sigmoid kernel is also known as hyperbolic tangent, or Multilayer Perceptron (because, in the neural network field, it is often used as neuron activation function). It is defined as: \\\\[k(x, y)  \\\\tanh( \\\\gamma x^\\\\top y + c_0)\\\\] where: , are the input vectors is known as slope is known as intercept ',\n",
       " 'sklearn_6_8_pairwise_metrics_affinities_and_kernels\\n6.8. Pairwise metrics, Affinities and Kernels\\nmodules/metrics.html\\n 6.8.5. RBF kernel  The function rbf_kernel computes the radial basis function (RBF) kernel between two vectors. This kernel is defined as: \\\\[k(x, y)  \\\\exp( -\\\\gamma \\\\| x-y \\\\|^2)\\\\] where and are the input vectors. If the kernel is known as the Gaussian kernel of variance . ',\n",
       " 'sklearn_6_8_pairwise_metrics_affinities_and_kernels\\n6.8. Pairwise metrics, Affinities and Kernels\\nmodules/metrics.html\\n 6.8.6. Laplacian kernel  The function laplacian_kernel is a variant on the radial basis function kernel defined as: \\\\[k(x, y)  \\\\exp( -\\\\gamma \\\\| x-y \\\\|_1)\\\\] where and are the input vectors and is the Manhattan distance between the input vectors. It has proven useful in ML applied to noiseless data. See e.g. Machine learning for quantum mechanics in a nutshell . ',\n",
       " 'sklearn_6_8_pairwise_metrics_affinities_and_kernels\\n6.8. Pairwise metrics, Affinities and Kernels\\nmodules/metrics.html\\n 6.8.7. Chi-squared kernel  The chi-squared kernel is a very popular choice for training non-linear SVMs in computer vision applications. It can be computed using chi2_kernel and then passed to an sklearn.svm.SVC with : It can also be directly used as the argument: The chi squared kernel is given by \\\\[k(x, y)  \\\\exp \\\\left (-\\\\gamma \\\\sum_i \\\\frac{(x[i] - y[i]) ^ 2}{x[i] + y[i]} \\\\right )\\\\] The data is assumed to be non-negative, and is often normalized to have an L1-norm of one. The normalization is rationalized with the connection to the chi squared distance, which is a distance between discrete probability distributions. The chi squared kernel is most commonly used on histograms (bags) of visual words. References: Zhang, J. and Marszalek, M. and Lazebnik, S. and Schmid, C. Local features and kernels for classification of texture and object categories: A comprehensive study International Journal of Computer Vision 2007 https://research.microsoft.com/en-us/um/people/manik/projects/trade-off/papers/ZhangIJCV06.pdf ',\n",
       " 'sklearn_6_9_transforming_the_prediction_target_y\\n6.9. Transforming the prediction target (y)\\nmodules/preprocessing_targets.html\\n 6.9.1. Label binarization  LabelBinarizer is a utility class to help create a label indicator matrix from a list of multi-class labels: For multiple labels per instance, use MultiLabelBinarizer : ',\n",
       " 'sklearn_6_9_transforming_the_prediction_target_y\\n6.9. Transforming the prediction target (y)\\nmodules/preprocessing_targets.html\\n 6.9.2. Label encoding  LabelEncoder is a utility class to help normalize labels such that they contain only values between 0 and n_classes-1. This is sometimes useful for writing efficient Cython routines. LabelEncoder can be used as follows: It can also be used to transform non-numerical labels (as long as they are hashable and comparable) to numerical labels: ',\n",
       " 'sklearn_7_1_general_dataset_api\\n7.1. General dataset API\\ndatasets/index.html#general-dataset-api\\n 7.1. General dataset API  There are three main kinds of dataset interfaces that can be used to get datasets depending on the desired type of dataset. The dataset loaders. They can be used to load small standard datasets, described in the Toy datasets section. The dataset fetchers. They can be used to download and load larger datasets, described in the Real world datasets section. Both loaders and fetchers functions return a sklearn.utils.Bunch object holding at least two items: an array of shape * with key (except for 20newsgroups) and a numpy array of length , containing the target values, with key . The Bunch object is a dictionary that exposes its keys are attributes. For more information about Bunch object, see sklearn.utils.Bunch : It’s also possible for almost all of these function to constrain the output to be a tuple containing only the data and the target, by setting the parameter to . The datasets also contain a full description in their attribute and some contain and . See the dataset descriptions below for details. The dataset generation functions. They can be used to generate controlled synthetic datasets, described in the Generated datasets section. These functions return a tuple consisting of a * numpy array and an array of length containing the targets . In addition, there are also miscellaneous tools to load datasets of other formats or from other locations, described in the Loading other datasets section. ',\n",
       " 'sklearn_7_1_general_dataset_api\\n7.1. General dataset API\\ndatasets/index.html#general-dataset-api\\n 7.2. Toy datasets  scikit-learn comes with a few small standard datasets that do not require to download any file from some external website. They can be loaded using the following functions: load_boston (*[,\\xa0return_X_y]) Load and return the boston house-prices dataset (regression). load_iris (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the iris dataset (classification). load_diabetes (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the diabetes dataset (regression). load_digits (*[,\\xa0n_class,\\xa0return_X_y,\\xa0as_frame]) Load and return the digits dataset (classification). load_linnerud (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the physical excercise linnerud dataset. load_wine (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the wine dataset (classification). load_breast_cancer (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the breast cancer wisconsin dataset (classification). These datasets are useful to quickly illustrate the behavior of the various algorithms implemented in scikit-learn. They are however often too small to be representative of real world machine learning tasks. 7.2.1. Boston house prices dataset  Data Set Characteristics: Number of Instances 506 Number of Attributes 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target. Attribute Information (in order) CRIM per capita crime rate by town ZN proportion of residential land zoned for lots over 25,000 sq.ft. INDUS proportion of non-retail business acres per town CHAS Charles River dummy variable ( 1 if tract bounds river; 0 otherwise) NOX nitric oxides concentration (parts per 10 million) RM average number of rooms per dwelling AGE proportion of owner-occupied units built prior to 1940 DIS weighted distances to five Boston employment centres RAD index of accessibility to radial highways TAX full-value property-tax rate per $10,000 PTRATIO pupil-teacher ratio by town B 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town LSTAT % lower status of the population MEDV Median value of owner-occupied homes in $1000’s Missing Attribute Values None Creator Harrison, D. and Rubinfeld, D.L. This is a copy of UCI ML housing dataset. https://archive.ics.uci.edu/ml/machine-learning-databases/housing/ This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University. The Boston house-price data of Harrison, D. and Rubinfeld, D.L. ‘Hedonic prices and the demand for clean air’, J. Environ. Economics & Management, vol.5, 81-102, 1978. Used in Belsley, Kuh & Welsch, ‘Regression diagnostics …’, Wiley, 1980. N.B. Various transformations are used in the table on pages 244-261 of the latter. The Boston house-price data has been used in many machine learning papers that address regression problems. References Belsley, Kuh & Welsch, ‘Regression diagnostics: Identifying Influential Data and Sources of Collinearity’, Wiley, 1980. 244-261. Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning. In Proceedings on the Tenth International Conference of Machine Learning, 236-243, University of Massachusetts, Amherst. Morgan Kaufmann. 7.2.2. Iris plants dataset  Data Set Characteristics: Number of Instances 150 (50 in each of three classes) Number of Attributes 4 numeric, predictive attributes and the class Attribute Information sepal length in cm sepal width in cm petal length in cm petal width in cm class: Iris-Setosa Iris-Versicolour Iris-Virginica Summary Statistics sepal length: 4.3 7.9 5.84 0.83 0.7826 sepal width: 2.0 4.4 3.05 0.43 -0.4194 petal length: 1.0 6.9 3.76 1.76 0.9490 (high!) petal width: 0.1 2.5 1.20 0.76 0.9565 (high!) Missing Attribute Values None Class Distribution 33.3% for each of 3 classes. Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 The famous Iris database, first used by Sir R.A. Fisher. The dataset is taken from Fisher’s paper. Note that it’s the same as in R, but not as in the UCI Machine Learning Repository, which has two wrong data points. This is perhaps the best known database to be found in the pattern recognition literature. Fisher’s paper is a classic in the field and is referenced frequently to this day. (See Duda & Hart, for example.) The data set contains 3 classes of 50 instances each, where each class refers to a type of iris plant. One class is linearly separable from the other 2; the latter are NOT linearly separable from each other. References Fisher, R.A. “The use of multiple measurements in taxonomic problems” Annual Eugenics, 7, Part II, 179-188 (1936); also in “Contributions to Mathematical Statistics” (John Wiley, NY, 1950). Duda, R.O., & Hart, P.E. (1973) Pattern Classification and Scene Analysis. (Q327.D83) John Wiley & Sons. ISBN 0-471-22361-1. See page 218. Dasarathy, B.V. (1980) “Nosing Around the Neighborhood: A New System Structure and Classification Rule for Recognition in Partially Exposed Environments”. IEEE Transactions on Pattern Analysis and Machine Intelligence, Vol. PAMI-2, No. 1, 67-71. Gates, G.W. (1972) “The Reduced Nearest Neighbor Rule”. IEEE Transactions on Information Theory, May 1972, 431-433. See also: 1988 MLC Proceedings, 54-64. Cheeseman et al”s AUTOCLASS II conceptual clustering system finds 3 classes in the data. Many, many more … 7.2.3. Diabetes dataset  Ten baseline variables, age, sex, body mass index, average blood pressure, and six blood serum measurements were obtained for each of n  442 diabetes patients, as well as the response of interest, a quantitative measure of disease progression one year after baseline. Data Set Characteristics: Number of Instances 442 Number of Attributes First 10 columns are numeric predictive values Target Column 11 is a quantitative measure of disease progression one year after baseline Attribute Information age age in years sex bmi body mass index bp average blood pressure s1 tc, T-Cells (a type of white blood cells) s2 ldl, low-density lipoproteins s3 hdl, high-density lipoproteins s4 tch, thyroid stimulating hormone s5 ltg, lamotrigine s6 glu, blood sugar level Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times (i.e. the sum of squares of each column totals 1). Source URL: https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html For more information see: Bradley Efron, Trevor Hastie, Iain Johnstone and Robert Tibshirani (2004) “Least Angle Regression,” Annals of Statistics (with discussion), 407-499. ( https://web.stanford.edu/~hastie/Papers/LARS/LeastAngle_2002.pdf ) 7.2.4. Optical recognition of handwritten digits dataset  Data Set Characteristics: Number of Instances 5620 Number of Attributes 64 Attribute Information 8x8 image of integer pixels in the range 0..16. Missing Attribute Values None Creator Alpaydin (alpaydin ‘@’ boun.edu.tr) Date July; 1998 This is a copy of the test set of the UCI ML hand-written digits datasets https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits The data set contains images of hand-written digits: 10 classes where each class refers to a digit. Preprocessing programs made available by NIST were used to extract normalized bitmaps of handwritten digits from a preprinted form. From a total of 43 people, 30 contributed to the training set and different 13 to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of 4x4 and the number of on pixels are counted in each block. This generates an input matrix of 8x8 where each element is an integer in the range 0..16. This reduces dimensionality and gives invariance to small distortions. For info on NIST preprocessing routines, see M. D. Garris, J. L. Blue, G. T. Candela, D. L. Dimmick, J. Geist, P. J. Grother, S. A. Janet, and C. L. Wilson, NIST Form-Based Handprint Recognition System, NISTIR 5469, 1994. References C. Kaynak (1995) Methods of Combining Multiple Classifiers and Their Applications to Handwritten Digit Recognition, MSc Thesis, Institute of Graduate Studies in Science and Engineering, Bogazici University. Alpaydin, C. Kaynak (1998) Cascading Classifiers, Kybernetika. Ken Tang and Ponnuthurai N. Suganthan and Xi Yao and A. Kai Qin. Linear dimensionalityreduction using relevance weighted LDA. School of Electrical and Electronic Engineering Nanyang Technological University. 2005. Claudio Gentile. A New Approximate Maximal Margin Classification Algorithm. NIPS. 2000. 7.2.5. Linnerrud dataset  Data Set Characteristics: Number of Instances 20 Number of Attributes 3 Missing Attribute Values None The Linnerud dataset is a multi-output regression dataset. It consists of three excercise (data) and three physiological (target) variables collected from twenty middle-aged men in a fitness club: physiological - CSV containing 20 observations on 3 physiological variables: Weight, Waist and Pulse. exercise - CSV containing 20 observations on 3 exercise variables: Chins, Situps and Jumps. References Tenenhaus, M. (1998). La regression PLS: theorie et pratique. Paris: Editions Technic. 7.2.6. Wine recognition dataset  Data Set Characteristics: Number of Instances 178 (50 in each of three classes) Number of Attributes 13 numeric, predictive attributes and the class Attribute Information Alcohol Malic acid Ash Alcalinity of ash Magnesium Total phenols Flavanoids Nonflavanoid phenols Proanthocyanins Color intensity Hue OD280/OD315 of diluted wines Proline class: class_0 class_1 class_2 Summary Statistics Alcohol: 11.0 14.8 13.0 0.8 Malic Acid: 0.74 5.80 2.34 1.12 Ash: 1.36 3.23 2.36 0.27 Alcalinity of Ash: 10.6 30.0 19.5 3.3 Magnesium: 70.0 162.0 99.7 14.3 Total Phenols: 0.98 3.88 2.29 0.63 Flavanoids: 0.34 5.08 2.03 1.00 Nonflavanoid Phenols: 0.13 0.66 0.36 0.12 Proanthocyanins: 0.41 3.58 1.59 0.57 Colour Intensity: 1.3 13.0 5.1 2.3 Hue: 0.48 1.71 0.96 0.23 OD280/OD315 of diluted wines: 1.27 4.00 2.61 0.71 Proline: 278 1680 746 315 Missing Attribute Values None Class Distribution class_0 (59), class_1 (71), class_2 (48) Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 This is a copy of UCI ML Wine recognition datasets. https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data The data is the results of a chemical analysis of wines grown in the same region in Italy by three different cultivators. There are thirteen different measurements taken for different constituents found in the three types of wine. Original Owners: Forina, M. et al, PARVUS - An Extendible Package for Data Exploration, Classification and Correlation. Institute of Pharmaceutical and Food Analysis and Technologies, Via Brigata Salerno, 16147 Genoa, Italy. Citation: Lichman, M. (2013). UCI Machine Learning Repository [ https://archive.ics.uci.edu/ml ]. Irvine, CA: University of California, School of Information and Computer Science. References (1) S. Aeberhard, D. Coomans and O. de Vel, Comparison of Classifiers in High Dimensional Settings, Tech. Rep. no. 92-02, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Technometrics). The data was used with many others for comparing various classifiers. The classes are separable, though only RDA has achieved 100% correct classification. (RDA : 100%, QDA 99.4%, LDA 98.9%, 1NN 96.1% (z-transformed data)) (All results using the leave-one-out technique) (2) S. Aeberhard, D. Coomans and O. de Vel, “THE CLASSIFICATION PERFORMANCE OF RDA” Tech. Rep. no. 92-01, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Journal of Chemometrics). 7.2.7. Breast cancer wisconsin (diagnostic) dataset  Data Set Characteristics: Number of Instances 569 Number of Attributes 30 numeric, predictive attributes and the class Attribute Information radius (mean of distances from center to points on the perimeter) texture (standard deviation of gray-scale values) perimeter area smoothness (local variation in radius lengths) compactness (perimeter^2 / area - 1.0) concavity (severity of concave portions of the contour) concave points (number of concave portions of the contour) symmetry fractal dimension (“coastline approximation” - 1) The mean, standard error, and “worst” or largest (mean of the three worst/largest values) of these features were computed for each image, resulting in 30 features. For instance, field 0 is Mean Radius, field 10 is Radius SE, field 20 is Worst Radius. class: WDBC-Malignant WDBC-Benign Summary Statistics radius (mean): 6.981 28.11 texture (mean): 9.71 39.28 perimeter (mean): 43.79 188.5 area (mean): 143.5 2501.0 smoothness (mean): 0.053 0.163 compactness (mean): 0.019 0.345 concavity (mean): 0.0 0.427 concave points (mean): 0.0 0.201 symmetry (mean): 0.106 0.304 fractal dimension (mean): 0.05 0.097 radius (standard error): 0.112 2.873 texture (standard error): 0.36 4.885 perimeter (standard error): 0.757 21.98 area (standard error): 6.802 542.2 smoothness (standard error): 0.002 0.031 compactness (standard error): 0.002 0.135 concavity (standard error): 0.0 0.396 concave points (standard error): 0.0 0.053 symmetry (standard error): 0.008 0.079 fractal dimension (standard error): 0.001 0.03 radius (worst): 7.93 36.04 texture (worst): 12.02 49.54 perimeter (worst): 50.41 251.2 area (worst): 185.2 4254.0 smoothness (worst): 0.071 0.223 compactness (worst): 0.027 1.058 concavity (worst): 0.0 1.252 concave points (worst): 0.0 0.291 symmetry (worst): 0.156 0.664 fractal dimension (worst): 0.055 0.208 Missing Attribute Values None Class Distribution 212 - Malignant, 357 - Benign Creator Dr. William H. Wolberg, W. Nick Street, Olvi L. Mangasarian Donor Nick Street Date November, 1995 This is a copy of UCI ML Breast Cancer Wisconsin (Diagnostic) datasets. https://goo.gl/U2Uwz2 Features are computed from a digitized image of a fine needle aspirate (FNA) of a breast mass. They describe characteristics of the cell nuclei present in the image. Separating plane described above was obtained using Multisurface Method-Tree (MSM-T) [K. P. Bennett, “Decision Tree Construction Via Linear Programming.” Proceedings of the 4th Midwest Artificial Intelligence and Cognitive Science Society, pp. 97-101, 1992], a classification method which uses linear programming to construct a decision tree. Relevant features were selected using an exhaustive search in the space of 1-4 features and 1-3 separating planes. The actual linear program used to obtain the separating plane in the 3-dimensional space is that described in: [K. P. Bennett and O. L. Mangasarian: “Robust Linear Programming Discrimination of Two Linearly Inseparable Sets”, Optimization Methods and Software 1, 1992, 23-34]. This database is also available through the UW CS ftp server: ftp ftp.cs.wisc.edu cd math-prog/cpo-dataset/machine-learn/WDBC/ References W.N. Street, W.H. Wolberg and O.L. Mangasarian. Nuclear feature extraction for breast tumor diagnosis. IS&T/SPIE 1993 International Symposium on Electronic Imaging: Science and Technology, volume 1905, pages 861-870, San Jose, CA, 1993. O.L. Mangasarian, W.N. Street and W.H. Wolberg. Breast cancer diagnosis and prognosis via linear programming. Operations Research, 43(4), pages 570-577, July-August 1995. W.H. Wolberg, W.N. Street, and O.L. Mangasarian. Machine learning techniques to diagnose breast cancer from fine-needle aspirates. Cancer Letters 77 (1994) 163-171. ',\n",
       " 'sklearn_7_1_general_dataset_api\\n7.1. General dataset API\\ndatasets/index.html#general-dataset-api\\n 7.3. Real world datasets  scikit-learn provides tools to load larger datasets, downloading them if necessary. They can be loaded using the following functions: fetch_olivetti_faces (*[,\\xa0data_home,\\xa0…]) Load the Olivetti faces data-set from AT&T (classification). fetch_20newsgroups (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the filenames and data from the 20 newsgroups dataset (classification). fetch_20newsgroups_vectorized (*[,\\xa0subset,\\xa0…]) Load the 20 newsgroups dataset and vectorize it into token counts (classification). fetch_lfw_people (*[,\\xa0data_home,\\xa0funneled,\\xa0…]) Load the Labeled Faces in the Wild (LFW) people dataset (classification). fetch_lfw_pairs (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the Labeled Faces in the Wild (LFW) pairs dataset (classification). fetch_covtype (*[,\\xa0data_home,\\xa0…]) Load the covertype dataset (classification). fetch_rcv1 (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the RCV1 multilabel dataset (classification). fetch_kddcup99 (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the kddcup99 dataset (classification). fetch_california_housing (*[,\\xa0data_home,\\xa0…]) Load the California housing dataset (regression). 7.3.1. The Olivetti faces dataset  This dataset contains a set of face images taken between April 1992 and April 1994 at AT&T Laboratories Cambridge. The sklearn.datasets.fetch_olivetti_faces function is the data fetching / caching function that downloads the data archive from AT&T. As described on the original website: There are ten different images of each of 40 distinct subjects. For some subjects, the images were taken at different times, varying the lighting, facial expressions (open / closed eyes, smiling / not smiling) and facial details (glasses / no glasses). All the images were taken against a dark homogeneous background with the subjects in an upright, frontal position (with tolerance for some side movement). Data Set Characteristics: Classes 40 Samples total 400 Dimensionality 4096 Features real, between 0 and 1 The image is quantized to 256 grey levels and stored as unsigned 8-bit integers; the loader will convert these to floating point values on the interval [0, 1], which are easier to work with for many algorithms. The “target” for this database is an integer from 0 to 39 indicating the identity of the person pictured; however, with only 10 examples per class, this relatively small dataset is more interesting from an unsupervised or semi-supervised perspective. The original dataset consisted of 92 x 112, while the version available here consists of 64x64 images. When using these images, please give credit to AT&T Laboratories Cambridge. 7.3.2. The 20 newsgroups text dataset  The 20 newsgroups dataset comprises around 18000 newsgroups posts on 20 topics split in two subsets: one for training (or development) and the other one for testing (or for performance evaluation). The split between the train and test set is based upon a messages posted before and after a specific date. This module contains two loaders. The first one, sklearn.datasets.fetch_20newsgroups , returns a list of the raw texts that can be fed to text feature extractors such as sklearn.feature_extraction.text.CountVectorizer with custom parameters so as to extract feature vectors. The second one, sklearn.datasets.fetch_20newsgroups_vectorized , returns ready-to-use features, i.e., it is not necessary to use a feature extractor. Data Set Characteristics: Classes 20 Samples total 18846 Dimensionality 1 Features text 7.3.2.1. Usage  The sklearn.datasets.fetch_20newsgroups function is a data fetching / caching functions that downloads the data archive from the original 20 newsgroups website , extracts the archive contents in the folder and calls the sklearn.datasets.load_files on either the training or testing set folder, or both of them: The real data lies in the and attributes. The target attribute is the integer index of the category: It is possible to load only a sub-selection of the categories by passing the list of the categories to load to the sklearn.datasets.fetch_20newsgroups function: 7.3.2.2. Converting text to vectors  In order to feed predictive or clustering models with the text data, one first need to turn the text into vectors of numerical values suitable for statistical analysis. This can be achieved with the utilities of the as demonstrated in the following example that extract TF-IDF vectors of unigram tokens from a subset of 20news: The extracted TF-IDF vectors are very sparse, with an average of 159 non-zero components by sample in a more than 30000-dimensional space (less than .5% non-zero features): sklearn.datasets.fetch_20newsgroups_vectorized is a function which returns ready-to-use token counts features instead of file names. 7.3.2.3. Filtering text for more realistic training  It is easy for a classifier to overfit on particular things that appear in the 20 Newsgroups data, such as newsgroup headers. Many classifiers achieve very high F-scores, but their results would not generalize to other documents that aren’t from this window of time. For example, let’s look at the results of a multinomial Naive Bayes classifier, which is fast to train and achieves a decent F-score: (The example Classification of text documents using sparse features shuffles the training and test data, instead of segmenting by time, and in that case multinomial Naive Bayes gets a much higher F-score of 0.88. Are you suspicious yet of what’s going on inside this classifier?) Let’s take a look at what the most informative features are: You can now see many things that these features have overfit to: Almost every group is distinguished by whether headers such as and appear more or less often. Another significant feature involves whether the sender is affiliated with a university, as indicated either by their headers or their signature. The word “article” is a significant feature, based on how often people quote previous posts like this: “In article [article ID], [name] <[e-mail address]> wrote:” Other features match the names and e-mail addresses of particular people who were posting at the time. With such an abundance of clues that distinguish newsgroups, the classifiers barely have to identify topics from text at all, and they all perform at the same high level. For this reason, the functions that load 20 Newsgroups data provide a parameter called remove , telling it what kinds of information to strip out of each file. remove should be a tuple containing any subset of , telling it to remove headers, signature blocks, and quotation blocks respectively. This classifier lost over a lot of its F-score, just because we removed metadata that has little to do with topic classification. It loses even more if we also strip this metadata from the training data: Some other classifiers cope better with this harder version of the task. Try running Sample pipeline for text feature extraction and evaluation with and without the option to compare the results. Recommendation When evaluating text classifiers on the 20 Newsgroups data, you should strip newsgroup-related metadata. In scikit-learn, you can do this by setting . The F-score will be lower because it is more realistic. Examples Sample pipeline for text feature extraction and evaluation Classification of text documents using sparse features 7.3.3. The Labeled Faces in the Wild face recognition dataset  This dataset is a collection of JPEG pictures of famous people collected over the internet, all details are available on the official website: http://vis-www.cs.umass.edu/lfw/ Each picture is centered on a single face. The typical task is called Face Verification: given a pair of two pictures, a binary classifier must predict whether the two images are from the same person. An alternative task, Face Recognition or Face Identification is: given the picture of the face of an unknown person, identify the name of the person by referring to a gallery of previously seen pictures of identified persons. Both Face Verification and Face Recognition are tasks that are typically performed on the output of a model trained to perform Face Detection. The most popular model for Face Detection is called Viola-Jones and is implemented in the OpenCV library. The LFW faces were extracted by this face detector from various online websites. Data Set Characteristics: Classes 5749 Samples total 13233 Dimensionality 5828 Features real, between 0 and 255 7.3.3.1. Usage  provides two loaders that will automatically download, cache, parse the metadata files, decode the jpeg and convert the interesting slices into memmapped numpy arrays. This dataset size is more than 200 MB. The first load typically takes more than a couple of minutes to fully decode the relevant part of the JPEG files into numpy arrays. If the dataset has been loaded once, the following times the loading times less than 200ms by using a memmapped version memoized on the disk in the folder using . The first loader is used for the Face Identification task: a multi-class classification task (hence supervised learning): The default slice is a rectangular shape around the face, removing most of the background: Each of the faces is assigned to a single person id in the array: The second loader is typically used for the face verification task: each sample is a pair of two picture belonging or not to the same person: Both for the sklearn.datasets.fetch_lfw_people and sklearn.datasets.fetch_lfw_pairs function it is possible to get an additional dimension with the RGB color channels by passing , in that case the shape will be . The sklearn.datasets.fetch_lfw_pairs datasets is subdivided into 3 subsets: the development set, the development set and an evaluation set meant to compute performance metrics using a 10-folds cross validation scheme. References: Labeled Faces in the Wild: A Database for Studying Face Recognition in Unconstrained Environments. Gary B. Huang, Manu Ramesh, Tamara Berg, and Erik Learned-Miller. University of Massachusetts, Amherst, Technical Report 07-49, October, 2007. 7.3.3.2. Examples  Faces recognition example using eigenfaces and SVMs 7.3.4. Forest covertypes  The samples in this dataset correspond to 30×30m patches of forest in the US, collected for the task of predicting each patch’s cover type, i.e. the dominant species of tree. There are seven covertypes, making this a multiclass classification problem. Each sample has 54 features, described on the dataset’s homepage . Some of the features are boolean indicators, while others are discrete or continuous measurements. Data Set Characteristics: Classes 7 Samples total 581012 Dimensionality 54 Features int sklearn.datasets.fetch_covtype will load the covertype dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.5. RCV1 dataset  Reuters Corpus Volume I (RCV1) is an archive of over 800,000 manually categorized newswire stories made available by Reuters, Ltd. for research purposes. The dataset is extensively described in 1 . Data Set Characteristics: Classes 103 Samples total 804414 Dimensionality 47236 Features real, between 0 and 1 sklearn.datasets.fetch_rcv1 will load the following version: RCV1-v2, vectors, full sets, topics multilabels: It returns a dictionary-like object, with the following attributes: : The feature matrix is a scipy CSR sparse matrix, with 804414 samples and 47236 features. Non-zero values contains cosine-normalized, log TF-IDF vectors. A nearly chronological split is proposed in 1 : The first 23149 samples are the training set. The last 781265 samples are the testing set. This follows the official LYRL2004 chronological split. The array has 0.16% of non zero values: : The target values are stored in a scipy CSR sparse matrix, with 804414 samples and 103 categories. Each sample has a value of 1 in its categories, and 0 in others. The array has 3.15% of non zero values: : Each sample can be identified by its ID, ranging (with gaps) from 2286 to 810596: : The target values are the topics of each sample. Each sample belongs to at least one topic, and to up to 17 topics. There are 103 topics, each represented by a string. Their corpus frequencies span five orders of magnitude, from 5 occurrences for ‘GMIL’, to 381327 for ‘CCAT’: The dataset will be downloaded from the rcv1 homepage if necessary. The compressed size is about 656 MB. References 1 ( 1 , 2 ) Lewis, D. D., Yang, Y., Rose, T. G., & Li, F. (2004). RCV1: A new benchmark collection for text categorization research. The Journal of Machine Learning Research, 5, 361-397. 7.3.6. Kddcup 99 dataset  The KDD Cup ‘99 dataset was created by processing the tcpdump portions of the 1998 DARPA Intrusion Detection System (IDS) Evaluation dataset, created by MIT Lincoln Lab [1]. The artificial data (described on the dataset’s homepage ) was generated using a closed network and hand-injected attacks to produce a large number of different types of attack with normal activity in the background. As the initial goal was to produce a large training set for supervised learning algorithms, there is a large proportion (80.1%) of abnormal data which is unrealistic in real world, and inappropriate for unsupervised anomaly detection which aims at detecting ‘abnormal’ data, ie qualitatively different from normal data in large minority among the observations. We thus transform the KDD Data set into two different data sets: SA and SF. -SA is obtained by simply selecting all the normal data, and a small proportion of abnormal data to gives an anomaly proportion of 1%. -SF is obtained as in [2] by simply picking up the data whose attribute logged_in is positive, thus focusing on the intrusion attack, which gives a proportion of 0.3% of attack. -http and smtp are two subsets of SF corresponding with third feature equal to ‘http’ (resp. to ‘smtp’) General KDD structure : Samples total 4898431 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SA structure : Samples total 976158 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SF structure : Samples total 699691 Dimensionality 4 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type http structure : Samples total 619052 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type smtp structure : Samples total 95373 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type sklearn.datasets.fetch_kddcup99 will load the kddcup99 dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.7. California Housing dataset  Data Set Characteristics: Number of Instances 20640 Number of Attributes 8 numeric, predictive attributes and the target Attribute Information MedInc median income in block HouseAge median house age in block AveRooms average number of rooms AveBedrms average number of bedrooms Population block population AveOccup average house occupancy Latitude house block latitude Longitude house block longitude Missing Attribute Values None This dataset was obtained from the StatLib repository. http://lib.stat.cmu.edu/datasets/ The target variable is the median house value for California districts. This dataset was derived from the 1990 U.S. census, using one row per census block group. A block group is the smallest geographical unit for which the U.S. Census Bureau publishes sample data (a block group typically has a population of 600 to 3,000 people). It can be downloaded/loaded using the sklearn.datasets.fetch_california_housing function. References Pace, R. Kelley and Ronald Barry, Sparse Spatial Autoregressions, Statistics and Probability Letters, 33 (1997) 291-297 ',\n",
       " 'sklearn_7_1_general_dataset_api\\n7.1. General dataset API\\ndatasets/index.html#general-dataset-api\\n 7.4. Generated datasets  In addition, scikit-learn includes various random sample generators that can be used to build artificial datasets of controlled size and complexity. 7.4.1. Generators for classification and clustering  These generators produce a matrix of features and corresponding discrete targets. 7.4.1.1. Single label  Both make_blobs and make_classification create multiclass datasets by allocating each class one or more normally-distributed clusters of points. make_blobs provides greater control regarding the centers and standard deviations of each cluster, and is used to demonstrate clustering. make_classification specialises in introducing noise by way of: correlated, redundant and uninformative features; multiple Gaussian clusters per class; and linear transformations of the feature space. make_gaussian_quantiles divides a single Gaussian cluster into near-equal-size classes separated by concentric hyperspheres. make_hastie_10_2 generates a similar binary, 10-dimensional problem. make_circles and make_moons generate 2d binary classification datasets that are challenging to certain algorithms (e.g. centroid-based clustering or linear classification), including optional Gaussian noise. They are useful for visualisation. make_circles produces Gaussian data with a spherical decision boundary for binary classification, while make_moons produces two interleaving half circles. 7.4.1.2. Multilabel  make_multilabel_classification generates random samples with multiple labels, reflecting a bag of words drawn from a mixture of topics. The number of topics for each document is drawn from a Poisson distribution, and the topics themselves are drawn from a fixed random distribution. Similarly, the number of words is drawn from Poisson, with words drawn from a multinomial, where each topic defines a probability distribution over words. Simplifications with respect to true bag-of-words mixtures include: Per-topic word distributions are independently drawn, where in reality all would be affected by a sparse base distribution, and would be correlated. For a document generated from multiple topics, all topics are weighted equally in generating its bag of words. Documents without labels words at random, rather than from a base distribution. 7.4.1.3. Biclustering  make_biclusters (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with constant block diagonal structure for biclustering. make_checkerboard (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with block checkerboard structure for biclustering. 7.4.2. Generators for regression  make_regression produces regression targets as an optionally-sparse random linear combination of random features, with noise. Its informative features may be uncorrelated, or low rank (few features account for most of the variance). Other regression generators generate functions deterministically from randomized features. make_sparse_uncorrelated produces a target as a linear combination of four features with fixed coefficients. Others encode explicitly non-linear relations: make_friedman1 is related by polynomial and sine transforms; make_friedman2 includes feature multiplication and reciprocation; and make_friedman3 is similar with an arctan transformation on the target. 7.4.3. Generators for manifold learning  make_s_curve ([n_samples,\\xa0noise,\\xa0random_state]) Generate an S curve dataset. make_swiss_roll ([n_samples,\\xa0noise,\\xa0random_state]) Generate a swiss roll dataset. 7.4.4. Generators for decomposition  make_low_rank_matrix ([n_samples,\\xa0…]) Generate a mostly low rank matrix with bell-shaped singular values make_sparse_coded_signal (n_samples,\\xa0*,\\xa0…) Generate a signal as a sparse combination of dictionary elements. make_spd_matrix (n_dim,\\xa0*[,\\xa0random_state]) Generate a random symmetric, positive-definite matrix. make_sparse_spd_matrix ([dim,\\xa0alpha,\\xa0…]) Generate a sparse symmetric definite positive matrix. ',\n",
       " 'sklearn_7_1_general_dataset_api\\n7.1. General dataset API\\ndatasets/index.html#general-dataset-api\\n 7.5. Loading other datasets  7.5.1. Sample images  Scikit-learn also embed a couple of sample JPEG images published under Creative Commons license by their authors. Those images can be useful to test algorithms and pipeline on 2D data. load_sample_images () Load sample images for image manipulation. load_sample_image (image_name) Load the numpy array of a single sample image Warning The default coding of images is based on the dtype to spare memory. Often machine learning algorithms work best if the input is converted to a floating point representation first. Also, if you plan to use don’t forget to scale to the range 0 - 1 as done in the following example. Examples: Color Quantization using K-Means 7.5.2. Datasets in svmlight / libsvm format  scikit-learn includes utility functions for loading datasets in the svmlight / libsvm format. In this format, each line takes the form . This format is especially suitable for sparse datasets. In this module, scipy sparse CSR matrices are used for and numpy arrays are used for . You may load a dataset like as follows: You may also load two (or more) datasets at once: In this case, and are guaranteed to have the same number of features. Another way to achieve the same result is to fix the number of features: Related links: Public datasets in svmlight / libsvm format : https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets Faster API-compatible implementation : https://github.com/mblondel/svmlight-loader 7.5.3. Downloading datasets from the openml.org repository  openml.org is a public repository for machine learning data and experiments, that allows everybody to upload open datasets. The package is able to download datasets from the repository using the function sklearn.datasets.fetch_openml . For example, to download a dataset of gene expressions in mice brains: To fully specify a dataset, you need to provide a name and a version, though the version is optional, see Dataset Versions below. The dataset contains a total of 1080 examples belonging to 8 different classes: You can get more information on the dataset by looking at the and attributes: The contains a free-text description of the data, while contains a dictionary of meta-data stored by openml, like the dataset id. For more details, see the OpenML documentation The of the mice protein dataset is 40966, and you can use this (or the name) to get more information on the dataset on the openml website: The also uniquely identifies a dataset from OpenML: 7.5.3.1. Dataset Versions  A dataset is uniquely specified by its , but not necessarily by its name. Several different “versions” of a dataset with the same name can exist which can contain entirely different datasets. If a particular version of a dataset has been found to contain significant issues, it might be deactivated. Using a name to specify a dataset will yield the earliest version of a dataset that is still active. That means that can yield different results at different times if earlier versions become inactive. You can see that the dataset with 40966 that we fetched above is the version 1 of the “miceprotein” dataset: In fact, this dataset only has one version. The iris dataset on the other hand has multiple versions: Specifying the dataset by the name “iris” yields the lowest version, version 1, with the 61. To make sure you always get this exact dataset, it is safest to specify it by the dataset . The other dataset, with 969, is version 3 (version 2 has become inactive), and contains a binarized version of the data: You can also specify both the name and the version, which also uniquely identifies the dataset: References: Vanschoren, van Rijn, Bischl and Torgo “OpenML: networked science in machine learning” , ACM SIGKDD Explorations Newsletter, 15(2), 49-60, 2014. 7.5.4. Loading from external datasets  scikit-learn works on any numeric data stored as numpy arrays or scipy sparse matrices. Other types that are convertible to numeric arrays such as pandas DataFrame are also acceptable. Here are some recommended ways to load standard columnar data into a format usable by scikit-learn: pandas.io provides tools to read data from common formats including CSV, Excel, JSON and SQL. DataFrames may also be constructed from lists of tuples or dicts. Pandas handles heterogeneous data smoothly and provides tools for manipulation and conversion into a numeric array suitable for scikit-learn. scipy.io specializes in binary formats often used in scientific computing context such as .mat and .arff numpy/routines.io for standard loading of columnar data into numpy arrays scikit-learn’s datasets.load_svmlight_file for the svmlight or libSVM sparse format scikit-learn’s datasets.load_files for directories of text files where the name of each directory is the name of each category and each file inside of each directory corresponds to one sample from that category For some miscellaneous data such as images, videos, and audio, you may wish to refer to: skimage.io or Imageio for loading images and videos into numpy arrays scipy.io.wavfile.read for reading WAV files into a numpy array Categorical (or nominal) features stored as strings (common in pandas DataFrames) will need converting to numerical features using sklearn.preprocessing.OneHotEncoder or sklearn.preprocessing.OrdinalEncoder or similar. See Preprocessing data . Note: if you manage your own numerical data it is recommended to use an optimized file format such as HDF5 to reduce data load times. Various libraries such as H5Py, PyTables and pandas provides a Python interface for reading and writing data in that format. ',\n",
       " 'sklearn_7_2_toy_datasets\\n7.2. Toy datasets\\ndatasets/index.html#toy-datasets\\n 7.1. General dataset API  There are three main kinds of dataset interfaces that can be used to get datasets depending on the desired type of dataset. The dataset loaders. They can be used to load small standard datasets, described in the Toy datasets section. The dataset fetchers. They can be used to download and load larger datasets, described in the Real world datasets section. Both loaders and fetchers functions return a sklearn.utils.Bunch object holding at least two items: an array of shape * with key (except for 20newsgroups) and a numpy array of length , containing the target values, with key . The Bunch object is a dictionary that exposes its keys are attributes. For more information about Bunch object, see sklearn.utils.Bunch : It’s also possible for almost all of these function to constrain the output to be a tuple containing only the data and the target, by setting the parameter to . The datasets also contain a full description in their attribute and some contain and . See the dataset descriptions below for details. The dataset generation functions. They can be used to generate controlled synthetic datasets, described in the Generated datasets section. These functions return a tuple consisting of a * numpy array and an array of length containing the targets . In addition, there are also miscellaneous tools to load datasets of other formats or from other locations, described in the Loading other datasets section. ',\n",
       " 'sklearn_7_2_toy_datasets\\n7.2. Toy datasets\\ndatasets/index.html#toy-datasets\\n 7.2. Toy datasets  scikit-learn comes with a few small standard datasets that do not require to download any file from some external website. They can be loaded using the following functions: load_boston (*[,\\xa0return_X_y]) Load and return the boston house-prices dataset (regression). load_iris (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the iris dataset (classification). load_diabetes (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the diabetes dataset (regression). load_digits (*[,\\xa0n_class,\\xa0return_X_y,\\xa0as_frame]) Load and return the digits dataset (classification). load_linnerud (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the physical excercise linnerud dataset. load_wine (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the wine dataset (classification). load_breast_cancer (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the breast cancer wisconsin dataset (classification). These datasets are useful to quickly illustrate the behavior of the various algorithms implemented in scikit-learn. They are however often too small to be representative of real world machine learning tasks. 7.2.1. Boston house prices dataset  Data Set Characteristics: Number of Instances 506 Number of Attributes 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target. Attribute Information (in order) CRIM per capita crime rate by town ZN proportion of residential land zoned for lots over 25,000 sq.ft. INDUS proportion of non-retail business acres per town CHAS Charles River dummy variable ( 1 if tract bounds river; 0 otherwise) NOX nitric oxides concentration (parts per 10 million) RM average number of rooms per dwelling AGE proportion of owner-occupied units built prior to 1940 DIS weighted distances to five Boston employment centres RAD index of accessibility to radial highways TAX full-value property-tax rate per $10,000 PTRATIO pupil-teacher ratio by town B 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town LSTAT % lower status of the population MEDV Median value of owner-occupied homes in $1000’s Missing Attribute Values None Creator Harrison, D. and Rubinfeld, D.L. This is a copy of UCI ML housing dataset. https://archive.ics.uci.edu/ml/machine-learning-databases/housing/ This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University. The Boston house-price data of Harrison, D. and Rubinfeld, D.L. ‘Hedonic prices and the demand for clean air’, J. Environ. Economics & Management, vol.5, 81-102, 1978. Used in Belsley, Kuh & Welsch, ‘Regression diagnostics …’, Wiley, 1980. N.B. Various transformations are used in the table on pages 244-261 of the latter. The Boston house-price data has been used in many machine learning papers that address regression problems. References Belsley, Kuh & Welsch, ‘Regression diagnostics: Identifying Influential Data and Sources of Collinearity’, Wiley, 1980. 244-261. Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning. In Proceedings on the Tenth International Conference of Machine Learning, 236-243, University of Massachusetts, Amherst. Morgan Kaufmann. 7.2.2. Iris plants dataset  Data Set Characteristics: Number of Instances 150 (50 in each of three classes) Number of Attributes 4 numeric, predictive attributes and the class Attribute Information sepal length in cm sepal width in cm petal length in cm petal width in cm class: Iris-Setosa Iris-Versicolour Iris-Virginica Summary Statistics sepal length: 4.3 7.9 5.84 0.83 0.7826 sepal width: 2.0 4.4 3.05 0.43 -0.4194 petal length: 1.0 6.9 3.76 1.76 0.9490 (high!) petal width: 0.1 2.5 1.20 0.76 0.9565 (high!) Missing Attribute Values None Class Distribution 33.3% for each of 3 classes. Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 The famous Iris database, first used by Sir R.A. Fisher. The dataset is taken from Fisher’s paper. Note that it’s the same as in R, but not as in the UCI Machine Learning Repository, which has two wrong data points. This is perhaps the best known database to be found in the pattern recognition literature. Fisher’s paper is a classic in the field and is referenced frequently to this day. (See Duda & Hart, for example.) The data set contains 3 classes of 50 instances each, where each class refers to a type of iris plant. One class is linearly separable from the other 2; the latter are NOT linearly separable from each other. References Fisher, R.A. “The use of multiple measurements in taxonomic problems” Annual Eugenics, 7, Part II, 179-188 (1936); also in “Contributions to Mathematical Statistics” (John Wiley, NY, 1950). Duda, R.O., & Hart, P.E. (1973) Pattern Classification and Scene Analysis. (Q327.D83) John Wiley & Sons. ISBN 0-471-22361-1. See page 218. Dasarathy, B.V. (1980) “Nosing Around the Neighborhood: A New System Structure and Classification Rule for Recognition in Partially Exposed Environments”. IEEE Transactions on Pattern Analysis and Machine Intelligence, Vol. PAMI-2, No. 1, 67-71. Gates, G.W. (1972) “The Reduced Nearest Neighbor Rule”. IEEE Transactions on Information Theory, May 1972, 431-433. See also: 1988 MLC Proceedings, 54-64. Cheeseman et al”s AUTOCLASS II conceptual clustering system finds 3 classes in the data. Many, many more … 7.2.3. Diabetes dataset  Ten baseline variables, age, sex, body mass index, average blood pressure, and six blood serum measurements were obtained for each of n  442 diabetes patients, as well as the response of interest, a quantitative measure of disease progression one year after baseline. Data Set Characteristics: Number of Instances 442 Number of Attributes First 10 columns are numeric predictive values Target Column 11 is a quantitative measure of disease progression one year after baseline Attribute Information age age in years sex bmi body mass index bp average blood pressure s1 tc, T-Cells (a type of white blood cells) s2 ldl, low-density lipoproteins s3 hdl, high-density lipoproteins s4 tch, thyroid stimulating hormone s5 ltg, lamotrigine s6 glu, blood sugar level Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times (i.e. the sum of squares of each column totals 1). Source URL: https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html For more information see: Bradley Efron, Trevor Hastie, Iain Johnstone and Robert Tibshirani (2004) “Least Angle Regression,” Annals of Statistics (with discussion), 407-499. ( https://web.stanford.edu/~hastie/Papers/LARS/LeastAngle_2002.pdf ) 7.2.4. Optical recognition of handwritten digits dataset  Data Set Characteristics: Number of Instances 5620 Number of Attributes 64 Attribute Information 8x8 image of integer pixels in the range 0..16. Missing Attribute Values None Creator Alpaydin (alpaydin ‘@’ boun.edu.tr) Date July; 1998 This is a copy of the test set of the UCI ML hand-written digits datasets https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits The data set contains images of hand-written digits: 10 classes where each class refers to a digit. Preprocessing programs made available by NIST were used to extract normalized bitmaps of handwritten digits from a preprinted form. From a total of 43 people, 30 contributed to the training set and different 13 to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of 4x4 and the number of on pixels are counted in each block. This generates an input matrix of 8x8 where each element is an integer in the range 0..16. This reduces dimensionality and gives invariance to small distortions. For info on NIST preprocessing routines, see M. D. Garris, J. L. Blue, G. T. Candela, D. L. Dimmick, J. Geist, P. J. Grother, S. A. Janet, and C. L. Wilson, NIST Form-Based Handprint Recognition System, NISTIR 5469, 1994. References C. Kaynak (1995) Methods of Combining Multiple Classifiers and Their Applications to Handwritten Digit Recognition, MSc Thesis, Institute of Graduate Studies in Science and Engineering, Bogazici University. Alpaydin, C. Kaynak (1998) Cascading Classifiers, Kybernetika. Ken Tang and Ponnuthurai N. Suganthan and Xi Yao and A. Kai Qin. Linear dimensionalityreduction using relevance weighted LDA. School of Electrical and Electronic Engineering Nanyang Technological University. 2005. Claudio Gentile. A New Approximate Maximal Margin Classification Algorithm. NIPS. 2000. 7.2.5. Linnerrud dataset  Data Set Characteristics: Number of Instances 20 Number of Attributes 3 Missing Attribute Values None The Linnerud dataset is a multi-output regression dataset. It consists of three excercise (data) and three physiological (target) variables collected from twenty middle-aged men in a fitness club: physiological - CSV containing 20 observations on 3 physiological variables: Weight, Waist and Pulse. exercise - CSV containing 20 observations on 3 exercise variables: Chins, Situps and Jumps. References Tenenhaus, M. (1998). La regression PLS: theorie et pratique. Paris: Editions Technic. 7.2.6. Wine recognition dataset  Data Set Characteristics: Number of Instances 178 (50 in each of three classes) Number of Attributes 13 numeric, predictive attributes and the class Attribute Information Alcohol Malic acid Ash Alcalinity of ash Magnesium Total phenols Flavanoids Nonflavanoid phenols Proanthocyanins Color intensity Hue OD280/OD315 of diluted wines Proline class: class_0 class_1 class_2 Summary Statistics Alcohol: 11.0 14.8 13.0 0.8 Malic Acid: 0.74 5.80 2.34 1.12 Ash: 1.36 3.23 2.36 0.27 Alcalinity of Ash: 10.6 30.0 19.5 3.3 Magnesium: 70.0 162.0 99.7 14.3 Total Phenols: 0.98 3.88 2.29 0.63 Flavanoids: 0.34 5.08 2.03 1.00 Nonflavanoid Phenols: 0.13 0.66 0.36 0.12 Proanthocyanins: 0.41 3.58 1.59 0.57 Colour Intensity: 1.3 13.0 5.1 2.3 Hue: 0.48 1.71 0.96 0.23 OD280/OD315 of diluted wines: 1.27 4.00 2.61 0.71 Proline: 278 1680 746 315 Missing Attribute Values None Class Distribution class_0 (59), class_1 (71), class_2 (48) Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 This is a copy of UCI ML Wine recognition datasets. https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data The data is the results of a chemical analysis of wines grown in the same region in Italy by three different cultivators. There are thirteen different measurements taken for different constituents found in the three types of wine. Original Owners: Forina, M. et al, PARVUS - An Extendible Package for Data Exploration, Classification and Correlation. Institute of Pharmaceutical and Food Analysis and Technologies, Via Brigata Salerno, 16147 Genoa, Italy. Citation: Lichman, M. (2013). UCI Machine Learning Repository [ https://archive.ics.uci.edu/ml ]. Irvine, CA: University of California, School of Information and Computer Science. References (1) S. Aeberhard, D. Coomans and O. de Vel, Comparison of Classifiers in High Dimensional Settings, Tech. Rep. no. 92-02, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Technometrics). The data was used with many others for comparing various classifiers. The classes are separable, though only RDA has achieved 100% correct classification. (RDA : 100%, QDA 99.4%, LDA 98.9%, 1NN 96.1% (z-transformed data)) (All results using the leave-one-out technique) (2) S. Aeberhard, D. Coomans and O. de Vel, “THE CLASSIFICATION PERFORMANCE OF RDA” Tech. Rep. no. 92-01, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Journal of Chemometrics). 7.2.7. Breast cancer wisconsin (diagnostic) dataset  Data Set Characteristics: Number of Instances 569 Number of Attributes 30 numeric, predictive attributes and the class Attribute Information radius (mean of distances from center to points on the perimeter) texture (standard deviation of gray-scale values) perimeter area smoothness (local variation in radius lengths) compactness (perimeter^2 / area - 1.0) concavity (severity of concave portions of the contour) concave points (number of concave portions of the contour) symmetry fractal dimension (“coastline approximation” - 1) The mean, standard error, and “worst” or largest (mean of the three worst/largest values) of these features were computed for each image, resulting in 30 features. For instance, field 0 is Mean Radius, field 10 is Radius SE, field 20 is Worst Radius. class: WDBC-Malignant WDBC-Benign Summary Statistics radius (mean): 6.981 28.11 texture (mean): 9.71 39.28 perimeter (mean): 43.79 188.5 area (mean): 143.5 2501.0 smoothness (mean): 0.053 0.163 compactness (mean): 0.019 0.345 concavity (mean): 0.0 0.427 concave points (mean): 0.0 0.201 symmetry (mean): 0.106 0.304 fractal dimension (mean): 0.05 0.097 radius (standard error): 0.112 2.873 texture (standard error): 0.36 4.885 perimeter (standard error): 0.757 21.98 area (standard error): 6.802 542.2 smoothness (standard error): 0.002 0.031 compactness (standard error): 0.002 0.135 concavity (standard error): 0.0 0.396 concave points (standard error): 0.0 0.053 symmetry (standard error): 0.008 0.079 fractal dimension (standard error): 0.001 0.03 radius (worst): 7.93 36.04 texture (worst): 12.02 49.54 perimeter (worst): 50.41 251.2 area (worst): 185.2 4254.0 smoothness (worst): 0.071 0.223 compactness (worst): 0.027 1.058 concavity (worst): 0.0 1.252 concave points (worst): 0.0 0.291 symmetry (worst): 0.156 0.664 fractal dimension (worst): 0.055 0.208 Missing Attribute Values None Class Distribution 212 - Malignant, 357 - Benign Creator Dr. William H. Wolberg, W. Nick Street, Olvi L. Mangasarian Donor Nick Street Date November, 1995 This is a copy of UCI ML Breast Cancer Wisconsin (Diagnostic) datasets. https://goo.gl/U2Uwz2 Features are computed from a digitized image of a fine needle aspirate (FNA) of a breast mass. They describe characteristics of the cell nuclei present in the image. Separating plane described above was obtained using Multisurface Method-Tree (MSM-T) [K. P. Bennett, “Decision Tree Construction Via Linear Programming.” Proceedings of the 4th Midwest Artificial Intelligence and Cognitive Science Society, pp. 97-101, 1992], a classification method which uses linear programming to construct a decision tree. Relevant features were selected using an exhaustive search in the space of 1-4 features and 1-3 separating planes. The actual linear program used to obtain the separating plane in the 3-dimensional space is that described in: [K. P. Bennett and O. L. Mangasarian: “Robust Linear Programming Discrimination of Two Linearly Inseparable Sets”, Optimization Methods and Software 1, 1992, 23-34]. This database is also available through the UW CS ftp server: ftp ftp.cs.wisc.edu cd math-prog/cpo-dataset/machine-learn/WDBC/ References W.N. Street, W.H. Wolberg and O.L. Mangasarian. Nuclear feature extraction for breast tumor diagnosis. IS&T/SPIE 1993 International Symposium on Electronic Imaging: Science and Technology, volume 1905, pages 861-870, San Jose, CA, 1993. O.L. Mangasarian, W.N. Street and W.H. Wolberg. Breast cancer diagnosis and prognosis via linear programming. Operations Research, 43(4), pages 570-577, July-August 1995. W.H. Wolberg, W.N. Street, and O.L. Mangasarian. Machine learning techniques to diagnose breast cancer from fine-needle aspirates. Cancer Letters 77 (1994) 163-171. ',\n",
       " 'sklearn_7_2_toy_datasets\\n7.2. Toy datasets\\ndatasets/index.html#toy-datasets\\n 7.3. Real world datasets  scikit-learn provides tools to load larger datasets, downloading them if necessary. They can be loaded using the following functions: fetch_olivetti_faces (*[,\\xa0data_home,\\xa0…]) Load the Olivetti faces data-set from AT&T (classification). fetch_20newsgroups (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the filenames and data from the 20 newsgroups dataset (classification). fetch_20newsgroups_vectorized (*[,\\xa0subset,\\xa0…]) Load the 20 newsgroups dataset and vectorize it into token counts (classification). fetch_lfw_people (*[,\\xa0data_home,\\xa0funneled,\\xa0…]) Load the Labeled Faces in the Wild (LFW) people dataset (classification). fetch_lfw_pairs (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the Labeled Faces in the Wild (LFW) pairs dataset (classification). fetch_covtype (*[,\\xa0data_home,\\xa0…]) Load the covertype dataset (classification). fetch_rcv1 (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the RCV1 multilabel dataset (classification). fetch_kddcup99 (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the kddcup99 dataset (classification). fetch_california_housing (*[,\\xa0data_home,\\xa0…]) Load the California housing dataset (regression). 7.3.1. The Olivetti faces dataset  This dataset contains a set of face images taken between April 1992 and April 1994 at AT&T Laboratories Cambridge. The sklearn.datasets.fetch_olivetti_faces function is the data fetching / caching function that downloads the data archive from AT&T. As described on the original website: There are ten different images of each of 40 distinct subjects. For some subjects, the images were taken at different times, varying the lighting, facial expressions (open / closed eyes, smiling / not smiling) and facial details (glasses / no glasses). All the images were taken against a dark homogeneous background with the subjects in an upright, frontal position (with tolerance for some side movement). Data Set Characteristics: Classes 40 Samples total 400 Dimensionality 4096 Features real, between 0 and 1 The image is quantized to 256 grey levels and stored as unsigned 8-bit integers; the loader will convert these to floating point values on the interval [0, 1], which are easier to work with for many algorithms. The “target” for this database is an integer from 0 to 39 indicating the identity of the person pictured; however, with only 10 examples per class, this relatively small dataset is more interesting from an unsupervised or semi-supervised perspective. The original dataset consisted of 92 x 112, while the version available here consists of 64x64 images. When using these images, please give credit to AT&T Laboratories Cambridge. 7.3.2. The 20 newsgroups text dataset  The 20 newsgroups dataset comprises around 18000 newsgroups posts on 20 topics split in two subsets: one for training (or development) and the other one for testing (or for performance evaluation). The split between the train and test set is based upon a messages posted before and after a specific date. This module contains two loaders. The first one, sklearn.datasets.fetch_20newsgroups , returns a list of the raw texts that can be fed to text feature extractors such as sklearn.feature_extraction.text.CountVectorizer with custom parameters so as to extract feature vectors. The second one, sklearn.datasets.fetch_20newsgroups_vectorized , returns ready-to-use features, i.e., it is not necessary to use a feature extractor. Data Set Characteristics: Classes 20 Samples total 18846 Dimensionality 1 Features text 7.3.2.1. Usage  The sklearn.datasets.fetch_20newsgroups function is a data fetching / caching functions that downloads the data archive from the original 20 newsgroups website , extracts the archive contents in the folder and calls the sklearn.datasets.load_files on either the training or testing set folder, or both of them: The real data lies in the and attributes. The target attribute is the integer index of the category: It is possible to load only a sub-selection of the categories by passing the list of the categories to load to the sklearn.datasets.fetch_20newsgroups function: 7.3.2.2. Converting text to vectors  In order to feed predictive or clustering models with the text data, one first need to turn the text into vectors of numerical values suitable for statistical analysis. This can be achieved with the utilities of the as demonstrated in the following example that extract TF-IDF vectors of unigram tokens from a subset of 20news: The extracted TF-IDF vectors are very sparse, with an average of 159 non-zero components by sample in a more than 30000-dimensional space (less than .5% non-zero features): sklearn.datasets.fetch_20newsgroups_vectorized is a function which returns ready-to-use token counts features instead of file names. 7.3.2.3. Filtering text for more realistic training  It is easy for a classifier to overfit on particular things that appear in the 20 Newsgroups data, such as newsgroup headers. Many classifiers achieve very high F-scores, but their results would not generalize to other documents that aren’t from this window of time. For example, let’s look at the results of a multinomial Naive Bayes classifier, which is fast to train and achieves a decent F-score: (The example Classification of text documents using sparse features shuffles the training and test data, instead of segmenting by time, and in that case multinomial Naive Bayes gets a much higher F-score of 0.88. Are you suspicious yet of what’s going on inside this classifier?) Let’s take a look at what the most informative features are: You can now see many things that these features have overfit to: Almost every group is distinguished by whether headers such as and appear more or less often. Another significant feature involves whether the sender is affiliated with a university, as indicated either by their headers or their signature. The word “article” is a significant feature, based on how often people quote previous posts like this: “In article [article ID], [name] <[e-mail address]> wrote:” Other features match the names and e-mail addresses of particular people who were posting at the time. With such an abundance of clues that distinguish newsgroups, the classifiers barely have to identify topics from text at all, and they all perform at the same high level. For this reason, the functions that load 20 Newsgroups data provide a parameter called remove , telling it what kinds of information to strip out of each file. remove should be a tuple containing any subset of , telling it to remove headers, signature blocks, and quotation blocks respectively. This classifier lost over a lot of its F-score, just because we removed metadata that has little to do with topic classification. It loses even more if we also strip this metadata from the training data: Some other classifiers cope better with this harder version of the task. Try running Sample pipeline for text feature extraction and evaluation with and without the option to compare the results. Recommendation When evaluating text classifiers on the 20 Newsgroups data, you should strip newsgroup-related metadata. In scikit-learn, you can do this by setting . The F-score will be lower because it is more realistic. Examples Sample pipeline for text feature extraction and evaluation Classification of text documents using sparse features 7.3.3. The Labeled Faces in the Wild face recognition dataset  This dataset is a collection of JPEG pictures of famous people collected over the internet, all details are available on the official website: http://vis-www.cs.umass.edu/lfw/ Each picture is centered on a single face. The typical task is called Face Verification: given a pair of two pictures, a binary classifier must predict whether the two images are from the same person. An alternative task, Face Recognition or Face Identification is: given the picture of the face of an unknown person, identify the name of the person by referring to a gallery of previously seen pictures of identified persons. Both Face Verification and Face Recognition are tasks that are typically performed on the output of a model trained to perform Face Detection. The most popular model for Face Detection is called Viola-Jones and is implemented in the OpenCV library. The LFW faces were extracted by this face detector from various online websites. Data Set Characteristics: Classes 5749 Samples total 13233 Dimensionality 5828 Features real, between 0 and 255 7.3.3.1. Usage  provides two loaders that will automatically download, cache, parse the metadata files, decode the jpeg and convert the interesting slices into memmapped numpy arrays. This dataset size is more than 200 MB. The first load typically takes more than a couple of minutes to fully decode the relevant part of the JPEG files into numpy arrays. If the dataset has been loaded once, the following times the loading times less than 200ms by using a memmapped version memoized on the disk in the folder using . The first loader is used for the Face Identification task: a multi-class classification task (hence supervised learning): The default slice is a rectangular shape around the face, removing most of the background: Each of the faces is assigned to a single person id in the array: The second loader is typically used for the face verification task: each sample is a pair of two picture belonging or not to the same person: Both for the sklearn.datasets.fetch_lfw_people and sklearn.datasets.fetch_lfw_pairs function it is possible to get an additional dimension with the RGB color channels by passing , in that case the shape will be . The sklearn.datasets.fetch_lfw_pairs datasets is subdivided into 3 subsets: the development set, the development set and an evaluation set meant to compute performance metrics using a 10-folds cross validation scheme. References: Labeled Faces in the Wild: A Database for Studying Face Recognition in Unconstrained Environments. Gary B. Huang, Manu Ramesh, Tamara Berg, and Erik Learned-Miller. University of Massachusetts, Amherst, Technical Report 07-49, October, 2007. 7.3.3.2. Examples  Faces recognition example using eigenfaces and SVMs 7.3.4. Forest covertypes  The samples in this dataset correspond to 30×30m patches of forest in the US, collected for the task of predicting each patch’s cover type, i.e. the dominant species of tree. There are seven covertypes, making this a multiclass classification problem. Each sample has 54 features, described on the dataset’s homepage . Some of the features are boolean indicators, while others are discrete or continuous measurements. Data Set Characteristics: Classes 7 Samples total 581012 Dimensionality 54 Features int sklearn.datasets.fetch_covtype will load the covertype dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.5. RCV1 dataset  Reuters Corpus Volume I (RCV1) is an archive of over 800,000 manually categorized newswire stories made available by Reuters, Ltd. for research purposes. The dataset is extensively described in 1 . Data Set Characteristics: Classes 103 Samples total 804414 Dimensionality 47236 Features real, between 0 and 1 sklearn.datasets.fetch_rcv1 will load the following version: RCV1-v2, vectors, full sets, topics multilabels: It returns a dictionary-like object, with the following attributes: : The feature matrix is a scipy CSR sparse matrix, with 804414 samples and 47236 features. Non-zero values contains cosine-normalized, log TF-IDF vectors. A nearly chronological split is proposed in 1 : The first 23149 samples are the training set. The last 781265 samples are the testing set. This follows the official LYRL2004 chronological split. The array has 0.16% of non zero values: : The target values are stored in a scipy CSR sparse matrix, with 804414 samples and 103 categories. Each sample has a value of 1 in its categories, and 0 in others. The array has 3.15% of non zero values: : Each sample can be identified by its ID, ranging (with gaps) from 2286 to 810596: : The target values are the topics of each sample. Each sample belongs to at least one topic, and to up to 17 topics. There are 103 topics, each represented by a string. Their corpus frequencies span five orders of magnitude, from 5 occurrences for ‘GMIL’, to 381327 for ‘CCAT’: The dataset will be downloaded from the rcv1 homepage if necessary. The compressed size is about 656 MB. References 1 ( 1 , 2 ) Lewis, D. D., Yang, Y., Rose, T. G., & Li, F. (2004). RCV1: A new benchmark collection for text categorization research. The Journal of Machine Learning Research, 5, 361-397. 7.3.6. Kddcup 99 dataset  The KDD Cup ‘99 dataset was created by processing the tcpdump portions of the 1998 DARPA Intrusion Detection System (IDS) Evaluation dataset, created by MIT Lincoln Lab [1]. The artificial data (described on the dataset’s homepage ) was generated using a closed network and hand-injected attacks to produce a large number of different types of attack with normal activity in the background. As the initial goal was to produce a large training set for supervised learning algorithms, there is a large proportion (80.1%) of abnormal data which is unrealistic in real world, and inappropriate for unsupervised anomaly detection which aims at detecting ‘abnormal’ data, ie qualitatively different from normal data in large minority among the observations. We thus transform the KDD Data set into two different data sets: SA and SF. -SA is obtained by simply selecting all the normal data, and a small proportion of abnormal data to gives an anomaly proportion of 1%. -SF is obtained as in [2] by simply picking up the data whose attribute logged_in is positive, thus focusing on the intrusion attack, which gives a proportion of 0.3% of attack. -http and smtp are two subsets of SF corresponding with third feature equal to ‘http’ (resp. to ‘smtp’) General KDD structure : Samples total 4898431 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SA structure : Samples total 976158 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SF structure : Samples total 699691 Dimensionality 4 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type http structure : Samples total 619052 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type smtp structure : Samples total 95373 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type sklearn.datasets.fetch_kddcup99 will load the kddcup99 dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.7. California Housing dataset  Data Set Characteristics: Number of Instances 20640 Number of Attributes 8 numeric, predictive attributes and the target Attribute Information MedInc median income in block HouseAge median house age in block AveRooms average number of rooms AveBedrms average number of bedrooms Population block population AveOccup average house occupancy Latitude house block latitude Longitude house block longitude Missing Attribute Values None This dataset was obtained from the StatLib repository. http://lib.stat.cmu.edu/datasets/ The target variable is the median house value for California districts. This dataset was derived from the 1990 U.S. census, using one row per census block group. A block group is the smallest geographical unit for which the U.S. Census Bureau publishes sample data (a block group typically has a population of 600 to 3,000 people). It can be downloaded/loaded using the sklearn.datasets.fetch_california_housing function. References Pace, R. Kelley and Ronald Barry, Sparse Spatial Autoregressions, Statistics and Probability Letters, 33 (1997) 291-297 ',\n",
       " 'sklearn_7_2_toy_datasets\\n7.2. Toy datasets\\ndatasets/index.html#toy-datasets\\n 7.4. Generated datasets  In addition, scikit-learn includes various random sample generators that can be used to build artificial datasets of controlled size and complexity. 7.4.1. Generators for classification and clustering  These generators produce a matrix of features and corresponding discrete targets. 7.4.1.1. Single label  Both make_blobs and make_classification create multiclass datasets by allocating each class one or more normally-distributed clusters of points. make_blobs provides greater control regarding the centers and standard deviations of each cluster, and is used to demonstrate clustering. make_classification specialises in introducing noise by way of: correlated, redundant and uninformative features; multiple Gaussian clusters per class; and linear transformations of the feature space. make_gaussian_quantiles divides a single Gaussian cluster into near-equal-size classes separated by concentric hyperspheres. make_hastie_10_2 generates a similar binary, 10-dimensional problem. make_circles and make_moons generate 2d binary classification datasets that are challenging to certain algorithms (e.g. centroid-based clustering or linear classification), including optional Gaussian noise. They are useful for visualisation. make_circles produces Gaussian data with a spherical decision boundary for binary classification, while make_moons produces two interleaving half circles. 7.4.1.2. Multilabel  make_multilabel_classification generates random samples with multiple labels, reflecting a bag of words drawn from a mixture of topics. The number of topics for each document is drawn from a Poisson distribution, and the topics themselves are drawn from a fixed random distribution. Similarly, the number of words is drawn from Poisson, with words drawn from a multinomial, where each topic defines a probability distribution over words. Simplifications with respect to true bag-of-words mixtures include: Per-topic word distributions are independently drawn, where in reality all would be affected by a sparse base distribution, and would be correlated. For a document generated from multiple topics, all topics are weighted equally in generating its bag of words. Documents without labels words at random, rather than from a base distribution. 7.4.1.3. Biclustering  make_biclusters (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with constant block diagonal structure for biclustering. make_checkerboard (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with block checkerboard structure for biclustering. 7.4.2. Generators for regression  make_regression produces regression targets as an optionally-sparse random linear combination of random features, with noise. Its informative features may be uncorrelated, or low rank (few features account for most of the variance). Other regression generators generate functions deterministically from randomized features. make_sparse_uncorrelated produces a target as a linear combination of four features with fixed coefficients. Others encode explicitly non-linear relations: make_friedman1 is related by polynomial and sine transforms; make_friedman2 includes feature multiplication and reciprocation; and make_friedman3 is similar with an arctan transformation on the target. 7.4.3. Generators for manifold learning  make_s_curve ([n_samples,\\xa0noise,\\xa0random_state]) Generate an S curve dataset. make_swiss_roll ([n_samples,\\xa0noise,\\xa0random_state]) Generate a swiss roll dataset. 7.4.4. Generators for decomposition  make_low_rank_matrix ([n_samples,\\xa0…]) Generate a mostly low rank matrix with bell-shaped singular values make_sparse_coded_signal (n_samples,\\xa0*,\\xa0…) Generate a signal as a sparse combination of dictionary elements. make_spd_matrix (n_dim,\\xa0*[,\\xa0random_state]) Generate a random symmetric, positive-definite matrix. make_sparse_spd_matrix ([dim,\\xa0alpha,\\xa0…]) Generate a sparse symmetric definite positive matrix. ',\n",
       " 'sklearn_7_2_toy_datasets\\n7.2. Toy datasets\\ndatasets/index.html#toy-datasets\\n 7.5. Loading other datasets  7.5.1. Sample images  Scikit-learn also embed a couple of sample JPEG images published under Creative Commons license by their authors. Those images can be useful to test algorithms and pipeline on 2D data. load_sample_images () Load sample images for image manipulation. load_sample_image (image_name) Load the numpy array of a single sample image Warning The default coding of images is based on the dtype to spare memory. Often machine learning algorithms work best if the input is converted to a floating point representation first. Also, if you plan to use don’t forget to scale to the range 0 - 1 as done in the following example. Examples: Color Quantization using K-Means 7.5.2. Datasets in svmlight / libsvm format  scikit-learn includes utility functions for loading datasets in the svmlight / libsvm format. In this format, each line takes the form . This format is especially suitable for sparse datasets. In this module, scipy sparse CSR matrices are used for and numpy arrays are used for . You may load a dataset like as follows: You may also load two (or more) datasets at once: In this case, and are guaranteed to have the same number of features. Another way to achieve the same result is to fix the number of features: Related links: Public datasets in svmlight / libsvm format : https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets Faster API-compatible implementation : https://github.com/mblondel/svmlight-loader 7.5.3. Downloading datasets from the openml.org repository  openml.org is a public repository for machine learning data and experiments, that allows everybody to upload open datasets. The package is able to download datasets from the repository using the function sklearn.datasets.fetch_openml . For example, to download a dataset of gene expressions in mice brains: To fully specify a dataset, you need to provide a name and a version, though the version is optional, see Dataset Versions below. The dataset contains a total of 1080 examples belonging to 8 different classes: You can get more information on the dataset by looking at the and attributes: The contains a free-text description of the data, while contains a dictionary of meta-data stored by openml, like the dataset id. For more details, see the OpenML documentation The of the mice protein dataset is 40966, and you can use this (or the name) to get more information on the dataset on the openml website: The also uniquely identifies a dataset from OpenML: 7.5.3.1. Dataset Versions  A dataset is uniquely specified by its , but not necessarily by its name. Several different “versions” of a dataset with the same name can exist which can contain entirely different datasets. If a particular version of a dataset has been found to contain significant issues, it might be deactivated. Using a name to specify a dataset will yield the earliest version of a dataset that is still active. That means that can yield different results at different times if earlier versions become inactive. You can see that the dataset with 40966 that we fetched above is the version 1 of the “miceprotein” dataset: In fact, this dataset only has one version. The iris dataset on the other hand has multiple versions: Specifying the dataset by the name “iris” yields the lowest version, version 1, with the 61. To make sure you always get this exact dataset, it is safest to specify it by the dataset . The other dataset, with 969, is version 3 (version 2 has become inactive), and contains a binarized version of the data: You can also specify both the name and the version, which also uniquely identifies the dataset: References: Vanschoren, van Rijn, Bischl and Torgo “OpenML: networked science in machine learning” , ACM SIGKDD Explorations Newsletter, 15(2), 49-60, 2014. 7.5.4. Loading from external datasets  scikit-learn works on any numeric data stored as numpy arrays or scipy sparse matrices. Other types that are convertible to numeric arrays such as pandas DataFrame are also acceptable. Here are some recommended ways to load standard columnar data into a format usable by scikit-learn: pandas.io provides tools to read data from common formats including CSV, Excel, JSON and SQL. DataFrames may also be constructed from lists of tuples or dicts. Pandas handles heterogeneous data smoothly and provides tools for manipulation and conversion into a numeric array suitable for scikit-learn. scipy.io specializes in binary formats often used in scientific computing context such as .mat and .arff numpy/routines.io for standard loading of columnar data into numpy arrays scikit-learn’s datasets.load_svmlight_file for the svmlight or libSVM sparse format scikit-learn’s datasets.load_files for directories of text files where the name of each directory is the name of each category and each file inside of each directory corresponds to one sample from that category For some miscellaneous data such as images, videos, and audio, you may wish to refer to: skimage.io or Imageio for loading images and videos into numpy arrays scipy.io.wavfile.read for reading WAV files into a numpy array Categorical (or nominal) features stored as strings (common in pandas DataFrames) will need converting to numerical features using sklearn.preprocessing.OneHotEncoder or sklearn.preprocessing.OrdinalEncoder or similar. See Preprocessing data . Note: if you manage your own numerical data it is recommended to use an optimized file format such as HDF5 to reduce data load times. Various libraries such as H5Py, PyTables and pandas provides a Python interface for reading and writing data in that format. ',\n",
       " 'sklearn_7_3_real_world_datasets\\n7.3. Real world datasets\\ndatasets/index.html#real-world-datasets\\n 7.1. General dataset API  There are three main kinds of dataset interfaces that can be used to get datasets depending on the desired type of dataset. The dataset loaders. They can be used to load small standard datasets, described in the Toy datasets section. The dataset fetchers. They can be used to download and load larger datasets, described in the Real world datasets section. Both loaders and fetchers functions return a sklearn.utils.Bunch object holding at least two items: an array of shape * with key (except for 20newsgroups) and a numpy array of length , containing the target values, with key . The Bunch object is a dictionary that exposes its keys are attributes. For more information about Bunch object, see sklearn.utils.Bunch : It’s also possible for almost all of these function to constrain the output to be a tuple containing only the data and the target, by setting the parameter to . The datasets also contain a full description in their attribute and some contain and . See the dataset descriptions below for details. The dataset generation functions. They can be used to generate controlled synthetic datasets, described in the Generated datasets section. These functions return a tuple consisting of a * numpy array and an array of length containing the targets . In addition, there are also miscellaneous tools to load datasets of other formats or from other locations, described in the Loading other datasets section. ',\n",
       " 'sklearn_7_3_real_world_datasets\\n7.3. Real world datasets\\ndatasets/index.html#real-world-datasets\\n 7.2. Toy datasets  scikit-learn comes with a few small standard datasets that do not require to download any file from some external website. They can be loaded using the following functions: load_boston (*[,\\xa0return_X_y]) Load and return the boston house-prices dataset (regression). load_iris (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the iris dataset (classification). load_diabetes (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the diabetes dataset (regression). load_digits (*[,\\xa0n_class,\\xa0return_X_y,\\xa0as_frame]) Load and return the digits dataset (classification). load_linnerud (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the physical excercise linnerud dataset. load_wine (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the wine dataset (classification). load_breast_cancer (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the breast cancer wisconsin dataset (classification). These datasets are useful to quickly illustrate the behavior of the various algorithms implemented in scikit-learn. They are however often too small to be representative of real world machine learning tasks. 7.2.1. Boston house prices dataset  Data Set Characteristics: Number of Instances 506 Number of Attributes 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target. Attribute Information (in order) CRIM per capita crime rate by town ZN proportion of residential land zoned for lots over 25,000 sq.ft. INDUS proportion of non-retail business acres per town CHAS Charles River dummy variable ( 1 if tract bounds river; 0 otherwise) NOX nitric oxides concentration (parts per 10 million) RM average number of rooms per dwelling AGE proportion of owner-occupied units built prior to 1940 DIS weighted distances to five Boston employment centres RAD index of accessibility to radial highways TAX full-value property-tax rate per $10,000 PTRATIO pupil-teacher ratio by town B 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town LSTAT % lower status of the population MEDV Median value of owner-occupied homes in $1000’s Missing Attribute Values None Creator Harrison, D. and Rubinfeld, D.L. This is a copy of UCI ML housing dataset. https://archive.ics.uci.edu/ml/machine-learning-databases/housing/ This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University. The Boston house-price data of Harrison, D. and Rubinfeld, D.L. ‘Hedonic prices and the demand for clean air’, J. Environ. Economics & Management, vol.5, 81-102, 1978. Used in Belsley, Kuh & Welsch, ‘Regression diagnostics …’, Wiley, 1980. N.B. Various transformations are used in the table on pages 244-261 of the latter. The Boston house-price data has been used in many machine learning papers that address regression problems. References Belsley, Kuh & Welsch, ‘Regression diagnostics: Identifying Influential Data and Sources of Collinearity’, Wiley, 1980. 244-261. Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning. In Proceedings on the Tenth International Conference of Machine Learning, 236-243, University of Massachusetts, Amherst. Morgan Kaufmann. 7.2.2. Iris plants dataset  Data Set Characteristics: Number of Instances 150 (50 in each of three classes) Number of Attributes 4 numeric, predictive attributes and the class Attribute Information sepal length in cm sepal width in cm petal length in cm petal width in cm class: Iris-Setosa Iris-Versicolour Iris-Virginica Summary Statistics sepal length: 4.3 7.9 5.84 0.83 0.7826 sepal width: 2.0 4.4 3.05 0.43 -0.4194 petal length: 1.0 6.9 3.76 1.76 0.9490 (high!) petal width: 0.1 2.5 1.20 0.76 0.9565 (high!) Missing Attribute Values None Class Distribution 33.3% for each of 3 classes. Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 The famous Iris database, first used by Sir R.A. Fisher. The dataset is taken from Fisher’s paper. Note that it’s the same as in R, but not as in the UCI Machine Learning Repository, which has two wrong data points. This is perhaps the best known database to be found in the pattern recognition literature. Fisher’s paper is a classic in the field and is referenced frequently to this day. (See Duda & Hart, for example.) The data set contains 3 classes of 50 instances each, where each class refers to a type of iris plant. One class is linearly separable from the other 2; the latter are NOT linearly separable from each other. References Fisher, R.A. “The use of multiple measurements in taxonomic problems” Annual Eugenics, 7, Part II, 179-188 (1936); also in “Contributions to Mathematical Statistics” (John Wiley, NY, 1950). Duda, R.O., & Hart, P.E. (1973) Pattern Classification and Scene Analysis. (Q327.D83) John Wiley & Sons. ISBN 0-471-22361-1. See page 218. Dasarathy, B.V. (1980) “Nosing Around the Neighborhood: A New System Structure and Classification Rule for Recognition in Partially Exposed Environments”. IEEE Transactions on Pattern Analysis and Machine Intelligence, Vol. PAMI-2, No. 1, 67-71. Gates, G.W. (1972) “The Reduced Nearest Neighbor Rule”. IEEE Transactions on Information Theory, May 1972, 431-433. See also: 1988 MLC Proceedings, 54-64. Cheeseman et al”s AUTOCLASS II conceptual clustering system finds 3 classes in the data. Many, many more … 7.2.3. Diabetes dataset  Ten baseline variables, age, sex, body mass index, average blood pressure, and six blood serum measurements were obtained for each of n  442 diabetes patients, as well as the response of interest, a quantitative measure of disease progression one year after baseline. Data Set Characteristics: Number of Instances 442 Number of Attributes First 10 columns are numeric predictive values Target Column 11 is a quantitative measure of disease progression one year after baseline Attribute Information age age in years sex bmi body mass index bp average blood pressure s1 tc, T-Cells (a type of white blood cells) s2 ldl, low-density lipoproteins s3 hdl, high-density lipoproteins s4 tch, thyroid stimulating hormone s5 ltg, lamotrigine s6 glu, blood sugar level Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times (i.e. the sum of squares of each column totals 1). Source URL: https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html For more information see: Bradley Efron, Trevor Hastie, Iain Johnstone and Robert Tibshirani (2004) “Least Angle Regression,” Annals of Statistics (with discussion), 407-499. ( https://web.stanford.edu/~hastie/Papers/LARS/LeastAngle_2002.pdf ) 7.2.4. Optical recognition of handwritten digits dataset  Data Set Characteristics: Number of Instances 5620 Number of Attributes 64 Attribute Information 8x8 image of integer pixels in the range 0..16. Missing Attribute Values None Creator Alpaydin (alpaydin ‘@’ boun.edu.tr) Date July; 1998 This is a copy of the test set of the UCI ML hand-written digits datasets https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits The data set contains images of hand-written digits: 10 classes where each class refers to a digit. Preprocessing programs made available by NIST were used to extract normalized bitmaps of handwritten digits from a preprinted form. From a total of 43 people, 30 contributed to the training set and different 13 to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of 4x4 and the number of on pixels are counted in each block. This generates an input matrix of 8x8 where each element is an integer in the range 0..16. This reduces dimensionality and gives invariance to small distortions. For info on NIST preprocessing routines, see M. D. Garris, J. L. Blue, G. T. Candela, D. L. Dimmick, J. Geist, P. J. Grother, S. A. Janet, and C. L. Wilson, NIST Form-Based Handprint Recognition System, NISTIR 5469, 1994. References C. Kaynak (1995) Methods of Combining Multiple Classifiers and Their Applications to Handwritten Digit Recognition, MSc Thesis, Institute of Graduate Studies in Science and Engineering, Bogazici University. Alpaydin, C. Kaynak (1998) Cascading Classifiers, Kybernetika. Ken Tang and Ponnuthurai N. Suganthan and Xi Yao and A. Kai Qin. Linear dimensionalityreduction using relevance weighted LDA. School of Electrical and Electronic Engineering Nanyang Technological University. 2005. Claudio Gentile. A New Approximate Maximal Margin Classification Algorithm. NIPS. 2000. 7.2.5. Linnerrud dataset  Data Set Characteristics: Number of Instances 20 Number of Attributes 3 Missing Attribute Values None The Linnerud dataset is a multi-output regression dataset. It consists of three excercise (data) and three physiological (target) variables collected from twenty middle-aged men in a fitness club: physiological - CSV containing 20 observations on 3 physiological variables: Weight, Waist and Pulse. exercise - CSV containing 20 observations on 3 exercise variables: Chins, Situps and Jumps. References Tenenhaus, M. (1998). La regression PLS: theorie et pratique. Paris: Editions Technic. 7.2.6. Wine recognition dataset  Data Set Characteristics: Number of Instances 178 (50 in each of three classes) Number of Attributes 13 numeric, predictive attributes and the class Attribute Information Alcohol Malic acid Ash Alcalinity of ash Magnesium Total phenols Flavanoids Nonflavanoid phenols Proanthocyanins Color intensity Hue OD280/OD315 of diluted wines Proline class: class_0 class_1 class_2 Summary Statistics Alcohol: 11.0 14.8 13.0 0.8 Malic Acid: 0.74 5.80 2.34 1.12 Ash: 1.36 3.23 2.36 0.27 Alcalinity of Ash: 10.6 30.0 19.5 3.3 Magnesium: 70.0 162.0 99.7 14.3 Total Phenols: 0.98 3.88 2.29 0.63 Flavanoids: 0.34 5.08 2.03 1.00 Nonflavanoid Phenols: 0.13 0.66 0.36 0.12 Proanthocyanins: 0.41 3.58 1.59 0.57 Colour Intensity: 1.3 13.0 5.1 2.3 Hue: 0.48 1.71 0.96 0.23 OD280/OD315 of diluted wines: 1.27 4.00 2.61 0.71 Proline: 278 1680 746 315 Missing Attribute Values None Class Distribution class_0 (59), class_1 (71), class_2 (48) Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 This is a copy of UCI ML Wine recognition datasets. https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data The data is the results of a chemical analysis of wines grown in the same region in Italy by three different cultivators. There are thirteen different measurements taken for different constituents found in the three types of wine. Original Owners: Forina, M. et al, PARVUS - An Extendible Package for Data Exploration, Classification and Correlation. Institute of Pharmaceutical and Food Analysis and Technologies, Via Brigata Salerno, 16147 Genoa, Italy. Citation: Lichman, M. (2013). UCI Machine Learning Repository [ https://archive.ics.uci.edu/ml ]. Irvine, CA: University of California, School of Information and Computer Science. References (1) S. Aeberhard, D. Coomans and O. de Vel, Comparison of Classifiers in High Dimensional Settings, Tech. Rep. no. 92-02, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Technometrics). The data was used with many others for comparing various classifiers. The classes are separable, though only RDA has achieved 100% correct classification. (RDA : 100%, QDA 99.4%, LDA 98.9%, 1NN 96.1% (z-transformed data)) (All results using the leave-one-out technique) (2) S. Aeberhard, D. Coomans and O. de Vel, “THE CLASSIFICATION PERFORMANCE OF RDA” Tech. Rep. no. 92-01, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Journal of Chemometrics). 7.2.7. Breast cancer wisconsin (diagnostic) dataset  Data Set Characteristics: Number of Instances 569 Number of Attributes 30 numeric, predictive attributes and the class Attribute Information radius (mean of distances from center to points on the perimeter) texture (standard deviation of gray-scale values) perimeter area smoothness (local variation in radius lengths) compactness (perimeter^2 / area - 1.0) concavity (severity of concave portions of the contour) concave points (number of concave portions of the contour) symmetry fractal dimension (“coastline approximation” - 1) The mean, standard error, and “worst” or largest (mean of the three worst/largest values) of these features were computed for each image, resulting in 30 features. For instance, field 0 is Mean Radius, field 10 is Radius SE, field 20 is Worst Radius. class: WDBC-Malignant WDBC-Benign Summary Statistics radius (mean): 6.981 28.11 texture (mean): 9.71 39.28 perimeter (mean): 43.79 188.5 area (mean): 143.5 2501.0 smoothness (mean): 0.053 0.163 compactness (mean): 0.019 0.345 concavity (mean): 0.0 0.427 concave points (mean): 0.0 0.201 symmetry (mean): 0.106 0.304 fractal dimension (mean): 0.05 0.097 radius (standard error): 0.112 2.873 texture (standard error): 0.36 4.885 perimeter (standard error): 0.757 21.98 area (standard error): 6.802 542.2 smoothness (standard error): 0.002 0.031 compactness (standard error): 0.002 0.135 concavity (standard error): 0.0 0.396 concave points (standard error): 0.0 0.053 symmetry (standard error): 0.008 0.079 fractal dimension (standard error): 0.001 0.03 radius (worst): 7.93 36.04 texture (worst): 12.02 49.54 perimeter (worst): 50.41 251.2 area (worst): 185.2 4254.0 smoothness (worst): 0.071 0.223 compactness (worst): 0.027 1.058 concavity (worst): 0.0 1.252 concave points (worst): 0.0 0.291 symmetry (worst): 0.156 0.664 fractal dimension (worst): 0.055 0.208 Missing Attribute Values None Class Distribution 212 - Malignant, 357 - Benign Creator Dr. William H. Wolberg, W. Nick Street, Olvi L. Mangasarian Donor Nick Street Date November, 1995 This is a copy of UCI ML Breast Cancer Wisconsin (Diagnostic) datasets. https://goo.gl/U2Uwz2 Features are computed from a digitized image of a fine needle aspirate (FNA) of a breast mass. They describe characteristics of the cell nuclei present in the image. Separating plane described above was obtained using Multisurface Method-Tree (MSM-T) [K. P. Bennett, “Decision Tree Construction Via Linear Programming.” Proceedings of the 4th Midwest Artificial Intelligence and Cognitive Science Society, pp. 97-101, 1992], a classification method which uses linear programming to construct a decision tree. Relevant features were selected using an exhaustive search in the space of 1-4 features and 1-3 separating planes. The actual linear program used to obtain the separating plane in the 3-dimensional space is that described in: [K. P. Bennett and O. L. Mangasarian: “Robust Linear Programming Discrimination of Two Linearly Inseparable Sets”, Optimization Methods and Software 1, 1992, 23-34]. This database is also available through the UW CS ftp server: ftp ftp.cs.wisc.edu cd math-prog/cpo-dataset/machine-learn/WDBC/ References W.N. Street, W.H. Wolberg and O.L. Mangasarian. Nuclear feature extraction for breast tumor diagnosis. IS&T/SPIE 1993 International Symposium on Electronic Imaging: Science and Technology, volume 1905, pages 861-870, San Jose, CA, 1993. O.L. Mangasarian, W.N. Street and W.H. Wolberg. Breast cancer diagnosis and prognosis via linear programming. Operations Research, 43(4), pages 570-577, July-August 1995. W.H. Wolberg, W.N. Street, and O.L. Mangasarian. Machine learning techniques to diagnose breast cancer from fine-needle aspirates. Cancer Letters 77 (1994) 163-171. ',\n",
       " 'sklearn_7_3_real_world_datasets\\n7.3. Real world datasets\\ndatasets/index.html#real-world-datasets\\n 7.3. Real world datasets  scikit-learn provides tools to load larger datasets, downloading them if necessary. They can be loaded using the following functions: fetch_olivetti_faces (*[,\\xa0data_home,\\xa0…]) Load the Olivetti faces data-set from AT&T (classification). fetch_20newsgroups (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the filenames and data from the 20 newsgroups dataset (classification). fetch_20newsgroups_vectorized (*[,\\xa0subset,\\xa0…]) Load the 20 newsgroups dataset and vectorize it into token counts (classification). fetch_lfw_people (*[,\\xa0data_home,\\xa0funneled,\\xa0…]) Load the Labeled Faces in the Wild (LFW) people dataset (classification). fetch_lfw_pairs (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the Labeled Faces in the Wild (LFW) pairs dataset (classification). fetch_covtype (*[,\\xa0data_home,\\xa0…]) Load the covertype dataset (classification). fetch_rcv1 (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the RCV1 multilabel dataset (classification). fetch_kddcup99 (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the kddcup99 dataset (classification). fetch_california_housing (*[,\\xa0data_home,\\xa0…]) Load the California housing dataset (regression). 7.3.1. The Olivetti faces dataset  This dataset contains a set of face images taken between April 1992 and April 1994 at AT&T Laboratories Cambridge. The sklearn.datasets.fetch_olivetti_faces function is the data fetching / caching function that downloads the data archive from AT&T. As described on the original website: There are ten different images of each of 40 distinct subjects. For some subjects, the images were taken at different times, varying the lighting, facial expressions (open / closed eyes, smiling / not smiling) and facial details (glasses / no glasses). All the images were taken against a dark homogeneous background with the subjects in an upright, frontal position (with tolerance for some side movement). Data Set Characteristics: Classes 40 Samples total 400 Dimensionality 4096 Features real, between 0 and 1 The image is quantized to 256 grey levels and stored as unsigned 8-bit integers; the loader will convert these to floating point values on the interval [0, 1], which are easier to work with for many algorithms. The “target” for this database is an integer from 0 to 39 indicating the identity of the person pictured; however, with only 10 examples per class, this relatively small dataset is more interesting from an unsupervised or semi-supervised perspective. The original dataset consisted of 92 x 112, while the version available here consists of 64x64 images. When using these images, please give credit to AT&T Laboratories Cambridge. 7.3.2. The 20 newsgroups text dataset  The 20 newsgroups dataset comprises around 18000 newsgroups posts on 20 topics split in two subsets: one for training (or development) and the other one for testing (or for performance evaluation). The split between the train and test set is based upon a messages posted before and after a specific date. This module contains two loaders. The first one, sklearn.datasets.fetch_20newsgroups , returns a list of the raw texts that can be fed to text feature extractors such as sklearn.feature_extraction.text.CountVectorizer with custom parameters so as to extract feature vectors. The second one, sklearn.datasets.fetch_20newsgroups_vectorized , returns ready-to-use features, i.e., it is not necessary to use a feature extractor. Data Set Characteristics: Classes 20 Samples total 18846 Dimensionality 1 Features text 7.3.2.1. Usage  The sklearn.datasets.fetch_20newsgroups function is a data fetching / caching functions that downloads the data archive from the original 20 newsgroups website , extracts the archive contents in the folder and calls the sklearn.datasets.load_files on either the training or testing set folder, or both of them: The real data lies in the and attributes. The target attribute is the integer index of the category: It is possible to load only a sub-selection of the categories by passing the list of the categories to load to the sklearn.datasets.fetch_20newsgroups function: 7.3.2.2. Converting text to vectors  In order to feed predictive or clustering models with the text data, one first need to turn the text into vectors of numerical values suitable for statistical analysis. This can be achieved with the utilities of the as demonstrated in the following example that extract TF-IDF vectors of unigram tokens from a subset of 20news: The extracted TF-IDF vectors are very sparse, with an average of 159 non-zero components by sample in a more than 30000-dimensional space (less than .5% non-zero features): sklearn.datasets.fetch_20newsgroups_vectorized is a function which returns ready-to-use token counts features instead of file names. 7.3.2.3. Filtering text for more realistic training  It is easy for a classifier to overfit on particular things that appear in the 20 Newsgroups data, such as newsgroup headers. Many classifiers achieve very high F-scores, but their results would not generalize to other documents that aren’t from this window of time. For example, let’s look at the results of a multinomial Naive Bayes classifier, which is fast to train and achieves a decent F-score: (The example Classification of text documents using sparse features shuffles the training and test data, instead of segmenting by time, and in that case multinomial Naive Bayes gets a much higher F-score of 0.88. Are you suspicious yet of what’s going on inside this classifier?) Let’s take a look at what the most informative features are: You can now see many things that these features have overfit to: Almost every group is distinguished by whether headers such as and appear more or less often. Another significant feature involves whether the sender is affiliated with a university, as indicated either by their headers or their signature. The word “article” is a significant feature, based on how often people quote previous posts like this: “In article [article ID], [name] <[e-mail address]> wrote:” Other features match the names and e-mail addresses of particular people who were posting at the time. With such an abundance of clues that distinguish newsgroups, the classifiers barely have to identify topics from text at all, and they all perform at the same high level. For this reason, the functions that load 20 Newsgroups data provide a parameter called remove , telling it what kinds of information to strip out of each file. remove should be a tuple containing any subset of , telling it to remove headers, signature blocks, and quotation blocks respectively. This classifier lost over a lot of its F-score, just because we removed metadata that has little to do with topic classification. It loses even more if we also strip this metadata from the training data: Some other classifiers cope better with this harder version of the task. Try running Sample pipeline for text feature extraction and evaluation with and without the option to compare the results. Recommendation When evaluating text classifiers on the 20 Newsgroups data, you should strip newsgroup-related metadata. In scikit-learn, you can do this by setting . The F-score will be lower because it is more realistic. Examples Sample pipeline for text feature extraction and evaluation Classification of text documents using sparse features 7.3.3. The Labeled Faces in the Wild face recognition dataset  This dataset is a collection of JPEG pictures of famous people collected over the internet, all details are available on the official website: http://vis-www.cs.umass.edu/lfw/ Each picture is centered on a single face. The typical task is called Face Verification: given a pair of two pictures, a binary classifier must predict whether the two images are from the same person. An alternative task, Face Recognition or Face Identification is: given the picture of the face of an unknown person, identify the name of the person by referring to a gallery of previously seen pictures of identified persons. Both Face Verification and Face Recognition are tasks that are typically performed on the output of a model trained to perform Face Detection. The most popular model for Face Detection is called Viola-Jones and is implemented in the OpenCV library. The LFW faces were extracted by this face detector from various online websites. Data Set Characteristics: Classes 5749 Samples total 13233 Dimensionality 5828 Features real, between 0 and 255 7.3.3.1. Usage  provides two loaders that will automatically download, cache, parse the metadata files, decode the jpeg and convert the interesting slices into memmapped numpy arrays. This dataset size is more than 200 MB. The first load typically takes more than a couple of minutes to fully decode the relevant part of the JPEG files into numpy arrays. If the dataset has been loaded once, the following times the loading times less than 200ms by using a memmapped version memoized on the disk in the folder using . The first loader is used for the Face Identification task: a multi-class classification task (hence supervised learning): The default slice is a rectangular shape around the face, removing most of the background: Each of the faces is assigned to a single person id in the array: The second loader is typically used for the face verification task: each sample is a pair of two picture belonging or not to the same person: Both for the sklearn.datasets.fetch_lfw_people and sklearn.datasets.fetch_lfw_pairs function it is possible to get an additional dimension with the RGB color channels by passing , in that case the shape will be . The sklearn.datasets.fetch_lfw_pairs datasets is subdivided into 3 subsets: the development set, the development set and an evaluation set meant to compute performance metrics using a 10-folds cross validation scheme. References: Labeled Faces in the Wild: A Database for Studying Face Recognition in Unconstrained Environments. Gary B. Huang, Manu Ramesh, Tamara Berg, and Erik Learned-Miller. University of Massachusetts, Amherst, Technical Report 07-49, October, 2007. 7.3.3.2. Examples  Faces recognition example using eigenfaces and SVMs 7.3.4. Forest covertypes  The samples in this dataset correspond to 30×30m patches of forest in the US, collected for the task of predicting each patch’s cover type, i.e. the dominant species of tree. There are seven covertypes, making this a multiclass classification problem. Each sample has 54 features, described on the dataset’s homepage . Some of the features are boolean indicators, while others are discrete or continuous measurements. Data Set Characteristics: Classes 7 Samples total 581012 Dimensionality 54 Features int sklearn.datasets.fetch_covtype will load the covertype dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.5. RCV1 dataset  Reuters Corpus Volume I (RCV1) is an archive of over 800,000 manually categorized newswire stories made available by Reuters, Ltd. for research purposes. The dataset is extensively described in 1 . Data Set Characteristics: Classes 103 Samples total 804414 Dimensionality 47236 Features real, between 0 and 1 sklearn.datasets.fetch_rcv1 will load the following version: RCV1-v2, vectors, full sets, topics multilabels: It returns a dictionary-like object, with the following attributes: : The feature matrix is a scipy CSR sparse matrix, with 804414 samples and 47236 features. Non-zero values contains cosine-normalized, log TF-IDF vectors. A nearly chronological split is proposed in 1 : The first 23149 samples are the training set. The last 781265 samples are the testing set. This follows the official LYRL2004 chronological split. The array has 0.16% of non zero values: : The target values are stored in a scipy CSR sparse matrix, with 804414 samples and 103 categories. Each sample has a value of 1 in its categories, and 0 in others. The array has 3.15% of non zero values: : Each sample can be identified by its ID, ranging (with gaps) from 2286 to 810596: : The target values are the topics of each sample. Each sample belongs to at least one topic, and to up to 17 topics. There are 103 topics, each represented by a string. Their corpus frequencies span five orders of magnitude, from 5 occurrences for ‘GMIL’, to 381327 for ‘CCAT’: The dataset will be downloaded from the rcv1 homepage if necessary. The compressed size is about 656 MB. References 1 ( 1 , 2 ) Lewis, D. D., Yang, Y., Rose, T. G., & Li, F. (2004). RCV1: A new benchmark collection for text categorization research. The Journal of Machine Learning Research, 5, 361-397. 7.3.6. Kddcup 99 dataset  The KDD Cup ‘99 dataset was created by processing the tcpdump portions of the 1998 DARPA Intrusion Detection System (IDS) Evaluation dataset, created by MIT Lincoln Lab [1]. The artificial data (described on the dataset’s homepage ) was generated using a closed network and hand-injected attacks to produce a large number of different types of attack with normal activity in the background. As the initial goal was to produce a large training set for supervised learning algorithms, there is a large proportion (80.1%) of abnormal data which is unrealistic in real world, and inappropriate for unsupervised anomaly detection which aims at detecting ‘abnormal’ data, ie qualitatively different from normal data in large minority among the observations. We thus transform the KDD Data set into two different data sets: SA and SF. -SA is obtained by simply selecting all the normal data, and a small proportion of abnormal data to gives an anomaly proportion of 1%. -SF is obtained as in [2] by simply picking up the data whose attribute logged_in is positive, thus focusing on the intrusion attack, which gives a proportion of 0.3% of attack. -http and smtp are two subsets of SF corresponding with third feature equal to ‘http’ (resp. to ‘smtp’) General KDD structure : Samples total 4898431 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SA structure : Samples total 976158 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SF structure : Samples total 699691 Dimensionality 4 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type http structure : Samples total 619052 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type smtp structure : Samples total 95373 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type sklearn.datasets.fetch_kddcup99 will load the kddcup99 dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.7. California Housing dataset  Data Set Characteristics: Number of Instances 20640 Number of Attributes 8 numeric, predictive attributes and the target Attribute Information MedInc median income in block HouseAge median house age in block AveRooms average number of rooms AveBedrms average number of bedrooms Population block population AveOccup average house occupancy Latitude house block latitude Longitude house block longitude Missing Attribute Values None This dataset was obtained from the StatLib repository. http://lib.stat.cmu.edu/datasets/ The target variable is the median house value for California districts. This dataset was derived from the 1990 U.S. census, using one row per census block group. A block group is the smallest geographical unit for which the U.S. Census Bureau publishes sample data (a block group typically has a population of 600 to 3,000 people). It can be downloaded/loaded using the sklearn.datasets.fetch_california_housing function. References Pace, R. Kelley and Ronald Barry, Sparse Spatial Autoregressions, Statistics and Probability Letters, 33 (1997) 291-297 ',\n",
       " 'sklearn_7_3_real_world_datasets\\n7.3. Real world datasets\\ndatasets/index.html#real-world-datasets\\n 7.4. Generated datasets  In addition, scikit-learn includes various random sample generators that can be used to build artificial datasets of controlled size and complexity. 7.4.1. Generators for classification and clustering  These generators produce a matrix of features and corresponding discrete targets. 7.4.1.1. Single label  Both make_blobs and make_classification create multiclass datasets by allocating each class one or more normally-distributed clusters of points. make_blobs provides greater control regarding the centers and standard deviations of each cluster, and is used to demonstrate clustering. make_classification specialises in introducing noise by way of: correlated, redundant and uninformative features; multiple Gaussian clusters per class; and linear transformations of the feature space. make_gaussian_quantiles divides a single Gaussian cluster into near-equal-size classes separated by concentric hyperspheres. make_hastie_10_2 generates a similar binary, 10-dimensional problem. make_circles and make_moons generate 2d binary classification datasets that are challenging to certain algorithms (e.g. centroid-based clustering or linear classification), including optional Gaussian noise. They are useful for visualisation. make_circles produces Gaussian data with a spherical decision boundary for binary classification, while make_moons produces two interleaving half circles. 7.4.1.2. Multilabel  make_multilabel_classification generates random samples with multiple labels, reflecting a bag of words drawn from a mixture of topics. The number of topics for each document is drawn from a Poisson distribution, and the topics themselves are drawn from a fixed random distribution. Similarly, the number of words is drawn from Poisson, with words drawn from a multinomial, where each topic defines a probability distribution over words. Simplifications with respect to true bag-of-words mixtures include: Per-topic word distributions are independently drawn, where in reality all would be affected by a sparse base distribution, and would be correlated. For a document generated from multiple topics, all topics are weighted equally in generating its bag of words. Documents without labels words at random, rather than from a base distribution. 7.4.1.3. Biclustering  make_biclusters (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with constant block diagonal structure for biclustering. make_checkerboard (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with block checkerboard structure for biclustering. 7.4.2. Generators for regression  make_regression produces regression targets as an optionally-sparse random linear combination of random features, with noise. Its informative features may be uncorrelated, or low rank (few features account for most of the variance). Other regression generators generate functions deterministically from randomized features. make_sparse_uncorrelated produces a target as a linear combination of four features with fixed coefficients. Others encode explicitly non-linear relations: make_friedman1 is related by polynomial and sine transforms; make_friedman2 includes feature multiplication and reciprocation; and make_friedman3 is similar with an arctan transformation on the target. 7.4.3. Generators for manifold learning  make_s_curve ([n_samples,\\xa0noise,\\xa0random_state]) Generate an S curve dataset. make_swiss_roll ([n_samples,\\xa0noise,\\xa0random_state]) Generate a swiss roll dataset. 7.4.4. Generators for decomposition  make_low_rank_matrix ([n_samples,\\xa0…]) Generate a mostly low rank matrix with bell-shaped singular values make_sparse_coded_signal (n_samples,\\xa0*,\\xa0…) Generate a signal as a sparse combination of dictionary elements. make_spd_matrix (n_dim,\\xa0*[,\\xa0random_state]) Generate a random symmetric, positive-definite matrix. make_sparse_spd_matrix ([dim,\\xa0alpha,\\xa0…]) Generate a sparse symmetric definite positive matrix. ',\n",
       " 'sklearn_7_3_real_world_datasets\\n7.3. Real world datasets\\ndatasets/index.html#real-world-datasets\\n 7.5. Loading other datasets  7.5.1. Sample images  Scikit-learn also embed a couple of sample JPEG images published under Creative Commons license by their authors. Those images can be useful to test algorithms and pipeline on 2D data. load_sample_images () Load sample images for image manipulation. load_sample_image (image_name) Load the numpy array of a single sample image Warning The default coding of images is based on the dtype to spare memory. Often machine learning algorithms work best if the input is converted to a floating point representation first. Also, if you plan to use don’t forget to scale to the range 0 - 1 as done in the following example. Examples: Color Quantization using K-Means 7.5.2. Datasets in svmlight / libsvm format  scikit-learn includes utility functions for loading datasets in the svmlight / libsvm format. In this format, each line takes the form . This format is especially suitable for sparse datasets. In this module, scipy sparse CSR matrices are used for and numpy arrays are used for . You may load a dataset like as follows: You may also load two (or more) datasets at once: In this case, and are guaranteed to have the same number of features. Another way to achieve the same result is to fix the number of features: Related links: Public datasets in svmlight / libsvm format : https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets Faster API-compatible implementation : https://github.com/mblondel/svmlight-loader 7.5.3. Downloading datasets from the openml.org repository  openml.org is a public repository for machine learning data and experiments, that allows everybody to upload open datasets. The package is able to download datasets from the repository using the function sklearn.datasets.fetch_openml . For example, to download a dataset of gene expressions in mice brains: To fully specify a dataset, you need to provide a name and a version, though the version is optional, see Dataset Versions below. The dataset contains a total of 1080 examples belonging to 8 different classes: You can get more information on the dataset by looking at the and attributes: The contains a free-text description of the data, while contains a dictionary of meta-data stored by openml, like the dataset id. For more details, see the OpenML documentation The of the mice protein dataset is 40966, and you can use this (or the name) to get more information on the dataset on the openml website: The also uniquely identifies a dataset from OpenML: 7.5.3.1. Dataset Versions  A dataset is uniquely specified by its , but not necessarily by its name. Several different “versions” of a dataset with the same name can exist which can contain entirely different datasets. If a particular version of a dataset has been found to contain significant issues, it might be deactivated. Using a name to specify a dataset will yield the earliest version of a dataset that is still active. That means that can yield different results at different times if earlier versions become inactive. You can see that the dataset with 40966 that we fetched above is the version 1 of the “miceprotein” dataset: In fact, this dataset only has one version. The iris dataset on the other hand has multiple versions: Specifying the dataset by the name “iris” yields the lowest version, version 1, with the 61. To make sure you always get this exact dataset, it is safest to specify it by the dataset . The other dataset, with 969, is version 3 (version 2 has become inactive), and contains a binarized version of the data: You can also specify both the name and the version, which also uniquely identifies the dataset: References: Vanschoren, van Rijn, Bischl and Torgo “OpenML: networked science in machine learning” , ACM SIGKDD Explorations Newsletter, 15(2), 49-60, 2014. 7.5.4. Loading from external datasets  scikit-learn works on any numeric data stored as numpy arrays or scipy sparse matrices. Other types that are convertible to numeric arrays such as pandas DataFrame are also acceptable. Here are some recommended ways to load standard columnar data into a format usable by scikit-learn: pandas.io provides tools to read data from common formats including CSV, Excel, JSON and SQL. DataFrames may also be constructed from lists of tuples or dicts. Pandas handles heterogeneous data smoothly and provides tools for manipulation and conversion into a numeric array suitable for scikit-learn. scipy.io specializes in binary formats often used in scientific computing context such as .mat and .arff numpy/routines.io for standard loading of columnar data into numpy arrays scikit-learn’s datasets.load_svmlight_file for the svmlight or libSVM sparse format scikit-learn’s datasets.load_files for directories of text files where the name of each directory is the name of each category and each file inside of each directory corresponds to one sample from that category For some miscellaneous data such as images, videos, and audio, you may wish to refer to: skimage.io or Imageio for loading images and videos into numpy arrays scipy.io.wavfile.read for reading WAV files into a numpy array Categorical (or nominal) features stored as strings (common in pandas DataFrames) will need converting to numerical features using sklearn.preprocessing.OneHotEncoder or sklearn.preprocessing.OrdinalEncoder or similar. See Preprocessing data . Note: if you manage your own numerical data it is recommended to use an optimized file format such as HDF5 to reduce data load times. Various libraries such as H5Py, PyTables and pandas provides a Python interface for reading and writing data in that format. ',\n",
       " 'sklearn_7_4_generated_datasets\\n7.4. Generated datasets\\ndatasets/index.html#generated-datasets\\n 7.1. General dataset API  There are three main kinds of dataset interfaces that can be used to get datasets depending on the desired type of dataset. The dataset loaders. They can be used to load small standard datasets, described in the Toy datasets section. The dataset fetchers. They can be used to download and load larger datasets, described in the Real world datasets section. Both loaders and fetchers functions return a sklearn.utils.Bunch object holding at least two items: an array of shape * with key (except for 20newsgroups) and a numpy array of length , containing the target values, with key . The Bunch object is a dictionary that exposes its keys are attributes. For more information about Bunch object, see sklearn.utils.Bunch : It’s also possible for almost all of these function to constrain the output to be a tuple containing only the data and the target, by setting the parameter to . The datasets also contain a full description in their attribute and some contain and . See the dataset descriptions below for details. The dataset generation functions. They can be used to generate controlled synthetic datasets, described in the Generated datasets section. These functions return a tuple consisting of a * numpy array and an array of length containing the targets . In addition, there are also miscellaneous tools to load datasets of other formats or from other locations, described in the Loading other datasets section. ',\n",
       " 'sklearn_7_4_generated_datasets\\n7.4. Generated datasets\\ndatasets/index.html#generated-datasets\\n 7.2. Toy datasets  scikit-learn comes with a few small standard datasets that do not require to download any file from some external website. They can be loaded using the following functions: load_boston (*[,\\xa0return_X_y]) Load and return the boston house-prices dataset (regression). load_iris (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the iris dataset (classification). load_diabetes (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the diabetes dataset (regression). load_digits (*[,\\xa0n_class,\\xa0return_X_y,\\xa0as_frame]) Load and return the digits dataset (classification). load_linnerud (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the physical excercise linnerud dataset. load_wine (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the wine dataset (classification). load_breast_cancer (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the breast cancer wisconsin dataset (classification). These datasets are useful to quickly illustrate the behavior of the various algorithms implemented in scikit-learn. They are however often too small to be representative of real world machine learning tasks. 7.2.1. Boston house prices dataset  Data Set Characteristics: Number of Instances 506 Number of Attributes 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target. Attribute Information (in order) CRIM per capita crime rate by town ZN proportion of residential land zoned for lots over 25,000 sq.ft. INDUS proportion of non-retail business acres per town CHAS Charles River dummy variable ( 1 if tract bounds river; 0 otherwise) NOX nitric oxides concentration (parts per 10 million) RM average number of rooms per dwelling AGE proportion of owner-occupied units built prior to 1940 DIS weighted distances to five Boston employment centres RAD index of accessibility to radial highways TAX full-value property-tax rate per $10,000 PTRATIO pupil-teacher ratio by town B 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town LSTAT % lower status of the population MEDV Median value of owner-occupied homes in $1000’s Missing Attribute Values None Creator Harrison, D. and Rubinfeld, D.L. This is a copy of UCI ML housing dataset. https://archive.ics.uci.edu/ml/machine-learning-databases/housing/ This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University. The Boston house-price data of Harrison, D. and Rubinfeld, D.L. ‘Hedonic prices and the demand for clean air’, J. Environ. Economics & Management, vol.5, 81-102, 1978. Used in Belsley, Kuh & Welsch, ‘Regression diagnostics …’, Wiley, 1980. N.B. Various transformations are used in the table on pages 244-261 of the latter. The Boston house-price data has been used in many machine learning papers that address regression problems. References Belsley, Kuh & Welsch, ‘Regression diagnostics: Identifying Influential Data and Sources of Collinearity’, Wiley, 1980. 244-261. Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning. In Proceedings on the Tenth International Conference of Machine Learning, 236-243, University of Massachusetts, Amherst. Morgan Kaufmann. 7.2.2. Iris plants dataset  Data Set Characteristics: Number of Instances 150 (50 in each of three classes) Number of Attributes 4 numeric, predictive attributes and the class Attribute Information sepal length in cm sepal width in cm petal length in cm petal width in cm class: Iris-Setosa Iris-Versicolour Iris-Virginica Summary Statistics sepal length: 4.3 7.9 5.84 0.83 0.7826 sepal width: 2.0 4.4 3.05 0.43 -0.4194 petal length: 1.0 6.9 3.76 1.76 0.9490 (high!) petal width: 0.1 2.5 1.20 0.76 0.9565 (high!) Missing Attribute Values None Class Distribution 33.3% for each of 3 classes. Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 The famous Iris database, first used by Sir R.A. Fisher. The dataset is taken from Fisher’s paper. Note that it’s the same as in R, but not as in the UCI Machine Learning Repository, which has two wrong data points. This is perhaps the best known database to be found in the pattern recognition literature. Fisher’s paper is a classic in the field and is referenced frequently to this day. (See Duda & Hart, for example.) The data set contains 3 classes of 50 instances each, where each class refers to a type of iris plant. One class is linearly separable from the other 2; the latter are NOT linearly separable from each other. References Fisher, R.A. “The use of multiple measurements in taxonomic problems” Annual Eugenics, 7, Part II, 179-188 (1936); also in “Contributions to Mathematical Statistics” (John Wiley, NY, 1950). Duda, R.O., & Hart, P.E. (1973) Pattern Classification and Scene Analysis. (Q327.D83) John Wiley & Sons. ISBN 0-471-22361-1. See page 218. Dasarathy, B.V. (1980) “Nosing Around the Neighborhood: A New System Structure and Classification Rule for Recognition in Partially Exposed Environments”. IEEE Transactions on Pattern Analysis and Machine Intelligence, Vol. PAMI-2, No. 1, 67-71. Gates, G.W. (1972) “The Reduced Nearest Neighbor Rule”. IEEE Transactions on Information Theory, May 1972, 431-433. See also: 1988 MLC Proceedings, 54-64. Cheeseman et al”s AUTOCLASS II conceptual clustering system finds 3 classes in the data. Many, many more … 7.2.3. Diabetes dataset  Ten baseline variables, age, sex, body mass index, average blood pressure, and six blood serum measurements were obtained for each of n  442 diabetes patients, as well as the response of interest, a quantitative measure of disease progression one year after baseline. Data Set Characteristics: Number of Instances 442 Number of Attributes First 10 columns are numeric predictive values Target Column 11 is a quantitative measure of disease progression one year after baseline Attribute Information age age in years sex bmi body mass index bp average blood pressure s1 tc, T-Cells (a type of white blood cells) s2 ldl, low-density lipoproteins s3 hdl, high-density lipoproteins s4 tch, thyroid stimulating hormone s5 ltg, lamotrigine s6 glu, blood sugar level Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times (i.e. the sum of squares of each column totals 1). Source URL: https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html For more information see: Bradley Efron, Trevor Hastie, Iain Johnstone and Robert Tibshirani (2004) “Least Angle Regression,” Annals of Statistics (with discussion), 407-499. ( https://web.stanford.edu/~hastie/Papers/LARS/LeastAngle_2002.pdf ) 7.2.4. Optical recognition of handwritten digits dataset  Data Set Characteristics: Number of Instances 5620 Number of Attributes 64 Attribute Information 8x8 image of integer pixels in the range 0..16. Missing Attribute Values None Creator Alpaydin (alpaydin ‘@’ boun.edu.tr) Date July; 1998 This is a copy of the test set of the UCI ML hand-written digits datasets https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits The data set contains images of hand-written digits: 10 classes where each class refers to a digit. Preprocessing programs made available by NIST were used to extract normalized bitmaps of handwritten digits from a preprinted form. From a total of 43 people, 30 contributed to the training set and different 13 to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of 4x4 and the number of on pixels are counted in each block. This generates an input matrix of 8x8 where each element is an integer in the range 0..16. This reduces dimensionality and gives invariance to small distortions. For info on NIST preprocessing routines, see M. D. Garris, J. L. Blue, G. T. Candela, D. L. Dimmick, J. Geist, P. J. Grother, S. A. Janet, and C. L. Wilson, NIST Form-Based Handprint Recognition System, NISTIR 5469, 1994. References C. Kaynak (1995) Methods of Combining Multiple Classifiers and Their Applications to Handwritten Digit Recognition, MSc Thesis, Institute of Graduate Studies in Science and Engineering, Bogazici University. Alpaydin, C. Kaynak (1998) Cascading Classifiers, Kybernetika. Ken Tang and Ponnuthurai N. Suganthan and Xi Yao and A. Kai Qin. Linear dimensionalityreduction using relevance weighted LDA. School of Electrical and Electronic Engineering Nanyang Technological University. 2005. Claudio Gentile. A New Approximate Maximal Margin Classification Algorithm. NIPS. 2000. 7.2.5. Linnerrud dataset  Data Set Characteristics: Number of Instances 20 Number of Attributes 3 Missing Attribute Values None The Linnerud dataset is a multi-output regression dataset. It consists of three excercise (data) and three physiological (target) variables collected from twenty middle-aged men in a fitness club: physiological - CSV containing 20 observations on 3 physiological variables: Weight, Waist and Pulse. exercise - CSV containing 20 observations on 3 exercise variables: Chins, Situps and Jumps. References Tenenhaus, M. (1998). La regression PLS: theorie et pratique. Paris: Editions Technic. 7.2.6. Wine recognition dataset  Data Set Characteristics: Number of Instances 178 (50 in each of three classes) Number of Attributes 13 numeric, predictive attributes and the class Attribute Information Alcohol Malic acid Ash Alcalinity of ash Magnesium Total phenols Flavanoids Nonflavanoid phenols Proanthocyanins Color intensity Hue OD280/OD315 of diluted wines Proline class: class_0 class_1 class_2 Summary Statistics Alcohol: 11.0 14.8 13.0 0.8 Malic Acid: 0.74 5.80 2.34 1.12 Ash: 1.36 3.23 2.36 0.27 Alcalinity of Ash: 10.6 30.0 19.5 3.3 Magnesium: 70.0 162.0 99.7 14.3 Total Phenols: 0.98 3.88 2.29 0.63 Flavanoids: 0.34 5.08 2.03 1.00 Nonflavanoid Phenols: 0.13 0.66 0.36 0.12 Proanthocyanins: 0.41 3.58 1.59 0.57 Colour Intensity: 1.3 13.0 5.1 2.3 Hue: 0.48 1.71 0.96 0.23 OD280/OD315 of diluted wines: 1.27 4.00 2.61 0.71 Proline: 278 1680 746 315 Missing Attribute Values None Class Distribution class_0 (59), class_1 (71), class_2 (48) Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 This is a copy of UCI ML Wine recognition datasets. https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data The data is the results of a chemical analysis of wines grown in the same region in Italy by three different cultivators. There are thirteen different measurements taken for different constituents found in the three types of wine. Original Owners: Forina, M. et al, PARVUS - An Extendible Package for Data Exploration, Classification and Correlation. Institute of Pharmaceutical and Food Analysis and Technologies, Via Brigata Salerno, 16147 Genoa, Italy. Citation: Lichman, M. (2013). UCI Machine Learning Repository [ https://archive.ics.uci.edu/ml ]. Irvine, CA: University of California, School of Information and Computer Science. References (1) S. Aeberhard, D. Coomans and O. de Vel, Comparison of Classifiers in High Dimensional Settings, Tech. Rep. no. 92-02, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Technometrics). The data was used with many others for comparing various classifiers. The classes are separable, though only RDA has achieved 100% correct classification. (RDA : 100%, QDA 99.4%, LDA 98.9%, 1NN 96.1% (z-transformed data)) (All results using the leave-one-out technique) (2) S. Aeberhard, D. Coomans and O. de Vel, “THE CLASSIFICATION PERFORMANCE OF RDA” Tech. Rep. no. 92-01, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Journal of Chemometrics). 7.2.7. Breast cancer wisconsin (diagnostic) dataset  Data Set Characteristics: Number of Instances 569 Number of Attributes 30 numeric, predictive attributes and the class Attribute Information radius (mean of distances from center to points on the perimeter) texture (standard deviation of gray-scale values) perimeter area smoothness (local variation in radius lengths) compactness (perimeter^2 / area - 1.0) concavity (severity of concave portions of the contour) concave points (number of concave portions of the contour) symmetry fractal dimension (“coastline approximation” - 1) The mean, standard error, and “worst” or largest (mean of the three worst/largest values) of these features were computed for each image, resulting in 30 features. For instance, field 0 is Mean Radius, field 10 is Radius SE, field 20 is Worst Radius. class: WDBC-Malignant WDBC-Benign Summary Statistics radius (mean): 6.981 28.11 texture (mean): 9.71 39.28 perimeter (mean): 43.79 188.5 area (mean): 143.5 2501.0 smoothness (mean): 0.053 0.163 compactness (mean): 0.019 0.345 concavity (mean): 0.0 0.427 concave points (mean): 0.0 0.201 symmetry (mean): 0.106 0.304 fractal dimension (mean): 0.05 0.097 radius (standard error): 0.112 2.873 texture (standard error): 0.36 4.885 perimeter (standard error): 0.757 21.98 area (standard error): 6.802 542.2 smoothness (standard error): 0.002 0.031 compactness (standard error): 0.002 0.135 concavity (standard error): 0.0 0.396 concave points (standard error): 0.0 0.053 symmetry (standard error): 0.008 0.079 fractal dimension (standard error): 0.001 0.03 radius (worst): 7.93 36.04 texture (worst): 12.02 49.54 perimeter (worst): 50.41 251.2 area (worst): 185.2 4254.0 smoothness (worst): 0.071 0.223 compactness (worst): 0.027 1.058 concavity (worst): 0.0 1.252 concave points (worst): 0.0 0.291 symmetry (worst): 0.156 0.664 fractal dimension (worst): 0.055 0.208 Missing Attribute Values None Class Distribution 212 - Malignant, 357 - Benign Creator Dr. William H. Wolberg, W. Nick Street, Olvi L. Mangasarian Donor Nick Street Date November, 1995 This is a copy of UCI ML Breast Cancer Wisconsin (Diagnostic) datasets. https://goo.gl/U2Uwz2 Features are computed from a digitized image of a fine needle aspirate (FNA) of a breast mass. They describe characteristics of the cell nuclei present in the image. Separating plane described above was obtained using Multisurface Method-Tree (MSM-T) [K. P. Bennett, “Decision Tree Construction Via Linear Programming.” Proceedings of the 4th Midwest Artificial Intelligence and Cognitive Science Society, pp. 97-101, 1992], a classification method which uses linear programming to construct a decision tree. Relevant features were selected using an exhaustive search in the space of 1-4 features and 1-3 separating planes. The actual linear program used to obtain the separating plane in the 3-dimensional space is that described in: [K. P. Bennett and O. L. Mangasarian: “Robust Linear Programming Discrimination of Two Linearly Inseparable Sets”, Optimization Methods and Software 1, 1992, 23-34]. This database is also available through the UW CS ftp server: ftp ftp.cs.wisc.edu cd math-prog/cpo-dataset/machine-learn/WDBC/ References W.N. Street, W.H. Wolberg and O.L. Mangasarian. Nuclear feature extraction for breast tumor diagnosis. IS&T/SPIE 1993 International Symposium on Electronic Imaging: Science and Technology, volume 1905, pages 861-870, San Jose, CA, 1993. O.L. Mangasarian, W.N. Street and W.H. Wolberg. Breast cancer diagnosis and prognosis via linear programming. Operations Research, 43(4), pages 570-577, July-August 1995. W.H. Wolberg, W.N. Street, and O.L. Mangasarian. Machine learning techniques to diagnose breast cancer from fine-needle aspirates. Cancer Letters 77 (1994) 163-171. ',\n",
       " 'sklearn_7_4_generated_datasets\\n7.4. Generated datasets\\ndatasets/index.html#generated-datasets\\n 7.3. Real world datasets  scikit-learn provides tools to load larger datasets, downloading them if necessary. They can be loaded using the following functions: fetch_olivetti_faces (*[,\\xa0data_home,\\xa0…]) Load the Olivetti faces data-set from AT&T (classification). fetch_20newsgroups (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the filenames and data from the 20 newsgroups dataset (classification). fetch_20newsgroups_vectorized (*[,\\xa0subset,\\xa0…]) Load the 20 newsgroups dataset and vectorize it into token counts (classification). fetch_lfw_people (*[,\\xa0data_home,\\xa0funneled,\\xa0…]) Load the Labeled Faces in the Wild (LFW) people dataset (classification). fetch_lfw_pairs (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the Labeled Faces in the Wild (LFW) pairs dataset (classification). fetch_covtype (*[,\\xa0data_home,\\xa0…]) Load the covertype dataset (classification). fetch_rcv1 (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the RCV1 multilabel dataset (classification). fetch_kddcup99 (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the kddcup99 dataset (classification). fetch_california_housing (*[,\\xa0data_home,\\xa0…]) Load the California housing dataset (regression). 7.3.1. The Olivetti faces dataset  This dataset contains a set of face images taken between April 1992 and April 1994 at AT&T Laboratories Cambridge. The sklearn.datasets.fetch_olivetti_faces function is the data fetching / caching function that downloads the data archive from AT&T. As described on the original website: There are ten different images of each of 40 distinct subjects. For some subjects, the images were taken at different times, varying the lighting, facial expressions (open / closed eyes, smiling / not smiling) and facial details (glasses / no glasses). All the images were taken against a dark homogeneous background with the subjects in an upright, frontal position (with tolerance for some side movement). Data Set Characteristics: Classes 40 Samples total 400 Dimensionality 4096 Features real, between 0 and 1 The image is quantized to 256 grey levels and stored as unsigned 8-bit integers; the loader will convert these to floating point values on the interval [0, 1], which are easier to work with for many algorithms. The “target” for this database is an integer from 0 to 39 indicating the identity of the person pictured; however, with only 10 examples per class, this relatively small dataset is more interesting from an unsupervised or semi-supervised perspective. The original dataset consisted of 92 x 112, while the version available here consists of 64x64 images. When using these images, please give credit to AT&T Laboratories Cambridge. 7.3.2. The 20 newsgroups text dataset  The 20 newsgroups dataset comprises around 18000 newsgroups posts on 20 topics split in two subsets: one for training (or development) and the other one for testing (or for performance evaluation). The split between the train and test set is based upon a messages posted before and after a specific date. This module contains two loaders. The first one, sklearn.datasets.fetch_20newsgroups , returns a list of the raw texts that can be fed to text feature extractors such as sklearn.feature_extraction.text.CountVectorizer with custom parameters so as to extract feature vectors. The second one, sklearn.datasets.fetch_20newsgroups_vectorized , returns ready-to-use features, i.e., it is not necessary to use a feature extractor. Data Set Characteristics: Classes 20 Samples total 18846 Dimensionality 1 Features text 7.3.2.1. Usage  The sklearn.datasets.fetch_20newsgroups function is a data fetching / caching functions that downloads the data archive from the original 20 newsgroups website , extracts the archive contents in the folder and calls the sklearn.datasets.load_files on either the training or testing set folder, or both of them: The real data lies in the and attributes. The target attribute is the integer index of the category: It is possible to load only a sub-selection of the categories by passing the list of the categories to load to the sklearn.datasets.fetch_20newsgroups function: 7.3.2.2. Converting text to vectors  In order to feed predictive or clustering models with the text data, one first need to turn the text into vectors of numerical values suitable for statistical analysis. This can be achieved with the utilities of the as demonstrated in the following example that extract TF-IDF vectors of unigram tokens from a subset of 20news: The extracted TF-IDF vectors are very sparse, with an average of 159 non-zero components by sample in a more than 30000-dimensional space (less than .5% non-zero features): sklearn.datasets.fetch_20newsgroups_vectorized is a function which returns ready-to-use token counts features instead of file names. 7.3.2.3. Filtering text for more realistic training  It is easy for a classifier to overfit on particular things that appear in the 20 Newsgroups data, such as newsgroup headers. Many classifiers achieve very high F-scores, but their results would not generalize to other documents that aren’t from this window of time. For example, let’s look at the results of a multinomial Naive Bayes classifier, which is fast to train and achieves a decent F-score: (The example Classification of text documents using sparse features shuffles the training and test data, instead of segmenting by time, and in that case multinomial Naive Bayes gets a much higher F-score of 0.88. Are you suspicious yet of what’s going on inside this classifier?) Let’s take a look at what the most informative features are: You can now see many things that these features have overfit to: Almost every group is distinguished by whether headers such as and appear more or less often. Another significant feature involves whether the sender is affiliated with a university, as indicated either by their headers or their signature. The word “article” is a significant feature, based on how often people quote previous posts like this: “In article [article ID], [name] <[e-mail address]> wrote:” Other features match the names and e-mail addresses of particular people who were posting at the time. With such an abundance of clues that distinguish newsgroups, the classifiers barely have to identify topics from text at all, and they all perform at the same high level. For this reason, the functions that load 20 Newsgroups data provide a parameter called remove , telling it what kinds of information to strip out of each file. remove should be a tuple containing any subset of , telling it to remove headers, signature blocks, and quotation blocks respectively. This classifier lost over a lot of its F-score, just because we removed metadata that has little to do with topic classification. It loses even more if we also strip this metadata from the training data: Some other classifiers cope better with this harder version of the task. Try running Sample pipeline for text feature extraction and evaluation with and without the option to compare the results. Recommendation When evaluating text classifiers on the 20 Newsgroups data, you should strip newsgroup-related metadata. In scikit-learn, you can do this by setting . The F-score will be lower because it is more realistic. Examples Sample pipeline for text feature extraction and evaluation Classification of text documents using sparse features 7.3.3. The Labeled Faces in the Wild face recognition dataset  This dataset is a collection of JPEG pictures of famous people collected over the internet, all details are available on the official website: http://vis-www.cs.umass.edu/lfw/ Each picture is centered on a single face. The typical task is called Face Verification: given a pair of two pictures, a binary classifier must predict whether the two images are from the same person. An alternative task, Face Recognition or Face Identification is: given the picture of the face of an unknown person, identify the name of the person by referring to a gallery of previously seen pictures of identified persons. Both Face Verification and Face Recognition are tasks that are typically performed on the output of a model trained to perform Face Detection. The most popular model for Face Detection is called Viola-Jones and is implemented in the OpenCV library. The LFW faces were extracted by this face detector from various online websites. Data Set Characteristics: Classes 5749 Samples total 13233 Dimensionality 5828 Features real, between 0 and 255 7.3.3.1. Usage  provides two loaders that will automatically download, cache, parse the metadata files, decode the jpeg and convert the interesting slices into memmapped numpy arrays. This dataset size is more than 200 MB. The first load typically takes more than a couple of minutes to fully decode the relevant part of the JPEG files into numpy arrays. If the dataset has been loaded once, the following times the loading times less than 200ms by using a memmapped version memoized on the disk in the folder using . The first loader is used for the Face Identification task: a multi-class classification task (hence supervised learning): The default slice is a rectangular shape around the face, removing most of the background: Each of the faces is assigned to a single person id in the array: The second loader is typically used for the face verification task: each sample is a pair of two picture belonging or not to the same person: Both for the sklearn.datasets.fetch_lfw_people and sklearn.datasets.fetch_lfw_pairs function it is possible to get an additional dimension with the RGB color channels by passing , in that case the shape will be . The sklearn.datasets.fetch_lfw_pairs datasets is subdivided into 3 subsets: the development set, the development set and an evaluation set meant to compute performance metrics using a 10-folds cross validation scheme. References: Labeled Faces in the Wild: A Database for Studying Face Recognition in Unconstrained Environments. Gary B. Huang, Manu Ramesh, Tamara Berg, and Erik Learned-Miller. University of Massachusetts, Amherst, Technical Report 07-49, October, 2007. 7.3.3.2. Examples  Faces recognition example using eigenfaces and SVMs 7.3.4. Forest covertypes  The samples in this dataset correspond to 30×30m patches of forest in the US, collected for the task of predicting each patch’s cover type, i.e. the dominant species of tree. There are seven covertypes, making this a multiclass classification problem. Each sample has 54 features, described on the dataset’s homepage . Some of the features are boolean indicators, while others are discrete or continuous measurements. Data Set Characteristics: Classes 7 Samples total 581012 Dimensionality 54 Features int sklearn.datasets.fetch_covtype will load the covertype dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.5. RCV1 dataset  Reuters Corpus Volume I (RCV1) is an archive of over 800,000 manually categorized newswire stories made available by Reuters, Ltd. for research purposes. The dataset is extensively described in 1 . Data Set Characteristics: Classes 103 Samples total 804414 Dimensionality 47236 Features real, between 0 and 1 sklearn.datasets.fetch_rcv1 will load the following version: RCV1-v2, vectors, full sets, topics multilabels: It returns a dictionary-like object, with the following attributes: : The feature matrix is a scipy CSR sparse matrix, with 804414 samples and 47236 features. Non-zero values contains cosine-normalized, log TF-IDF vectors. A nearly chronological split is proposed in 1 : The first 23149 samples are the training set. The last 781265 samples are the testing set. This follows the official LYRL2004 chronological split. The array has 0.16% of non zero values: : The target values are stored in a scipy CSR sparse matrix, with 804414 samples and 103 categories. Each sample has a value of 1 in its categories, and 0 in others. The array has 3.15% of non zero values: : Each sample can be identified by its ID, ranging (with gaps) from 2286 to 810596: : The target values are the topics of each sample. Each sample belongs to at least one topic, and to up to 17 topics. There are 103 topics, each represented by a string. Their corpus frequencies span five orders of magnitude, from 5 occurrences for ‘GMIL’, to 381327 for ‘CCAT’: The dataset will be downloaded from the rcv1 homepage if necessary. The compressed size is about 656 MB. References 1 ( 1 , 2 ) Lewis, D. D., Yang, Y., Rose, T. G., & Li, F. (2004). RCV1: A new benchmark collection for text categorization research. The Journal of Machine Learning Research, 5, 361-397. 7.3.6. Kddcup 99 dataset  The KDD Cup ‘99 dataset was created by processing the tcpdump portions of the 1998 DARPA Intrusion Detection System (IDS) Evaluation dataset, created by MIT Lincoln Lab [1]. The artificial data (described on the dataset’s homepage ) was generated using a closed network and hand-injected attacks to produce a large number of different types of attack with normal activity in the background. As the initial goal was to produce a large training set for supervised learning algorithms, there is a large proportion (80.1%) of abnormal data which is unrealistic in real world, and inappropriate for unsupervised anomaly detection which aims at detecting ‘abnormal’ data, ie qualitatively different from normal data in large minority among the observations. We thus transform the KDD Data set into two different data sets: SA and SF. -SA is obtained by simply selecting all the normal data, and a small proportion of abnormal data to gives an anomaly proportion of 1%. -SF is obtained as in [2] by simply picking up the data whose attribute logged_in is positive, thus focusing on the intrusion attack, which gives a proportion of 0.3% of attack. -http and smtp are two subsets of SF corresponding with third feature equal to ‘http’ (resp. to ‘smtp’) General KDD structure : Samples total 4898431 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SA structure : Samples total 976158 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SF structure : Samples total 699691 Dimensionality 4 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type http structure : Samples total 619052 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type smtp structure : Samples total 95373 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type sklearn.datasets.fetch_kddcup99 will load the kddcup99 dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.7. California Housing dataset  Data Set Characteristics: Number of Instances 20640 Number of Attributes 8 numeric, predictive attributes and the target Attribute Information MedInc median income in block HouseAge median house age in block AveRooms average number of rooms AveBedrms average number of bedrooms Population block population AveOccup average house occupancy Latitude house block latitude Longitude house block longitude Missing Attribute Values None This dataset was obtained from the StatLib repository. http://lib.stat.cmu.edu/datasets/ The target variable is the median house value for California districts. This dataset was derived from the 1990 U.S. census, using one row per census block group. A block group is the smallest geographical unit for which the U.S. Census Bureau publishes sample data (a block group typically has a population of 600 to 3,000 people). It can be downloaded/loaded using the sklearn.datasets.fetch_california_housing function. References Pace, R. Kelley and Ronald Barry, Sparse Spatial Autoregressions, Statistics and Probability Letters, 33 (1997) 291-297 ',\n",
       " 'sklearn_7_4_generated_datasets\\n7.4. Generated datasets\\ndatasets/index.html#generated-datasets\\n 7.4. Generated datasets  In addition, scikit-learn includes various random sample generators that can be used to build artificial datasets of controlled size and complexity. 7.4.1. Generators for classification and clustering  These generators produce a matrix of features and corresponding discrete targets. 7.4.1.1. Single label  Both make_blobs and make_classification create multiclass datasets by allocating each class one or more normally-distributed clusters of points. make_blobs provides greater control regarding the centers and standard deviations of each cluster, and is used to demonstrate clustering. make_classification specialises in introducing noise by way of: correlated, redundant and uninformative features; multiple Gaussian clusters per class; and linear transformations of the feature space. make_gaussian_quantiles divides a single Gaussian cluster into near-equal-size classes separated by concentric hyperspheres. make_hastie_10_2 generates a similar binary, 10-dimensional problem. make_circles and make_moons generate 2d binary classification datasets that are challenging to certain algorithms (e.g. centroid-based clustering or linear classification), including optional Gaussian noise. They are useful for visualisation. make_circles produces Gaussian data with a spherical decision boundary for binary classification, while make_moons produces two interleaving half circles. 7.4.1.2. Multilabel  make_multilabel_classification generates random samples with multiple labels, reflecting a bag of words drawn from a mixture of topics. The number of topics for each document is drawn from a Poisson distribution, and the topics themselves are drawn from a fixed random distribution. Similarly, the number of words is drawn from Poisson, with words drawn from a multinomial, where each topic defines a probability distribution over words. Simplifications with respect to true bag-of-words mixtures include: Per-topic word distributions are independently drawn, where in reality all would be affected by a sparse base distribution, and would be correlated. For a document generated from multiple topics, all topics are weighted equally in generating its bag of words. Documents without labels words at random, rather than from a base distribution. 7.4.1.3. Biclustering  make_biclusters (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with constant block diagonal structure for biclustering. make_checkerboard (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with block checkerboard structure for biclustering. 7.4.2. Generators for regression  make_regression produces regression targets as an optionally-sparse random linear combination of random features, with noise. Its informative features may be uncorrelated, or low rank (few features account for most of the variance). Other regression generators generate functions deterministically from randomized features. make_sparse_uncorrelated produces a target as a linear combination of four features with fixed coefficients. Others encode explicitly non-linear relations: make_friedman1 is related by polynomial and sine transforms; make_friedman2 includes feature multiplication and reciprocation; and make_friedman3 is similar with an arctan transformation on the target. 7.4.3. Generators for manifold learning  make_s_curve ([n_samples,\\xa0noise,\\xa0random_state]) Generate an S curve dataset. make_swiss_roll ([n_samples,\\xa0noise,\\xa0random_state]) Generate a swiss roll dataset. 7.4.4. Generators for decomposition  make_low_rank_matrix ([n_samples,\\xa0…]) Generate a mostly low rank matrix with bell-shaped singular values make_sparse_coded_signal (n_samples,\\xa0*,\\xa0…) Generate a signal as a sparse combination of dictionary elements. make_spd_matrix (n_dim,\\xa0*[,\\xa0random_state]) Generate a random symmetric, positive-definite matrix. make_sparse_spd_matrix ([dim,\\xa0alpha,\\xa0…]) Generate a sparse symmetric definite positive matrix. ',\n",
       " 'sklearn_7_4_generated_datasets\\n7.4. Generated datasets\\ndatasets/index.html#generated-datasets\\n 7.5. Loading other datasets  7.5.1. Sample images  Scikit-learn also embed a couple of sample JPEG images published under Creative Commons license by their authors. Those images can be useful to test algorithms and pipeline on 2D data. load_sample_images () Load sample images for image manipulation. load_sample_image (image_name) Load the numpy array of a single sample image Warning The default coding of images is based on the dtype to spare memory. Often machine learning algorithms work best if the input is converted to a floating point representation first. Also, if you plan to use don’t forget to scale to the range 0 - 1 as done in the following example. Examples: Color Quantization using K-Means 7.5.2. Datasets in svmlight / libsvm format  scikit-learn includes utility functions for loading datasets in the svmlight / libsvm format. In this format, each line takes the form . This format is especially suitable for sparse datasets. In this module, scipy sparse CSR matrices are used for and numpy arrays are used for . You may load a dataset like as follows: You may also load two (or more) datasets at once: In this case, and are guaranteed to have the same number of features. Another way to achieve the same result is to fix the number of features: Related links: Public datasets in svmlight / libsvm format : https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets Faster API-compatible implementation : https://github.com/mblondel/svmlight-loader 7.5.3. Downloading datasets from the openml.org repository  openml.org is a public repository for machine learning data and experiments, that allows everybody to upload open datasets. The package is able to download datasets from the repository using the function sklearn.datasets.fetch_openml . For example, to download a dataset of gene expressions in mice brains: To fully specify a dataset, you need to provide a name and a version, though the version is optional, see Dataset Versions below. The dataset contains a total of 1080 examples belonging to 8 different classes: You can get more information on the dataset by looking at the and attributes: The contains a free-text description of the data, while contains a dictionary of meta-data stored by openml, like the dataset id. For more details, see the OpenML documentation The of the mice protein dataset is 40966, and you can use this (or the name) to get more information on the dataset on the openml website: The also uniquely identifies a dataset from OpenML: 7.5.3.1. Dataset Versions  A dataset is uniquely specified by its , but not necessarily by its name. Several different “versions” of a dataset with the same name can exist which can contain entirely different datasets. If a particular version of a dataset has been found to contain significant issues, it might be deactivated. Using a name to specify a dataset will yield the earliest version of a dataset that is still active. That means that can yield different results at different times if earlier versions become inactive. You can see that the dataset with 40966 that we fetched above is the version 1 of the “miceprotein” dataset: In fact, this dataset only has one version. The iris dataset on the other hand has multiple versions: Specifying the dataset by the name “iris” yields the lowest version, version 1, with the 61. To make sure you always get this exact dataset, it is safest to specify it by the dataset . The other dataset, with 969, is version 3 (version 2 has become inactive), and contains a binarized version of the data: You can also specify both the name and the version, which also uniquely identifies the dataset: References: Vanschoren, van Rijn, Bischl and Torgo “OpenML: networked science in machine learning” , ACM SIGKDD Explorations Newsletter, 15(2), 49-60, 2014. 7.5.4. Loading from external datasets  scikit-learn works on any numeric data stored as numpy arrays or scipy sparse matrices. Other types that are convertible to numeric arrays such as pandas DataFrame are also acceptable. Here are some recommended ways to load standard columnar data into a format usable by scikit-learn: pandas.io provides tools to read data from common formats including CSV, Excel, JSON and SQL. DataFrames may also be constructed from lists of tuples or dicts. Pandas handles heterogeneous data smoothly and provides tools for manipulation and conversion into a numeric array suitable for scikit-learn. scipy.io specializes in binary formats often used in scientific computing context such as .mat and .arff numpy/routines.io for standard loading of columnar data into numpy arrays scikit-learn’s datasets.load_svmlight_file for the svmlight or libSVM sparse format scikit-learn’s datasets.load_files for directories of text files where the name of each directory is the name of each category and each file inside of each directory corresponds to one sample from that category For some miscellaneous data such as images, videos, and audio, you may wish to refer to: skimage.io or Imageio for loading images and videos into numpy arrays scipy.io.wavfile.read for reading WAV files into a numpy array Categorical (or nominal) features stored as strings (common in pandas DataFrames) will need converting to numerical features using sklearn.preprocessing.OneHotEncoder or sklearn.preprocessing.OrdinalEncoder or similar. See Preprocessing data . Note: if you manage your own numerical data it is recommended to use an optimized file format such as HDF5 to reduce data load times. Various libraries such as H5Py, PyTables and pandas provides a Python interface for reading and writing data in that format. ',\n",
       " 'sklearn_7_5_loading_other_datasets\\n7.5. Loading other datasets\\ndatasets/index.html#loading-other-datasets\\n 7.1. General dataset API  There are three main kinds of dataset interfaces that can be used to get datasets depending on the desired type of dataset. The dataset loaders. They can be used to load small standard datasets, described in the Toy datasets section. The dataset fetchers. They can be used to download and load larger datasets, described in the Real world datasets section. Both loaders and fetchers functions return a sklearn.utils.Bunch object holding at least two items: an array of shape * with key (except for 20newsgroups) and a numpy array of length , containing the target values, with key . The Bunch object is a dictionary that exposes its keys are attributes. For more information about Bunch object, see sklearn.utils.Bunch : It’s also possible for almost all of these function to constrain the output to be a tuple containing only the data and the target, by setting the parameter to . The datasets also contain a full description in their attribute and some contain and . See the dataset descriptions below for details. The dataset generation functions. They can be used to generate controlled synthetic datasets, described in the Generated datasets section. These functions return a tuple consisting of a * numpy array and an array of length containing the targets . In addition, there are also miscellaneous tools to load datasets of other formats or from other locations, described in the Loading other datasets section. ',\n",
       " 'sklearn_7_5_loading_other_datasets\\n7.5. Loading other datasets\\ndatasets/index.html#loading-other-datasets\\n 7.2. Toy datasets  scikit-learn comes with a few small standard datasets that do not require to download any file from some external website. They can be loaded using the following functions: load_boston (*[,\\xa0return_X_y]) Load and return the boston house-prices dataset (regression). load_iris (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the iris dataset (classification). load_diabetes (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the diabetes dataset (regression). load_digits (*[,\\xa0n_class,\\xa0return_X_y,\\xa0as_frame]) Load and return the digits dataset (classification). load_linnerud (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the physical excercise linnerud dataset. load_wine (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the wine dataset (classification). load_breast_cancer (*[,\\xa0return_X_y,\\xa0as_frame]) Load and return the breast cancer wisconsin dataset (classification). These datasets are useful to quickly illustrate the behavior of the various algorithms implemented in scikit-learn. They are however often too small to be representative of real world machine learning tasks. 7.2.1. Boston house prices dataset  Data Set Characteristics: Number of Instances 506 Number of Attributes 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target. Attribute Information (in order) CRIM per capita crime rate by town ZN proportion of residential land zoned for lots over 25,000 sq.ft. INDUS proportion of non-retail business acres per town CHAS Charles River dummy variable ( 1 if tract bounds river; 0 otherwise) NOX nitric oxides concentration (parts per 10 million) RM average number of rooms per dwelling AGE proportion of owner-occupied units built prior to 1940 DIS weighted distances to five Boston employment centres RAD index of accessibility to radial highways TAX full-value property-tax rate per $10,000 PTRATIO pupil-teacher ratio by town B 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town LSTAT % lower status of the population MEDV Median value of owner-occupied homes in $1000’s Missing Attribute Values None Creator Harrison, D. and Rubinfeld, D.L. This is a copy of UCI ML housing dataset. https://archive.ics.uci.edu/ml/machine-learning-databases/housing/ This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University. The Boston house-price data of Harrison, D. and Rubinfeld, D.L. ‘Hedonic prices and the demand for clean air’, J. Environ. Economics & Management, vol.5, 81-102, 1978. Used in Belsley, Kuh & Welsch, ‘Regression diagnostics …’, Wiley, 1980. N.B. Various transformations are used in the table on pages 244-261 of the latter. The Boston house-price data has been used in many machine learning papers that address regression problems. References Belsley, Kuh & Welsch, ‘Regression diagnostics: Identifying Influential Data and Sources of Collinearity’, Wiley, 1980. 244-261. Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning. In Proceedings on the Tenth International Conference of Machine Learning, 236-243, University of Massachusetts, Amherst. Morgan Kaufmann. 7.2.2. Iris plants dataset  Data Set Characteristics: Number of Instances 150 (50 in each of three classes) Number of Attributes 4 numeric, predictive attributes and the class Attribute Information sepal length in cm sepal width in cm petal length in cm petal width in cm class: Iris-Setosa Iris-Versicolour Iris-Virginica Summary Statistics sepal length: 4.3 7.9 5.84 0.83 0.7826 sepal width: 2.0 4.4 3.05 0.43 -0.4194 petal length: 1.0 6.9 3.76 1.76 0.9490 (high!) petal width: 0.1 2.5 1.20 0.76 0.9565 (high!) Missing Attribute Values None Class Distribution 33.3% for each of 3 classes. Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 The famous Iris database, first used by Sir R.A. Fisher. The dataset is taken from Fisher’s paper. Note that it’s the same as in R, but not as in the UCI Machine Learning Repository, which has two wrong data points. This is perhaps the best known database to be found in the pattern recognition literature. Fisher’s paper is a classic in the field and is referenced frequently to this day. (See Duda & Hart, for example.) The data set contains 3 classes of 50 instances each, where each class refers to a type of iris plant. One class is linearly separable from the other 2; the latter are NOT linearly separable from each other. References Fisher, R.A. “The use of multiple measurements in taxonomic problems” Annual Eugenics, 7, Part II, 179-188 (1936); also in “Contributions to Mathematical Statistics” (John Wiley, NY, 1950). Duda, R.O., & Hart, P.E. (1973) Pattern Classification and Scene Analysis. (Q327.D83) John Wiley & Sons. ISBN 0-471-22361-1. See page 218. Dasarathy, B.V. (1980) “Nosing Around the Neighborhood: A New System Structure and Classification Rule for Recognition in Partially Exposed Environments”. IEEE Transactions on Pattern Analysis and Machine Intelligence, Vol. PAMI-2, No. 1, 67-71. Gates, G.W. (1972) “The Reduced Nearest Neighbor Rule”. IEEE Transactions on Information Theory, May 1972, 431-433. See also: 1988 MLC Proceedings, 54-64. Cheeseman et al”s AUTOCLASS II conceptual clustering system finds 3 classes in the data. Many, many more … 7.2.3. Diabetes dataset  Ten baseline variables, age, sex, body mass index, average blood pressure, and six blood serum measurements were obtained for each of n  442 diabetes patients, as well as the response of interest, a quantitative measure of disease progression one year after baseline. Data Set Characteristics: Number of Instances 442 Number of Attributes First 10 columns are numeric predictive values Target Column 11 is a quantitative measure of disease progression one year after baseline Attribute Information age age in years sex bmi body mass index bp average blood pressure s1 tc, T-Cells (a type of white blood cells) s2 ldl, low-density lipoproteins s3 hdl, high-density lipoproteins s4 tch, thyroid stimulating hormone s5 ltg, lamotrigine s6 glu, blood sugar level Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times (i.e. the sum of squares of each column totals 1). Source URL: https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html For more information see: Bradley Efron, Trevor Hastie, Iain Johnstone and Robert Tibshirani (2004) “Least Angle Regression,” Annals of Statistics (with discussion), 407-499. ( https://web.stanford.edu/~hastie/Papers/LARS/LeastAngle_2002.pdf ) 7.2.4. Optical recognition of handwritten digits dataset  Data Set Characteristics: Number of Instances 5620 Number of Attributes 64 Attribute Information 8x8 image of integer pixels in the range 0..16. Missing Attribute Values None Creator Alpaydin (alpaydin ‘@’ boun.edu.tr) Date July; 1998 This is a copy of the test set of the UCI ML hand-written digits datasets https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits The data set contains images of hand-written digits: 10 classes where each class refers to a digit. Preprocessing programs made available by NIST were used to extract normalized bitmaps of handwritten digits from a preprinted form. From a total of 43 people, 30 contributed to the training set and different 13 to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of 4x4 and the number of on pixels are counted in each block. This generates an input matrix of 8x8 where each element is an integer in the range 0..16. This reduces dimensionality and gives invariance to small distortions. For info on NIST preprocessing routines, see M. D. Garris, J. L. Blue, G. T. Candela, D. L. Dimmick, J. Geist, P. J. Grother, S. A. Janet, and C. L. Wilson, NIST Form-Based Handprint Recognition System, NISTIR 5469, 1994. References C. Kaynak (1995) Methods of Combining Multiple Classifiers and Their Applications to Handwritten Digit Recognition, MSc Thesis, Institute of Graduate Studies in Science and Engineering, Bogazici University. Alpaydin, C. Kaynak (1998) Cascading Classifiers, Kybernetika. Ken Tang and Ponnuthurai N. Suganthan and Xi Yao and A. Kai Qin. Linear dimensionalityreduction using relevance weighted LDA. School of Electrical and Electronic Engineering Nanyang Technological University. 2005. Claudio Gentile. A New Approximate Maximal Margin Classification Algorithm. NIPS. 2000. 7.2.5. Linnerrud dataset  Data Set Characteristics: Number of Instances 20 Number of Attributes 3 Missing Attribute Values None The Linnerud dataset is a multi-output regression dataset. It consists of three excercise (data) and three physiological (target) variables collected from twenty middle-aged men in a fitness club: physiological - CSV containing 20 observations on 3 physiological variables: Weight, Waist and Pulse. exercise - CSV containing 20 observations on 3 exercise variables: Chins, Situps and Jumps. References Tenenhaus, M. (1998). La regression PLS: theorie et pratique. Paris: Editions Technic. 7.2.6. Wine recognition dataset  Data Set Characteristics: Number of Instances 178 (50 in each of three classes) Number of Attributes 13 numeric, predictive attributes and the class Attribute Information Alcohol Malic acid Ash Alcalinity of ash Magnesium Total phenols Flavanoids Nonflavanoid phenols Proanthocyanins Color intensity Hue OD280/OD315 of diluted wines Proline class: class_0 class_1 class_2 Summary Statistics Alcohol: 11.0 14.8 13.0 0.8 Malic Acid: 0.74 5.80 2.34 1.12 Ash: 1.36 3.23 2.36 0.27 Alcalinity of Ash: 10.6 30.0 19.5 3.3 Magnesium: 70.0 162.0 99.7 14.3 Total Phenols: 0.98 3.88 2.29 0.63 Flavanoids: 0.34 5.08 2.03 1.00 Nonflavanoid Phenols: 0.13 0.66 0.36 0.12 Proanthocyanins: 0.41 3.58 1.59 0.57 Colour Intensity: 1.3 13.0 5.1 2.3 Hue: 0.48 1.71 0.96 0.23 OD280/OD315 of diluted wines: 1.27 4.00 2.61 0.71 Proline: 278 1680 746 315 Missing Attribute Values None Class Distribution class_0 (59), class_1 (71), class_2 (48) Creator R.A. Fisher Donor Michael Marshall ( MARSHALL%PLU @ io . arc . nasa . gov ) Date July, 1988 This is a copy of UCI ML Wine recognition datasets. https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data The data is the results of a chemical analysis of wines grown in the same region in Italy by three different cultivators. There are thirteen different measurements taken for different constituents found in the three types of wine. Original Owners: Forina, M. et al, PARVUS - An Extendible Package for Data Exploration, Classification and Correlation. Institute of Pharmaceutical and Food Analysis and Technologies, Via Brigata Salerno, 16147 Genoa, Italy. Citation: Lichman, M. (2013). UCI Machine Learning Repository [ https://archive.ics.uci.edu/ml ]. Irvine, CA: University of California, School of Information and Computer Science. References (1) S. Aeberhard, D. Coomans and O. de Vel, Comparison of Classifiers in High Dimensional Settings, Tech. Rep. no. 92-02, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Technometrics). The data was used with many others for comparing various classifiers. The classes are separable, though only RDA has achieved 100% correct classification. (RDA : 100%, QDA 99.4%, LDA 98.9%, 1NN 96.1% (z-transformed data)) (All results using the leave-one-out technique) (2) S. Aeberhard, D. Coomans and O. de Vel, “THE CLASSIFICATION PERFORMANCE OF RDA” Tech. Rep. no. 92-01, (1992), Dept. of Computer Science and Dept. of Mathematics and Statistics, James Cook University of North Queensland. (Also submitted to Journal of Chemometrics). 7.2.7. Breast cancer wisconsin (diagnostic) dataset  Data Set Characteristics: Number of Instances 569 Number of Attributes 30 numeric, predictive attributes and the class Attribute Information radius (mean of distances from center to points on the perimeter) texture (standard deviation of gray-scale values) perimeter area smoothness (local variation in radius lengths) compactness (perimeter^2 / area - 1.0) concavity (severity of concave portions of the contour) concave points (number of concave portions of the contour) symmetry fractal dimension (“coastline approximation” - 1) The mean, standard error, and “worst” or largest (mean of the three worst/largest values) of these features were computed for each image, resulting in 30 features. For instance, field 0 is Mean Radius, field 10 is Radius SE, field 20 is Worst Radius. class: WDBC-Malignant WDBC-Benign Summary Statistics radius (mean): 6.981 28.11 texture (mean): 9.71 39.28 perimeter (mean): 43.79 188.5 area (mean): 143.5 2501.0 smoothness (mean): 0.053 0.163 compactness (mean): 0.019 0.345 concavity (mean): 0.0 0.427 concave points (mean): 0.0 0.201 symmetry (mean): 0.106 0.304 fractal dimension (mean): 0.05 0.097 radius (standard error): 0.112 2.873 texture (standard error): 0.36 4.885 perimeter (standard error): 0.757 21.98 area (standard error): 6.802 542.2 smoothness (standard error): 0.002 0.031 compactness (standard error): 0.002 0.135 concavity (standard error): 0.0 0.396 concave points (standard error): 0.0 0.053 symmetry (standard error): 0.008 0.079 fractal dimension (standard error): 0.001 0.03 radius (worst): 7.93 36.04 texture (worst): 12.02 49.54 perimeter (worst): 50.41 251.2 area (worst): 185.2 4254.0 smoothness (worst): 0.071 0.223 compactness (worst): 0.027 1.058 concavity (worst): 0.0 1.252 concave points (worst): 0.0 0.291 symmetry (worst): 0.156 0.664 fractal dimension (worst): 0.055 0.208 Missing Attribute Values None Class Distribution 212 - Malignant, 357 - Benign Creator Dr. William H. Wolberg, W. Nick Street, Olvi L. Mangasarian Donor Nick Street Date November, 1995 This is a copy of UCI ML Breast Cancer Wisconsin (Diagnostic) datasets. https://goo.gl/U2Uwz2 Features are computed from a digitized image of a fine needle aspirate (FNA) of a breast mass. They describe characteristics of the cell nuclei present in the image. Separating plane described above was obtained using Multisurface Method-Tree (MSM-T) [K. P. Bennett, “Decision Tree Construction Via Linear Programming.” Proceedings of the 4th Midwest Artificial Intelligence and Cognitive Science Society, pp. 97-101, 1992], a classification method which uses linear programming to construct a decision tree. Relevant features were selected using an exhaustive search in the space of 1-4 features and 1-3 separating planes. The actual linear program used to obtain the separating plane in the 3-dimensional space is that described in: [K. P. Bennett and O. L. Mangasarian: “Robust Linear Programming Discrimination of Two Linearly Inseparable Sets”, Optimization Methods and Software 1, 1992, 23-34]. This database is also available through the UW CS ftp server: ftp ftp.cs.wisc.edu cd math-prog/cpo-dataset/machine-learn/WDBC/ References W.N. Street, W.H. Wolberg and O.L. Mangasarian. Nuclear feature extraction for breast tumor diagnosis. IS&T/SPIE 1993 International Symposium on Electronic Imaging: Science and Technology, volume 1905, pages 861-870, San Jose, CA, 1993. O.L. Mangasarian, W.N. Street and W.H. Wolberg. Breast cancer diagnosis and prognosis via linear programming. Operations Research, 43(4), pages 570-577, July-August 1995. W.H. Wolberg, W.N. Street, and O.L. Mangasarian. Machine learning techniques to diagnose breast cancer from fine-needle aspirates. Cancer Letters 77 (1994) 163-171. ',\n",
       " 'sklearn_7_5_loading_other_datasets\\n7.5. Loading other datasets\\ndatasets/index.html#loading-other-datasets\\n 7.3. Real world datasets  scikit-learn provides tools to load larger datasets, downloading them if necessary. They can be loaded using the following functions: fetch_olivetti_faces (*[,\\xa0data_home,\\xa0…]) Load the Olivetti faces data-set from AT&T (classification). fetch_20newsgroups (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the filenames and data from the 20 newsgroups dataset (classification). fetch_20newsgroups_vectorized (*[,\\xa0subset,\\xa0…]) Load the 20 newsgroups dataset and vectorize it into token counts (classification). fetch_lfw_people (*[,\\xa0data_home,\\xa0funneled,\\xa0…]) Load the Labeled Faces in the Wild (LFW) people dataset (classification). fetch_lfw_pairs (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the Labeled Faces in the Wild (LFW) pairs dataset (classification). fetch_covtype (*[,\\xa0data_home,\\xa0…]) Load the covertype dataset (classification). fetch_rcv1 (*[,\\xa0data_home,\\xa0subset,\\xa0…]) Load the RCV1 multilabel dataset (classification). fetch_kddcup99 (*[,\\xa0subset,\\xa0data_home,\\xa0…]) Load the kddcup99 dataset (classification). fetch_california_housing (*[,\\xa0data_home,\\xa0…]) Load the California housing dataset (regression). 7.3.1. The Olivetti faces dataset  This dataset contains a set of face images taken between April 1992 and April 1994 at AT&T Laboratories Cambridge. The sklearn.datasets.fetch_olivetti_faces function is the data fetching / caching function that downloads the data archive from AT&T. As described on the original website: There are ten different images of each of 40 distinct subjects. For some subjects, the images were taken at different times, varying the lighting, facial expressions (open / closed eyes, smiling / not smiling) and facial details (glasses / no glasses). All the images were taken against a dark homogeneous background with the subjects in an upright, frontal position (with tolerance for some side movement). Data Set Characteristics: Classes 40 Samples total 400 Dimensionality 4096 Features real, between 0 and 1 The image is quantized to 256 grey levels and stored as unsigned 8-bit integers; the loader will convert these to floating point values on the interval [0, 1], which are easier to work with for many algorithms. The “target” for this database is an integer from 0 to 39 indicating the identity of the person pictured; however, with only 10 examples per class, this relatively small dataset is more interesting from an unsupervised or semi-supervised perspective. The original dataset consisted of 92 x 112, while the version available here consists of 64x64 images. When using these images, please give credit to AT&T Laboratories Cambridge. 7.3.2. The 20 newsgroups text dataset  The 20 newsgroups dataset comprises around 18000 newsgroups posts on 20 topics split in two subsets: one for training (or development) and the other one for testing (or for performance evaluation). The split between the train and test set is based upon a messages posted before and after a specific date. This module contains two loaders. The first one, sklearn.datasets.fetch_20newsgroups , returns a list of the raw texts that can be fed to text feature extractors such as sklearn.feature_extraction.text.CountVectorizer with custom parameters so as to extract feature vectors. The second one, sklearn.datasets.fetch_20newsgroups_vectorized , returns ready-to-use features, i.e., it is not necessary to use a feature extractor. Data Set Characteristics: Classes 20 Samples total 18846 Dimensionality 1 Features text 7.3.2.1. Usage  The sklearn.datasets.fetch_20newsgroups function is a data fetching / caching functions that downloads the data archive from the original 20 newsgroups website , extracts the archive contents in the folder and calls the sklearn.datasets.load_files on either the training or testing set folder, or both of them: The real data lies in the and attributes. The target attribute is the integer index of the category: It is possible to load only a sub-selection of the categories by passing the list of the categories to load to the sklearn.datasets.fetch_20newsgroups function: 7.3.2.2. Converting text to vectors  In order to feed predictive or clustering models with the text data, one first need to turn the text into vectors of numerical values suitable for statistical analysis. This can be achieved with the utilities of the as demonstrated in the following example that extract TF-IDF vectors of unigram tokens from a subset of 20news: The extracted TF-IDF vectors are very sparse, with an average of 159 non-zero components by sample in a more than 30000-dimensional space (less than .5% non-zero features): sklearn.datasets.fetch_20newsgroups_vectorized is a function which returns ready-to-use token counts features instead of file names. 7.3.2.3. Filtering text for more realistic training  It is easy for a classifier to overfit on particular things that appear in the 20 Newsgroups data, such as newsgroup headers. Many classifiers achieve very high F-scores, but their results would not generalize to other documents that aren’t from this window of time. For example, let’s look at the results of a multinomial Naive Bayes classifier, which is fast to train and achieves a decent F-score: (The example Classification of text documents using sparse features shuffles the training and test data, instead of segmenting by time, and in that case multinomial Naive Bayes gets a much higher F-score of 0.88. Are you suspicious yet of what’s going on inside this classifier?) Let’s take a look at what the most informative features are: You can now see many things that these features have overfit to: Almost every group is distinguished by whether headers such as and appear more or less often. Another significant feature involves whether the sender is affiliated with a university, as indicated either by their headers or their signature. The word “article” is a significant feature, based on how often people quote previous posts like this: “In article [article ID], [name] <[e-mail address]> wrote:” Other features match the names and e-mail addresses of particular people who were posting at the time. With such an abundance of clues that distinguish newsgroups, the classifiers barely have to identify topics from text at all, and they all perform at the same high level. For this reason, the functions that load 20 Newsgroups data provide a parameter called remove , telling it what kinds of information to strip out of each file. remove should be a tuple containing any subset of , telling it to remove headers, signature blocks, and quotation blocks respectively. This classifier lost over a lot of its F-score, just because we removed metadata that has little to do with topic classification. It loses even more if we also strip this metadata from the training data: Some other classifiers cope better with this harder version of the task. Try running Sample pipeline for text feature extraction and evaluation with and without the option to compare the results. Recommendation When evaluating text classifiers on the 20 Newsgroups data, you should strip newsgroup-related metadata. In scikit-learn, you can do this by setting . The F-score will be lower because it is more realistic. Examples Sample pipeline for text feature extraction and evaluation Classification of text documents using sparse features 7.3.3. The Labeled Faces in the Wild face recognition dataset  This dataset is a collection of JPEG pictures of famous people collected over the internet, all details are available on the official website: http://vis-www.cs.umass.edu/lfw/ Each picture is centered on a single face. The typical task is called Face Verification: given a pair of two pictures, a binary classifier must predict whether the two images are from the same person. An alternative task, Face Recognition or Face Identification is: given the picture of the face of an unknown person, identify the name of the person by referring to a gallery of previously seen pictures of identified persons. Both Face Verification and Face Recognition are tasks that are typically performed on the output of a model trained to perform Face Detection. The most popular model for Face Detection is called Viola-Jones and is implemented in the OpenCV library. The LFW faces were extracted by this face detector from various online websites. Data Set Characteristics: Classes 5749 Samples total 13233 Dimensionality 5828 Features real, between 0 and 255 7.3.3.1. Usage  provides two loaders that will automatically download, cache, parse the metadata files, decode the jpeg and convert the interesting slices into memmapped numpy arrays. This dataset size is more than 200 MB. The first load typically takes more than a couple of minutes to fully decode the relevant part of the JPEG files into numpy arrays. If the dataset has been loaded once, the following times the loading times less than 200ms by using a memmapped version memoized on the disk in the folder using . The first loader is used for the Face Identification task: a multi-class classification task (hence supervised learning): The default slice is a rectangular shape around the face, removing most of the background: Each of the faces is assigned to a single person id in the array: The second loader is typically used for the face verification task: each sample is a pair of two picture belonging or not to the same person: Both for the sklearn.datasets.fetch_lfw_people and sklearn.datasets.fetch_lfw_pairs function it is possible to get an additional dimension with the RGB color channels by passing , in that case the shape will be . The sklearn.datasets.fetch_lfw_pairs datasets is subdivided into 3 subsets: the development set, the development set and an evaluation set meant to compute performance metrics using a 10-folds cross validation scheme. References: Labeled Faces in the Wild: A Database for Studying Face Recognition in Unconstrained Environments. Gary B. Huang, Manu Ramesh, Tamara Berg, and Erik Learned-Miller. University of Massachusetts, Amherst, Technical Report 07-49, October, 2007. 7.3.3.2. Examples  Faces recognition example using eigenfaces and SVMs 7.3.4. Forest covertypes  The samples in this dataset correspond to 30×30m patches of forest in the US, collected for the task of predicting each patch’s cover type, i.e. the dominant species of tree. There are seven covertypes, making this a multiclass classification problem. Each sample has 54 features, described on the dataset’s homepage . Some of the features are boolean indicators, while others are discrete or continuous measurements. Data Set Characteristics: Classes 7 Samples total 581012 Dimensionality 54 Features int sklearn.datasets.fetch_covtype will load the covertype dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.5. RCV1 dataset  Reuters Corpus Volume I (RCV1) is an archive of over 800,000 manually categorized newswire stories made available by Reuters, Ltd. for research purposes. The dataset is extensively described in 1 . Data Set Characteristics: Classes 103 Samples total 804414 Dimensionality 47236 Features real, between 0 and 1 sklearn.datasets.fetch_rcv1 will load the following version: RCV1-v2, vectors, full sets, topics multilabels: It returns a dictionary-like object, with the following attributes: : The feature matrix is a scipy CSR sparse matrix, with 804414 samples and 47236 features. Non-zero values contains cosine-normalized, log TF-IDF vectors. A nearly chronological split is proposed in 1 : The first 23149 samples are the training set. The last 781265 samples are the testing set. This follows the official LYRL2004 chronological split. The array has 0.16% of non zero values: : The target values are stored in a scipy CSR sparse matrix, with 804414 samples and 103 categories. Each sample has a value of 1 in its categories, and 0 in others. The array has 3.15% of non zero values: : Each sample can be identified by its ID, ranging (with gaps) from 2286 to 810596: : The target values are the topics of each sample. Each sample belongs to at least one topic, and to up to 17 topics. There are 103 topics, each represented by a string. Their corpus frequencies span five orders of magnitude, from 5 occurrences for ‘GMIL’, to 381327 for ‘CCAT’: The dataset will be downloaded from the rcv1 homepage if necessary. The compressed size is about 656 MB. References 1 ( 1 , 2 ) Lewis, D. D., Yang, Y., Rose, T. G., & Li, F. (2004). RCV1: A new benchmark collection for text categorization research. The Journal of Machine Learning Research, 5, 361-397. 7.3.6. Kddcup 99 dataset  The KDD Cup ‘99 dataset was created by processing the tcpdump portions of the 1998 DARPA Intrusion Detection System (IDS) Evaluation dataset, created by MIT Lincoln Lab [1]. The artificial data (described on the dataset’s homepage ) was generated using a closed network and hand-injected attacks to produce a large number of different types of attack with normal activity in the background. As the initial goal was to produce a large training set for supervised learning algorithms, there is a large proportion (80.1%) of abnormal data which is unrealistic in real world, and inappropriate for unsupervised anomaly detection which aims at detecting ‘abnormal’ data, ie qualitatively different from normal data in large minority among the observations. We thus transform the KDD Data set into two different data sets: SA and SF. -SA is obtained by simply selecting all the normal data, and a small proportion of abnormal data to gives an anomaly proportion of 1%. -SF is obtained as in [2] by simply picking up the data whose attribute logged_in is positive, thus focusing on the intrusion attack, which gives a proportion of 0.3% of attack. -http and smtp are two subsets of SF corresponding with third feature equal to ‘http’ (resp. to ‘smtp’) General KDD structure : Samples total 4898431 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SA structure : Samples total 976158 Dimensionality 41 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type SF structure : Samples total 699691 Dimensionality 4 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type http structure : Samples total 619052 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type smtp structure : Samples total 95373 Dimensionality 3 Features discrete (int) or continuous (float) Targets str, ‘normal.’ or name of the anomaly type sklearn.datasets.fetch_kddcup99 will load the kddcup99 dataset; it returns a dictionary-like object with the feature matrix in the member and the target values in . The dataset will be downloaded from the web if necessary. 7.3.7. California Housing dataset  Data Set Characteristics: Number of Instances 20640 Number of Attributes 8 numeric, predictive attributes and the target Attribute Information MedInc median income in block HouseAge median house age in block AveRooms average number of rooms AveBedrms average number of bedrooms Population block population AveOccup average house occupancy Latitude house block latitude Longitude house block longitude Missing Attribute Values None This dataset was obtained from the StatLib repository. http://lib.stat.cmu.edu/datasets/ The target variable is the median house value for California districts. This dataset was derived from the 1990 U.S. census, using one row per census block group. A block group is the smallest geographical unit for which the U.S. Census Bureau publishes sample data (a block group typically has a population of 600 to 3,000 people). It can be downloaded/loaded using the sklearn.datasets.fetch_california_housing function. References Pace, R. Kelley and Ronald Barry, Sparse Spatial Autoregressions, Statistics and Probability Letters, 33 (1997) 291-297 ',\n",
       " 'sklearn_7_5_loading_other_datasets\\n7.5. Loading other datasets\\ndatasets/index.html#loading-other-datasets\\n 7.4. Generated datasets  In addition, scikit-learn includes various random sample generators that can be used to build artificial datasets of controlled size and complexity. 7.4.1. Generators for classification and clustering  These generators produce a matrix of features and corresponding discrete targets. 7.4.1.1. Single label  Both make_blobs and make_classification create multiclass datasets by allocating each class one or more normally-distributed clusters of points. make_blobs provides greater control regarding the centers and standard deviations of each cluster, and is used to demonstrate clustering. make_classification specialises in introducing noise by way of: correlated, redundant and uninformative features; multiple Gaussian clusters per class; and linear transformations of the feature space. make_gaussian_quantiles divides a single Gaussian cluster into near-equal-size classes separated by concentric hyperspheres. make_hastie_10_2 generates a similar binary, 10-dimensional problem. make_circles and make_moons generate 2d binary classification datasets that are challenging to certain algorithms (e.g. centroid-based clustering or linear classification), including optional Gaussian noise. They are useful for visualisation. make_circles produces Gaussian data with a spherical decision boundary for binary classification, while make_moons produces two interleaving half circles. 7.4.1.2. Multilabel  make_multilabel_classification generates random samples with multiple labels, reflecting a bag of words drawn from a mixture of topics. The number of topics for each document is drawn from a Poisson distribution, and the topics themselves are drawn from a fixed random distribution. Similarly, the number of words is drawn from Poisson, with words drawn from a multinomial, where each topic defines a probability distribution over words. Simplifications with respect to true bag-of-words mixtures include: Per-topic word distributions are independently drawn, where in reality all would be affected by a sparse base distribution, and would be correlated. For a document generated from multiple topics, all topics are weighted equally in generating its bag of words. Documents without labels words at random, rather than from a base distribution. 7.4.1.3. Biclustering  make_biclusters (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with constant block diagonal structure for biclustering. make_checkerboard (shape,\\xa0n_clusters,\\xa0*[,\\xa0…]) Generate an array with block checkerboard structure for biclustering. 7.4.2. Generators for regression  make_regression produces regression targets as an optionally-sparse random linear combination of random features, with noise. Its informative features may be uncorrelated, or low rank (few features account for most of the variance). Other regression generators generate functions deterministically from randomized features. make_sparse_uncorrelated produces a target as a linear combination of four features with fixed coefficients. Others encode explicitly non-linear relations: make_friedman1 is related by polynomial and sine transforms; make_friedman2 includes feature multiplication and reciprocation; and make_friedman3 is similar with an arctan transformation on the target. 7.4.3. Generators for manifold learning  make_s_curve ([n_samples,\\xa0noise,\\xa0random_state]) Generate an S curve dataset. make_swiss_roll ([n_samples,\\xa0noise,\\xa0random_state]) Generate a swiss roll dataset. 7.4.4. Generators for decomposition  make_low_rank_matrix ([n_samples,\\xa0…]) Generate a mostly low rank matrix with bell-shaped singular values make_sparse_coded_signal (n_samples,\\xa0*,\\xa0…) Generate a signal as a sparse combination of dictionary elements. make_spd_matrix (n_dim,\\xa0*[,\\xa0random_state]) Generate a random symmetric, positive-definite matrix. make_sparse_spd_matrix ([dim,\\xa0alpha,\\xa0…]) Generate a sparse symmetric definite positive matrix. ',\n",
       " 'sklearn_7_5_loading_other_datasets\\n7.5. Loading other datasets\\ndatasets/index.html#loading-other-datasets\\n 7.5. Loading other datasets  7.5.1. Sample images  Scikit-learn also embed a couple of sample JPEG images published under Creative Commons license by their authors. Those images can be useful to test algorithms and pipeline on 2D data. load_sample_images () Load sample images for image manipulation. load_sample_image (image_name) Load the numpy array of a single sample image Warning The default coding of images is based on the dtype to spare memory. Often machine learning algorithms work best if the input is converted to a floating point representation first. Also, if you plan to use don’t forget to scale to the range 0 - 1 as done in the following example. Examples: Color Quantization using K-Means 7.5.2. Datasets in svmlight / libsvm format  scikit-learn includes utility functions for loading datasets in the svmlight / libsvm format. In this format, each line takes the form . This format is especially suitable for sparse datasets. In this module, scipy sparse CSR matrices are used for and numpy arrays are used for . You may load a dataset like as follows: You may also load two (or more) datasets at once: In this case, and are guaranteed to have the same number of features. Another way to achieve the same result is to fix the number of features: Related links: Public datasets in svmlight / libsvm format : https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets Faster API-compatible implementation : https://github.com/mblondel/svmlight-loader 7.5.3. Downloading datasets from the openml.org repository  openml.org is a public repository for machine learning data and experiments, that allows everybody to upload open datasets. The package is able to download datasets from the repository using the function sklearn.datasets.fetch_openml . For example, to download a dataset of gene expressions in mice brains: To fully specify a dataset, you need to provide a name and a version, though the version is optional, see Dataset Versions below. The dataset contains a total of 1080 examples belonging to 8 different classes: You can get more information on the dataset by looking at the and attributes: The contains a free-text description of the data, while contains a dictionary of meta-data stored by openml, like the dataset id. For more details, see the OpenML documentation The of the mice protein dataset is 40966, and you can use this (or the name) to get more information on the dataset on the openml website: The also uniquely identifies a dataset from OpenML: 7.5.3.1. Dataset Versions  A dataset is uniquely specified by its , but not necessarily by its name. Several different “versions” of a dataset with the same name can exist which can contain entirely different datasets. If a particular version of a dataset has been found to contain significant issues, it might be deactivated. Using a name to specify a dataset will yield the earliest version of a dataset that is still active. That means that can yield different results at different times if earlier versions become inactive. You can see that the dataset with 40966 that we fetched above is the version 1 of the “miceprotein” dataset: In fact, this dataset only has one version. The iris dataset on the other hand has multiple versions: Specifying the dataset by the name “iris” yields the lowest version, version 1, with the 61. To make sure you always get this exact dataset, it is safest to specify it by the dataset . The other dataset, with 969, is version 3 (version 2 has become inactive), and contains a binarized version of the data: You can also specify both the name and the version, which also uniquely identifies the dataset: References: Vanschoren, van Rijn, Bischl and Torgo “OpenML: networked science in machine learning” , ACM SIGKDD Explorations Newsletter, 15(2), 49-60, 2014. 7.5.4. Loading from external datasets  scikit-learn works on any numeric data stored as numpy arrays or scipy sparse matrices. Other types that are convertible to numeric arrays such as pandas DataFrame are also acceptable. Here are some recommended ways to load standard columnar data into a format usable by scikit-learn: pandas.io provides tools to read data from common formats including CSV, Excel, JSON and SQL. DataFrames may also be constructed from lists of tuples or dicts. Pandas handles heterogeneous data smoothly and provides tools for manipulation and conversion into a numeric array suitable for scikit-learn. scipy.io specializes in binary formats often used in scientific computing context such as .mat and .arff numpy/routines.io for standard loading of columnar data into numpy arrays scikit-learn’s datasets.load_svmlight_file for the svmlight or libSVM sparse format scikit-learn’s datasets.load_files for directories of text files where the name of each directory is the name of each category and each file inside of each directory corresponds to one sample from that category For some miscellaneous data such as images, videos, and audio, you may wish to refer to: skimage.io or Imageio for loading images and videos into numpy arrays scipy.io.wavfile.read for reading WAV files into a numpy array Categorical (or nominal) features stored as strings (common in pandas DataFrames) will need converting to numerical features using sklearn.preprocessing.OneHotEncoder or sklearn.preprocessing.OrdinalEncoder or similar. See Preprocessing data . Note: if you manage your own numerical data it is recommended to use an optimized file format such as HDF5 to reduce data load times. Various libraries such as H5Py, PyTables and pandas provides a Python interface for reading and writing data in that format. ',\n",
       " 'sklearn_8_1_strategies_to_scale_computationally_bigger_data\\n8.1. Strategies to scale computationally: bigger data\\nmodules/computing.html#strategies-to-scale-computationally-bigger-data\\n 8.1. Strategies to scale computationally: bigger data  For some applications the amount of examples, features (or both) and/or the speed at which they need to be processed are challenging for traditional approaches. In these cases scikit-learn has a number of options you can consider to make your system scale. 8.1.1. Scaling with instances using out-of-core learning  Out-of-core (or “external memory”) learning is a technique used to learn from data that cannot fit in a computer’s main memory (RAM). Here is a sketch of a system designed to achieve this goal: a way to stream instances a way to extract features from instances an incremental algorithm 8.1.1.1. Streaming instances  Basically, 1. may be a reader that yields instances from files on a hard drive, a database, from a network stream etc. However, details on how to achieve this are beyond the scope of this documentation. 8.1.1.2. Extracting features  2. could be any relevant way to extract features among the different feature extraction methods supported by scikit-learn. However, when working with data that needs vectorization and where the set of features or values is not known in advance one should take explicit care. A good example is text classification where unknown terms are likely to be found during training. It is possible to use a stateful vectorizer if making multiple passes over the data is reasonable from an application point of view. Otherwise, one can turn up the difficulty by using a stateless feature extractor. Currently the preferred way to do this is to use the so-called hashing trick as implemented by sklearn.feature_extraction.FeatureHasher for datasets with categorical variables represented as list of Python dicts or sklearn.feature_extraction.text.HashingVectorizer for text documents. 8.1.1.3. Incremental learning  Finally, for 3. we have a number of options inside scikit-learn. Although not all algorithms can learn incrementally (i.e. without seeing all the instances at once), all estimators implementing the API are candidates. Actually, the ability to learn incrementally from a mini-batch of instances (sometimes called “online learning”) is key to out-of-core learning as it guarantees that at any given time there will be only a small amount of instances in the main memory. Choosing a good size for the mini-batch that balances relevancy and memory footprint could involve some tuning 1 . Here is a list of incremental estimators for different tasks: Classification sklearn.naive_bayes.MultinomialNB sklearn.naive_bayes.BernoulliNB sklearn.linear_model.Perceptron sklearn.linear_model.SGDClassifier sklearn.linear_model.PassiveAggressiveClassifier sklearn.neural_network.MLPClassifier Regression sklearn.linear_model.SGDRegressor sklearn.linear_model.PassiveAggressiveRegressor sklearn.neural_network.MLPRegressor Clustering sklearn.cluster.MiniBatchKMeans sklearn.cluster.Birch Decomposition / feature Extraction sklearn.decomposition.MiniBatchDictionaryLearning sklearn.decomposition.IncrementalPCA sklearn.decomposition.LatentDirichletAllocation Preprocessing sklearn.preprocessing.StandardScaler sklearn.preprocessing.MinMaxScaler sklearn.preprocessing.MaxAbsScaler For classification, a somewhat important thing to note is that although a stateless feature extraction routine may be able to cope with new/unseen attributes, the incremental learner itself may be unable to cope with new/unseen targets classes. In this case you have to pass all the possible classes to the first call using the parameter. Another aspect to consider when choosing a proper algorithm is that not all of them put the same importance on each example over time. Namely, the is still sensitive to badly labeled examples even after many examples whereas the and families are more robust to this kind of artifacts. Conversely, the latter also tend to give less importance to remarkably different, yet properly labeled examples when they come late in the stream as their learning rate decreases over time. 8.1.1.4. Examples  Finally, we have a full-fledged example of Out-of-core classification of text documents . It is aimed at providing a starting point for people wanting to build out-of-core learning systems and demonstrates most of the notions discussed above. Furthermore, it also shows the evolution of the performance of different algorithms with the number of processed examples. Now looking at the computation time of the different parts, we see that the vectorization is much more expensive than learning itself. From the different algorithms, is the most expensive, but its overhead can be mitigated by increasing the size of the mini-batches (exercise: change to 100 and 10000 in the program and compare). 8.1.1.5. Notes  1 Depending on the algorithm the mini-batch size can influence results or not. SGD*, PassiveAggressive*, and discrete NaiveBayes are truly online and are not affected by batch size. Conversely, MiniBatchKMeans convergence rate is affected by the batch size. Also, its memory footprint can vary dramatically with batch size. ',\n",
       " 'sklearn_8_1_strategies_to_scale_computationally_bigger_data\\n8.1. Strategies to scale computationally: bigger data\\nmodules/computing.html#strategies-to-scale-computationally-bigger-data\\n 8.2. Computational Performance  For some applications the performance (mainly latency and throughput at prediction time) of estimators is crucial. It may also be of interest to consider the training throughput but this is often less important in a production setup (where it often takes place offline). We will review here the orders of magnitude you can expect from a number of scikit-learn estimators in different contexts and provide some tips and tricks for overcoming performance bottlenecks. Prediction latency is measured as the elapsed time necessary to make a prediction (e.g. in micro-seconds). Latency is often viewed as a distribution and operations engineers often focus on the latency at a given percentile of this distribution (e.g. the 90 percentile). Prediction throughput is defined as the number of predictions the software can deliver in a given amount of time (e.g. in predictions per second). An important aspect of performance optimization is also that it can hurt prediction accuracy. Indeed, simpler models (e.g. linear instead of non-linear, or with fewer parameters) often run faster but are not always able to take into account the same exact properties of the data as more complex ones. 8.2.1. Prediction Latency  One of the most straight-forward concerns one may have when using/choosing a machine learning toolkit is the latency at which predictions can be made in a production environment. The main factors that influence the prediction latency are Number of features Input data representation and sparsity Model complexity Feature extraction A last major parameter is also the possibility to do predictions in bulk or one-at-a-time mode. 8.2.1.1. Bulk versus Atomic mode  In general doing predictions in bulk (many instances at the same time) is more efficient for a number of reasons (branching predictability, CPU cache, linear algebra libraries optimizations etc.). Here we see on a setting with few features that independently of estimator choice the bulk mode is always faster, and for some of them by 1 to 2 orders of magnitude: To benchmark different estimators for your case you can simply change the parameter in this example: Prediction Latency . This should give you an estimate of the order of magnitude of the prediction latency. 8.2.1.2. Configuring Scikit-learn for reduced validation overhead  Scikit-learn does some validation on data that increases the overhead per call to and similar functions. In particular, checking that features are finite (not NaN or infinite) involves a full pass over the data. If you ensure that your data is acceptable, you may suppress checking for finiteness by setting the environment variable to a non-empty string before importing scikit-learn, or configure it in Python with sklearn.set_config . For more control than these global settings, a config_context allows you to set this configuration within a specified context: Note that this will affect all uses of sklearn.utils.assert_all_finite within the context. 8.2.1.3. Influence of the Number of Features  Obviously when the number of features increases so does the memory consumption of each example. Indeed, for a matrix of instances with features, the space complexity is in . From a computing perspective it also means that the number of basic operations (e.g., multiplications for vector-matrix products in linear models) increases too. Here is a graph of the evolution of the prediction latency with the number of features: Overall you can expect the prediction time to increase at least linearly with the number of features (non-linear cases can happen depending on the global memory footprint and estimator). 8.2.1.4. Influence of the Input Data Representation  Scipy provides sparse matrix data structures which are optimized for storing sparse data. The main feature of sparse formats is that you don’t store zeros so if your data is sparse then you use much less memory. A non-zero value in a sparse ( CSR or CSC ) representation will only take on average one 32bit integer position + the 64 bit floating point value + an additional 32bit per row or column in the matrix. Using sparse input on a dense (or sparse) linear model can speedup prediction by quite a bit as only the non zero valued features impact the dot product and thus the model predictions. Hence if you have 100 non zeros in 1e6 dimensional space, you only need 100 multiply and add operation instead of 1e6. Calculation over a dense representation, however, may leverage highly optimised vector operations and multithreading in BLAS, and tends to result in fewer CPU cache misses. So the sparsity should typically be quite high (10% non-zeros max, to be checked depending on the hardware) for the sparse input representation to be faster than the dense input representation on a machine with many CPUs and an optimized BLAS implementation. Here is sample code to test the sparsity of your input: As a rule of thumb you can consider that if the sparsity ratio is greater than 90% you can probably benefit from sparse formats. Check Scipy’s sparse matrix formats documentation for more information on how to build (or convert your data to) sparse matrix formats. Most of the time the and formats work best. 8.2.1.5. Influence of the Model Complexity  Generally speaking, when model complexity increases, predictive power and latency are supposed to increase. Increasing predictive power is usually interesting, but for many applications we would better not increase prediction latency too much. We will now review this idea for different families of supervised models. For sklearn.linear_model (e.g. Lasso, ElasticNet, SGDClassifier/Regressor, Ridge & RidgeClassifier, PassiveAggressiveClassifier/Regressor, LinearSVC, LogisticRegression…) the decision function that is applied at prediction time is the same (a dot product) , so latency should be equivalent. Here is an example using sklearn.linear_model.SGDClassifier with the penalty. The regularization strength is globally controlled by the parameter. With a sufficiently high , one can then increase the parameter of to enforce various levels of sparsity in the model coefficients. Higher sparsity here is interpreted as less model complexity as we need fewer coefficients to describe it fully. Of course sparsity influences in turn the prediction time as the sparse dot-product takes time roughly proportional to the number of non-zero coefficients. For the sklearn.svm family of algorithms with a non-linear kernel, the latency is tied to the number of support vectors (the fewer the faster). Latency and throughput should (asymptotically) grow linearly with the number of support vectors in a SVC or SVR model. The kernel will also influence the latency as it is used to compute the projection of the input vector once per support vector. In the following graph the parameter of sklearn.svm.NuSVR was used to influence the number of support vectors. For sklearn.ensemble of trees (e.g. RandomForest, GBT, ExtraTrees etc) the number of trees and their depth play the most important role. Latency and throughput should scale linearly with the number of trees. In this case we used directly the parameter of sklearn.ensemble.gradient_boosting.GradientBoostingRegressor . In any case be warned that decreasing model complexity can hurt accuracy as mentioned above. For instance a non-linearly separable problem can be handled with a speedy linear model but prediction power will very likely suffer in the process. 8.2.1.6. Feature Extraction Latency  Most scikit-learn models are usually pretty fast as they are implemented either with compiled Cython extensions or optimized computing libraries. On the other hand, in many real world applications the feature extraction process (i.e. turning raw data like database rows or network packets into numpy arrays) governs the overall prediction time. For example on the Reuters text classification task the whole preparation (reading and parsing SGML files, tokenizing the text and hashing it into a common vector space) is taking 100 to 500 times more time than the actual prediction code, depending on the chosen model. In many cases it is thus recommended to carefully time and profile your feature extraction code as it may be a good place to start optimizing when your overall latency is too slow for your application. 8.2.2. Prediction Throughput  Another important metric to care about when sizing production systems is the throughput i.e. the number of predictions you can make in a given amount of time. Here is a benchmark from the Prediction Latency example that measures this quantity for a number of estimators on synthetic data: These throughputs are achieved on a single process. An obvious way to increase the throughput of your application is to spawn additional instances (usually processes in Python because of the GIL ) that share the same model. One might also add machines to spread the load. A detailed explanation on how to achieve this is beyond the scope of this documentation though. 8.2.3. Tips and Tricks  8.2.3.1. Linear algebra libraries  As scikit-learn relies heavily on Numpy/Scipy and linear algebra in general it makes sense to take explicit care of the versions of these libraries. Basically, you ought to make sure that Numpy is built using an optimized BLAS / LAPACK library. Not all models benefit from optimized BLAS and Lapack implementations. For instance models based on (randomized) decision trees typically do not rely on BLAS calls in their inner loops, nor do kernel SVMs ( , , , ). On the other hand a linear model implemented with a BLAS DGEMM call (via ) will typically benefit hugely from a tuned BLAS implementation and lead to orders of magnitude speedup over a non-optimized BLAS. You can display the BLAS / LAPACK implementation used by your NumPy / SciPy / scikit-learn install with the following commands: Optimized BLAS / LAPACK implementations include: Atlas (need hardware specific tuning by rebuilding on the target machine) OpenBLAS MKL Apple Accelerate and vecLib frameworks (OSX only) More information can be found on the Scipy install page and in this blog post from Daniel Nouri which has some nice step by step install instructions for Debian / Ubuntu. 8.2.3.2. Limiting Working Memory  Some calculations when implemented using standard numpy vectorized operations involve using a large amount of temporary memory. This may potentially exhaust system memory. Where computations can be performed in fixed-memory chunks, we attempt to do so, and allow the user to hint at the maximum size of this working memory (defaulting to 1GB) using sklearn.set_config or config_context . The following suggests to limit temporary working memory to 128 MiB: An example of a chunked operation adhering to this setting is metric.pairwise_distances_chunked , which facilitates computing row-wise reductions of a pairwise distance matrix. 8.2.3.3. Model Compression  Model compression in scikit-learn only concerns linear models for the moment. In this context it means that we want to control the model sparsity (i.e. the number of non-zero coordinates in the model vectors). It is generally a good idea to combine model sparsity with sparse input data representation. Here is sample code that illustrates the use of the method: In this example we prefer the penalty as it is often a good compromise between model compactness and prediction power. One can also further tune the parameter (in combination with the regularization strength ) to control this tradeoff. A typical benchmark on synthetic data yields a >30% decrease in latency when both the model and input are sparse (with 0.000024 and 0.027400 non-zero coefficients ratio respectively). Your mileage may vary depending on the sparsity and size of your data and model. Furthermore, sparsifying can be very useful to reduce the memory usage of predictive models deployed on production servers. 8.2.3.4. Model Reshaping  Model reshaping consists in selecting only a portion of the available features to fit a model. In other words, if a model discards features during the learning phase we can then strip those from the input. This has several benefits. Firstly it reduces memory (and therefore time) overhead of the model itself. It also allows to discard explicit feature selection components in a pipeline once we know which features to keep from a previous run. Finally, it can help reduce processing time and I/O usage upstream in the data access and feature extraction layers by not collecting and building features that are discarded by the model. For instance if the raw data come from a database, it can make it possible to write simpler and faster queries or reduce I/O usage by making the queries return lighter records. At the moment, reshaping needs to be performed manually in scikit-learn. In the case of sparse input (particularly in format), it is generally sufficient to not generate the relevant features, leaving their columns empty. 8.2.3.5. Links  scikit-learn developer performance documentation Scipy sparse matrix formats documentation ',\n",
       " 'sklearn_8_1_strategies_to_scale_computationally_bigger_data\\n8.1. Strategies to scale computationally: bigger data\\nmodules/computing.html#strategies-to-scale-computationally-bigger-data\\n 8.3. Parallelism, resource management, and configuration  8.3.1. Parallelism  Some scikit-learn estimators and utilities can parallelize costly operations using multiple CPU cores, thanks to the following components: via the joblib library. In this case the number of threads or processes can be controlled with the parameter. via OpenMP, used in C or Cython code. In addition, some of the numpy routines that are used internally by scikit-learn may also be parallelized if numpy is installed with specific numerical libraries such as MKL, OpenBLAS, or BLIS. We describe these 3 scenarios in the following subsections. 8.3.1.1. Joblib-based parallelism  When the underlying implementation uses joblib, the number of workers (threads or processes) that are spawned in parallel can be controlled via the parameter. Note Where (and how) parallelization happens in the estimators is currently poorly documented. Please help us by improving our docs and tackle issue 14228 ! Joblib is able to support both multi-processing and multi-threading. Whether joblib chooses to spawn a thread or a process depends on the backend that it’s using. Scikit-learn generally relies on the backend, which is joblib’s default backend. Loky is a multi-processing backend. When doing multi-processing, in order to avoid duplicating the memory in each process (which isn’t reasonable with big datasets), joblib will create a memmap that all processes can share, when the data is bigger than 1MB. In some specific cases (when the code that is run in parallel releases the GIL), scikit-learn will indicate to that a multi-threading backend is preferable. As a user, you may control the backend that joblib will use (regardless of what scikit-learn recommends) by using a context manager: Please refer to the joblib’s docs for more details. In practice, whether parallelism is helpful at improving runtime depends on many factors. It is usually a good idea to experiment rather than assuming that increasing the number of workers is always a good thing. In some cases it can be highly detrimental to performance to run multiple copies of some estimators or functions in parallel (see oversubscription below). 8.3.1.2. OpenMP-based parallelism  OpenMP is used to parallelize code written in Cython or C, relying on multi-threading exclusively. By default (and unless joblib is trying to avoid oversubscription), the implementation will use as many threads as possible. You can control the exact number of threads that are used via the environment variable: 8.3.1.3. Parallel Numpy routines from numerical libraries  Scikit-learn relies heavily on NumPy and SciPy, which internally call multi-threaded linear algebra routines implemented in libraries such as MKL, OpenBLAS or BLIS. The number of threads used by the OpenBLAS, MKL or BLIS libraries can be set via the , , and environment variables. Please note that scikit-learn has no direct control over these implementations. Scikit-learn solely relies on Numpy and Scipy. Note At the time of writing (2019), NumPy and SciPy packages distributed on pypi.org (used by ) and on the conda-forge channel are linked with OpenBLAS, while conda packages shipped on the “defaults” channel from anaconda.org are linked by default with MKL. 8.3.1.4. Oversubscription: spawning too many threads  It is generally recommended to avoid using significantly more processes or threads than the number of CPUs on a machine. Over-subscription happens when a program is running too many threads at the same time. Suppose you have a machine with 8 CPUs. Consider a case where you’re running a GridSearchCV (parallelized with joblib) with over a HistGradientBoostingClassifier (parallelized with OpenMP). Each instance of HistGradientBoostingClassifier will spawn 8 threads (since you have 8 CPUs). That’s a total of threads, which leads to oversubscription of physical CPU resources and to scheduling overhead. Oversubscription can arise in the exact same fashion with parallelized routines from MKL, OpenBLAS or BLIS that are nested in joblib calls. Starting from , when the backend is used (which is the default), joblib will tell its child processes to limit the number of threads they can use, so as to avoid oversubscription. In practice the heuristic that joblib uses is to tell the processes to use , via their corresponding environment variable. Back to our example from above, since the joblib backend of GridSearchCV is , each process will only be able to use 1 thread instead of 8, thus mitigating the oversubscription issue. Note that: Manually setting one of the environment variables ( , , , or ) will take precedence over what joblib tries to do. The total number of threads will be . Note that setting this limit will also impact your computations in the main process, which will only use . Joblib exposes a context manager for finer control over the number of threads in its workers (see joblib docs linked below). Joblib is currently unable to avoid oversubscription in a multi-threading context. It can only do so with the backend (which spawns processes). You will find additional details about joblib mitigation of oversubscription in joblib documentation . 8.3.2. Configuration switches  8.3.2.1. Python runtime  sklearn.set_config controls the following behaviors: assume_finite used to skip validation, which enables faster computations but may lead to segmentation faults if the data contains NaNs. working_memory the optimal size of temporary arrays used by some algorithms. 8.3.2.2. Environment variables  These environment variables should be set before importing scikit-learn. SKLEARN_SITE_JOBLIB When this environment variable is set to a non zero value, scikit-learn uses the site joblib rather than its vendored version. Consequently, joblib must be installed for scikit-learn to run. Note that using the site joblib is at your own risks: the versions of scikit-learn and joblib need to be compatible. Currently, joblib 0.11+ is supported. In addition, dumps from joblib.Memory might be incompatible, and you might loose some caches and have to redownload some datasets. Deprecated since version 0.21: As of version 0.21 this parameter has no effect, vendored joblib was removed and site joblib is always used. SKLEARN_ASSUME_FINITE Sets the default value for the argument of sklearn.set_config . SKLEARN_WORKING_MEMORY Sets the default value for the argument of sklearn.set_config . SKLEARN_SEED Sets the seed of the global random generator when running the tests, for reproducibility. SKLEARN_SKIP_NETWORK_TESTS When this environment variable is set to a non zero value, the tests that need network access are skipped. ',\n",
       " 'sklearn_8_2_computational_performance\\n8.2. Computational Performance\\nmodules/computing.html#computational-performance\\n 8.1. Strategies to scale computationally: bigger data  For some applications the amount of examples, features (or both) and/or the speed at which they need to be processed are challenging for traditional approaches. In these cases scikit-learn has a number of options you can consider to make your system scale. 8.1.1. Scaling with instances using out-of-core learning  Out-of-core (or “external memory”) learning is a technique used to learn from data that cannot fit in a computer’s main memory (RAM). Here is a sketch of a system designed to achieve this goal: a way to stream instances a way to extract features from instances an incremental algorithm 8.1.1.1. Streaming instances  Basically, 1. may be a reader that yields instances from files on a hard drive, a database, from a network stream etc. However, details on how to achieve this are beyond the scope of this documentation. 8.1.1.2. Extracting features  2. could be any relevant way to extract features among the different feature extraction methods supported by scikit-learn. However, when working with data that needs vectorization and where the set of features or values is not known in advance one should take explicit care. A good example is text classification where unknown terms are likely to be found during training. It is possible to use a stateful vectorizer if making multiple passes over the data is reasonable from an application point of view. Otherwise, one can turn up the difficulty by using a stateless feature extractor. Currently the preferred way to do this is to use the so-called hashing trick as implemented by sklearn.feature_extraction.FeatureHasher for datasets with categorical variables represented as list of Python dicts or sklearn.feature_extraction.text.HashingVectorizer for text documents. 8.1.1.3. Incremental learning  Finally, for 3. we have a number of options inside scikit-learn. Although not all algorithms can learn incrementally (i.e. without seeing all the instances at once), all estimators implementing the API are candidates. Actually, the ability to learn incrementally from a mini-batch of instances (sometimes called “online learning”) is key to out-of-core learning as it guarantees that at any given time there will be only a small amount of instances in the main memory. Choosing a good size for the mini-batch that balances relevancy and memory footprint could involve some tuning 1 . Here is a list of incremental estimators for different tasks: Classification sklearn.naive_bayes.MultinomialNB sklearn.naive_bayes.BernoulliNB sklearn.linear_model.Perceptron sklearn.linear_model.SGDClassifier sklearn.linear_model.PassiveAggressiveClassifier sklearn.neural_network.MLPClassifier Regression sklearn.linear_model.SGDRegressor sklearn.linear_model.PassiveAggressiveRegressor sklearn.neural_network.MLPRegressor Clustering sklearn.cluster.MiniBatchKMeans sklearn.cluster.Birch Decomposition / feature Extraction sklearn.decomposition.MiniBatchDictionaryLearning sklearn.decomposition.IncrementalPCA sklearn.decomposition.LatentDirichletAllocation Preprocessing sklearn.preprocessing.StandardScaler sklearn.preprocessing.MinMaxScaler sklearn.preprocessing.MaxAbsScaler For classification, a somewhat important thing to note is that although a stateless feature extraction routine may be able to cope with new/unseen attributes, the incremental learner itself may be unable to cope with new/unseen targets classes. In this case you have to pass all the possible classes to the first call using the parameter. Another aspect to consider when choosing a proper algorithm is that not all of them put the same importance on each example over time. Namely, the is still sensitive to badly labeled examples even after many examples whereas the and families are more robust to this kind of artifacts. Conversely, the latter also tend to give less importance to remarkably different, yet properly labeled examples when they come late in the stream as their learning rate decreases over time. 8.1.1.4. Examples  Finally, we have a full-fledged example of Out-of-core classification of text documents . It is aimed at providing a starting point for people wanting to build out-of-core learning systems and demonstrates most of the notions discussed above. Furthermore, it also shows the evolution of the performance of different algorithms with the number of processed examples. Now looking at the computation time of the different parts, we see that the vectorization is much more expensive than learning itself. From the different algorithms, is the most expensive, but its overhead can be mitigated by increasing the size of the mini-batches (exercise: change to 100 and 10000 in the program and compare). 8.1.1.5. Notes  1 Depending on the algorithm the mini-batch size can influence results or not. SGD*, PassiveAggressive*, and discrete NaiveBayes are truly online and are not affected by batch size. Conversely, MiniBatchKMeans convergence rate is affected by the batch size. Also, its memory footprint can vary dramatically with batch size. ',\n",
       " 'sklearn_8_2_computational_performance\\n8.2. Computational Performance\\nmodules/computing.html#computational-performance\\n 8.2. Computational Performance  For some applications the performance (mainly latency and throughput at prediction time) of estimators is crucial. It may also be of interest to consider the training throughput but this is often less important in a production setup (where it often takes place offline). We will review here the orders of magnitude you can expect from a number of scikit-learn estimators in different contexts and provide some tips and tricks for overcoming performance bottlenecks. Prediction latency is measured as the elapsed time necessary to make a prediction (e.g. in micro-seconds). Latency is often viewed as a distribution and operations engineers often focus on the latency at a given percentile of this distribution (e.g. the 90 percentile). Prediction throughput is defined as the number of predictions the software can deliver in a given amount of time (e.g. in predictions per second). An important aspect of performance optimization is also that it can hurt prediction accuracy. Indeed, simpler models (e.g. linear instead of non-linear, or with fewer parameters) often run faster but are not always able to take into account the same exact properties of the data as more complex ones. 8.2.1. Prediction Latency  One of the most straight-forward concerns one may have when using/choosing a machine learning toolkit is the latency at which predictions can be made in a production environment. The main factors that influence the prediction latency are Number of features Input data representation and sparsity Model complexity Feature extraction A last major parameter is also the possibility to do predictions in bulk or one-at-a-time mode. 8.2.1.1. Bulk versus Atomic mode  In general doing predictions in bulk (many instances at the same time) is more efficient for a number of reasons (branching predictability, CPU cache, linear algebra libraries optimizations etc.). Here we see on a setting with few features that independently of estimator choice the bulk mode is always faster, and for some of them by 1 to 2 orders of magnitude: To benchmark different estimators for your case you can simply change the parameter in this example: Prediction Latency . This should give you an estimate of the order of magnitude of the prediction latency. 8.2.1.2. Configuring Scikit-learn for reduced validation overhead  Scikit-learn does some validation on data that increases the overhead per call to and similar functions. In particular, checking that features are finite (not NaN or infinite) involves a full pass over the data. If you ensure that your data is acceptable, you may suppress checking for finiteness by setting the environment variable to a non-empty string before importing scikit-learn, or configure it in Python with sklearn.set_config . For more control than these global settings, a config_context allows you to set this configuration within a specified context: Note that this will affect all uses of sklearn.utils.assert_all_finite within the context. 8.2.1.3. Influence of the Number of Features  Obviously when the number of features increases so does the memory consumption of each example. Indeed, for a matrix of instances with features, the space complexity is in . From a computing perspective it also means that the number of basic operations (e.g., multiplications for vector-matrix products in linear models) increases too. Here is a graph of the evolution of the prediction latency with the number of features: Overall you can expect the prediction time to increase at least linearly with the number of features (non-linear cases can happen depending on the global memory footprint and estimator). 8.2.1.4. Influence of the Input Data Representation  Scipy provides sparse matrix data structures which are optimized for storing sparse data. The main feature of sparse formats is that you don’t store zeros so if your data is sparse then you use much less memory. A non-zero value in a sparse ( CSR or CSC ) representation will only take on average one 32bit integer position + the 64 bit floating point value + an additional 32bit per row or column in the matrix. Using sparse input on a dense (or sparse) linear model can speedup prediction by quite a bit as only the non zero valued features impact the dot product and thus the model predictions. Hence if you have 100 non zeros in 1e6 dimensional space, you only need 100 multiply and add operation instead of 1e6. Calculation over a dense representation, however, may leverage highly optimised vector operations and multithreading in BLAS, and tends to result in fewer CPU cache misses. So the sparsity should typically be quite high (10% non-zeros max, to be checked depending on the hardware) for the sparse input representation to be faster than the dense input representation on a machine with many CPUs and an optimized BLAS implementation. Here is sample code to test the sparsity of your input: As a rule of thumb you can consider that if the sparsity ratio is greater than 90% you can probably benefit from sparse formats. Check Scipy’s sparse matrix formats documentation for more information on how to build (or convert your data to) sparse matrix formats. Most of the time the and formats work best. 8.2.1.5. Influence of the Model Complexity  Generally speaking, when model complexity increases, predictive power and latency are supposed to increase. Increasing predictive power is usually interesting, but for many applications we would better not increase prediction latency too much. We will now review this idea for different families of supervised models. For sklearn.linear_model (e.g. Lasso, ElasticNet, SGDClassifier/Regressor, Ridge & RidgeClassifier, PassiveAggressiveClassifier/Regressor, LinearSVC, LogisticRegression…) the decision function that is applied at prediction time is the same (a dot product) , so latency should be equivalent. Here is an example using sklearn.linear_model.SGDClassifier with the penalty. The regularization strength is globally controlled by the parameter. With a sufficiently high , one can then increase the parameter of to enforce various levels of sparsity in the model coefficients. Higher sparsity here is interpreted as less model complexity as we need fewer coefficients to describe it fully. Of course sparsity influences in turn the prediction time as the sparse dot-product takes time roughly proportional to the number of non-zero coefficients. For the sklearn.svm family of algorithms with a non-linear kernel, the latency is tied to the number of support vectors (the fewer the faster). Latency and throughput should (asymptotically) grow linearly with the number of support vectors in a SVC or SVR model. The kernel will also influence the latency as it is used to compute the projection of the input vector once per support vector. In the following graph the parameter of sklearn.svm.NuSVR was used to influence the number of support vectors. For sklearn.ensemble of trees (e.g. RandomForest, GBT, ExtraTrees etc) the number of trees and their depth play the most important role. Latency and throughput should scale linearly with the number of trees. In this case we used directly the parameter of sklearn.ensemble.gradient_boosting.GradientBoostingRegressor . In any case be warned that decreasing model complexity can hurt accuracy as mentioned above. For instance a non-linearly separable problem can be handled with a speedy linear model but prediction power will very likely suffer in the process. 8.2.1.6. Feature Extraction Latency  Most scikit-learn models are usually pretty fast as they are implemented either with compiled Cython extensions or optimized computing libraries. On the other hand, in many real world applications the feature extraction process (i.e. turning raw data like database rows or network packets into numpy arrays) governs the overall prediction time. For example on the Reuters text classification task the whole preparation (reading and parsing SGML files, tokenizing the text and hashing it into a common vector space) is taking 100 to 500 times more time than the actual prediction code, depending on the chosen model. In many cases it is thus recommended to carefully time and profile your feature extraction code as it may be a good place to start optimizing when your overall latency is too slow for your application. 8.2.2. Prediction Throughput  Another important metric to care about when sizing production systems is the throughput i.e. the number of predictions you can make in a given amount of time. Here is a benchmark from the Prediction Latency example that measures this quantity for a number of estimators on synthetic data: These throughputs are achieved on a single process. An obvious way to increase the throughput of your application is to spawn additional instances (usually processes in Python because of the GIL ) that share the same model. One might also add machines to spread the load. A detailed explanation on how to achieve this is beyond the scope of this documentation though. 8.2.3. Tips and Tricks  8.2.3.1. Linear algebra libraries  As scikit-learn relies heavily on Numpy/Scipy and linear algebra in general it makes sense to take explicit care of the versions of these libraries. Basically, you ought to make sure that Numpy is built using an optimized BLAS / LAPACK library. Not all models benefit from optimized BLAS and Lapack implementations. For instance models based on (randomized) decision trees typically do not rely on BLAS calls in their inner loops, nor do kernel SVMs ( , , , ). On the other hand a linear model implemented with a BLAS DGEMM call (via ) will typically benefit hugely from a tuned BLAS implementation and lead to orders of magnitude speedup over a non-optimized BLAS. You can display the BLAS / LAPACK implementation used by your NumPy / SciPy / scikit-learn install with the following commands: Optimized BLAS / LAPACK implementations include: Atlas (need hardware specific tuning by rebuilding on the target machine) OpenBLAS MKL Apple Accelerate and vecLib frameworks (OSX only) More information can be found on the Scipy install page and in this blog post from Daniel Nouri which has some nice step by step install instructions for Debian / Ubuntu. 8.2.3.2. Limiting Working Memory  Some calculations when implemented using standard numpy vectorized operations involve using a large amount of temporary memory. This may potentially exhaust system memory. Where computations can be performed in fixed-memory chunks, we attempt to do so, and allow the user to hint at the maximum size of this working memory (defaulting to 1GB) using sklearn.set_config or config_context . The following suggests to limit temporary working memory to 128 MiB: An example of a chunked operation adhering to this setting is metric.pairwise_distances_chunked , which facilitates computing row-wise reductions of a pairwise distance matrix. 8.2.3.3. Model Compression  Model compression in scikit-learn only concerns linear models for the moment. In this context it means that we want to control the model sparsity (i.e. the number of non-zero coordinates in the model vectors). It is generally a good idea to combine model sparsity with sparse input data representation. Here is sample code that illustrates the use of the method: In this example we prefer the penalty as it is often a good compromise between model compactness and prediction power. One can also further tune the parameter (in combination with the regularization strength ) to control this tradeoff. A typical benchmark on synthetic data yields a >30% decrease in latency when both the model and input are sparse (with 0.000024 and 0.027400 non-zero coefficients ratio respectively). Your mileage may vary depending on the sparsity and size of your data and model. Furthermore, sparsifying can be very useful to reduce the memory usage of predictive models deployed on production servers. 8.2.3.4. Model Reshaping  Model reshaping consists in selecting only a portion of the available features to fit a model. In other words, if a model discards features during the learning phase we can then strip those from the input. This has several benefits. Firstly it reduces memory (and therefore time) overhead of the model itself. It also allows to discard explicit feature selection components in a pipeline once we know which features to keep from a previous run. Finally, it can help reduce processing time and I/O usage upstream in the data access and feature extraction layers by not collecting and building features that are discarded by the model. For instance if the raw data come from a database, it can make it possible to write simpler and faster queries or reduce I/O usage by making the queries return lighter records. At the moment, reshaping needs to be performed manually in scikit-learn. In the case of sparse input (particularly in format), it is generally sufficient to not generate the relevant features, leaving their columns empty. 8.2.3.5. Links  scikit-learn developer performance documentation Scipy sparse matrix formats documentation ',\n",
       " 'sklearn_8_2_computational_performance\\n8.2. Computational Performance\\nmodules/computing.html#computational-performance\\n 8.3. Parallelism, resource management, and configuration  8.3.1. Parallelism  Some scikit-learn estimators and utilities can parallelize costly operations using multiple CPU cores, thanks to the following components: via the joblib library. In this case the number of threads or processes can be controlled with the parameter. via OpenMP, used in C or Cython code. In addition, some of the numpy routines that are used internally by scikit-learn may also be parallelized if numpy is installed with specific numerical libraries such as MKL, OpenBLAS, or BLIS. We describe these 3 scenarios in the following subsections. 8.3.1.1. Joblib-based parallelism  When the underlying implementation uses joblib, the number of workers (threads or processes) that are spawned in parallel can be controlled via the parameter. Note Where (and how) parallelization happens in the estimators is currently poorly documented. Please help us by improving our docs and tackle issue 14228 ! Joblib is able to support both multi-processing and multi-threading. Whether joblib chooses to spawn a thread or a process depends on the backend that it’s using. Scikit-learn generally relies on the backend, which is joblib’s default backend. Loky is a multi-processing backend. When doing multi-processing, in order to avoid duplicating the memory in each process (which isn’t reasonable with big datasets), joblib will create a memmap that all processes can share, when the data is bigger than 1MB. In some specific cases (when the code that is run in parallel releases the GIL), scikit-learn will indicate to that a multi-threading backend is preferable. As a user, you may control the backend that joblib will use (regardless of what scikit-learn recommends) by using a context manager: Please refer to the joblib’s docs for more details. In practice, whether parallelism is helpful at improving runtime depends on many factors. It is usually a good idea to experiment rather than assuming that increasing the number of workers is always a good thing. In some cases it can be highly detrimental to performance to run multiple copies of some estimators or functions in parallel (see oversubscription below). 8.3.1.2. OpenMP-based parallelism  OpenMP is used to parallelize code written in Cython or C, relying on multi-threading exclusively. By default (and unless joblib is trying to avoid oversubscription), the implementation will use as many threads as possible. You can control the exact number of threads that are used via the environment variable: 8.3.1.3. Parallel Numpy routines from numerical libraries  Scikit-learn relies heavily on NumPy and SciPy, which internally call multi-threaded linear algebra routines implemented in libraries such as MKL, OpenBLAS or BLIS. The number of threads used by the OpenBLAS, MKL or BLIS libraries can be set via the , , and environment variables. Please note that scikit-learn has no direct control over these implementations. Scikit-learn solely relies on Numpy and Scipy. Note At the time of writing (2019), NumPy and SciPy packages distributed on pypi.org (used by ) and on the conda-forge channel are linked with OpenBLAS, while conda packages shipped on the “defaults” channel from anaconda.org are linked by default with MKL. 8.3.1.4. Oversubscription: spawning too many threads  It is generally recommended to avoid using significantly more processes or threads than the number of CPUs on a machine. Over-subscription happens when a program is running too many threads at the same time. Suppose you have a machine with 8 CPUs. Consider a case where you’re running a GridSearchCV (parallelized with joblib) with over a HistGradientBoostingClassifier (parallelized with OpenMP). Each instance of HistGradientBoostingClassifier will spawn 8 threads (since you have 8 CPUs). That’s a total of threads, which leads to oversubscription of physical CPU resources and to scheduling overhead. Oversubscription can arise in the exact same fashion with parallelized routines from MKL, OpenBLAS or BLIS that are nested in joblib calls. Starting from , when the backend is used (which is the default), joblib will tell its child processes to limit the number of threads they can use, so as to avoid oversubscription. In practice the heuristic that joblib uses is to tell the processes to use , via their corresponding environment variable. Back to our example from above, since the joblib backend of GridSearchCV is , each process will only be able to use 1 thread instead of 8, thus mitigating the oversubscription issue. Note that: Manually setting one of the environment variables ( , , , or ) will take precedence over what joblib tries to do. The total number of threads will be . Note that setting this limit will also impact your computations in the main process, which will only use . Joblib exposes a context manager for finer control over the number of threads in its workers (see joblib docs linked below). Joblib is currently unable to avoid oversubscription in a multi-threading context. It can only do so with the backend (which spawns processes). You will find additional details about joblib mitigation of oversubscription in joblib documentation . 8.3.2. Configuration switches  8.3.2.1. Python runtime  sklearn.set_config controls the following behaviors: assume_finite used to skip validation, which enables faster computations but may lead to segmentation faults if the data contains NaNs. working_memory the optimal size of temporary arrays used by some algorithms. 8.3.2.2. Environment variables  These environment variables should be set before importing scikit-learn. SKLEARN_SITE_JOBLIB When this environment variable is set to a non zero value, scikit-learn uses the site joblib rather than its vendored version. Consequently, joblib must be installed for scikit-learn to run. Note that using the site joblib is at your own risks: the versions of scikit-learn and joblib need to be compatible. Currently, joblib 0.11+ is supported. In addition, dumps from joblib.Memory might be incompatible, and you might loose some caches and have to redownload some datasets. Deprecated since version 0.21: As of version 0.21 this parameter has no effect, vendored joblib was removed and site joblib is always used. SKLEARN_ASSUME_FINITE Sets the default value for the argument of sklearn.set_config . SKLEARN_WORKING_MEMORY Sets the default value for the argument of sklearn.set_config . SKLEARN_SEED Sets the seed of the global random generator when running the tests, for reproducibility. SKLEARN_SKIP_NETWORK_TESTS When this environment variable is set to a non zero value, the tests that need network access are skipped. ',\n",
       " 'sklearn_8_3_parallelism_resource_management_and_configuration\\n8.3. Parallelism, resource management, and configuration\\nmodules/computing.html#parallelism-resource-management-and-configuration\\n 8.1. Strategies to scale computationally: bigger data  For some applications the amount of examples, features (or both) and/or the speed at which they need to be processed are challenging for traditional approaches. In these cases scikit-learn has a number of options you can consider to make your system scale. 8.1.1. Scaling with instances using out-of-core learning  Out-of-core (or “external memory”) learning is a technique used to learn from data that cannot fit in a computer’s main memory (RAM). Here is a sketch of a system designed to achieve this goal: a way to stream instances a way to extract features from instances an incremental algorithm 8.1.1.1. Streaming instances  Basically, 1. may be a reader that yields instances from files on a hard drive, a database, from a network stream etc. However, details on how to achieve this are beyond the scope of this documentation. 8.1.1.2. Extracting features  2. could be any relevant way to extract features among the different feature extraction methods supported by scikit-learn. However, when working with data that needs vectorization and where the set of features or values is not known in advance one should take explicit care. A good example is text classification where unknown terms are likely to be found during training. It is possible to use a stateful vectorizer if making multiple passes over the data is reasonable from an application point of view. Otherwise, one can turn up the difficulty by using a stateless feature extractor. Currently the preferred way to do this is to use the so-called hashing trick as implemented by sklearn.feature_extraction.FeatureHasher for datasets with categorical variables represented as list of Python dicts or sklearn.feature_extraction.text.HashingVectorizer for text documents. 8.1.1.3. Incremental learning  Finally, for 3. we have a number of options inside scikit-learn. Although not all algorithms can learn incrementally (i.e. without seeing all the instances at once), all estimators implementing the API are candidates. Actually, the ability to learn incrementally from a mini-batch of instances (sometimes called “online learning”) is key to out-of-core learning as it guarantees that at any given time there will be only a small amount of instances in the main memory. Choosing a good size for the mini-batch that balances relevancy and memory footprint could involve some tuning 1 . Here is a list of incremental estimators for different tasks: Classification sklearn.naive_bayes.MultinomialNB sklearn.naive_bayes.BernoulliNB sklearn.linear_model.Perceptron sklearn.linear_model.SGDClassifier sklearn.linear_model.PassiveAggressiveClassifier sklearn.neural_network.MLPClassifier Regression sklearn.linear_model.SGDRegressor sklearn.linear_model.PassiveAggressiveRegressor sklearn.neural_network.MLPRegressor Clustering sklearn.cluster.MiniBatchKMeans sklearn.cluster.Birch Decomposition / feature Extraction sklearn.decomposition.MiniBatchDictionaryLearning sklearn.decomposition.IncrementalPCA sklearn.decomposition.LatentDirichletAllocation Preprocessing sklearn.preprocessing.StandardScaler sklearn.preprocessing.MinMaxScaler sklearn.preprocessing.MaxAbsScaler For classification, a somewhat important thing to note is that although a stateless feature extraction routine may be able to cope with new/unseen attributes, the incremental learner itself may be unable to cope with new/unseen targets classes. In this case you have to pass all the possible classes to the first call using the parameter. Another aspect to consider when choosing a proper algorithm is that not all of them put the same importance on each example over time. Namely, the is still sensitive to badly labeled examples even after many examples whereas the and families are more robust to this kind of artifacts. Conversely, the latter also tend to give less importance to remarkably different, yet properly labeled examples when they come late in the stream as their learning rate decreases over time. 8.1.1.4. Examples  Finally, we have a full-fledged example of Out-of-core classification of text documents . It is aimed at providing a starting point for people wanting to build out-of-core learning systems and demonstrates most of the notions discussed above. Furthermore, it also shows the evolution of the performance of different algorithms with the number of processed examples. Now looking at the computation time of the different parts, we see that the vectorization is much more expensive than learning itself. From the different algorithms, is the most expensive, but its overhead can be mitigated by increasing the size of the mini-batches (exercise: change to 100 and 10000 in the program and compare). 8.1.1.5. Notes  1 Depending on the algorithm the mini-batch size can influence results or not. SGD*, PassiveAggressive*, and discrete NaiveBayes are truly online and are not affected by batch size. Conversely, MiniBatchKMeans convergence rate is affected by the batch size. Also, its memory footprint can vary dramatically with batch size. ',\n",
       " 'sklearn_8_3_parallelism_resource_management_and_configuration\\n8.3. Parallelism, resource management, and configuration\\nmodules/computing.html#parallelism-resource-management-and-configuration\\n 8.2. Computational Performance  For some applications the performance (mainly latency and throughput at prediction time) of estimators is crucial. It may also be of interest to consider the training throughput but this is often less important in a production setup (where it often takes place offline). We will review here the orders of magnitude you can expect from a number of scikit-learn estimators in different contexts and provide some tips and tricks for overcoming performance bottlenecks. Prediction latency is measured as the elapsed time necessary to make a prediction (e.g. in micro-seconds). Latency is often viewed as a distribution and operations engineers often focus on the latency at a given percentile of this distribution (e.g. the 90 percentile). Prediction throughput is defined as the number of predictions the software can deliver in a given amount of time (e.g. in predictions per second). An important aspect of performance optimization is also that it can hurt prediction accuracy. Indeed, simpler models (e.g. linear instead of non-linear, or with fewer parameters) often run faster but are not always able to take into account the same exact properties of the data as more complex ones. 8.2.1. Prediction Latency  One of the most straight-forward concerns one may have when using/choosing a machine learning toolkit is the latency at which predictions can be made in a production environment. The main factors that influence the prediction latency are Number of features Input data representation and sparsity Model complexity Feature extraction A last major parameter is also the possibility to do predictions in bulk or one-at-a-time mode. 8.2.1.1. Bulk versus Atomic mode  In general doing predictions in bulk (many instances at the same time) is more efficient for a number of reasons (branching predictability, CPU cache, linear algebra libraries optimizations etc.). Here we see on a setting with few features that independently of estimator choice the bulk mode is always faster, and for some of them by 1 to 2 orders of magnitude: To benchmark different estimators for your case you can simply change the parameter in this example: Prediction Latency . This should give you an estimate of the order of magnitude of the prediction latency. 8.2.1.2. Configuring Scikit-learn for reduced validation overhead  Scikit-learn does some validation on data that increases the overhead per call to and similar functions. In particular, checking that features are finite (not NaN or infinite) involves a full pass over the data. If you ensure that your data is acceptable, you may suppress checking for finiteness by setting the environment variable to a non-empty string before importing scikit-learn, or configure it in Python with sklearn.set_config . For more control than these global settings, a config_context allows you to set this configuration within a specified context: Note that this will affect all uses of sklearn.utils.assert_all_finite within the context. 8.2.1.3. Influence of the Number of Features  Obviously when the number of features increases so does the memory consumption of each example. Indeed, for a matrix of instances with features, the space complexity is in . From a computing perspective it also means that the number of basic operations (e.g., multiplications for vector-matrix products in linear models) increases too. Here is a graph of the evolution of the prediction latency with the number of features: Overall you can expect the prediction time to increase at least linearly with the number of features (non-linear cases can happen depending on the global memory footprint and estimator). 8.2.1.4. Influence of the Input Data Representation  Scipy provides sparse matrix data structures which are optimized for storing sparse data. The main feature of sparse formats is that you don’t store zeros so if your data is sparse then you use much less memory. A non-zero value in a sparse ( CSR or CSC ) representation will only take on average one 32bit integer position + the 64 bit floating point value + an additional 32bit per row or column in the matrix. Using sparse input on a dense (or sparse) linear model can speedup prediction by quite a bit as only the non zero valued features impact the dot product and thus the model predictions. Hence if you have 100 non zeros in 1e6 dimensional space, you only need 100 multiply and add operation instead of 1e6. Calculation over a dense representation, however, may leverage highly optimised vector operations and multithreading in BLAS, and tends to result in fewer CPU cache misses. So the sparsity should typically be quite high (10% non-zeros max, to be checked depending on the hardware) for the sparse input representation to be faster than the dense input representation on a machine with many CPUs and an optimized BLAS implementation. Here is sample code to test the sparsity of your input: As a rule of thumb you can consider that if the sparsity ratio is greater than 90% you can probably benefit from sparse formats. Check Scipy’s sparse matrix formats documentation for more information on how to build (or convert your data to) sparse matrix formats. Most of the time the and formats work best. 8.2.1.5. Influence of the Model Complexity  Generally speaking, when model complexity increases, predictive power and latency are supposed to increase. Increasing predictive power is usually interesting, but for many applications we would better not increase prediction latency too much. We will now review this idea for different families of supervised models. For sklearn.linear_model (e.g. Lasso, ElasticNet, SGDClassifier/Regressor, Ridge & RidgeClassifier, PassiveAggressiveClassifier/Regressor, LinearSVC, LogisticRegression…) the decision function that is applied at prediction time is the same (a dot product) , so latency should be equivalent. Here is an example using sklearn.linear_model.SGDClassifier with the penalty. The regularization strength is globally controlled by the parameter. With a sufficiently high , one can then increase the parameter of to enforce various levels of sparsity in the model coefficients. Higher sparsity here is interpreted as less model complexity as we need fewer coefficients to describe it fully. Of course sparsity influences in turn the prediction time as the sparse dot-product takes time roughly proportional to the number of non-zero coefficients. For the sklearn.svm family of algorithms with a non-linear kernel, the latency is tied to the number of support vectors (the fewer the faster). Latency and throughput should (asymptotically) grow linearly with the number of support vectors in a SVC or SVR model. The kernel will also influence the latency as it is used to compute the projection of the input vector once per support vector. In the following graph the parameter of sklearn.svm.NuSVR was used to influence the number of support vectors. For sklearn.ensemble of trees (e.g. RandomForest, GBT, ExtraTrees etc) the number of trees and their depth play the most important role. Latency and throughput should scale linearly with the number of trees. In this case we used directly the parameter of sklearn.ensemble.gradient_boosting.GradientBoostingRegressor . In any case be warned that decreasing model complexity can hurt accuracy as mentioned above. For instance a non-linearly separable problem can be handled with a speedy linear model but prediction power will very likely suffer in the process. 8.2.1.6. Feature Extraction Latency  Most scikit-learn models are usually pretty fast as they are implemented either with compiled Cython extensions or optimized computing libraries. On the other hand, in many real world applications the feature extraction process (i.e. turning raw data like database rows or network packets into numpy arrays) governs the overall prediction time. For example on the Reuters text classification task the whole preparation (reading and parsing SGML files, tokenizing the text and hashing it into a common vector space) is taking 100 to 500 times more time than the actual prediction code, depending on the chosen model. In many cases it is thus recommended to carefully time and profile your feature extraction code as it may be a good place to start optimizing when your overall latency is too slow for your application. 8.2.2. Prediction Throughput  Another important metric to care about when sizing production systems is the throughput i.e. the number of predictions you can make in a given amount of time. Here is a benchmark from the Prediction Latency example that measures this quantity for a number of estimators on synthetic data: These throughputs are achieved on a single process. An obvious way to increase the throughput of your application is to spawn additional instances (usually processes in Python because of the GIL ) that share the same model. One might also add machines to spread the load. A detailed explanation on how to achieve this is beyond the scope of this documentation though. 8.2.3. Tips and Tricks  8.2.3.1. Linear algebra libraries  As scikit-learn relies heavily on Numpy/Scipy and linear algebra in general it makes sense to take explicit care of the versions of these libraries. Basically, you ought to make sure that Numpy is built using an optimized BLAS / LAPACK library. Not all models benefit from optimized BLAS and Lapack implementations. For instance models based on (randomized) decision trees typically do not rely on BLAS calls in their inner loops, nor do kernel SVMs ( , , , ). On the other hand a linear model implemented with a BLAS DGEMM call (via ) will typically benefit hugely from a tuned BLAS implementation and lead to orders of magnitude speedup over a non-optimized BLAS. You can display the BLAS / LAPACK implementation used by your NumPy / SciPy / scikit-learn install with the following commands: Optimized BLAS / LAPACK implementations include: Atlas (need hardware specific tuning by rebuilding on the target machine) OpenBLAS MKL Apple Accelerate and vecLib frameworks (OSX only) More information can be found on the Scipy install page and in this blog post from Daniel Nouri which has some nice step by step install instructions for Debian / Ubuntu. 8.2.3.2. Limiting Working Memory  Some calculations when implemented using standard numpy vectorized operations involve using a large amount of temporary memory. This may potentially exhaust system memory. Where computations can be performed in fixed-memory chunks, we attempt to do so, and allow the user to hint at the maximum size of this working memory (defaulting to 1GB) using sklearn.set_config or config_context . The following suggests to limit temporary working memory to 128 MiB: An example of a chunked operation adhering to this setting is metric.pairwise_distances_chunked , which facilitates computing row-wise reductions of a pairwise distance matrix. 8.2.3.3. Model Compression  Model compression in scikit-learn only concerns linear models for the moment. In this context it means that we want to control the model sparsity (i.e. the number of non-zero coordinates in the model vectors). It is generally a good idea to combine model sparsity with sparse input data representation. Here is sample code that illustrates the use of the method: In this example we prefer the penalty as it is often a good compromise between model compactness and prediction power. One can also further tune the parameter (in combination with the regularization strength ) to control this tradeoff. A typical benchmark on synthetic data yields a >30% decrease in latency when both the model and input are sparse (with 0.000024 and 0.027400 non-zero coefficients ratio respectively). Your mileage may vary depending on the sparsity and size of your data and model. Furthermore, sparsifying can be very useful to reduce the memory usage of predictive models deployed on production servers. 8.2.3.4. Model Reshaping  Model reshaping consists in selecting only a portion of the available features to fit a model. In other words, if a model discards features during the learning phase we can then strip those from the input. This has several benefits. Firstly it reduces memory (and therefore time) overhead of the model itself. It also allows to discard explicit feature selection components in a pipeline once we know which features to keep from a previous run. Finally, it can help reduce processing time and I/O usage upstream in the data access and feature extraction layers by not collecting and building features that are discarded by the model. For instance if the raw data come from a database, it can make it possible to write simpler and faster queries or reduce I/O usage by making the queries return lighter records. At the moment, reshaping needs to be performed manually in scikit-learn. In the case of sparse input (particularly in format), it is generally sufficient to not generate the relevant features, leaving their columns empty. 8.2.3.5. Links  scikit-learn developer performance documentation Scipy sparse matrix formats documentation ',\n",
       " 'sklearn_8_3_parallelism_resource_management_and_configuration\\n8.3. Parallelism, resource management, and configuration\\nmodules/computing.html#parallelism-resource-management-and-configuration\\n 8.3. Parallelism, resource management, and configuration  8.3.1. Parallelism  Some scikit-learn estimators and utilities can parallelize costly operations using multiple CPU cores, thanks to the following components: via the joblib library. In this case the number of threads or processes can be controlled with the parameter. via OpenMP, used in C or Cython code. In addition, some of the numpy routines that are used internally by scikit-learn may also be parallelized if numpy is installed with specific numerical libraries such as MKL, OpenBLAS, or BLIS. We describe these 3 scenarios in the following subsections. 8.3.1.1. Joblib-based parallelism  When the underlying implementation uses joblib, the number of workers (threads or processes) that are spawned in parallel can be controlled via the parameter. Note Where (and how) parallelization happens in the estimators is currently poorly documented. Please help us by improving our docs and tackle issue 14228 ! Joblib is able to support both multi-processing and multi-threading. Whether joblib chooses to spawn a thread or a process depends on the backend that it’s using. Scikit-learn generally relies on the backend, which is joblib’s default backend. Loky is a multi-processing backend. When doing multi-processing, in order to avoid duplicating the memory in each process (which isn’t reasonable with big datasets), joblib will create a memmap that all processes can share, when the data is bigger than 1MB. In some specific cases (when the code that is run in parallel releases the GIL), scikit-learn will indicate to that a multi-threading backend is preferable. As a user, you may control the backend that joblib will use (regardless of what scikit-learn recommends) by using a context manager: Please refer to the joblib’s docs for more details. In practice, whether parallelism is helpful at improving runtime depends on many factors. It is usually a good idea to experiment rather than assuming that increasing the number of workers is always a good thing. In some cases it can be highly detrimental to performance to run multiple copies of some estimators or functions in parallel (see oversubscription below). 8.3.1.2. OpenMP-based parallelism  OpenMP is used to parallelize code written in Cython or C, relying on multi-threading exclusively. By default (and unless joblib is trying to avoid oversubscription), the implementation will use as many threads as possible. You can control the exact number of threads that are used via the environment variable: 8.3.1.3. Parallel Numpy routines from numerical libraries  Scikit-learn relies heavily on NumPy and SciPy, which internally call multi-threaded linear algebra routines implemented in libraries such as MKL, OpenBLAS or BLIS. The number of threads used by the OpenBLAS, MKL or BLIS libraries can be set via the , , and environment variables. Please note that scikit-learn has no direct control over these implementations. Scikit-learn solely relies on Numpy and Scipy. Note At the time of writing (2019), NumPy and SciPy packages distributed on pypi.org (used by ) and on the conda-forge channel are linked with OpenBLAS, while conda packages shipped on the “defaults” channel from anaconda.org are linked by default with MKL. 8.3.1.4. Oversubscription: spawning too many threads  It is generally recommended to avoid using significantly more processes or threads than the number of CPUs on a machine. Over-subscription happens when a program is running too many threads at the same time. Suppose you have a machine with 8 CPUs. Consider a case where you’re running a GridSearchCV (parallelized with joblib) with over a HistGradientBoostingClassifier (parallelized with OpenMP). Each instance of HistGradientBoostingClassifier will spawn 8 threads (since you have 8 CPUs). That’s a total of threads, which leads to oversubscription of physical CPU resources and to scheduling overhead. Oversubscription can arise in the exact same fashion with parallelized routines from MKL, OpenBLAS or BLIS that are nested in joblib calls. Starting from , when the backend is used (which is the default), joblib will tell its child processes to limit the number of threads they can use, so as to avoid oversubscription. In practice the heuristic that joblib uses is to tell the processes to use , via their corresponding environment variable. Back to our example from above, since the joblib backend of GridSearchCV is , each process will only be able to use 1 thread instead of 8, thus mitigating the oversubscription issue. Note that: Manually setting one of the environment variables ( , , , or ) will take precedence over what joblib tries to do. The total number of threads will be . Note that setting this limit will also impact your computations in the main process, which will only use . Joblib exposes a context manager for finer control over the number of threads in its workers (see joblib docs linked below). Joblib is currently unable to avoid oversubscription in a multi-threading context. It can only do so with the backend (which spawns processes). You will find additional details about joblib mitigation of oversubscription in joblib documentation . 8.3.2. Configuration switches  8.3.2.1. Python runtime  sklearn.set_config controls the following behaviors: assume_finite used to skip validation, which enables faster computations but may lead to segmentation faults if the data contains NaNs. working_memory the optimal size of temporary arrays used by some algorithms. 8.3.2.2. Environment variables  These environment variables should be set before importing scikit-learn. SKLEARN_SITE_JOBLIB When this environment variable is set to a non zero value, scikit-learn uses the site joblib rather than its vendored version. Consequently, joblib must be installed for scikit-learn to run. Note that using the site joblib is at your own risks: the versions of scikit-learn and joblib need to be compatible. Currently, joblib 0.11+ is supported. In addition, dumps from joblib.Memory might be incompatible, and you might loose some caches and have to redownload some datasets. Deprecated since version 0.21: As of version 0.21 this parameter has no effect, vendored joblib was removed and site joblib is always used. SKLEARN_ASSUME_FINITE Sets the default value for the argument of sklearn.set_config . SKLEARN_WORKING_MEMORY Sets the default value for the argument of sklearn.set_config . SKLEARN_SEED Sets the seed of the global random generator when running the tests, for reproducibility. SKLEARN_SKIP_NETWORK_TESTS When this environment variable is set to a non zero value, the tests that need network access are skipped. ']"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extractLabels(samples, package, language):\n",
    "    reList = []\n",
    "    for x in samples:\n",
    "        if x.strip() == 'text':\n",
    "            continue\n",
    "        y = x.split(\"\\n\")\n",
    "        page_id = y[0]\n",
    "        section = y[1]\n",
    "        url = y[2]\n",
    "        text = y[3]\n",
    "        try:\n",
    "            field_name = re.search(\"^[0-9.]*[ ][A-Z][-a-zA-Z]*(?:\\s+[A-Z][-a-zA-Z]*)*\", text.strip()).group(0)\n",
    "            if len(field_name) > 30:\n",
    "                field_name = \" \".join(text.strip().split(\" \")[0:4])\n",
    "        except AttributeError:\n",
    "            field_name = \" \".join(text.strip().split(\" \")[0:4])\n",
    "        reList.append((package, language, page_id, section, url, field_name))\n",
    "    \n",
    "    return reList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{base_path}sklearn_labels.csv', 'w') as f:\n",
    "    for tup in extractLabels(samples, 'sklearn', 'Python'):\n",
    "        f.write('|'.join(tup) + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Caret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "caret_glob = sorted(glob.glob('caret/*'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "caret_samples = []\n",
    "for file in caret_glob:\n",
    "    with open(file) as f:\n",
    "        caret_samples.append(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "141"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(caret_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open(f'{base_path}caret.csv', 'w') as f:\n",
    "    doc_text = \"text\\n\" + \"\\n\".join(['\"' + sample.replace('\\n', ' ') + '\"' for sample in caret_samples])\n",
    "    f.write(doc_text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{base_path}caret_labels.csv', 'w') as f:\n",
    "    for tup in extractLabels(caret_samples, 'caret', 'R'):\n",
    "        f.write('|'.join(tup) + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "numpy_glob = sorted(glob.glob('numpy/*'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "numpy_samples = []\n",
    "for file in numpy_glob:\n",
    "    with open(file) as f:\n",
    "        numpy_samples.append(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(numpy_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open(f'{base_path}numpy.csv', 'w') as f:\n",
    "    doc_text = \"text\\n\" + \"\\n\".join(['\"' + sample.replace('\\n', ' ') + '\"' for sample in numpy_samples])\n",
    "    f.write(doc_text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{base_path}numpy_labels.csv', 'w') as f:\n",
    "    for tup in extractLabels(numpy_samples, 'numpy', 'Python'):\n",
    "        f.write('|'.join(tup) + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SciPy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy_glob = sorted(glob.glob('scipy/*'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy_samples = []\n",
    "for file in scipy_glob:\n",
    "    with open(file) as f:\n",
    "        scipy_samples.append(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(scipy_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "with open(f'{base_path}scipy.csv', 'w') as f:\n",
    "    doc_text = \"text\\n\" + \"\\n\".join(['\"' + sample.replace('\\n', ' ') + '\"' for sample in scipy_samples])\n",
    "    f.write(doc_text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{base_path}scipy_labels.csv', 'w') as f:\n",
    "    for tup in extractLabels(scipy_samples, 'scipy', 'Python'):\n",
    "        f.write('|'.join(tup) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['all_data_model.dbow_numnoisewords.2_vecdim.100_batchsize.32_lr.0.001000_epoch.95_loss.0.755801.csv',\n",
       " 'caret.csv',\n",
       " 'numpy.csv',\n",
       " 'sklearn.csv',\n",
       " 'all_data.csv',\n",
       " 'scipy.csv',\n",
       " 'all_data_model.dbow_numnoisewords.2_vecdim.100_batchsize.32_lr.0.001000_epoch.97_loss.0.750086.csv']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(base_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "caret.csv\n",
      "text\n",
      "141\n",
      "numpy.csv\n",
      "text\n",
      "22\n",
      "sklearn.csv\n",
      "text\n",
      "242\n",
      "scipy.csv\n",
      "text\n",
      "70\n",
      "476\n"
     ]
    }
   ],
   "source": [
    "filenames = [\n",
    "    'caret.csv',\n",
    "    'numpy.csv',\n",
    "    'sklearn.csv',\n",
    "    'scipy.csv'\n",
    "]\n",
    "with open(base_path + 'all_data2.csv', 'w') as outfile:\n",
    "    j = 0\n",
    "    outfile.write(\"text\\n\")\n",
    "    j +=1 \n",
    "    for fname in filenames:\n",
    "        i = 0\n",
    "        print(fname)\n",
    "        with open(base_path + fname) as infile:\n",
    "            for line in infile:\n",
    "                \n",
    "                if line.strip() == 'text':\n",
    "                    print(\"text\")\n",
    "                    continue\n",
    "                outfile.write(line.strip() + '\\n')\n",
    "                i += 1\n",
    "                j +=1\n",
    "        print(i)\n",
    "    print(j)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "caret_labels.csv\n",
      "141\n",
      "numpy_labels.csv\n",
      "22\n",
      "sklearn_labels.csv\n",
      "242\n",
      "scipy_labels.csv\n",
      "70\n"
     ]
    }
   ],
   "source": [
    "filenames = [\n",
    "    'caret_labels.csv',\n",
    "    'numpy_labels.csv',\n",
    "    'sklearn_labels.csv',\n",
    "    'scipy_labels.csv'\n",
    "]\n",
    "with open(base_path + 'all_labels.csv', 'w') as outfile:\n",
    "    outfile.write(\"text\\n\")\n",
    "    for fname in filenames:\n",
    "        i = 0\n",
    "        print(fname)\n",
    "        with open(base_path + fname) as infile:\n",
    "            for line in infile:\n",
    "                \n",
    "                if not line: continue\n",
    "                if line.strip() == 'text':\n",
    "                    print(\"text\")\n",
    "                    continue\n",
    "                i += 1\n",
    "                outfile.write(line)\n",
    "            print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_line = []\n",
    "with open(base_path + 'all_data2.csv') as f:\n",
    "    for line in f:\n",
    "        data_line.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "label_line = []\n",
    "with open(base_path + 'all_labels.csv') as f:\n",
    "    for line in f:\n",
    "        label_line.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[5,\n",
       " 65,\n",
       " 69,\n",
       " 71,\n",
       " 75,\n",
       " 77,\n",
       " 80,\n",
       " 81,\n",
       " 81,\n",
       " 81,\n",
       " 82,\n",
       " 82,\n",
       " 83,\n",
       " 83,\n",
       " 83,\n",
       " 83,\n",
       " 83,\n",
       " 83,\n",
       " 84,\n",
       " 84,\n",
       " 84,\n",
       " 85,\n",
       " 85,\n",
       " 86,\n",
       " 86,\n",
       " 87,\n",
       " 87,\n",
       " 87,\n",
       " 88,\n",
       " 88,\n",
       " 89,\n",
       " 89,\n",
       " 89,\n",
       " 90,\n",
       " 90,\n",
       " 92,\n",
       " 92,\n",
       " 94,\n",
       " 94,\n",
       " 95,\n",
       " 95,\n",
       " 95,\n",
       " 95,\n",
       " 96,\n",
       " 97,\n",
       " 97,\n",
       " 98,\n",
       " 98,\n",
       " 98,\n",
       " 99,\n",
       " 99,\n",
       " 99,\n",
       " 100,\n",
       " 100,\n",
       " 100,\n",
       " 100,\n",
       " 100,\n",
       " 101,\n",
       " 101,\n",
       " 101,\n",
       " 101,\n",
       " 102,\n",
       " 102,\n",
       " 102,\n",
       " 103,\n",
       " 103,\n",
       " 103,\n",
       " 103,\n",
       " 103,\n",
       " 103,\n",
       " 103,\n",
       " 103,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 104,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 105,\n",
       " 106,\n",
       " 106,\n",
       " 106,\n",
       " 106,\n",
       " 107,\n",
       " 107,\n",
       " 107,\n",
       " 107,\n",
       " 107,\n",
       " 107,\n",
       " 107,\n",
       " 107,\n",
       " 107,\n",
       " 107,\n",
       " 107,\n",
       " 108,\n",
       " 108,\n",
       " 109,\n",
       " 109,\n",
       " 109,\n",
       " 109,\n",
       " 109,\n",
       " 110,\n",
       " 110,\n",
       " 110,\n",
       " 110,\n",
       " 110,\n",
       " 110,\n",
       " 111,\n",
       " 111,\n",
       " 111,\n",
       " 111,\n",
       " 111,\n",
       " 111,\n",
       " 111,\n",
       " 111,\n",
       " 111,\n",
       " 111,\n",
       " 111,\n",
       " 112,\n",
       " 112,\n",
       " 112,\n",
       " 112,\n",
       " 112,\n",
       " 112,\n",
       " 112,\n",
       " 112,\n",
       " 112,\n",
       " 113,\n",
       " 113,\n",
       " 113,\n",
       " 113,\n",
       " 113,\n",
       " 113,\n",
       " 113,\n",
       " 113,\n",
       " 113,\n",
       " 113,\n",
       " 114,\n",
       " 114,\n",
       " 114,\n",
       " 114,\n",
       " 114,\n",
       " 114,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 115,\n",
       " 116,\n",
       " 116,\n",
       " 116,\n",
       " 116,\n",
       " 116,\n",
       " 116,\n",
       " 116,\n",
       " 116,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 117,\n",
       " 118,\n",
       " 118,\n",
       " 118,\n",
       " 118,\n",
       " 119,\n",
       " 119,\n",
       " 119,\n",
       " 119,\n",
       " 119,\n",
       " 119,\n",
       " 119,\n",
       " 119,\n",
       " 119,\n",
       " 119,\n",
       " 119,\n",
       " 120,\n",
       " 120,\n",
       " 120,\n",
       " 120,\n",
       " 120,\n",
       " 120,\n",
       " 120,\n",
       " 120,\n",
       " 121,\n",
       " 121,\n",
       " 121,\n",
       " 121,\n",
       " 121,\n",
       " 121,\n",
       " 121,\n",
       " 121,\n",
       " 122,\n",
       " 122,\n",
       " 122,\n",
       " 122,\n",
       " 122,\n",
       " 122,\n",
       " 122,\n",
       " 122,\n",
       " 122,\n",
       " 123,\n",
       " 123,\n",
       " 123,\n",
       " 123,\n",
       " 123,\n",
       " 123,\n",
       " 124,\n",
       " 124,\n",
       " 124,\n",
       " 124,\n",
       " 124,\n",
       " 124,\n",
       " 124,\n",
       " 124,\n",
       " 124,\n",
       " 124,\n",
       " 125,\n",
       " 125,\n",
       " 125,\n",
       " 125,\n",
       " 125,\n",
       " 125,\n",
       " 125,\n",
       " 125,\n",
       " 126,\n",
       " 126,\n",
       " 126,\n",
       " 126,\n",
       " 126,\n",
       " 126,\n",
       " 126,\n",
       " 126,\n",
       " 126,\n",
       " 127,\n",
       " 127,\n",
       " 127,\n",
       " 127,\n",
       " 127,\n",
       " 127,\n",
       " 127,\n",
       " 127,\n",
       " 128,\n",
       " 128,\n",
       " 128,\n",
       " 128,\n",
       " 128,\n",
       " 128,\n",
       " 128,\n",
       " 128,\n",
       " 128,\n",
       " 129,\n",
       " 129,\n",
       " 129,\n",
       " 129,\n",
       " 129,\n",
       " 129,\n",
       " 130,\n",
       " 130,\n",
       " 130,\n",
       " 131,\n",
       " 131,\n",
       " 131,\n",
       " 131,\n",
       " 131,\n",
       " 131,\n",
       " 131,\n",
       " 131,\n",
       " 131,\n",
       " 132,\n",
       " 132,\n",
       " 132,\n",
       " 132,\n",
       " 132,\n",
       " 132,\n",
       " 132,\n",
       " 132,\n",
       " 133,\n",
       " 133,\n",
       " 133,\n",
       " 134,\n",
       " 134,\n",
       " 134,\n",
       " 134,\n",
       " 134,\n",
       " 134,\n",
       " 134,\n",
       " 135,\n",
       " 135,\n",
       " 136,\n",
       " 136,\n",
       " 136,\n",
       " 136,\n",
       " 136,\n",
       " 136,\n",
       " 137,\n",
       " 137,\n",
       " 137,\n",
       " 137,\n",
       " 137,\n",
       " 138,\n",
       " 138,\n",
       " 138,\n",
       " 139,\n",
       " 139,\n",
       " 139,\n",
       " 140,\n",
       " 140,\n",
       " 141,\n",
       " 141,\n",
       " 141,\n",
       " 141,\n",
       " 141,\n",
       " 141,\n",
       " 141,\n",
       " 141,\n",
       " 141,\n",
       " 142,\n",
       " 142,\n",
       " 143,\n",
       " 143,\n",
       " 145,\n",
       " 145,\n",
       " 146,\n",
       " 146,\n",
       " 146,\n",
       " 147,\n",
       " 147,\n",
       " 148,\n",
       " 148,\n",
       " 148,\n",
       " 148,\n",
       " 149,\n",
       " 149,\n",
       " 149,\n",
       " 149,\n",
       " 150,\n",
       " 150,\n",
       " 151,\n",
       " 151,\n",
       " 152,\n",
       " 152,\n",
       " 152,\n",
       " 152,\n",
       " 153,\n",
       " 153,\n",
       " 153,\n",
       " 154,\n",
       " 155,\n",
       " 155,\n",
       " 155,\n",
       " 155,\n",
       " 155,\n",
       " 156,\n",
       " 156,\n",
       " 156,\n",
       " 157,\n",
       " 157,\n",
       " 157,\n",
       " 157,\n",
       " 157,\n",
       " 158,\n",
       " 158,\n",
       " 159,\n",
       " 159,\n",
       " 159,\n",
       " 159,\n",
       " 159,\n",
       " 160,\n",
       " 160,\n",
       " 160,\n",
       " 160,\n",
       " 161,\n",
       " 161,\n",
       " 162,\n",
       " 163,\n",
       " 163,\n",
       " 165,\n",
       " 165,\n",
       " 165,\n",
       " 165,\n",
       " 166,\n",
       " 167,\n",
       " 168,\n",
       " 168,\n",
       " 169,\n",
       " 170,\n",
       " 171,\n",
       " 171,\n",
       " 171,\n",
       " 171,\n",
       " 171,\n",
       " 172,\n",
       " 172,\n",
       " 173,\n",
       " 174,\n",
       " 174,\n",
       " 174,\n",
       " 175,\n",
       " 176,\n",
       " 176,\n",
       " 176,\n",
       " 178,\n",
       " 179,\n",
       " 180,\n",
       " 180,\n",
       " 181,\n",
       " 183,\n",
       " 184,\n",
       " 185,\n",
       " 185,\n",
       " 185,\n",
       " 186,\n",
       " 186,\n",
       " 186,\n",
       " 190,\n",
       " 192,\n",
       " 192,\n",
       " 194,\n",
       " 199,\n",
       " 199,\n",
       " 199,\n",
       " 203,\n",
       " 206,\n",
       " 206,\n",
       " 208,\n",
       " 209,\n",
       " 216,\n",
       " 216,\n",
       " 216,\n",
       " 217,\n",
       " 223,\n",
       " 224,\n",
       " 224,\n",
       " 232,\n",
       " 239]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted([len(x) for x in label_line])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "476\n",
      "473\n"
     ]
    }
   ],
   "source": [
    "print(len(label_line)); print(len(data_line))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output vs Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_lines = []\n",
    "with open('model_data/data/all_data.csv') as f:\n",
    "    for line in f:\n",
    "        input_lines.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "474"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(input_lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'text\\n'"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_lines[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_lines = []\n",
    "with open('model_data/data/all_data_model.dbow_numnoisewords.2_vecdim.100_batchsize.32_lr.0.001000_epoch.97_loss.0.750086.csv') as f:\n",
    "    for line in f:\n",
    "        output_lines.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "486"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(output_lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'d0,d1,d2,d3,d4,d5,d6,d7,d8,d9,d10,d11,d12,d13,d14,d15,d16,d17,d18,d19,d20,d21,d22,d23,d24,d25,d26,d27,d28,d29,d30,d31,d32,d33,d34,d35,d36,d37,d38,d39,d40,d41,d42,d43,d44,d45,d46,d47,d48,d49,d50,d51,d52,d53,d54,d55,d56,d57,d58,d59,d60,d61,d62,d63,d64,d65,d66,d67,d68,d69,d70,d71,d72,d73,d74,d75,d76,d77,d78,d79,d80,d81,d82,d83,d84,d85,d86,d87,d88,d89,d90,d91,d92,d93,d94,d95,d96,d97,d98,d99\\n'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_lines[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['sklearn_2_1_gaussian_mixture_models',\n",
       " '2.1.',\n",
       " 'Gaussian',\n",
       " 'mixture',\n",
       " 'models',\n",
       " 'modules/mixture.html',\n",
       " '',\n",
       " '2.1.2.',\n",
       " 'Variational',\n",
       " 'Bayesian',\n",
       " 'Gaussian',\n",
       " 'Mixture',\n",
       " '',\n",
       " 'The',\n",
       " 'BayesianGaussianMixture']"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_lines[100].strip('\"').split(\" \")[:15]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "sklearn_index = []\n",
    "for sample in samples:\n",
    "\n",
    "    sample_split = sample.split(\"\\n\")\n",
    "    page_id = sample_split[0]\n",
    "    page_name = sample_split[1]\n",
    "    link = sample_split[2]\n",
    "    text = sample_split[3]\n",
    "    try:\n",
    "        field_name = re.search(\"^[0-9.]*[ ][A-Z][-a-zA-Z]*(?:\\s+[A-Z][-a-zA-Z]*)*\", text.strip()).group(0)\n",
    "    except AttributeError:\n",
    "        field_name = \" \".join(text.strip().split(\" \")[0:10])\n",
    "    sklearn_index.append((page_id, page_name, link, field_name))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "243"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sklearn_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "caret_index = []\n",
    "for caret_sample in caret_samples:\n",
    "    sample_split = caret_sample.split(\"\\n\")\n",
    "    page_id = sample_split[0]\n",
    "    page_name = sample_split[1]\n",
    "    link = sample_split[2]\n",
    "    text = sample_split[3]\n",
    "\n",
    "    field_name = re.search(\"^[0-9.]*[ ][A-Z][-a-zA-Z]*(?:\\s+[A-Z][-a-zA-Z]*)*\", text.strip()).group(0)\n",
    "    caret_index.append((page_id, page_name, link, field_name))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('caret_11_subsampling_for_class_imbalances',\n",
       " '11 Subsampling For Class Imbalances',\n",
       " 'subsampling-for-class-imbalances.html',\n",
       " '11.1 Subsampling')"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'7.0.9 Distance Weighted Discrimination'"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "re.search(\"^[0-9.]*[ ][A-Z][-a-zA-Z]*(?:\\s+[A-Z][-a-zA-Z]*)*\", text.strip()).group(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched = []\n",
    "for line_idx in range(len(output_lines)):\n",
    "    matched.append((input_lines[line_idx], output_lines[line_idx]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_clean = [x for x in matched if not re.match(\".*[A-Z].*\", x[1])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "matched_clean_vectors = [x[1].split(\",\") for x in matched_clean[1:]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "vec_matrix = np.array(matched_clean_vectors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(481, 100)"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vec_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial import distance\n",
    "\n",
    "distances = distance.cdist([vec_matrix[5]], vec_matrix, \"cosine\")[0]\n",
    "# min_index = np.argmin(distances)\n",
    "ind = np.argpartition(a, 5)[:5]\n",
    "sorted_ind = ind[np.argsort(distances[ind])]\n",
    "min_distances = distances[sorted_ind]\n",
    "max_similarity = [1 - x for x in min_distances]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.6670743594734331,\n",
       " 0.6652293382107732,\n",
       " 0.6196681421232006,\n",
       " 0.5960319745742784,\n",
       " 0.5836655087197274]"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv3",
   "language": "python",
   "name": ".venv3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
